// igniteui-react-charts/ES5/AnchoredRadialBucketCalculator_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { RadialBucketCalculator } from "./RadialBucketCalculator";
import { Type, markType, Point_$type, enumGetBox, typeCast } from "igniteui-react-core/ES5/type";
import { isNaN_, intDivide, isInfinity } from "igniteui-react-core/ES5/number";
import { RadialBase } from "./RadialBase";
import { CategoryLineRasterizer } from "./CategoryLineRasterizer";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { TrendLineType_$type } from "igniteui-react-core/ES5/TrendLineType";
import { Brush } from "igniteui-react-core/ES5/Brush";
import { PenLineCap_$type } from "igniteui-react-core/ES5/PenLineCap";
import { DoubleCollection } from "igniteui-react-core/ES5/DoubleCollection";
import { RadialFrame } from "./RadialFrame";
import { ScalerParams } from "./ScalerParams";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { AxisRange } from "./AxisRange";
import { NumericAxisBase } from "./NumericAxisBase";
import { IFastItemsSource_$type } from "igniteui-react-core/ES5/IFastItemsSource";
import { RadialBaseView } from "./RadialBaseView";
import { CollisionAvoider } from "./CollisionAvoider";
import { Clipper } from "igniteui-react-core/ES5/Clipper";
import { TrendResolutionParams } from "./TrendResolutionParams";
import { RadialTrendLineManager } from "./RadialTrendLineManager";
import { CategoryMarkerManager } from "./CategoryMarkerManager";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { stringIsNullOrEmpty } from "igniteui-react-core/ES5/string";
/**
 * @hidden
 */
var AnchoredRadialBucketCalculator = /** @class */ (function (_super) {
    tslib_1.__extends(AnchoredRadialBucketCalculator, _super);
    function AnchoredRadialBucketCalculator(a) {
        var _this = _super.call(this, a) || this;
        _this.f = null;
        _this.f = a;
        return _this;
    }
    AnchoredRadialBucketCalculator.prototype.getBucket = function (a) {
        var b = this.f.cs.nv;
        var c = Math.min(a * this.b, b.count - 1);
        var d = Math.min(c + this.b - 1, b.count - 1);
        var e = NaN;
        var f = NaN;
        for (var g = c; g <= d; ++g) {
            var h = b.item(g);
            if (!isNaN_(e)) {
                if (!isNaN_(h)) {
                    e = Math.min(e, h);
                    f = Math.max(f, h);
                }
            }
            else {
                e = h;
                f = h;
            }
        }
        if (!isNaN_(e)) {
            return [0.5 * (c + d), e, f];
        }
        return [0.5 * (c + d), NaN, NaN];
    };
    AnchoredRadialBucketCalculator.$t = markType(AnchoredRadialBucketCalculator, 'AnchoredRadialBucketCalculator', RadialBucketCalculator.$);
    return AnchoredRadialBucketCalculator;
}(RadialBucketCalculator));
export { AnchoredRadialBucketCalculator };
/**
 * @hidden
 */
var AnchoredRadialSeries = /** @class */ (function (_super) {
    tslib_1.__extends(AnchoredRadialSeries, _super);
    function AnchoredRadialSeries() {
        var _this = _super.call(this) || this;
        _this.nq = null;
        _this.nu = null;
        _this.oo = { $type: Point_$type, x: 0, y: 0 };
        _this.ns = null;
        _this.nr = null;
        _this.ns = new RadialFrame(3);
        _this.nq = new CategoryLineRasterizer();
        return _this;
    }
    AnchoredRadialSeries.prototype.bf = function () {
        return new AnchoredRadialSeriesView(this);
    };
    AnchoredRadialSeries.prototype.ht = function () {
        _super.prototype.ht.call(this);
        this.mz = null;
        this.m1 = null;
    };
    AnchoredRadialSeries.prototype.fs = function (a, b) {
        var c = this.gu(b);
        if (c == AnchoredRadialSeries.$$p[0]) {
            return this.ft(a, b, this.n7);
        }
        return _super.prototype.fs.call(this, a, b);
    };
    AnchoredRadialSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.np = a;
    };
    AnchoredRadialSeries.prototype.fe = function (a, b) {
        return this.ff(a, b, this.mz, null, this.nv);
    };
    AnchoredRadialSeries.prototype.fc = function (a, b) {
        return this.fd(a, b, this.mz, null, this.nv);
    };
    AnchoredRadialSeries.prototype.d9 = function (a, b, c, d, e) {
        if (this.nv == null) {
            return Number.POSITIVE_INFINITY;
        }
        return this.ea(a, b, this.mz, d, e, this.nv.count, null);
    };
    AnchoredRadialSeries.prototype.en = function (a, b, c) {
        var d = this.getEffectiveViewport1(this.bi);
        var e = this.n0(this.mz, this.bi.bs, this.bi.br, d);
        var f = new ScalerParams(0, this.bi.bs, this.bi.br, this.mz.b8, d);
        f.c = this.bs.m3;
        return this.ep(this.nv, a, this.mz, f, e, null, b, c);
    };
    AnchoredRadialSeries.prototype.lk = function (a, b, c) {
        if (this.m1 == null || this.mz == null || this.bi == null || this.bi.bs.isEmpty || this.bi.br.isEmpty) {
            return { $type: Point_$type, x: NaN, y: NaN };
        }
        var d = this.getEffectiveViewport1(this.bi);
        var e = new ScalerParams(0, this.bi.bs, this.bi.br, this.m1.b8, d);
        e.c = this.getEffectiveViewport1(this.bi);
        var f = this.en(a, b, c);
        f = this.m1.dv(f, e);
        var g = this.m2.c(a);
        if (!b && this.mz != null) {
            var h = this.fe(a, c);
            var i = this.fc(a, c);
            var j = this.n0(this.mz, this.bi.bs, this.bi.br, d);
            var k = new ScalerParams(0, this.bi.bs, this.bi.br, this.mz.b8, d);
            k.c = this.bs.m3;
            var l = this.d9(a, h, this.mz, k, j);
            var m = this.d9(a, i, this.mz, k, j);
            if (m <= l) {
                g = this.mz.dv(i, k) + j;
            }
            else {
                g = this.mz.dv(h, k) + j;
            }
        }
        var n = this.m2.d(g, f, this.bi.bs, this.bi.br);
        var o = this.m2.e(g, f, this.bi.bs, this.bi.br);
        return { $type: Point_$type, x: n, y: o };
    };
    Object.defineProperty(AnchoredRadialSeries.prototype, "nv", {
        get: function () {
            return this.nu;
        },
        set: function (a) {
            if (this.nu != a) {
                var b = this.nu;
                this.nu = a;
                this.raisePropertyChanged("ValueColumn", b, this.nu);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeries.prototype, "ok", {
        get: function () {
            return this.c(AnchoredRadialSeries.ob);
        },
        set: function (a) {
            this.h(AnchoredRadialSeries.ob, a);
        },
        enumerable: true,
        configurable: true
    });
    AnchoredRadialSeries.prototype.an = function (a) {
        if (a != null && a == this.mz && this.nv != null && this.nv.count > 0) {
            return new AxisRange(0, this.nv.count - 1);
        }
        if (a != null && a == this.m1 && this.nv != null && this.nv.count > 0) {
            return new AxisRange(this.nv.minimum, this.nv.maximum);
        }
        return null;
    };
    AnchoredRadialSeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        if (this.np.ct.g$b(a, b, c, d)) {
            this.jc(false);
            this.ij();
        }
        var e = typeCast(NumericAxisBase.$, this.m1);
        switch (b) {
            case "FastItemsSource":
                if (typeCast(IFastItemsSource_$type, c) !== null) {
                    c.deregisterColumn(this.nv);
                    this.nv = null;
                }
                if (typeCast(IFastItemsSource_$type, d) !== null) {
                    this.nv = this.bw(this.n7);
                    this.np.co.e(this.eu);
                }
                if (e != null && !e.ck()) {
                    this.np.co.e(this.eu);
                    this.jc(false);
                }
                break;
            case AnchoredRadialSeries.$$p[0]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.nv);
                    this.nv = this.bw(this.n7);
                }
                break;
            case "ValueColumn":
                if (e != null && !e.ck()) {
                    this.np.co.e(this.eu);
                    this.jc(false);
                }
                break;
            case "TrendLineBrush":
                this.jn();
                break;
            case "TrendLineType":
                this.ij();
                break;
        }
    };
    AnchoredRadialSeries.prototype.hy = function (a, b, c, d) {
        switch (a) {
            case 4:
            case 1:
            case 0:
                this.np.co.e(this.eu);
                break;
        }
        this.np.ct.g$e(a, b, c, d);
        switch (a) {
            case 4:
                if (this.m1 != null && !this.m1.ck()) {
                    this.jc(true);
                }
                break;
            case 1:
                if (this.m1 != null && !this.m1.ck()) {
                    this.jc(true);
                }
                break;
            case 0:
                if (this.m1 != null && !this.m1.ck()) {
                    this.jc(true);
                }
                break;
            case 2:
                if (this.n7 != null && this.np.co.b > 0 && this.m1 != null && !this.m1.ck()) {
                    this.jc(true);
                }
                break;
            case 3:
                if (d == this.n7) {
                    if (this.m1 != null && !this.m1.ck()) {
                        this.jc(true);
                    }
                }
                break;
        }
    };
    AnchoredRadialSeries.prototype.dw = function (a, b, c) {
        var d = _super.prototype.dw.call(this, a, b, c);
        var e = c;
        if (this.bz == null || this.bz.count == 0 || this.mz == null || this.nv == null || this.mz.jx == 0 || this.m1 == null || isInfinity(this.m1.j3) || isInfinity(this.m1.j2) || e.co.b < 1) {
            d = false;
        }
        return d;
    };
    AnchoredRadialSeries.prototype.getOffsetValue = function () {
        return this.n0(this.mz, this.bi.bs, this.bi.br, this.getEffectiveViewport1(this.bi));
    };
    AnchoredRadialSeries.prototype.getCategoryWidth = function () {
        return this.mz.getCategorySize(this.bi.bs, this.bi.br, this.getEffectiveViewport1(this.bi));
    };
    AnchoredRadialSeries.prototype.get_ny = function () {
        return false;
    };
    Object.defineProperty(AnchoredRadialSeries.prototype, "ny", {
        get: function () {
            return this.get_ny();
        },
        enumerable: true,
        configurable: true
    });
    AnchoredRadialSeries.prototype.n0 = function (a, b, c, d) {
        var e = 0;
        var f = this.preferredCategoryMode(a);
        if (f == 0 && a.categoryMode != 0) {
            f = 1;
        }
        switch (f) {
            case 0:
                e = 0;
                break;
            case 1:
                e = 0.5 * a.getCategorySize(b, c, d);
                break;
            case 2:
                e = a.getGroupCenter(this.ne(), b, c, d);
                break;
        }
        if (a.b8) {
            e = -e;
        }
        return e;
    };
    AnchoredRadialSeries.prototype.nj = function (a, b) {
        var _this = this;
        var c = b.bs;
        var d = b.br;
        var e = this.getEffectiveViewport1(b);
        var f = this.mz;
        var g = this.m1;
        var h = g.j3;
        var i = g.j2;
        a.f.clear();
        a.m.clear();
        a.p.clear();
        var j = this.l2();
        var k = 0;
        var l = b;
        var m = new CollisionAvoider();
        var n = 0;
        var o = b.co.d;
        n = this.n0(f, c, d, e);
        l.ct.q = this.m1.li;
        l.ct.p = this.m1.lh;
        l.ct.r = function (p, q) { return _this.n1(p, q, c, d); };
        l.ct.s = function (p, q) { return _this.n2(p, q, c, d); };
        var p = ((function () {
            var $ret = new Clipper(0, d, false);
            $ret.i = a.p;
            return $ret;
        })());
        var q = ((function () {
            var $ret = new TrendResolutionParams();
            $ret.c = b.co.b;
            $ret.d = b.co.c;
            $ret.e = o;
            $ret.a = n;
            $ret.b = _this.eu;
            $ret.f = d;
            $ret.g = c;
            return $ret;
        })());
        l.ct.t(a.p, this.nt, this.nv, this.n4, function (r) { return _this.mz.getScaledAngle(r); }, function (r) { return _this.m1.lo(r); }, q, p);
        var r = true;
        if (this.nz(b)) {
            o--;
        }
        for (var s = b.co.c; s <= o; ++s) {
            var t = s;
            if (t * b.co.b >= this.mz.jx) {
                t -= intDivide((this.mz.jx), b.co.b);
            }
            var u = b.co.getBucket(t);
            if (!isNaN_(u[0])) {
                u[0] = f.getScaledAngle(u[0]) + n;
                if (u[1] < h || u[1] > i) {
                    continue;
                }
                u[1] = g.lo(u[1]);
                if (b.co.b > 1) {
                    if (u[2] < h || u[2] > i) {
                        continue;
                    }
                    u[2] = g.lo(u[2]);
                }
                else {
                    u[2] = u[1];
                }
                if ((isNaN_(u[1]) || isNaN_(u[2])) && r && this.ny && this.nx()) {
                    o++;
                }
                else {
                    r = false;
                }
                a.f.add(u);
                if (j) {
                    var v = Math.min(t * b.co.b, this.bz.count - 1);
                    var w = this.m2.d(u[0], u[1], c, d);
                    var x = this.m2.e(u[0], u[1], c, d);
                    var y = new Rect(0, w - 5, x - 5, 11, 11);
                    if (!isNaN_(w) && !isNaN_(x) && !isInfinity(w) && !isInfinity(x) && m.tryAdd(y)) {
                        a.m.add({ $type: Point_$type, x: w, y: x });
                        var z = b.cp.item(k);
                        z.ak = v;
                        z.content.item = this.bz.item(v);
                        ++k;
                    }
                }
            }
            else {
                if (r && this.ny && this.nx()) {
                    o++;
                }
            }
        }
        b.cp.count = k;
        return;
    };
    AnchoredRadialSeries.prototype.n1 = function (a, b, c, d) {
        return this.m2.d(a, b, c, d);
    };
    AnchoredRadialSeries.prototype.n2 = function (a, b, c, d) {
        return this.m2.e(a, b, c, d);
    };
    AnchoredRadialSeries.prototype.nz = function (a) {
        var b = a;
        return !this.ny && b.co.c == 0 && b.co.d == this.mz.jx;
    };
    AnchoredRadialSeries.prototype.no = function (a, b, c) {
        var d = null;
        var e = c.bs;
        var f = c.br;
        if (b > -1 && !e.isEmpty && !f.isEmpty) {
            var g = a._inner[0][0] < f.left - 2000 ? f.left - 10 : NaN;
            var h = f.bottom + 10;
            var i = a._inner[b][0] > f.right + 2000 ? f.right + 10 : NaN;
            var j = f.top - 10;
            d = new Clipper(1, g, h, i, j, false);
        }
        return d;
    };
    AnchoredRadialSeries.prototype.oa = function (a, b, c, d, e) {
        if (a.count > 0 && d.count > 0) {
            if (!e || this.nw()) {
                a.add(this.oo);
                a.add(a._inner[0]);
                d.add(this.oo);
                d.add(d._inner[0]);
            }
        }
    };
    AnchoredRadialSeries.prototype.nx = function () {
        return !this.nw();
    };
    AnchoredRadialSeries.prototype.nw = function () {
        var a = this.bi.bs;
        return !a.containsPoint({ $type: Point_$type, x: 0.5, y: 0.5 });
    };
    AnchoredRadialSeries.prototype.hv = function (a, b) {
        _super.prototype.hv.call(this, a, b);
        var c = b;
        if (a) {
            c.cp.clear();
        }
        c.ct.g$d();
    };
    AnchoredRadialSeries.prototype.nk = function (a, b) {
        var c = b.bs;
        var d = b.br;
        var e = this.getEffectiveViewport1(b);
        var f = this.lp(b);
        var g = b;
        g.ct.g$g(a.p);
        CategoryMarkerManager.b(this, a.m, g.cp, this.l4);
        g.cj();
        this.n9(d, c, e, f, g);
    };
    AnchoredRadialSeries.prototype.n9 = function (a, b, c, d, e) {
        e.cu(a, b, c, d);
    };
    AnchoredRadialSeries.prototype.jn = function () {
        _super.prototype.jn.call(this);
        if (this.fh < 0) {
            return;
        }
        this.np.cv();
    };
    AnchoredRadialSeries.prototype.fv = function (a, b) {
        if (a == this.np.ct.g$j) {
            return null;
        }
        return _super.prototype.fv.call(this, a, b);
    };
    AnchoredRadialSeries.prototype.c9 = function (a) {
        var b = this.bi.bs;
        b = b.copy();
        var c = this.bi.br;
        var d = !b.isEmpty && !c.isEmpty && this.bz != null ? this.bz.indexOf(a) : -1;
        if (this.mz == null || this.nv == null || this.m1 == null) {
            return false;
        }
        if (d < 0 || d > this.nv.count - 1) {
            return false;
        }
        var e = this.mz.getScaledAngle(d);
        var f = this.m1.lo(this.nv.item(d));
        if (isNaN_(f)) {
            f = (this.m1.lh + this.m1.li) / 2;
        }
        var g = 0.5 + (Math.cos(e) * f);
        var h = 0.5 + (Math.sin(e) * f);
        if (!isNaN_(g)) {
            if (g < b.left + 0.1 * b.width) {
                g = g + 0.4 * b.width;
                b.x = g - 0.5 * b.width;
            }
            if (g > b.right - 0.1 * b.width) {
                g = g - 0.4 * b.width;
                b.x = g - 0.5 * b.width;
            }
        }
        if (!isNaN_(h)) {
            if (h < b.top + 0.1 * b.height) {
                h = h + 0.4 * b.height;
                b.y = h - 0.5 * b.height;
            }
            if (h > b.bottom - 0.1 * b.height) {
                h = h - 0.4 * b.height;
                b.y = h - 0.5 * b.height;
            }
        }
        if (this.bn != null) {
            this.bn.ap(this.bs, b, true);
        }
        return d >= 0;
    };
    AnchoredRadialSeries.prototype.je = function (a, b) {
        _super.prototype.je.call(this, a, b);
        if (!this.dl) {
            this.bi.az(b);
            return;
        }
        this.bi.az(b);
        var c = typeCast(AnchoredRadialSeriesView.$, this.bh);
        c.co.e(this.eu);
        if (this.cc(this.bh)) {
            return;
        }
        if (!this.df) {
            this.ns = new RadialFrame(3);
            this.nj(this.ns, c);
        }
        this.df = false;
        this.nk(this.ns, c);
        this.dl = false;
    };
    AnchoredRadialSeries.prototype.jb = function (a, b, c, d, e) {
        _super.prototype.jb.call(this, a, b, c, d, e);
        var f = this.a0.a$h.item(d);
        var g = f;
        g.co.e(this.eu);
        f.ay(c);
        if (this.cc(g)) {
            return;
        }
        if (this.nr == null) {
            this.nr = new RadialFrame(3);
        }
        this.nr.u();
        this.nj(this.nr, g);
        this.nk(this.nr, g);
    };
    AnchoredRadialSeries.prototype.h4 = function (a) {
        _super.prototype.h4.call(this, a);
        this.a2.exportTrendlineData(this, a, this.np.ct.g$j);
    };
    AnchoredRadialSeries.$t = markType(AnchoredRadialSeries, 'AnchoredRadialSeries', RadialBase.$);
    AnchoredRadialSeries.ob = DependencyProperty.i("ActualTrendLineBrush", Brush.$, AnchoredRadialSeries.$, new PropertyMetadata(2, null, function (a, b) { return a.raisePropertyChanged("ActualTrendLineBrush", b.oldValue, b.newValue); }));
    return AnchoredRadialSeries;
}(RadialBase));
export { AnchoredRadialSeries };
Type.dep(DependencyProperty, PropertyMetadata, AnchoredRadialSeries, 'raisePropertyChanged', ['ValueMemberPath:n7:oj', [2, null], 'TrendLineType:nt:oh', [TrendLineType_$type, enumGetBox(TrendLineType_$type, 0)], 'TrendLineThickness:n3:og', [1, 1.5], 'TrendLinePeriod:n4:of', [1, 7], 'TrendLineZIndex:n5:oi', [1, 1], 'TrendLineBrush:ol:oc', [Brush.$, null], 'TrendLineDashArray:om:od', [DoubleCollection.$], 'TrendLineDashCap:on:oe', [PenLineCap_$type, enumGetBox(PenLineCap_$type, 0)]]);
/**
 * @hidden
 */
var AnchoredRadialSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(AnchoredRadialSeriesView, _super);
    function AnchoredRadialSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cs = null;
        _this.ct = null;
        _this.co = new AnchoredRadialBucketCalculator(_this);
        _this.cs = a;
        _this.ct = new RadialTrendLineManager();
        return _this;
    }
    AnchoredRadialSeriesView.prototype.cu = function (a, b, c, d) {
    };
    AnchoredRadialSeriesView.prototype.cv = function () {
        this.cs.ok = null;
        if (this.cs.ol != null) {
            this.cs.ok = this.cs.ol;
        }
        else {
            this.cs.ok = this.cs.k2;
        }
    };
    AnchoredRadialSeriesView.prototype.a2 = function (a, b) {
        if (a.d) {
            if (!b) {
            }
            if (this.cs.nt != 0 && !b) {
                var c = this.ct.g$j;
                c.ac = this.cs.n3;
                c._stroke = this.cs.ok;
                c.ai = this.cs.om;
                c.ad = this.cs.on;
                a.u(c);
            }
        }
        _super.prototype.a2.call(this, a, b);
    };
    AnchoredRadialSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'>";
        if (this.cs.mz.e6 != null) {
            a += "<span>${item." + this.cs.mz.e6 + "}</span><br/>";
        }
        a += "<span";
        var b = this.e.a0.a$q(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.cs.fx + ": </span><span class='ui-priority-primary'>" + "${item." + this.cs.n7 + "}</span></div>";
        return a;
    };
    AnchoredRadialSeriesView.$t = markType(AnchoredRadialSeriesView, 'AnchoredRadialSeriesView', RadialBaseView.$);
    return AnchoredRadialSeriesView;
}(RadialBaseView));
export { AnchoredRadialSeriesView };
