// igniteui-react-charts/ES5/AngleAxisLabelPanel_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { AxisLabelPanelBase } from "./AxisLabelPanelBase";
import { Boolean_$type, typeCast, Type, markType, Point_$type, typeGetValue, fromEnum, runOn, Number_$type } from "igniteui-react-core/ES5/type";
import { AxisLabelPanelBaseView } from "./AxisLabelPanelBaseView";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { Extensions } from "igniteui-react-core/ES5/Extensions";
import { Axis } from "./Axis";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { LabelPosition } from "./LabelPosition";
import { XamDataChart } from "./XamDataChart";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { IAngleScaler_$type } from "./IAngleScaler";
import { CategoryAxisRenderer } from "./CategoryAxisRenderer";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { PolarAxisRenderingManager } from "./PolarAxisRenderingManager";
import { AxisLabelManager } from "./AxisLabelManager";
import { ViewportUtils } from "./ViewportUtils";
import { RadialAxisRenderingParameters } from "./RadialAxisRenderingParameters";
import { CategoryTickmarkValues } from "./CategoryTickmarkValues";
import { MathUtil } from "igniteui-react-core/ES5/MathUtil";
import { SeriesViewer } from "./SeriesViewer";
import { ScalerParams } from "./ScalerParams";
import { AxisRangeChangedEventArgs } from "./AxisRangeChangedEventArgs";
import { LineGeometry } from "igniteui-react-core/ES5/LineGeometry";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { RangeInfo } from "./RangeInfo";
import { isNaN_, isInfinity } from "igniteui-react-core/ES5/number";
import { NumericAxisBase } from "./NumericAxisBase";
import { LogarithmicTickmarkValues } from "./LogarithmicTickmarkValues";
import { PolarAxisRenderingParameters } from "./PolarAxisRenderingParameters";
import { RadialAxisLabelPanel } from "./RadialAxisLabelPanel";
import { LinearTickmarkValues } from "./LinearTickmarkValues";
import { IPolarRadialRenderingParameters_$type } from "./IPolarRadialRenderingParameters";
import { PathGeometry } from "igniteui-react-core/ES5/PathGeometry";
import { PathFigure } from "igniteui-react-core/ES5/PathFigure";
import { ArcSegment } from "igniteui-react-core/ES5/ArcSegment";
import { Size } from "igniteui-react-core/ES5/Size";
import { CategoryAxisBaseView } from "./CategoryAxisBaseView";
import { NumericAxisBaseView } from "./NumericAxisBaseView";
/**
 * @hidden
 */
var AngleAxisLabelPanel = /** @class */ (function (_super) {
    tslib_1.__extends(AngleAxisLabelPanel, _super);
    function AngleAxisLabelPanel() {
        var _this = _super.call(this) || this;
        _this.ct = false;
        _this.cz = null;
        _this.cx = 0;
        _this.cu = 0;
        _this.cw = 0;
        _this.cv = 0;
        return _this;
    }
    AngleAxisLabelPanel.prototype.ad = function () {
        return new AngleAxisLabelPanelView(this);
    };
    AngleAxisLabelPanel.prototype.b8 = function (a) {
        _super.prototype.b8.call(this, a);
        this.cs = a;
    };
    AngleAxisLabelPanel.prototype.ao = function (a, b) {
        if (!Extensions.b(b)) {
            return false;
        }
        if (this.ac == null || this.ac.i0.isEmpty) {
            return _super.prototype.ao.call(this, a, b);
        }
        return true;
    };
    AngleAxisLabelPanel.prototype.ay = function () {
        var a = new List$1(Rect.$, 0);
        var b = new List$1(Boolean_$type, 0);
        if (this.a3.count != this.a2.count) {
            return a;
        }
        this.cs.ay(a, b);
        var c = false;
        this.am = this.al(a, b);
        if (!this.ac.bz()) {
            c = true;
            this.a9 = this.cx / 2;
            this.a9 = this.a9 + this.cy();
        }
        else {
            c = true;
            this.cs.ac();
        }
        if (c) {
            for (var d = 0; d < a.count; d++) {
                var e = this.a3._inner[d];
                var f = this.a2._inner[d];
                var g = a._inner[d];
                var h = this.cz(f.c);
                var i = h.x - this.bc(e) / 2;
                var j = h.y - this.bb(e) / 2;
                g.x = i;
                g.y = j;
                a._inner[d] = g;
            }
        }
        return a;
    };
    AngleAxisLabelPanel.prototype.cy = function () {
        var a;
        var b;
        b = this.ac;
        a = null;
        if (b != null) {
            a = typeCast(XamDataChart.$, b.bf);
        }
        if (a == null || b == null) {
            return 0;
        }
        var c = new List$1(Axis.$, 0);
        for (var d = 0; d < a.axes.count; d++) {
            var e = a.axes._inner[d];
            if ((typeCast(NumericAngleAxis.$, e) !== null || typeCast(CategoryAngleAxis.$, e) !== null) && typeCast(AngleAxisLabelPanel.$, e.at) !== null && e.at.a3.count > 0 && (!e.bz()) && !e.by() && (e.ag == null || this.ac.ag == null || e.ag.li == b.ag.li)) {
                c.add(e);
            }
        }
        var f = c.indexOf(b);
        if (f == -1) {
            return 0;
        }
        var g = 0;
        var h = 5;
        if (f == 0) {
            g += h;
        }
        else {
            g += c._inner[f - 1].at.a9;
            g += c._inner[f].at.a9 * 2 + h;
        }
        return g;
    };
    AngleAxisLabelPanel.prototype.ah = function () {
        return 4;
    };
    AngleAxisLabelPanel.prototype.as = function (a) {
        return a == 4 || a == 5;
    };
    AngleAxisLabelPanel.$t = markType(AngleAxisLabelPanel, 'AngleAxisLabelPanel', AxisLabelPanelBase.$);
    return AngleAxisLabelPanel;
}(AxisLabelPanelBase));
export { AngleAxisLabelPanel };
/**
 * @hidden
 */
var CategoryAngleAxis = /** @class */ (function (_super) {
    tslib_1.__extends(CategoryAngleAxis, _super);
    function CategoryAngleAxis() {
        var _this = _super.call(this) || this;
        _this.k1 = null;
        _this.ky = null;
        _this.k2 = false;
        _this.k6 = NaN;
        _this.kz = null;
        _this.k7 = 0;
        _this.k4 = 0;
        _this.k5 = 0;
        _this.li = 1;
        _this.ab = CategoryAngleAxis.$;
        _this.ky = _this.kx();
        _this.k1 = new PolarAxisRenderingManager();
        return _this;
    }
    CategoryAngleAxis.prototype.a1 = function () {
        return new CategoryAngleAxisView(this);
    };
    CategoryAngleAxis.prototype.gr = function (a) {
        _super.prototype.gr.call(this, a);
        this.kw = a;
    };
    CategoryAngleAxis.prototype.get_b0 = function () {
        return true;
    };
    CategoryAngleAxis.prototype.as = function () {
        var _this = this;
        var a = new AngleAxisLabelPanel();
        a.cz = function (b) {
            var c = _this.bf != null ? _this.bf.mq : Rect.empty;
            var d = !c.isEmpty ? _this.i0 : Rect.empty;
            return _this.lv(b, { $type: Point_$type, x: 0.5, y: 0.5 }, c, d, _this.at.a9);
        };
        return a;
    };
    CategoryAngleAxis.prototype.lg = function (a) {
        return Math.round(a * Math.pow(10, 10)) / Math.pow(10, 10);
    };
    CategoryAngleAxis.prototype.kx = function () {
        var _this = this;
        var a = ((function () {
            var $ret = new AxisLabelManager();
            $ret.a = _this;
            $ret.f = _this.cv;
            $ret.e = _this.cu;
            $ret.b = _this.at;
            return $ret;
        })());
        if (this.aw != null) {
            this.aw.registerAxis(this);
        }
        var b = new CategoryAxisRenderer(a);
        b.u = function () {
            var c = _this.a2.ay();
            var d = _this.a2.a1();
            var e = _this.a2.az();
            var f = _this.a2.a0();
            _this.hd();
            _this.gh(c);
            _this.gh(d);
            _this.gh(e);
            _this.gh(f);
        };
        b.o = function (c, d) { return !d.isEmpty && !c.isEmpty && _this.k0 != null; };
        b.d = function (c, d, e, f) { return _this.kv(c, d, e, f); };
        b.v = function () {
            if (!_this.k2) {
                _this.k2 = true;
                _this.k0.ck();
                _this.k2 = false;
            }
        };
        b.g = function (c) {
            var d = typeGetValue(c);
            if (d > _this.je.count - 1) {
                d -= _this.je.count;
            }
            var e = _this.je.item(d);
            return _this.e5(e);
        };
        b.b.c = function (c) {
            var e_1, _a;
            if ((_this.aw == null || _this.aw.visibility == 0) && _this.k0 != null && _this.k6 != c) {
                var d = typeCast(XamDataChart.$, _this.bf);
                if (d == null) {
                    return;
                }
                _this.k6 = c;
                _this.at.a8 = c;
                d.ht();
                try {
                    for (var _b = tslib_1.__values(fromEnum(d.axes)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var e = _c.value;
                        if (e != _this && typeCast(AngleAxisLabelPanel.$, e.at) !== null) {
                            e.a2.ad();
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        };
        b.k = function (c, d, e) {
            var f = typeCast(RadialAxisRenderingParameters.$, c);
            _this.k1.m(d, e, c.z, c.aa, f.minLength, f.maxLength, f.center);
        };
        b.l = function (c, d, e, f) {
            var g = typeCast(RadialAxisRenderingParameters.$, c);
            _this.k1.n(d, e, f, g.z, c.aa, g.minLength, g.maxLength, g.center);
        };
        b.m = function (c, d) { return _this.getScaledAngle(d); };
        b.q = function (c, d) {
            if (_this.lg(d - _this.k7) < 0) {
                return false;
            }
            if (_this.lg(d - _this.k7 - (2 * Math.PI)) > 0) {
                return false;
            }
            return true;
        };
        b.j = function (c) {
            var d = typeCast(RadialAxisRenderingParameters.$, c);
            if (d.e == d.k._inner[0]) {
                _this.k1.i(c.t, c.n, c.z, c.aa, d.center, d.minAngle, d.maxAngle);
            }
        };
        b.e = function (c) {
            c.n = _this.la();
            c.r = c.n;
        };
        b.p = function (c, d, e) {
            var f = typeCast(RadialAxisRenderingParameters.$, c);
            if (e) {
                return false;
            }
            var g = _this.lv(d, f.center, c.aa, c.z, 0);
            if (g.x < c.z.right && g.x >= c.z.left && g.y < c.z.bottom && g.y >= c.z.top) {
                return true;
            }
            return false;
        };
        b.a = function (c, d, e, f) {
            var g = c.y;
            var h = new ScalerParams(0, c.aa, c.z, _this.b8, g);
            var i = d;
            if (_this.categoryMode != 0) {
                var j = (e * f) + 1;
                j = Math.min(j, _this.jx);
                var k = _this.dv(j, h);
                i = (d + k) / 2;
            }
            return i;
        };
        b.f = runOn(this, this.getGroupCenter);
        b.i = runOn(this, this.ld);
        return b;
    };
    CategoryAngleAxis.prototype.lv = function (a, b, c, d, e) {
        var f = this.la();
        var g = ViewportUtils.b(e, c, d);
        if (this.aw != null && (this.aw.c == 5 || this.aw.c == 1)) {
            g *= -1;
        }
        var h = b.x + (f + g) * Math.cos(a);
        var i = b.y + (f + g) * Math.sin(a);
        h = ViewportUtils.c(h, c, d);
        i = ViewportUtils.f(i, c, d);
        return { $type: Point_$type, x: h, y: i };
    };
    CategoryAngleAxis.prototype.la = function () {
        if (this.k0 == null) {
            return 0;
        }
        if (!this.by()) {
            return this.k0.ln();
        }
        else {
            return this.k0.lo((this.e1));
        }
    };
    CategoryAngleAxis.prototype.kv = function (a, b, c, d) {
        if (this.je == null) {
            return null;
        }
        var e = new RadialAxisRenderingParameters();
        var f = this.je.count - 1;
        var g = this.a2.ay();
        var h = this.a2.a1();
        var i = this.a2.az();
        var j = this.a2.a0();
        e.t = g;
        e.w = h;
        e.u = i;
        e.v = j;
        e.l = f;
        e.m = 0;
        e.h = false;
        e.z = a;
        e.y = c;
        e.x = d;
        e.aa = b;
        e.g = this.k3();
        e.p = this.le;
        e.o = this.d8();
        e.q = this.lf;
        e.s = this.e6;
        var k = this.k1.f(b);
        var l = this.k1.g(b);
        var m = 0.5 * this.k0.li;
        var n = 0.5 * this.k0.lh;
        var o = k;
        var p = l;
        var q = this.k0.ln();
        if (isNaN_(q) || isInfinity(q)) {
            return null;
        }
        if (p >= m) {
            p = q;
        }
        if (o < n) {
            o = n;
        }
        var r = a.width;
        this.k1.k(b, e, 0, this.je.count, this.b8, runOn(this, this.getUnscaledAngle), r);
        var s = { $type: Point_$type, x: 0.5, y: 0.5 };
        e.center = s;
        e.maxLength = p;
        e.minLength = o;
        e.effectiveMaximum = q;
        e.ae = this.je.count;
        e.ab = this.categoryMode;
        e.ad = true;
        e.ac = this.b8;
        e.af = this.j0;
        e.f = new CategoryTickmarkValues();
        e.j = this.ci;
        return e;
    };
    CategoryAngleAxis.prototype.getMinMaxAngle = function (a, b, c) {
        var d = this.k1.l(a, b, c);
        b = d.p1;
        c = d.p2;
        return {
            p1: b,
            p2: c
        };
    };
    CategoryAngleAxis.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.g5(false);
    };
    Object.defineProperty(CategoryAngleAxis.prototype, "k0", {
        get: function () {
            if (this.kz != null) {
                return this.kz;
            }
            var a = typeCast(XamDataChart.$, this.bf);
            if (a != null) {
                for (var b = 0; b < a.axes.count; b++) {
                    if (typeCast(NumericRadiusAxis.$, a.axes._inner[b]) !== null) {
                        return a.axes._inner[b];
                    }
                }
            }
            return this.kz;
        },
        set: function (a) {
            this.kz = a;
        },
        enumerable: true,
        configurable: true
    });
    CategoryAngleAxis.prototype.ah = function () {
        return this.k0;
    };
    CategoryAngleAxis.prototype.getCategorySize = function (a, b, c) {
        return 2 * Math.PI / this.jx;
    };
    CategoryAngleAxis.prototype.jq = function (a, b, c) {
        var d = !isNaN_(this.jn) ? MathUtil.b(this.jn, 0, 1) : 0;
        var e = 0;
        if (!isNaN_(this.ju)) {
            e = Math.min(this.ju, 1);
        }
        var f = 1 - 0.5 * d;
        return this.getCategorySize(a, b, c) * f / (this.j0 - (this.j0 - 1) * e);
    };
    CategoryAngleAxis.prototype.getGroupCenter = function (a, b, c, d) {
        var e = 0.5;
        if (this.j0 > 1) {
            var f = !isNaN_(this.jn) ? MathUtil.b(this.jn, 0, 1) : 0;
            var g = 0;
            if (!isNaN_(this.ju)) {
                g = Math.min(this.ju, 1);
            }
            var h = 1 - 0.5 * f;
            var i = h / (this.j0 - (this.j0 - 1) * g);
            var j = (h - i) / (this.j0 - 1);
            e = 0.25 * f + 0.5 * i + a * j;
        }
        return this.getCategorySize(b, c, d) * e;
    };
    CategoryAngleAxis.prototype.ld = function (a) {
        var b = 0.5;
        if (this.j0 > 1) {
            var c = !isNaN_(this.jn) ? MathUtil.b(this.jn, 0, 1) : 0;
            var d = 0;
            if (!isNaN_(this.ju)) {
                d = Math.min(this.ju, 1);
            }
            var e = 1 - 0.5 * c;
            var f = e / (this.j0 - (this.j0 - 1) * d);
            var g = (e - f) / (this.j0 - 1);
            b = 0.25 * c + 0.5 * f + a * g;
        }
        return b;
    };
    CategoryAngleAxis.prototype.g6 = function (a) {
        _super.prototype.g6.call(this, a);
        var b = this.bf != null ? this.bf.mq : Rect.empty;
        var c = !b.isEmpty ? this.i0 : Rect.empty;
        var d = this.iv();
        var e = this.iu();
        var f = this.ky.c(a, c, b, d, e);
        if (f != null) {
            this.k8 = f.f.m;
            this.k9 = f.f.l;
        }
    };
    CategoryAngleAxis.prototype.lc = function (a, b) {
        var c;
        var d;
        var e = this.k1.l(b, c, d);
        c = e.p1;
        d = e.p2;
        if (c == 0) {
            if (this.b8) {
                return this.jx;
            }
            else {
                return 0;
            }
        }
        var f = this.getUnscaledAngle(c);
        if (f < 0 || f > this.jx) {
            f = this.getUnscaledAngle(c + Math.PI * 2);
        }
        return f;
    };
    CategoryAngleAxis.prototype.lb = function (a, b) {
        var c;
        var d;
        var e = this.k1.l(b, c, d);
        c = e.p1;
        d = e.p2;
        if (d > Math.PI * 2) {
            d = d - Math.PI * 2;
        }
        if (d == Math.PI * 2) {
            if (this.b8) {
                return 0;
            }
            else {
                return this.jx;
            }
        }
        var f = this.getUnscaledAngle(d);
        if (f < 0 || f > this.jx) {
            f = this.getUnscaledAngle(d + Math.PI * 2);
        }
        return f;
    };
    CategoryAngleAxis.prototype.getScaledAngle = function (a) {
        var b = this.jx;
        var c = b >= 2 ? (a) / (b) : b == 1 ? 0.5 : NaN;
        if (this.b9) {
            c = 1 - c;
        }
        return (c * 2 * Math.PI) + this.k7;
    };
    CategoryAngleAxis.prototype.getUnscaledAngle = function (a) {
        if (a < this.k7) {
            a += 2 * Math.PI;
        }
        var b = (a - this.k7) / (2 * Math.PI);
        if (this.b8) {
            b = 1 - b;
        }
        return b * (this.jx);
    };
    CategoryAngleAxis.prototype.dv = function (a, b) {
        return this.getScaledAngle(a);
    };
    CategoryAngleAxis.prototype.dy = function (a, b) {
        return this.getUnscaledAngle(a);
    };
    CategoryAngleAxis.prototype.gt = function (a, b, c, d) {
        var e_2, _a, e_3, _b, e_4, _c;
        _super.prototype.gt.call(this, a, b, c, d);
        var e = typeCast(XamDataChart.$, this.bf);
        switch (b) {
            case "CrossingAxis":
                var f = typeCast(NumericRadiusAxis.$, d);
                this.lr(f);
                if (f != null) {
                    f.lx(this);
                }
                this.cf = true;
                this.g5(false);
                break;
            case CategoryAngleAxis.$$p[0]:
                this.k7 = this.lh;
                while (this.k7 < 0) {
                    this.k7 += 360;
                }
                while (this.k7 >= 360) {
                    this.k7 -= 360;
                }
                this.k7 = (this.k7 * Math.PI) / 180;
                this.cf = true;
                this.g5(false);
                try {
                    for (var _d = tslib_1.__values(fromEnum(this.cw)), _e = _d.next(); !_e.done; _e = _d.next()) {
                        var g = _e.value;
                        g.jc(false);
                        g.ij();
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
                break;
            case "Label":
                if (e != null) {
                    try {
                        for (var _f = tslib_1.__values(fromEnum(e.axes)), _g = _f.next(); !_g.done; _g = _f.next()) {
                            var h = _g.value;
                            h.cf = true;
                            h.g4();
                        }
                    }
                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                    finally {
                        try {
                            if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                        }
                        finally { if (e_3) throw e_3.error; }
                    }
                }
                break;
            case "CrossingValue":
                if (e != null) {
                    try {
                        for (var _h = tslib_1.__values(fromEnum(e.axes)), _j = _h.next(); !_j.done; _j = _h.next()) {
                            var i = _j.value;
                            if (typeCast(NumericAngleAxis.$, i) !== null || typeCast(CategoryAngleAxis.$, i) !== null) {
                                i.cf = true;
                                i.g4();
                            }
                        }
                    }
                    catch (e_4_1) { e_4 = { error: e_4_1 }; }
                    finally {
                        try {
                            if (_j && !_j.done && (_c = _h.return)) _c.call(_h);
                        }
                        finally { if (e_4) throw e_4.error; }
                    }
                }
                break;
            case "LabelSettings":
                this.ky = this.kx();
                this.lp();
                this.cf = true;
                this.g5(false);
                break;
        }
    };
    CategoryAngleAxis.prototype.lp = function () {
        this.k6 = NaN;
    };
    CategoryAngleAxis.prototype.lr = function (a) {
        this.k0 = a;
    };
    Object.defineProperty(CategoryAngleAxis.prototype, "le", {
        get: function () {
            return this.c(CategoryAngleAxis.ls);
        },
        set: function (a) {
            this.h(CategoryAngleAxis.ls, a);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CategoryAngleAxis.prototype, "k8", {
        get: function () {
            return this.k4;
        },
        set: function (a) {
            if (this.k8 != a) {
                var b = this.k4;
                this.k4 = a;
                this.gz("ActualInterval", b, this.k8);
            }
        },
        enumerable: true,
        configurable: true
    });
    CategoryAngleAxis.prototype.d6 = function () {
        return this.k8;
    };
    Object.defineProperty(CategoryAngleAxis.prototype, "lf", {
        get: function () {
            return this.c(CategoryAngleAxis.lt);
        },
        set: function (a) {
            this.h(CategoryAngleAxis.lt, a);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CategoryAngleAxis.prototype, "k9", {
        get: function () {
            return this.k5;
        },
        set: function (a) {
            if (this.k9 != a) {
                var b = this.k5;
                this.k5 = a;
                this.gz("ActualMinorInterval", b, this.k9);
            }
        },
        enumerable: true,
        configurable: true
    });
    CategoryAngleAxis.prototype.k3 = function () {
        return false;
    };
    CategoryAngleAxis.prototype.get_az = function () {
        return 2;
    };
    Object.defineProperty(CategoryAngleAxis.prototype, "az", {
        get: function () {
            return this.get_az();
        },
        enumerable: true,
        configurable: true
    });
    CategoryAngleAxis.prototype.cm = function () {
        if (this.cf && this.cn) {
            this.a9 = this.a8();
        }
        if (this.je == null) {
            return false;
        }
        var a = this.je.count;
        if (a != this.lj) {
            var b = new AxisRangeChangedEventArgs(0, 1, 1, this.lj, a);
            this.lj = a;
            this.g0(b);
            return true;
        }
        return false;
    };
    Object.defineProperty(CategoryAngleAxis.prototype, "lj", {
        get: function () {
            return this.li;
        },
        set: function (a) {
            this.li = a;
        },
        enumerable: true,
        configurable: true
    });
    CategoryAngleAxis.prototype.is = function (a, b, c, d, e) {
        if (d === void 0) { d = null; }
        if (e === void 0) { e = null; }
        var f = typeCast(RadialAxisRenderingParameters.$, this.ky.d(c, b, d, e));
        if (f == null) {
            return null;
        }
        var g = Math.cos(a);
        var h = Math.sin(a);
        var i = f.center.x + g * f.minLength;
        var j = f.center.y + h * f.minLength;
        var k = f.center.x + g * f.maxLength;
        var l = f.center.y + h * f.maxLength;
        i = ViewportUtils.c(i, b, c);
        j = ViewportUtils.f(j, b, c);
        k = ViewportUtils.c(k, b, c);
        l = ViewportUtils.f(l, b, c);
        var m = new LineGeometry();
        m.c = { $type: Point_$type, x: i, y: j };
        m.b = { $type: Point_$type, x: k, y: l };
        return m;
    };
    Object.defineProperty(CategoryAngleAxis.prototype, "isVertical", {
        get: function () {
            return this.ce;
        },
        enumerable: true,
        configurable: true
    });
    CategoryAngleAxis.prototype.j$c = function (a, b) {
        return this.dy.apply(this, arguments);
    };
    CategoryAngleAxis.prototype.j$b = function (a, b) {
        return this.dv.apply(this, arguments);
    };
    CategoryAngleAxis.prototype.j$d = function (a, b, c, d, e) {
        this.gl.apply(this, arguments);
    };
    CategoryAngleAxis.prototype.j$e = function (a, b, c, d) {
        this.gm.apply(this, arguments);
    };
    CategoryAngleAxis.prototype.j$f = function (a, b, c, d) {
        this.gn.apply(this, arguments);
    };
    Object.defineProperty(CategoryAngleAxis.prototype, "j$a", {
        get: function () {
            return this.b8;
        },
        enumerable: true,
        configurable: true
    });
    CategoryAngleAxis.$t = markType(CategoryAngleAxis, 'CategoryAngleAxis', CategoryAxisBase.$, [IAngleScaler_$type]);
    CategoryAngleAxis.ls = DependencyProperty.i("Interval", Number_$type, CategoryAngleAxis.$, new PropertyMetadata(2, NaN, function (a, b) {
        a.gz("Interval", b.oldValue, b.newValue);
        a.g5(false);
    }));
    CategoryAngleAxis.lt = DependencyProperty.i("MinorInterval", Number_$type, CategoryAngleAxis.$, new PropertyMetadata(2, NaN, function (a, b) {
        (typeCast(CategoryAngleAxis.$, a)).gz("MinorInterval", b.oldValue, b.newValue);
        (typeCast(CategoryAngleAxis.$, a)).g5(false);
    }));
    return CategoryAngleAxis;
}(CategoryAxisBase));
export { CategoryAngleAxis };
Type.dep(DependencyProperty, PropertyMetadata, CategoryAngleAxis, 'gz', ['StartAngleOffset:lh:lu', [1, 0]]);
/**
 * @hidden
 */
var NumericAngleAxis = /** @class */ (function (_super) {
    tslib_1.__extends(NumericAngleAxis, _super);
    function NumericAngleAxis() {
        var _this = _super.call(this) || this;
        _this.le = null;
        _this.lf = false;
        _this.lg = NaN;
        _this.lh = 0;
        _this.lc = null;
        _this.ab = NumericAngleAxis.$;
        _this.le = new PolarAxisRenderingManager();
        _this.je = _this.jc();
        return _this;
    }
    NumericAngleAxis.prototype.a1 = function () {
        return new NumericAngleAxisView(this);
    };
    NumericAngleAxis.prototype.gr = function (a) {
        _super.prototype.gr.call(this, a);
        this.lb = a;
    };
    NumericAngleAxis.prototype.get_b0 = function () {
        return true;
    };
    NumericAngleAxis.prototype.as = function () {
        var _this = this;
        var a = new AngleAxisLabelPanel();
        a.cz = function (b) {
            var c = _this.bf != null ? _this.bf.mq : Rect.empty;
            var d = !c.isEmpty ? _this.i0 : Rect.empty;
            return _this.lr(b, { $type: Point_$type, x: 0.5, y: 0.5 }, c, d, _this.at.a9);
        };
        return a;
    };
    NumericAngleAxis.prototype.li = function () {
        if (this.ld == null) {
            return 0;
        }
        if (!this.by()) {
            return this.ld.ln();
        }
        else {
            return this.ld.lo((this.e1));
        }
    };
    NumericAngleAxis.prototype.lk = function (a) {
        return Math.round(a * Math.pow(10, 10)) / Math.pow(10, 10);
    };
    NumericAngleAxis.prototype.jc = function () {
        var _this = this;
        var a = _super.prototype.jc.call(this);
        a.b.c = function (b) {
            var e_5, _a;
            if ((_this.aw == null || _this.aw.visibility == 0) && _this.ld != null && _this.lg != b) {
                var c = typeCast(XamDataChart.$, _this.bf);
                if (c == null) {
                    return;
                }
                _this.lg = b;
                _this.at.a8 = b;
                c.ht();
                try {
                    for (var _b = tslib_1.__values(fromEnum(c.axes)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var d = _c.value;
                        if (d != _this && typeCast(AngleAxisLabelPanel.$, d.at) !== null) {
                            d.a2.ad();
                        }
                    }
                }
                catch (e_5_1) { e_5 = { error: e_5_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_5) throw e_5.error; }
                }
            }
        };
        a.e = function (b) {
            b.n = _this.li();
            b.r = b.n;
        };
        a.j = function (b) {
            var c = typeCast(PolarAxisRenderingParameters.$, b);
            if (c.e == c.k._inner[0]) {
                _this.le.i(b.t, b.n, b.z, b.aa, c.center, c.minAngle, c.maxAngle);
            }
        };
        a.k = function (b, c, d) {
            var e = typeCast(PolarAxisRenderingParameters.$, b);
            _this.le.m(c, d, b.z, b.aa, e.minLength, e.maxLength, e.center);
        };
        a.l = function (b, c, d, e) {
            var f = typeCast(PolarAxisRenderingParameters.$, b);
            _this.le.n(c, d, e, f.z, b.aa, f.minLength, f.maxLength, f.center);
        };
        a.d = function (b, c, d, e) {
            var f = typeCast(PolarAxisRenderingParameters.$, _this.jf(b, c, d, e));
            return f;
        };
        a.v = function () {
            if (!_this.lf) {
                _this.lf = true;
                _this.ld.ck();
                _this.lf = false;
            }
        };
        a.m = function (b, c) { return _this.getScaledAngle(c); };
        a.o = function (b, c) { return !c.isEmpty && !b.isEmpty && _this.ld != null; };
        a.q = function (b, c) {
            if (_this.lk(c - _this.lh) < 0) {
                return false;
            }
            if (_this.lk(c - _this.lh - (2 * Math.PI)) > 0) {
                return false;
            }
            return true;
        };
        a.p = function (b, c, d) {
            var e = typeCast(PolarAxisRenderingParameters.$, b);
            var f = _this.lr(_this.getScaledAngle(e.l), e.center, b.aa, b.z, 0);
            var g = _this.lr(c, e.center, b.aa, b.z, 0);
            if (d && MathUtil.e(f.x - g.x, f.y - g.y) < 2) {
                return false;
            }
            if (g.x < b.z.right && g.x >= b.z.left && g.y < b.z.bottom && g.y >= b.z.top) {
                return true;
            }
            return false;
        };
        a.r = function (b, c, d, e) {
            if (c < b.m && typeCast(LogarithmicTickmarkValues.$, b.f) !== null) {
                return b.m;
            }
            else if (c > b.l && (typeCast(LogarithmicTickmarkValues.$, b.f) !== null || b.h)) {
                return b.l;
            }
            return c;
        };
        return a;
    };
    NumericAngleAxis.prototype.lr = function (a, b, c, d, e) {
        var f = this.li();
        var g = ViewportUtils.b(e, c, d);
        if (this.aw != null && (this.aw.c == 5 || this.aw.c == 1)) {
            g *= -1;
        }
        var h = b.x + (f + g) * Math.cos(a);
        var i = b.y + (f + g) * Math.sin(a);
        h = ViewportUtils.c(h, c, d);
        i = ViewportUtils.f(i, c, d);
        return { $type: Point_$type, x: h, y: i };
    };
    NumericAngleAxis.prototype.dv = function (a, b) {
        return this.getScaledAngle(a);
    };
    NumericAngleAxis.prototype.getScaledAngle1 = function (a, b, c) {
        var d = 0;
        if (b) {
            d = (Math.log(a) - this.j9) / (this.j8 - this.j9);
        }
        else {
            d = (a - this.j3) / (this.j2 - this.j3);
        }
        if (c) {
            d = 1 - d;
        }
        return (d * 2 * Math.PI) + this.lh;
    };
    NumericAngleAxis.prototype.getScaledAngle = function (a) {
        return this.getScaledAngle1(a, this.jt, this.b9);
    };
    NumericAngleAxis.prototype.dy = function (a, b) {
        return this.getUnscaledAngle(a);
    };
    NumericAngleAxis.prototype.getUnscaledAngle = function (a) {
        var b = (a - this.lh) / (2 * Math.PI);
        if (this.b8) {
            b = 1 - b;
        }
        if (this.jt) {
            return Math.exp(b * (this.j8 - this.j9) + this.j9);
        }
        else {
            return this.j3 + b * (this.j2 - this.j3);
        }
    };
    NumericAngleAxis.prototype.gt = function (a, b, c, d) {
        var e_6, _a, e_7, _b, e_8, _c;
        _super.prototype.gt.call(this, a, b, c, d);
        var e = typeCast(XamDataChart.$, this.bf);
        switch (b) {
            case "CrossingAxis":
                var f = typeCast(NumericRadiusAxis.$, d);
                this.lp(f);
                if (f != null) {
                    f.lx(this);
                }
                this.g5(false);
                break;
            case NumericAngleAxis.$$p[0]:
                this.lh = this.ll;
                while (this.lh < 0) {
                    this.lh += 360;
                }
                while (this.lh >= 360) {
                    this.lh -= 360;
                }
                this.lh = (this.ll * Math.PI) / 180;
                this.g5(false);
                try {
                    for (var _d = tslib_1.__values(fromEnum(this.ct())), _e = _d.next(); !_e.done; _e = _d.next()) {
                        var g = _e.value;
                        g.jc(false);
                        g.ij();
                    }
                }
                catch (e_6_1) { e_6 = { error: e_6_1 }; }
                finally {
                    try {
                        if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                    }
                    finally { if (e_6) throw e_6.error; }
                }
                break;
            case "Label":
                if (e != null) {
                    try {
                        for (var _f = tslib_1.__values(fromEnum(e.axes)), _g = _f.next(); !_g.done; _g = _f.next()) {
                            var h = _g.value;
                            h.g4();
                        }
                    }
                    catch (e_7_1) { e_7 = { error: e_7_1 }; }
                    finally {
                        try {
                            if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                        }
                        finally { if (e_7) throw e_7.error; }
                    }
                }
                break;
            case "CrossingValue":
                if (e != null) {
                    try {
                        for (var _h = tslib_1.__values(fromEnum(e.axes)), _j = _h.next(); !_j.done; _j = _h.next()) {
                            var i = _j.value;
                            if (typeCast(NumericAngleAxis.$, i) !== null || typeCast(CategoryAngleAxis.$, i) !== null) {
                                i.g4();
                            }
                        }
                    }
                    catch (e_8_1) { e_8 = { error: e_8_1 }; }
                    finally {
                        try {
                            if (_j && !_j.done && (_c = _h.return)) _c.call(_h);
                        }
                        finally { if (e_8) throw e_8.error; }
                    }
                }
                break;
            case "LabelSettings":
                this.je = this.jc();
                this.ln();
                this.cf = true;
                this.g5(false);
                break;
        }
    };
    NumericAngleAxis.prototype.ln = function () {
        this.lg = NaN;
    };
    NumericAngleAxis.prototype.jg = function () {
        return new PolarAxisRenderingParameters();
    };
    NumericAngleAxis.prototype.jf = function (a, b, c, d) {
        var e = typeCast(PolarAxisRenderingParameters.$, _super.prototype.jf.call(this, a, b, c, d));
        var f = this.le.f(b);
        var g = this.le.g(b);
        var h = 0.5 * this.ld.li;
        var i = 0.5 * this.ld.lh;
        var j = f;
        var k = g;
        var l = this.ld.ln();
        if (isNaN_(l) || isInfinity(l)) {
            return null;
        }
        if (k >= h) {
            k = l;
        }
        if (j < i) {
            j = i;
        }
        var m = a.width;
        this.le.k(b, e, this.j3, this.j2, this.b8, runOn(this, this.getUnscaledAngle), m);
        var n = { $type: Point_$type, x: 0.5, y: 0.5 };
        e.center = n;
        e.maxLength = k;
        e.minLength = j;
        e.effectiveMaximum = l;
        return e;
    };
    NumericAngleAxis.prototype.getMinMaxAngle = function (a, b, c) {
        var d = this.le.l(a, b, c);
        b = d.p1;
        c = d.p2;
        return {
            p1: b,
            p2: c
        };
    };
    NumericAngleAxis.prototype.g6 = function (a) {
        _super.prototype.g6.call(this, a);
        var b = this.bf != null ? this.bf.mq : Rect.empty;
        var c = !b.isEmpty ? this.i0 : Rect.empty;
        var d = this.iv();
        var e = this.iu();
        var f = this.je.c(a, c, b, d, e);
        if (f != null) {
            this.j1 = f.f.m;
            this.j4 = f.f.l;
        }
    };
    Object.defineProperty(NumericAngleAxis.prototype, "ld", {
        get: function () {
            if (this.lc != null) {
                return this.lc;
            }
            var a = typeCast(XamDataChart.$, this.bf);
            if (a != null) {
                for (var b = 0; b < a.axes.count; b++) {
                    if (a.axes._inner[b].cb) {
                        return a.axes._inner[b];
                    }
                }
            }
            return this.lc;
        },
        set: function (a) {
            this.lc = a;
        },
        enumerable: true,
        configurable: true
    });
    NumericAngleAxis.prototype.ah = function () {
        return this.ld;
    };
    NumericAngleAxis.prototype.lp = function (a) {
        this.ld = a;
    };
    NumericAngleAxis.prototype.hg = function (a, b) {
        _super.prototype.hg.call(this, a, b);
        if (b.height != a.height || b.width != a.width) {
            this.ck();
        }
    };
    NumericAngleAxis.prototype.get_az = function () {
        return 2;
    };
    Object.defineProperty(NumericAngleAxis.prototype, "az", {
        get: function () {
            return this.get_az();
        },
        enumerable: true,
        configurable: true
    });
    NumericAngleAxis.prototype.is = function (a, b, c, d, e) {
        if (d === void 0) { d = null; }
        if (e === void 0) { e = null; }
        if (this.ld == null) {
            return null;
        }
        var f = typeCast(PolarAxisRenderingParameters.$, this.jf(c, b, d, e));
        if (f == null) {
            return null;
        }
        var g = Math.cos(a);
        var h = Math.sin(a);
        var i = f.center.x + g * f.minLength;
        var j = f.center.y + h * f.minLength;
        var k = f.center.x + g * f.maxLength;
        var l = f.center.y + h * f.maxLength;
        i = ViewportUtils.c(i, b, c);
        j = ViewportUtils.f(j, b, c);
        k = ViewportUtils.c(k, b, c);
        l = ViewportUtils.f(l, b, c);
        var m = new LineGeometry();
        m.c = { $type: Point_$type, x: i, y: j };
        m.b = { $type: Point_$type, x: k, y: l };
        return m;
    };
    NumericAngleAxis.$t = markType(NumericAngleAxis, 'NumericAngleAxis', NumericAxisBase.$, [IAngleScaler_$type]);
    return NumericAngleAxis;
}(NumericAxisBase));
export { NumericAngleAxis };
Type.dep(DependencyProperty, PropertyMetadata, NumericAngleAxis, 'gz', ['StartAngleOffset:ll:lq', [1, 0]]);
/**
 * @hidden
 */
var NumericRadiusAxis = /** @class */ (function (_super) {
    tslib_1.__extends(NumericRadiusAxis, _super);
    function NumericRadiusAxis() {
        var _this = _super.call(this) || this;
        _this.le = null;
        _this.lf = false;
        _this.lg = NaN;
        _this.li = 0;
        _this.lh = 0;
        _this.lk = 0;
        _this.lb = null;
        _this.ab = NumericRadiusAxis.$;
        _this.li = _this.ls;
        _this.lh = _this.lr;
        _this.le = new PolarAxisRenderingManager();
        _this.je = _this.jc();
        return _this;
    }
    NumericRadiusAxis.prototype.a1 = function () {
        return new NumericRadiusAxisView(this);
    };
    NumericRadiusAxis.prototype.gr = function (a) {
        _super.prototype.gr.call(this, a);
        this.ld = a;
    };
    NumericRadiusAxis.prototype.get_cb = function () {
        return true;
    };
    NumericRadiusAxis.prototype.du = function () {
        var a = this.li;
        var b = this.lh;
        var c = Math.min(this.lm(), this.ll()) * (a - b) / 2;
        c = Math.max(c, 14);
        return c;
    };
    NumericRadiusAxis.prototype.lm = function () {
        return this.i0.width;
    };
    NumericRadiusAxis.prototype.ll = function () {
        return this.i0.height;
    };
    NumericRadiusAxis.prototype.as = function () {
        return new RadialAxisLabelPanel();
    };
    NumericRadiusAxis.prototype.lj = function (a) {
        if (a == null) {
            return 0;
        }
        return a;
    };
    NumericRadiusAxis.prototype.jc = function () {
        var _this = this;
        var a = _super.prototype.jc.call(this);
        a.b.c = function (b) {
            if ((_this.aw == null || _this.aw.visibility == 0) && _this.lc != null) {
                if ((_this.aw == null || (_this.aw.c == 4 || _this.aw.c == 5)) && _this.lg != b) {
                    _this.lg = b;
                    _this.at.a8 = b;
                    _this.bf.ht();
                }
            }
        };
        a.k = function (b, c, d) {
            var e = typeCast(PolarAxisRenderingParameters.$, b);
            _this.le.i(c, d, e.z, e.aa, e.center, e.minAngle, e.maxAngle);
        };
        a.l = function (b, c, d, e) {
            if (d == e) {
                return;
            }
            var f = typeCast(PolarAxisRenderingParameters.$, b);
            _this.le.j(c, d, e, f.z, f.aa, f.center, f.minAngle, f.maxAngle);
        };
        a.m = function (b, c) { return _this.lo(c); };
        a.q = function (b, c) {
            var d = typeCast(PolarAxisRenderingParameters.$, b);
            return c <= d.effectiveMaximum;
        };
        a.n = function (b, c) {
            var d = typeCast(PolarAxisRenderingParameters.$, b);
            return c <= d.effectiveMaximum;
        };
        a.j = function (b) {
            var c = typeCast(PolarAxisRenderingParameters.$, b);
            _this.le.m(b.t, c.crossingAngleRadians, b.z, b.aa, c.minLength, c.maxLength, c.center);
        };
        a.e = function (b) {
            b.n = _this.aw == null || (_this.aw.c == 4 || _this.aw.c == 0) ? b.z.top : b.z.bottom;
            b.r = b.n;
            var c = typeCast(PolarAxisRenderingParameters.$, b);
            c.crossingAngleRadians = (_this.lj(_this.e1) * Math.PI) / 180;
            if (_this.aw == null || (_this.aw.c == 4 || _this.aw.c == 5)) {
                c.n = ViewportUtils.f(0.5, c.aa, c.z) - b.z.top;
                c.r = c.n;
                var d = typeCast(RadialAxisLabelPanel.$, _this.at);
                if (d != null) {
                    var e = 0;
                    if (_this.aw != null && _this.aw.c == 4) {
                        e = 1;
                    }
                    d.dh = { $type: Point_$type, x: ViewportUtils.c(0.5, c.aa, c.z), y: e };
                    d.de = c.crossingAngleRadians;
                }
            }
        };
        a.p = function (b, c, d) {
            var e = typeCast(PolarAxisRenderingParameters.$, b);
            if (_this.lc == null) {
                return false;
            }
            if (c > e.effectiveMaximum) {
                return false;
            }
            var f = false;
            f = _this.aw == null || (_this.aw.c == 4 || _this.aw.c == 5);
            var g = 0;
            if (f) {
                g = _this.lk;
            }
            var h = e.center.x + c * Math.cos(g);
            var i = e.center.y + c * Math.sin(g);
            h = ViewportUtils.c(h, e.aa, e.z);
            i = ViewportUtils.f(i, e.aa, e.z);
            if (h <= b.z.right && h >= b.z.left && ((i <= b.z.bottom && i >= b.z.top) || !f)) {
                return true;
            }
            return false;
        };
        a.h = function (b, c) {
            var d = typeCast(PolarAxisRenderingParameters.$, b);
            return new LabelPosition(ViewportUtils.c(d.center.x + c, d.aa, d.z));
        };
        a.r = function (b, c, d, e) {
            if (c < b.m) {
                return b.m;
            }
            else if (c > b.l) {
                return b.l;
            }
            return c;
        };
        return a;
    };
    NumericRadiusAxis.prototype.dv = function (a, b) {
        return this.lo(a);
    };
    NumericRadiusAxis.prototype.lp = function (a, b, c, d, e) {
        var f = 0;
        if (b) {
            if (a <= 0) {
                f = (Math.log(this.j3) - this.j9) / (this.j8 - this.j9);
            }
            else {
                f = (Math.log(a) - this.j9) / (this.j8 - this.j9);
            }
        }
        else {
            f = (a - this.j3) / (this.j2 - this.j3);
        }
        if (c) {
            f = 1 - f;
        }
        f = e + (f * (d - e));
        f /= 2;
        return f;
    };
    NumericRadiusAxis.prototype.lo = function (a) {
        return this.lp(a, this.jt, this.b9, this.li, this.lh);
    };
    NumericRadiusAxis.prototype.lq = function (a) {
        var b = a * 2;
        b = (b - this.lh) / (this.li - this.lh);
        if (this.b8) {
            b = 1 - b;
        }
        if (this.jt) {
            return Math.exp(b * (this.j8 - this.j9) + this.j9);
        }
        else {
            return this.j3 + b * (this.j2 - this.j3);
        }
    };
    NumericRadiusAxis.prototype.dy = function (a, b) {
        return this.lq(a);
    };
    NumericRadiusAxis.prototype.gt = function (a, b, c, d) {
        var e_9, _a, e_10, _b;
        switch (b) {
            case "CrossingValue":
                this.lk = this.lj(this.e1) * Math.PI / 180;
                break;
        }
        _super.prototype.gt.call(this, a, b, c, d);
        switch (b) {
            case NumericRadiusAxis.$$p[1]:
                this.li = this.ls;
                if (this.li < 0) {
                    this.li = 0.1;
                }
                if (this.li > 1) {
                    this.li = 1;
                }
                if (this.lh >= this.li) {
                    this.lh = this.li - 0.01;
                    if (this.lh < 0) {
                        this.lh = 0;
                        this.li = 0.01;
                    }
                }
                this.cf = true;
                this.g5(false);
                if (this.lc != null) {
                    this.lc.g4();
                }
                try {
                    for (var _c = tslib_1.__values(fromEnum(this.ct())), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var e = _d.value;
                        e.jc(false);
                    }
                }
                catch (e_9_1) { e_9 = { error: e_9_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                    }
                    finally { if (e_9) throw e_9.error; }
                }
                break;
            case NumericRadiusAxis.$$p[0]:
                this.lh = this.lr;
                if (this.lh < 0) {
                    this.lh = 0.1;
                }
                if (this.lh > 1) {
                    this.lh = 1;
                }
                if (this.lh >= this.li) {
                    this.lh = this.li - 0.01;
                    if (this.lh < 0) {
                        this.lh = 0;
                        this.li = 0.01;
                    }
                }
                this.cf = true;
                this.g5(false);
                if (this.lc != null) {
                    this.lc.g4();
                }
                try {
                    for (var _e = tslib_1.__values(fromEnum(this.ct())), _f = _e.next(); !_f.done; _f = _e.next()) {
                        var f = _f.value;
                        f.dl = true;
                        f.jc(false);
                    }
                }
                catch (e_10_1) { e_10 = { error: e_10_1 }; }
                finally {
                    try {
                        if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                    }
                    finally { if (e_10) throw e_10.error; }
                }
                if (this.bf != null) {
                    this.bf.h1();
                }
                break;
            case "CrossingAxis":
                var g = typeCast(NumericAngleAxis.$, d);
                var h = typeCast(CategoryAngleAxis.$, d);
                if (g == null && h == null) {
                    this.lx(null);
                }
                if (g != null) {
                    this.lx(g);
                    g.lp(this);
                }
                if (h != null) {
                    this.lx(h);
                    h.lr(this);
                }
                this.cf = true;
                this.g5(false);
                break;
            case "IsInverted":
                if (this.lc != null) {
                    this.lc.cf = true;
                    this.lc.g5(false);
                }
                break;
        }
    };
    NumericRadiusAxis.prototype.lx = function (a) {
        this.lc = a;
    };
    Object.defineProperty(NumericRadiusAxis.prototype, "lc", {
        get: function () {
            if (this.lb != null) {
                return this.lb;
            }
            var a = typeCast(XamDataChart.$, this.bf);
            if (a != null) {
                for (var b = 0; b < a.axes.count; b++) {
                    if (typeCast(NumericAngleAxis.$, a.axes._inner[b]) !== null || typeCast(CategoryAngleAxis.$, a.axes._inner[b]) !== null) {
                        return a.axes._inner[b];
                    }
                }
            }
            return this.lb;
        },
        set: function (a) {
            this.lb = a;
        },
        enumerable: true,
        configurable: true
    });
    NumericRadiusAxis.prototype.ah = function () {
        return this.lc;
    };
    NumericRadiusAxis.prototype.jg = function () {
        return new PolarAxisRenderingParameters();
    };
    NumericRadiusAxis.prototype.jf = function (a, b, c, d) {
        var e = typeCast(PolarAxisRenderingParameters.$, _super.prototype.jf.call(this, a, b, c, d));
        var f = this.le.f(b);
        var g = this.le.g(b);
        var h = 0.5 * this.li;
        var i = 0.5 * this.lh;
        var j, k;
        if (Rect.l_op_Equality(b, SeriesViewer.m0)) {
            k = this.j2;
            j = this.j3;
        }
        else {
            k = Math.min(g, h);
            j = this.lq(f);
            k = this.lq(k);
            var l = this.ly(a, b, j, k);
            j = l.p2;
            k = l.p3;
        }
        var m = { $type: Point_$type, x: 0.5, y: 0.5 };
        var n = f;
        var o = g;
        var p = Math.min(a.width, a.height) * (this.li - this.lh) / 2;
        e.center = m;
        var q = Math.max(o, n);
        var r = Math.min(n, o);
        e.maxLength = q;
        e.minLength = r;
        var s = Math.min(j, k);
        var t = Math.max(j, k);
        if (s < this.j3) {
            s = this.j3;
        }
        if (t > this.j2) {
            t = this.j2;
        }
        this.j6 = s;
        this.j5 = t;
        e.k.add(((function () {
            var $ret = new RangeInfo();
            $ret.d = s;
            $ret.c = t;
            $ret.b = p;
            return $ret;
        })()));
        var u = typeCast(IAngleScaler_$type, this.lc);
        if (u != null) {
            var v = void 0;
            var w = void 0;
            var x = u.getMinMaxAngle(b, v, w);
            v = x.p1;
            w = x.p2;
            e.minAngle = Math.min(v, w);
            e.maxAngle = Math.max(v, w);
        }
        var y = this.ln();
        if (o >= h) {
            o = y;
        }
        if (n < i) {
            n = i;
        }
        e.minLength = n;
        e.maxLength = o;
        e.effectiveMaximum = y;
        e.f = this.jj;
        var z = typeCast(LinearTickmarkValues.$, e.f);
        if (z != null) {
            z.y = this.lt(m, e.minLength, e.maxLength, b, a);
        }
        return e;
    };
    NumericRadiusAxis.prototype.ly = function (a, b, c, d) {
        var e = { $type: Point_$type, x: 0.5, y: 0.5 };
        var f = 0;
        if (this.b8) {
            f = this.lo(this.j3);
        }
        else {
            f = this.lo(this.j2);
        }
        var g = 0;
        if (this.e1 != null) {
            g = this.lk;
        }
        var h = e.x + f * Math.cos(g);
        var i = e.y + f * Math.sin(g);
        e.x = ViewportUtils.c(e.x, b, a);
        e.y = ViewportUtils.f(e.y, b, a);
        h = ViewportUtils.c(h, b, a);
        i = ViewportUtils.f(i, b, a);
        if (h >= a.left && h <= a.right && i >= a.top && i <= a.bottom) {
            if (this.b8) {
                d = this.j3;
            }
            else {
                d = this.j2;
            }
        }
        if (e.x >= a.left && e.x <= a.right && e.y >= a.top && e.y <= a.bottom) {
            if (this.b8) {
                c = this.j2;
            }
            else {
                c = this.j3;
            }
        }
        return {
            p2: c,
            p3: d
        };
    };
    NumericRadiusAxis.prototype.lt = function (a, b, c, d, e) {
        var f = ViewportUtils.d(c - b, d, e);
        var g = f / Math.min(e.width, e.height);
        if (g > 0.7) {
            return 10;
        }
        return 5;
    };
    NumericRadiusAxis.prototype.ln = function () {
        var a = 0;
        if (!this.b8) {
            a = this.lo(this.j2);
        }
        else {
            a = this.lo(this.j3);
        }
        return a;
    };
    NumericRadiusAxis.prototype.k1 = function (a) {
        if (this.lc != null) {
            this.lc.g4();
        }
    };
    NumericRadiusAxis.prototype.g6 = function (a) {
        _super.prototype.g6.call(this, a);
        var b = this.bf != null ? this.bf.mq : Rect.empty;
        var c = !b.isEmpty ? this.i0 : Rect.empty;
        var d = this.iv();
        var e = this.iu();
        var f = this.je.c(a, c, b, d, e);
        if (f != null) {
            this.j1 = f.f.m;
            this.j4 = f.f.l;
        }
    };
    NumericRadiusAxis.prototype.hg = function (a, b) {
        _super.prototype.hg.call(this, a, b);
        if (b.height != a.height || b.width != a.width) {
            this.ck();
        }
    };
    NumericRadiusAxis.prototype.lw = function (a, b, c, d, e) {
        var f = typeCast(IPolarRadialRenderingParameters_$type, this.jf(b, c, d, e));
        if (f == null) {
            return;
        }
        this.le.j(a.c, f.minLength, f.maxLength, b, c, f.center, f.minAngle, f.maxAngle);
    };
    NumericRadiusAxis.prototype.get_az = function () {
        return 3;
    };
    Object.defineProperty(NumericRadiusAxis.prototype, "az", {
        get: function () {
            return this.get_az();
        },
        enumerable: true,
        configurable: true
    });
    NumericRadiusAxis.prototype.is = function (a, b, c, d, e) {
        if (d === void 0) { d = null; }
        if (e === void 0) { e = null; }
        var f = typeCast(PolarAxisRenderingParameters.$, this.jf(c, b, d, e));
        var g = a;
        var h = new PathGeometry();
        var i = ViewportUtils.d(g, b, c);
        var j = ViewportUtils.g(g, b, c);
        if (i <= 0 || j <= 0) {
            return null;
        }
        var k = ViewportUtils.c(f.center.x, b, c);
        var l = ViewportUtils.f(f.center.y, b, c);
        if (f.maxAngle - f.minAngle < Math.PI && f.maxAngle - f.minAngle > 0) {
            var m = { $type: Point_$type, x: ViewportUtils.c(f.center.x + g * Math.cos(f.minAngle), b, c), y: ViewportUtils.f(f.center.y + g * Math.sin(f.minAngle), b, c) };
            var n_1 = { $type: Point_$type, x: ViewportUtils.c(f.center.x + g * Math.cos(f.maxAngle), b, c), y: ViewportUtils.f(f.center.y + g * Math.sin(f.maxAngle), b, c) };
            var o = new PathFigure();
            o._startPoint = m;
            o._isClosed = false;
            o._segments.add(((function () {
                var $ret = new ArcSegment();
                $ret.b = false;
                $ret.e = n_1;
                $ret.f = new Size(1, i, j);
                $ret.d = 1;
                return $ret;
            })()));
            h.b.add(o);
        }
        else {
            var p = new PathFigure();
            p._startPoint = { $type: Point_$type, x: k, y: l - j };
            p._isClosed = true;
            p._segments.add(((function () {
                var $ret = new ArcSegment();
                $ret.b = false;
                $ret.e = { $type: Point_$type, x: k, y: l + j };
                $ret.f = new Size(1, i, j);
                $ret.d = 1;
                return $ret;
            })()));
            p._segments.add(((function () {
                var $ret = new ArcSegment();
                $ret.b = false;
                $ret.e = { $type: Point_$type, x: k, y: l - j };
                $ret.f = new Size(1, i, j);
                $ret.d = 1;
                return $ret;
            })()));
            h.b.add(p);
        }
        return h;
    };
    NumericRadiusAxis.$t = markType(NumericRadiusAxis, 'NumericRadiusAxis', NumericAxisBase.$);
    return NumericRadiusAxis;
}(NumericAxisBase));
export { NumericRadiusAxis };
Type.dep(DependencyProperty, PropertyMetadata, NumericRadiusAxis, 'gz', ['InnerRadiusExtentScale:lr:lz', [1, 0], 'RadiusExtentScale:ls:l0', [1, 0.75]]);
/**
 * @hidden
 */
var AngleAxisLabelPanelView = /** @class */ (function (_super) {
    tslib_1.__extends(AngleAxisLabelPanelView, _super);
    function AngleAxisLabelPanelView(a) {
        var _this = _super.call(this, a) || this;
        _this.ax = null;
        _this.ax = a;
        return _this;
    }
    AngleAxisLabelPanelView.prototype.am = function () {
        _super.prototype.am.call(this);
        this.ax.ct = true;
    };
    AngleAxisLabelPanelView.prototype.ay = function (a, b) {
        this.ax.cx = -1.7976931348623157E+308;
        this.ax.cu = -1.7976931348623157E+308;
        this.ax.cw = -1.7976931348623157E+308;
        this.ax.cv = -1.7976931348623157E+308;
        for (var c = 0; c < this.a.a3.count; c++) {
            var d = this.a.a3._inner[c];
            var e = this.a.a2._inner[c];
            var f = e.a;
            var g = this.ax.cz(e.c);
            var h = this.m(d);
            var i = this.l(d);
            var j = h + this.p() + this.q();
            var k = i + this.r() + this.o();
            var l = g.x - j / 2;
            var m = g.y - k / 2;
            this.ax.cw = Math.max(h, this.ax.cw);
            this.ax.cv = Math.max(i, this.ax.cv);
            this.ax.cx = Math.max(j, this.ax.cx);
            this.ax.cu = Math.max(k, this.ax.cu);
            var n = new Rect(0, l, m, j, k);
            a.add(n);
            b.add(f);
        }
    };
    AngleAxisLabelPanelView.prototype.o = function () {
        return 0;
    };
    AngleAxisLabelPanelView.prototype.p = function () {
        return 0;
    };
    AngleAxisLabelPanelView.prototype.q = function () {
        return 0;
    };
    AngleAxisLabelPanelView.prototype.r = function () {
        return 0;
    };
    AngleAxisLabelPanelView.$t = markType(AngleAxisLabelPanelView, 'AngleAxisLabelPanelView', AxisLabelPanelBaseView.$);
    return AngleAxisLabelPanelView;
}(AxisLabelPanelBaseView));
export { AngleAxisLabelPanelView };
/**
 * @hidden
 */
var CategoryAngleAxisView = /** @class */ (function (_super) {
    tslib_1.__extends(CategoryAngleAxisView, _super);
    function CategoryAngleAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bg = null;
        _this.bg = a;
        return _this;
    }
    CategoryAngleAxisView.$t = markType(CategoryAngleAxisView, 'CategoryAngleAxisView', CategoryAxisBaseView.$);
    return CategoryAngleAxisView;
}(CategoryAxisBaseView));
export { CategoryAngleAxisView };
/**
 * @hidden
 */
var NumericAngleAxisView = /** @class */ (function (_super) {
    tslib_1.__extends(NumericAngleAxisView, _super);
    function NumericAngleAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bg = null;
        _this.bg = a;
        return _this;
    }
    NumericAngleAxisView.$t = markType(NumericAngleAxisView, 'NumericAngleAxisView', NumericAxisBaseView.$);
    return NumericAngleAxisView;
}(NumericAxisBaseView));
export { NumericAngleAxisView };
/**
 * @hidden
 */
var NumericRadiusAxisView = /** @class */ (function (_super) {
    tslib_1.__extends(NumericRadiusAxisView, _super);
    function NumericRadiusAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bg = null;
        _this.bg = a;
        return _this;
    }
    NumericRadiusAxisView.$t = markType(NumericRadiusAxisView, 'NumericRadiusAxisView', NumericAxisBaseView.$);
    return NumericRadiusAxisView;
}(NumericAxisBaseView));
export { NumericRadiusAxisView };
