// igniteui-react-charts/ES5/AverageDirectionalIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageDirectionalIndexIndicatorStrategy } from "./AverageDirectionalIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var AverageDirectionalIndexIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(AverageDirectionalIndexIndicator, _super);
    function AverageDirectionalIndexIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AverageDirectionalIndexIndicator.prototype.get_qb = function () {
        return new AverageDirectionalIndexIndicatorStrategy();
    };
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    AverageDirectionalIndexIndicator.prototype.get_qo = function () {
        return AverageDirectionalIndexIndicator.$;
    };
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(AverageDirectionalIndexIndicator.qx));
        },
        set: function (a) {
            this.h(AverageDirectionalIndexIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    AverageDirectionalIndexIndicator.prototype.qh = function () {
        return this.qw;
    };
    AverageDirectionalIndexIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    AverageDirectionalIndexIndicator.$t = markType(AverageDirectionalIndexIndicator, 'AverageDirectionalIndexIndicator', StrategyBasedIndicator.$);
    AverageDirectionalIndexIndicator.qx = StrategyBasedIndicator.qt(14, AverageDirectionalIndexIndicator.$);
    return AverageDirectionalIndexIndicator;
}(StrategyBasedIndicator));
export { AverageDirectionalIndexIndicator };
