// igniteui-react-charts/ES5/AverageTrueRangeIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageTrueRangeIndicatorStrategy } from "./AverageTrueRangeIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var AverageTrueRangeIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(AverageTrueRangeIndicator, _super);
    function AverageTrueRangeIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AverageTrueRangeIndicator.prototype.get_qb = function () {
        return new AverageTrueRangeIndicatorStrategy();
    };
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    AverageTrueRangeIndicator.prototype.get_qo = function () {
        return AverageTrueRangeIndicator.$;
    };
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(AverageTrueRangeIndicator.qx));
        },
        set: function (a) {
            this.h(AverageTrueRangeIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    AverageTrueRangeIndicator.prototype.qh = function () {
        return this.qw;
    };
    AverageTrueRangeIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    AverageTrueRangeIndicator.$t = markType(AverageTrueRangeIndicator, 'AverageTrueRangeIndicator', StrategyBasedIndicator.$);
    AverageTrueRangeIndicator.qx = StrategyBasedIndicator.qt(14, AverageTrueRangeIndicator.$);
    return AverageTrueRangeIndicator;
}(StrategyBasedIndicator));
export { AverageTrueRangeIndicator };
