// igniteui-react-charts/ES5/CalloutLayer_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { AnnotationLayer } from "./AnnotationLayer";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { Base, Type, String_$type, typeCast, enumGetBox, fromEnum, runOn, delegateCombine, delegateRemove, Date_$type, Point_$type, markType } from "igniteui-react-core/ES5/type";
import { Series } from "./Series";
import { Brush } from "igniteui-react-core/ES5/Brush";
import { Thickness } from "igniteui-react-core/ES5/Thickness";
import { CalloutCollisionMode_$type } from "./CalloutCollisionMode";
import { FontUtil } from "igniteui-react-core/ES5/FontUtil";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { CalloutLayerFrame } from "./CalloutLayerFrame";
import { HashSet$1 } from "igniteui-react-core/ES5/HashSet$1";
import { CalloutAnnotationWrapper } from "./CalloutAnnotationWrapper";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { LinkedList$1 } from "igniteui-react-core/ES5/LinkedList$1";
import { Dictionary$2 } from "igniteui-react-core/ES5/Dictionary$2";
import { LinkedListNode$1 } from "igniteui-react-core/ES5/LinkedListNode$1";
import { LabelCollisionManager } from "igniteui-react-core/ES5/LabelCollisionManager";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { CalloutAnnotation } from "./CalloutAnnotation";
import { IFastItemsSource_$type } from "igniteui-react-core/ES5/IFastItemsSource";
import { CalloutSeriesSelectingEventArgs } from "./CalloutSeriesSelectingEventArgs";
import { CalloutStyleUpdatingEventArgs } from "./CalloutStyleUpdatingEventArgs";
import { QuadTreeBoundingBox } from "igniteui-react-core/ES5/QuadTreeBoundingBox";
import { LabelCollisionInfo } from "igniteui-react-core/ES5/LabelCollisionInfo";
import { ScalerParams } from "./ScalerParams";
import { Size } from "igniteui-react-core/ES5/Size";
import { CalloutContentUpdatingEventArgs } from "./CalloutContentUpdatingEventArgs";
import { CalloutLabelUpdatingEventArgs } from "./CalloutLabelUpdatingEventArgs";
import { CalloutKey } from "./CalloutKey";
import { DataContext } from "igniteui-react-core/ES5/DataContext";
import { HashPool$2 } from "igniteui-react-core/ES5/HashPool$2";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { isNaN_ } from "igniteui-react-core/ES5/number";
import { stringIsNullOrEmpty } from "igniteui-react-core/ES5/string";
import { FastItemObjectColumn } from "igniteui-react-core/ES5/FastItemObjectColumn";
import { AnnotationLayerView } from "./AnnotationLayerView";
import { BrushUtil } from "igniteui-react-core/ES5/BrushUtil";
import { TextBlock } from "igniteui-react-core/ES5/TextBlock";
import { DOMExecutionContext } from "igniteui-react-core/ES5/DOMExecutionContext";
import { Color } from "igniteui-react-core/ES5/Color";
import { Line } from "igniteui-react-core/ES5/Line";
import { CalloutViewHelper } from "./CalloutViewHelper";
import { Rectangle } from "igniteui-react-core/ES5/Rectangle";
import { TranslateTransform } from "igniteui-react-core/ES5/TranslateTransform";
/**
 * @hidden
 */
var CalloutLayer = /** @class */ (function (_super) {
    tslib_1.__extends(CalloutLayer, _super);
    function CalloutLayer() {
        var _this = _super.call(this) || this;
        _this.calloutStyleUpdating = null;
        _this.ne = null;
        _this.nf = null;
        _this.ng = null;
        _this.nd = null;
        _this.nc = null;
        _this.n2 = new HashSet$1(CalloutAnnotationWrapper.$, 0);
        _this.n5 = new List$1(CalloutAnnotationWrapper.$, 0);
        _this.n6 = new List$1(CalloutAnnotationWrapper.$, 0);
        _this.na = null;
        _this.n3 = new LinkedList$1(CalloutAnnotationWrapper.$);
        _this.n0 = new Dictionary$2(CalloutAnnotationWrapper.$, LinkedListNode$1.$.specialize(CalloutAnnotationWrapper.$), 0);
        _this.nm = null;
        _this.qv = new Rect(0, 0, 0, 1, 1);
        _this.ns = false;
        _this.og = 0;
        _this.qu = Rect.empty;
        _this.nr = false;
        _this.calloutContentUpdating = null;
        _this.calloutLabelUpdating = null;
        _this.calloutSeriesSelecting = null;
        _this.n4 = new List$1(SeriesPointOfInterest.$, 0);
        _this.nu = false;
        _this.nv = false;
        _this.n8 = 0;
        _this.nt = false;
        _this.nq = true;
        FastItemObjectColumn.m();
        _this.l9 = true;
        var a = new CalloutLayerFrame();
        var b = new CalloutLayerFrame();
        var c = new CalloutLayerFrame();
        var d = _this.fi / 1000;
        _this.l1 = a;
        _this.l0 = b;
        _this.l2 = c;
        _this.ab = CalloutLayer.$;
        return _this;
    }
    CalloutLayer.prototype.pn = function (a, b) {
        this.n6.clear();
        this.jc(false);
    };
    CalloutLayer.prototype.hy = function (a, b, c, d) {
        _super.prototype.hy.call(this, a, b, c, d);
        var e = this.n1();
        switch (a) {
            case 3:
                this.pz(b, e);
                break;
            case 1:
                for (var f = b; f < b + c; f++) {
                    var g = new CalloutAnnotation();
                    var h = new CalloutAnnotationWrapper();
                    h.b = g;
                    h.b.series = this.m9;
                    this.n6.insert(f, h);
                    this.pz(f, e);
                }
                break;
            case 0:
                this.n6.v(b, c);
                break;
            case 2:
                for (var i = b; i < b + c; i++) {
                    this.pz(i, e);
                }
                break;
            case 4:
                this.n6.clear();
                for (var j = 0; j < this.bz.count; j++) {
                    var k = new CalloutAnnotation();
                    var l = new CalloutAnnotationWrapper();
                    l.b = k;
                    l.b.series = this.m9;
                    this.n6.add(l);
                    this.pz(j, e);
                }
                break;
        }
        this.jc(true);
    };
    CalloutLayer.prototype.da = function (a) {
        return false;
    };
    CalloutLayer.prototype.n1 = function () {
        var a = new Dictionary$2(String_$type, Series.$, 0);
        if (this.bs != null) {
            for (var b = 0; b < this.bs.series.count; b++) {
                var c = this.bs.series._inner[b];
                if (c.name != null && !a.containsKey(c.name)) {
                    a.addItem(c.name, c);
                }
            }
        }
        return a;
    };
    CalloutLayer.prototype.pz = function (a, b) {
        var _this = this;
        if (this.nl == null || this.nj == null) {
            return;
        }
        var c = a;
        if (this.nk != null) {
            c = this.nk.item(a);
        }
        var d = this.nl.item(a);
        var e = this.nj.item(a);
        var f = this.n6._inner[a];
        f.b.xValue = c;
        f.b.yValue = d;
        f.b.text = e != null ? e.toString() : "";
        if (this.ni != null) {
            f.b.key = this.ni.item(a);
        }
        if (this.nh != null) {
            f.b.content = this.nh.item(a);
        }
        else {
            f.b.content = this.bz.item(a);
        }
        if (this.calloutSeriesSelecting != null) {
            var g = ((function () {
                var $ret = new CalloutSeriesSelectingEventArgs();
                $ret.c = c;
                $ret.d = d;
                $ret.item = _this.bz.item(a);
                return $ret;
            })());
            this.calloutSeriesSelecting(this, g);
            if (g.a != null) {
                f.b.series = g.a;
            }
            else if (g.e != null) {
                if (b.containsKey(g.e)) {
                    f.b.series = b.item(g.e);
                }
            }
        }
        this.p0(f);
    };
    CalloutLayer.prototype.p0 = function (a) {
        var b = new Thickness(0, 3);
        if (isNaN_(this.qx.c) || isNaN_(this.qx.e) || isNaN_(this.qx.d) || isNaN_(this.qx.b)) {
            b = new Thickness(0, 3);
        }
        else {
            b = this.qx;
        }
        var c = this.qq != null ? this.qq : this.m6.ce(a.b.series);
        var d = this.qt != null ? this.qt : this.m6.ch();
        var e = this.qs != null ? this.qs : this.m6.cg();
        var f = this.qr != null ? this.qr : this.m6.cf(a.b.series);
        var g = isNaN_(this.oe) ? 0 : this.oe;
        if (this.ny && this.calloutStyleUpdating != null) {
            var h = new CalloutStyleUpdatingEventArgs();
            h.item = a.b.content;
            h.xValue = a.b.xValue;
            h.yValue = a.b.yValue;
            h.series = a.b.series;
            h.background = c;
            h.textColor = d;
            h.outline = e;
            h.leaderBrush = f;
            h.strokeThickness = g;
            this.calloutStyleUpdating(this, h);
            c = h.background;
            d = h.textColor;
            e = h.outline;
            f = h.leaderBrush;
            g = h.strokeThickness;
        }
        a.b.backgroundPadding = b;
        a.b.background = c;
        a.b.textColor = d;
        a.b.outline = e;
        a.b.leaderBrush = f;
        a.b.strokeThickness = g;
    };
    CalloutLayer.prototype.fs = function (a, b) {
        var c = this.gu(b);
        if (c == CalloutLayer.$$p[19]) {
            return this.ft(a, b, this.pb);
        }
        if (c == CalloutLayer.$$p[23]) {
            return this.ft(a, b, this.pi);
        }
        if (c == CalloutLayer.$$p[24]) {
            return this.ft(a, b, this.pl);
        }
        if (c == CalloutLayer.$$p[18]) {
            return this.ft(a, b, this.o8);
        }
        if (c == CalloutLayer.$$p[14]) {
            return this.ft(a, b, this.o2);
        }
        return _super.prototype.fs.call(this, a, b);
    };
    Object.defineProperty(CalloutLayer.prototype, "nj", {
        get: function () {
            return this.ne;
        },
        set: function (a) {
            if (this.ne != a) {
                var b = this.ne;
                this.ne = a;
                this.raisePropertyChanged("LabelColumn", b, this.ne);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutLayer.prototype, "nk", {
        get: function () {
            return this.nf;
        },
        set: function (a) {
            if (this.nf != a) {
                var b = this.nf;
                this.nf = a;
                this.raisePropertyChanged("XColumn", b, this.nf);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutLayer.prototype, "nl", {
        get: function () {
            return this.ng;
        },
        set: function (a) {
            if (this.ng != a) {
                var b = this.ng;
                this.ng = a;
                this.raisePropertyChanged("YColumn", b, this.ng);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutLayer.prototype, "ni", {
        get: function () {
            return this.nd;
        },
        set: function (a) {
            if (this.nd != a) {
                var b = this.nd;
                this.nd = a;
                this.raisePropertyChanged("KeyColumn", b, this.nd);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutLayer.prototype, "nh", {
        get: function () {
            return this.nc;
        },
        set: function (a) {
            if (this.nc != a) {
                var b = this.nc;
                this.nc = a;
                this.raisePropertyChanged("ContentColumn", b, this.nc);
            }
        },
        enumerable: true,
        configurable: true
    });
    CalloutLayer.prototype.pp = function (a, b) {
        if (b.propertyName == CalloutAnnotation.$$p[13] || b.propertyName == CalloutAnnotation.$$p[14]) {
            var c = a.k;
            if (!this.n2.contains(c)) {
                this.n5.add(c);
                this.jc(false);
            }
        }
    };
    CalloutLayer.prototype.bf = function () {
        return new CalloutLayerView(this);
    };
    CalloutLayer.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.m6 = a;
    };
    Object.defineProperty(CalloutLayer.prototype, "qx", {
        get: function () {
            return new Thickness(1, this.oa, this.oc, this.ob, this.n9);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutLayer.prototype, "nb", {
        get: function () {
            if (this.na == null) {
                this.na = FontUtil.toFontInfo(this.e4, this.pe);
            }
            return this.na;
        },
        enumerable: true,
        configurable: true
    });
    CalloutLayer.prototype.mk = function (a, b) {
        if (this.m9 == null || this.m9 == a) {
            if (!this.aq.c) {
                this.jc(b);
            }
        }
    };
    CalloutLayer.prototype.mm = function (a, b) {
        var e_1, _a;
        _super.prototype.mm.call(this, a, b);
        var c = a;
        c.p.clear();
        c.q.clear();
        c.m.clear();
        c.f.clear();
        c.h.clear();
        c.g.clear();
        c.j.clear();
        c.l.clear();
        c.k.clear();
        c.i.clear();
        c.n.clear();
        c.o.clear();
        c.t.clear();
        c.u.clear();
        c.r.clear();
        c.s.clear();
        var d = this.n7();
        var e = new List$1(CalloutAnnotationWrapper.$, 0);
        if (this.nw) {
            if (this.l7(this.m9)) {
                this.pw(this.m9, c, e);
            }
            else {
                try {
                    for (var _b = tslib_1.__values(fromEnum(this.bs.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (this.l7(f)) {
                            this.pw(f, c, e);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        }
        this.pv(c, d, e);
    };
    CalloutLayer.prototype.js = function (a, b) {
        _super.prototype.js.call(this, a, b);
        var c = b.width - this.qv.width;
        var d = b.height - this.qv.height;
        var e = 0;
        if (Math.abs(c) > Math.abs(d)) {
            e = c;
            this.n8 = c * this.lx.width;
        }
        else {
            e = d;
            this.n8 = d * this.lx.height;
        }
        if (e > 0 && this.n8 > 5) {
            this.nv = true;
        }
        if (e < 0 && this.n8 > 5) {
            this.nu = true;
        }
    };
    CalloutLayer.prototype.jr = function (a, b) {
        _super.prototype.jr.call(this, a, b);
        if (!a.equals1(b)) {
            this.nt = true;
        }
    };
    Object.defineProperty(CalloutLayer.prototype, "qw", {
        get: function () {
            return this.qu;
        },
        set: function (a) {
            this.qu = a;
        },
        enumerable: true,
        configurable: true
    });
    CalloutLayer.prototype.pv = function (a, b, c) {
        var e_2, _a;
        if (this.bs != null) {
            this.qv = this.bs.mq;
        }
        if (this.lx.isEmpty) {
            return;
        }
        this.qw = new Rect(0, this.lx.left, this.lx.top, this.lx.width, this.lx.height);
        if (this.nm != null && this.oz != null && this.oz != this.nm.ak) {
            this.pr();
        }
        if (this.nm == null) {
            var d = this.nn();
            if (d != null) {
                this.nm = d;
                var e = this.nm;
                e.labelsMoved = delegateCombine(e.labelsMoved, runOn(this, this.po));
            }
            else {
                var f = this.lp(this.bi);
                var g = 0;
                if (this.m9 != null) {
                    if (this.m9.isVertical) {
                        g = 2;
                    }
                }
                else {
                    var h = false;
                    if (this.bs != null && this.bs.series.count > 0) {
                        h = true;
                        try {
                            for (var _b = tslib_1.__values(fromEnum(this.bs.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                                var i = _c.value;
                                if (!i.cj) {
                                    if (!i.isVertical) {
                                        h = false;
                                    }
                                }
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                    }
                    if (h) {
                        g = 2;
                    }
                }
                this.nm = new LabelCollisionManager(this.lx.right, this.lx.bottom, this.lx, this.m6.b4(), this.od, this.no(), g, this.ok(), this.oj(), this.nx);
                if (this.oz != null) {
                    this.nm.ak = this.oz;
                    this.nm.ai = this;
                }
                var j = this.nm;
                j.labelsMoved = delegateCombine(j.labelsMoved, runOn(this, this.po));
            }
        }
        if (this.nu) {
            this.nu = false;
            if (this.nm.ag > 2) {
                this.ns = true;
            }
        }
        if (this.nv) {
            this.nv = false;
            if (this.n8 > 5) {
                this.ns = true;
            }
        }
        if (this.nt) {
            this.nt = false;
            this.nm.ax(this.lx.right, this.lx.bottom, this.lx);
            this.ns = true;
        }
        var k = new HashSet$1(CalloutAnnotationWrapper.$, 0);
        var l = new HashSet$1(CalloutAnnotationWrapper.$, 0);
        var m = new List$1(CalloutAnnotationWrapper.$, 0);
        var n = new List$1(CalloutAnnotationWrapper.$, 0);
        var o = new HashSet$1(CalloutAnnotationWrapper.$, 0);
        for (var p = 0; p < b.count; p++) {
            o.add_1(b._inner[p]);
        }
        for (var q = 0; q < c.count; q++) {
            o.add_1(c._inner[q]);
        }
        var r = this.n3.c;
        while (r != null) {
            var s = r.a;
            if (!o.contains(r.c)) {
                n.add(r.c);
                l.add_1(r.c);
                var t = r.c.b;
                t.propertyChanged = delegateRemove(t.propertyChanged, runOn(this, this.pp));
                this.n3.h(r);
                this.n0.removeItem(r.c);
                this.nm.aw(r.c.c);
            }
            r = s;
        }
        for (var u = 0; u < b.count; u++) {
            if (!this.n0.containsKey(b._inner[u])) {
                this.n3.f(b._inner[u]);
                var v = b._inner[u].b;
                v.propertyChanged = delegateCombine(v.propertyChanged, runOn(this, this.pp));
                this.n0.addItem(b._inner[u], this.n3.d);
                m.add(b._inner[u]);
                k.add_1(b._inner[u]);
                this.ns = true;
            }
        }
        for (var w = 0; w < c.count; w++) {
            if (!this.n0.containsKey(c._inner[w])) {
                this.n3.f(c._inner[w]);
                this.n0.addItem(c._inner[w], this.n3.d);
                var x = c._inner[w].b;
                x.propertyChanged = delegateCombine(x.propertyChanged, runOn(this, this.pp));
                m.add(c._inner[w]);
                k.add_1(c._inner[w]);
                this.ns = true;
            }
        }
        if (this.n5.count > 0) {
            this.ns = true;
            this.n5.clear();
            this.n2.clear();
        }
        r = this.n3.c;
        while (r != null) {
            var y = this.np(r.c);
            if (y == null) {
                return;
            }
            if (!Base.equalsStatic(y, r.c.c.c)) {
                var z = false;
                if (r.c.c.c == null) {
                    z = true;
                }
                var aa = r.c.c;
                aa.o = y.b;
                aa.p = y.c;
                if (!isNaN_(aa.o) && !isNaN_(aa.p)) {
                    aa.n = y.e;
                    aa.m = y.d;
                    if (aa.b == null) {
                        this.nm.p(aa);
                    }
                    if (z) {
                        this.nm.al(aa);
                    }
                    else {
                        this.nm.at(aa);
                    }
                }
            }
            r = r.a;
        }
        var ab = this.nm.ah();
        var ac = this.nm.q();
        if (this.ns && (ab != 0 || ac)) {
            this.ns = false;
            this.og = 0x7FFFFFFF;
            if (this.oz != null) {
                var ad = true;
                for (var ae = 0; ae < this.bs.series.count; ae++) {
                    var af = this.bs.series._inner[ae];
                    if (af.cj && typeCast(CalloutLayer.$, af) !== null) {
                        var ag = af;
                        if (ag.oz == this.oz) {
                            if (!ag.qw.equals1(this.nm.az)) {
                                ad = false;
                                break;
                            }
                        }
                    }
                }
                if (ad) {
                    this.nm.ay();
                }
                else {
                    this.nm.am();
                }
            }
            else {
                this.nm.ay();
            }
        }
        var ah = this.n3.c;
        while (ah != null) {
            var ai = ah.c;
            a.m.add1(ai);
            a.f.add(ai.b.background);
            a.h.add(ai.b.outline);
            a.g.add(ai.b.leaderBrush);
            a.j.add(ai.b.backgroundPadding.c);
            a.l.add(ai.b.backgroundPadding.e);
            a.k.add(ai.b.backgroundPadding.d);
            a.i.add(ai.b.backgroundPadding.b);
            a.n.add(ai.b.strokeThickness);
            a.o.add(ai.b.textColor);
            a.p.add1(ai.b.text);
            a.q.add1(ai.b.content);
            var aj = ai.c;
            a.t.add(ai.c.o - (ai.c.g - ai.c.l));
            a.u.add(ai.c.p - (ai.c.h - ai.c.k));
            a.r.add(ai.c.g - ai.c.l);
            a.s.add(ai.c.h - ai.c.k);
            ah = ah.a;
        }
    };
    CalloutLayer.prototype.pr = function () {
        if (this.nm != null) {
            var a = this.n3.c;
            while (a != null) {
                var b = a.c.c;
                if (b != null) {
                    this.nm.aw(b);
                }
                a = a.a;
            }
            var c = this.nm;
            c.labelsMoved = delegateRemove(c.labelsMoved, runOn(this, this.po));
            this.nm = null;
        }
        this.pq();
    };
    CalloutLayer.prototype.nn = function () {
        if (this.bs == null) {
            return null;
        }
        for (var a = 0; a < this.bs.series.count; a++) {
            var b = this.bs.series._inner[a];
            if (b.cj && typeCast(CalloutLayer.$, b) !== null) {
                var c = b;
                if (c != this) {
                    if (c.oz != null && c.oz == this.oz) {
                        if (c.nm != null) {
                            return c.nm;
                        }
                    }
                }
            }
        }
        return null;
    };
    CalloutLayer.prototype.oj = function () {
        return 20000;
    };
    CalloutLayer.prototype.ok = function () {
        return 300;
    };
    CalloutLayer.prototype.no = function () {
        switch (this.m3) {
            case 0: return 2;
            case 2: return 1;
            case 3: return 2;
            case 1: return 0;
        }
        return 2;
    };
    CalloutLayer.prototype.po = function (a) {
        var b = this.nm.ah();
        if (b < this.og) {
            this.og = b;
            this.jc(false);
        }
    };
    CalloutLayer.prototype.np = function (a) {
        var b = this.m9;
        if (a.b.series != null) {
            b = a.b.series;
        }
        if (b == null) {
            for (var c = 0; c < this.bs.series.count; c++) {
                var d = this.bs.series._inner[c];
                if (!d.cj) {
                    b = d;
                    break;
                }
            }
        }
        if (b == null) {
            return null;
        }
        var e = b.ag();
        var f = b.eh();
        var g = null;
        var h = null;
        for (var i = 0; i < e.length; i++) {
            if (e[i].ce) {
                h = e[i];
            }
            else {
                g = e[i];
            }
        }
        if (g == null || h == null) {
            return null;
        }
        if (!g.cd() || !h.cd()) {
            return null;
        }
        var j = 0;
        var k = 0;
        if (g.b2 && g.categoryMode != 0) {
            if (b.cr) {
                j = b.getOffsetValue();
            }
            else if (b.cl) {
                j = b.getOffsetValue();
            }
        }
        if (h.b2 && h.categoryMode != 0) {
            if (b.cl) {
                k = b.getOffsetValue();
            }
        }
        var l = new ScalerParams(0, this.bs.mq, g.i0, g.b8, g.iv());
        l.b = f;
        var m = new ScalerParams(0, this.bs.mq, h.i0, h.b8, h.iv());
        m.b = f;
        var n = g.dv(this.of(a.b.xValue), l) + j;
        var o = h.dv(this.of(a.b.yValue), m) + k;
        var p = new QuadTreeBoundingBox();
        p.b = n;
        p.c = o;
        p.e = 3;
        p.d = 3;
        p.g = a.b;
        return p;
    };
    CalloutLayer.prototype.of = function (a) {
        if (typeCast(Date_$type, a) !== null) {
            return a.getTime();
        }
        if (a == null) {
            return 0;
        }
        return a;
    };
    CalloutLayer.prototype.n7 = function () {
        var a = new List$1(CalloutAnnotationWrapper.$, 0);
        for (var b = 0; b < this.n6.count; b++) {
            if (stringIsNullOrEmpty(this.n6._inner[b].b.text)) {
                continue;
            }
            if (this.n6._inner[b].c == null) {
                var c = new LabelCollisionInfo();
                this.n6._inner[b].c = c;
                var d = this.m6.cj(this.ol(this.n6._inner[b]), this.n6._inner[b].b.backgroundPadding);
                c.k = d.height / 2;
                c.l = d.width / 2;
            }
            var e = this.np(this.n6._inner[b]);
            if (e == null || e.b < this.lx.left || e.b > this.lx.right || e.c < this.lx.top || e.c > this.lx.bottom) {
                continue;
            }
            a.add(this.n6._inner[b]);
        }
        return a;
    };
    CalloutLayer.prototype.ol = function (a) {
        return a.b.text;
    };
    CalloutLayer.prototype.l7 = function (a) {
        if (!_super.prototype.l7.call(this, a)) {
            return false;
        }
        return true;
    };
    CalloutLayer.prototype.im = function () {
        _super.prototype.im.call(this);
        if (!this.aq.c && !this.ca()) {
            return;
        }
        if (this.nq) {
            this.nq = false;
            this.pt();
        }
        if (this.ca()) {
            this.py(this.l2, false);
        }
        else {
            this.py(this.l0, false);
        }
    };
    CalloutLayer.prototype.pu = function () {
        if (!this.nr) {
            this.nr = true;
            this.jc(true);
        }
    };
    CalloutLayer.prototype.pw = function (a, b, c) {
        if (a.ch(this.n4, this.nz, this.oh)) {
            var d = this.n4;
            var _loop_1 = function (e) {
                var f = d._inner[e];
                var g = f.a.item != null ? f.a.item : f.a;
                var h = a;
                var i = new CalloutKey(g, h);
                var j = new CalloutAnnotationWrapper();
                var k = new CalloutAnnotation();
                j.b = k;
                k.key = i;
                var l = false;
                var m = false;
                if (this_1.nr) {
                    this_1.nr = false;
                    m = true;
                }
                if (this_1.n0.containsKey(j)) {
                    j = this_1.n0.item(j).c;
                    k = j.b;
                }
                else {
                    if (!m) {
                        l = true;
                    }
                }
                var n = k.xValue;
                var o = k.yValue;
                if (this_1.of(n) != this_1.of(f.b)) {
                    k.xValue = f.b;
                }
                if (this_1.of(o) != this_1.of(f.c)) {
                    k.yValue = f.c;
                }
                k.series = a;
                if (l && this_1.calloutLabelUpdating != null) {
                    var p = ((function () {
                        var $ret = new CalloutLabelUpdatingEventArgs();
                        $ret.c = k.xValue;
                        $ret.d = k.yValue;
                        $ret.item = g;
                        return $ret;
                    })());
                    this_1.calloutLabelUpdating(this_1, p);
                    if (p.b != null && !Base.equalsStatic(p.b, j.b.text)) {
                        var q = p.b;
                        j.b.text = q != null ? q.toString() : "";
                    }
                }
                else {
                    if (f.d != null) {
                        k.text = f.d;
                    }
                }
                if (k.text == null) {
                    return "continue";
                }
                if (l) {
                    if (this_1.calloutContentUpdating != null) {
                        var r = ((function () {
                            var $ret = new CalloutContentUpdatingEventArgs();
                            $ret.c = k.xValue;
                            $ret.d = k.yValue;
                            $ret.item = g;
                            $ret.a = g;
                            return $ret;
                        })());
                        this_1.calloutContentUpdating(this_1, r);
                        if (r.a != null && r.a != j.b.content) {
                            j.b.content = r.a;
                        }
                    }
                    else {
                        if (g != null && g != j.b.content) {
                            j.b.content = g;
                        }
                    }
                }
                if (l) {
                    this_1.p0(j);
                }
                if (j.c == null) {
                    var s = new LabelCollisionInfo();
                    j.c = s;
                    var t = this_1.m6.cj(this_1.ol(j), j.b.backgroundPadding);
                    s.k = t.height / 2;
                    s.l = t.width / 2;
                }
                var u = this_1.np(j);
                if (u == null || u.b < this_1.lx.left || u.b > this_1.lx.right || u.c < this_1.lx.top || u.c > this_1.lx.bottom) {
                    return "continue";
                }
                c.add(j);
            };
            var this_1 = this;
            for (var e = 0; e < this.n4.count; e++) {
                _loop_1(e);
            }
        }
    };
    CalloutLayer.prototype.bd = function () {
        return 1;
    };
    CalloutLayer.prototype.ar = function (a, b) {
        var c = this.nm.d(a, b);
        if (c == null) {
            return null;
        }
        return c.u;
    };
    CalloutLayer.prototype.dh = function (a, b) {
        if (this.nm == null) {
            return false;
        }
        return this.nm.d(a, b) != null;
    };
    CalloutLayer.prototype.py = function (a, b) {
        var e_3, _a;
        var c = 0;
        var d = new HashSet$1(CalloutAnnotationWrapper.$, 0);
        for (c = 0; c < a.p.count; c++) {
            var e = a.m._inner[c];
            if (isNaN_(a.t._inner[c]) || isNaN_(a.u._inner[c])) {
                continue;
            }
            var f = this.m6.b6.item(e);
            d.add_1(e);
            if (f.f != a.p._inner[c]) {
                f.f = a.p._inner[c];
                var g = f.s();
            }
            f.c = a.n._inner[c];
            f.k = a.f._inner[c];
            f.n = a.o._inner[c];
            f.m = f.c > 0 ? a.h._inner[c] : this.bi.bn();
            f.l = a.g._inner[c];
            f.v = new Thickness(1, a.j._inner[c], a.l._inner[c], a.k._inner[c], a.i._inner[c]);
            this.m6.cc(f, Math.min(a.r._inner[c], a.r._inner[c] + a.t._inner[c]), Math.min(a.s._inner[c], a.s._inner[c] + a.u._inner[c]));
            this.m6.cd(f, e, a.q._inner[c]);
            f.p = { $type: Point_$type, x: a.t._inner[c], y: a.u._inner[c] };
        }
        var h = new List$1(CalloutAnnotationWrapper.$, 0);
        try {
            for (var _b = tslib_1.__values(fromEnum(this.m6.b6.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var i = _c.value;
                if (!d.contains(i)) {
                    h.add(i);
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        for (var j = 0; j < h.count; j++) {
            this.m6.b6.g(h._inner[j]);
        }
    };
    CalloutLayer.prototype.mp = function (a, b) {
        _super.prototype.mp.call(this, a, b);
        if (this.ca()) {
            this.py(this.l2, true);
        }
        else {
            this.py(this.l0, false);
        }
    };
    CalloutLayer.prototype.c8 = function () {
        return true;
    };
    CalloutLayer.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case CalloutLayer.$$p[21]:
                this.na = null;
                this.jc(true);
                break;
            case CalloutLayer.$$p[20]:
                this.jc(true);
                break;
            case CalloutLayer.$$p[13]:
                this.jc(true);
                break;
            case "SeriesViewer":
                if (c != null) {
                    var e = c;
                    e.seriesAnimating = delegateRemove(e.seriesAnimating, runOn(this, this.px));
                }
                if (d != null) {
                    var f = d;
                    f.seriesAnimating = delegateCombine(f.seriesAnimating, runOn(this, this.px));
                }
                if (d == null) {
                    this.pr();
                }
                break;
            case CalloutLayer.$$p[15]:
            case CalloutLayer.$$p[16]:
            case CalloutLayer.$$p[1]:
            case CalloutLayer.$$p[5]:
            case CalloutLayer.$$p[4]:
            case CalloutLayer.$$p[7]:
            case CalloutLayer.$$p[9]:
            case CalloutLayer.$$p[8]:
            case CalloutLayer.$$p[6]:
            case CalloutLayer.$$p[11]:
            case CalloutLayer.$$p[12]:
                this.ps();
                break;
            case CalloutLayer.$$p[19]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.nj);
                    this.nj = this.by(this.pb);
                }
                break;
            case "LabelColumn":
                this.ps();
                break;
            case CalloutLayer.$$p[14]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.nh);
                    this.nh = this.by(this.o2);
                }
                break;
            case "ContentColumn":
                this.ps();
                break;
            case CalloutLayer.$$p[23]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.nk);
                    this.nk = stringIsNullOrEmpty(this.pi) == false ? this.by(this.pi) : null;
                }
                break;
            case "XColumn":
                this.ps();
                break;
            case CalloutLayer.$$p[24]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.nl);
                    this.nl = this.by(this.pl);
                }
                break;
            case "YColumn":
                this.ps();
                break;
            case CalloutLayer.$$p[18]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.ni);
                    this.ni = this.by(this.o8);
                }
                break;
            case "KeyColumn":
                this.ps();
                break;
            case "FastItemsSource":
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    c.deregisterColumn(this.nj);
                    c.deregisterColumn(this.nh);
                    c.deregisterColumn(this.nk);
                    c.deregisterColumn(this.nl);
                    c.deregisterColumn(this.ni);
                    this.nj = null;
                    this.nh = null;
                    this.nk = null;
                    this.nl = null;
                    this.ni = null;
                    this.n6.clear();
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    this.nj = this.by(this.pb);
                    this.nh = this.by(this.o2);
                    if (stringIsNullOrEmpty(this.pi) == false) {
                        this.nk = this.by(this.pi);
                    }
                    this.nl = this.by(this.pl);
                    this.ni = this.by(this.o8);
                }
                this.ps();
                break;
        }
    };
    CalloutLayer.prototype.px = function (a) {
        if (a.cj) {
            return;
        }
        if ((a == this.m9 || this.m9 == null) && this.nw) {
            this.jc(false);
        }
    };
    CalloutLayer.prototype.ps = function () {
        if (this.nq == false) {
            this.nq = true;
            this.jc(false);
        }
    };
    CalloutLayer.prototype.pq = function () {
        this.m6.b6.clear();
        this.n3.g();
        this.n0.clear();
    };
    CalloutLayer.prototype.pt = function () {
        if (this.nl == null || this.nj == null) {
            var a = this.n3.c;
            while (a != null) {
                this.p0(a.c);
                a = a.a;
            }
            return;
        }
        var b = this.n1();
        for (var c = 0; c < this.nl.count; c++) {
            var d = void 0;
            if (this.n6.count <= c) {
                var e = new CalloutAnnotation();
                d = new CalloutAnnotationWrapper();
                d.b = e;
                d.b.series = this.m9;
                this.n6.add(d);
            }
            else {
                d = this.n6._inner[c];
            }
            this.pz(c, b);
        }
    };
    CalloutLayer.$t = markType(CalloutLayer, 'CalloutLayer', AnnotationLayer.$);
    return CalloutLayer;
}(AnnotationLayer));
export { CalloutLayer };
Type.dep(DependencyProperty, PropertyMetadata, CalloutLayer, 'raisePropertyChanged', ['AutoCalloutLabelPrecision:oh:p1', [1, 5], 'CalloutBackground:qq:p2', [Brush.$, null], 'CalloutCollisionMode:m3:p3', [CalloutCollisionMode_$type, enumGetBox(CalloutCollisionMode_$type, 3)], 'CalloutInterpolatedValuePrecision:oi:p4', [1, -1], 'CalloutLeaderBrush:qr:p5', [Brush.$, null], 'CalloutOutline:qs:p6', [Brush.$, null], 'CalloutPaddingBottom:n9:p7', [1, NaN], 'CalloutPaddingLeft:oa:p8', [1, NaN], 'CalloutPaddingRight:ob:p9', [1, NaN], 'CalloutPaddingTop:oc:qa', [1, NaN], 'CalloutPositionPadding:od:qb', [1, 10], 'CalloutStrokeThickness:oe:qc', [1, NaN], 'CalloutTextColor:qt:qd', [Brush.$, null], 'CollisionChannel:oz:qe', [2, null], 'ContentMemberPath:o2:qf', [2, null], 'IsAutoCalloutBehaviorEnabled:nw:qg', [0, false], 'IsCalloutOffsettingEnabled:nx:qh', [0, true], 'IsCustomCalloutStyleEnabled:ny:qi', [0, false], 'KeyMemberPath:o8:qj', [2, null], 'LabelMemberPath:pb:qk', [2, null], 'TargetSeries:m9:ql', [Series.$, null], 'TextStyle:pe:qm', [2, null], 'UseValueForAutoCalloutLabels:nz:qn', [0, true], 'XMemberPath:pi:qo', [2, null], 'YMemberPath:pl:qp', [2, null]]);
/**
 * @hidden
 */
var CalloutLayerView = /** @class */ (function (_super) {
    tslib_1.__extends(CalloutLayerView, _super);
    function CalloutLayerView(a) {
        var _this = _super.call(this, a) || this;
        _this.b1 = null;
        _this.b7 = null;
        _this.b6 = null;
        _this.b8 = NaN;
        _this.b2 = null;
        _this.b1 = a;
        _this.b7 = new List$1(CalloutView.$, 0);
        return _this;
    }
    CalloutLayerView.prototype.ap = function () {
        var _this = this;
        _super.prototype.ap.call(this);
        this.b6 = ((function () {
            var $ret = new HashPool$2(CalloutAnnotationWrapper.$, CalloutView.$);
            $ret.create = runOn(_this, _this.b3);
            $ret.activate = runOn(_this, _this.b9);
            $ret.disactivate = runOn(_this, _this.cb);
            $ret.destroy = runOn(_this, _this.ca);
            return $ret;
        })());
    };
    CalloutLayerView.prototype.b3 = function () {
        var a = new CalloutView(this);
        this.b7.add(a);
        return a;
    };
    CalloutLayerView.prototype.b9 = function (a) {
        a.w = 0;
    };
    CalloutLayerView.prototype.cb = function (a) {
        a.w = 1;
    };
    CalloutLayerView.prototype.ca = function (a) {
        this.b7.remove(a);
    };
    CalloutLayerView.prototype.ci = function (a) {
        if (a == null) {
            return a;
        }
        return BrushUtil.h(a, 0.1);
    };
    CalloutLayerView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        a.ac(this.b5());
        if (a.d && !b) {
            for (var c = 0; c < this.b7.count; c++) {
                var d = this.b7._inner[c];
                if (d.w == 0) {
                    d.h(a);
                }
            }
        }
    };
    CalloutLayerView.prototype.b5 = function () {
        if (this.b1.nb != null) {
            return this.b1.nb;
        }
        return this.e.bs.getFontInfo();
    };
    CalloutLayerView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        for (var b = 0; b < this.b7.count; b++) {
            var c = this.b7._inner[b];
            if (c.w != 0) {
                continue;
            }
            c.g(this.b1, this.b5(), a);
        }
    };
    CalloutLayerView.prototype.ck = function (a) {
        if (isNaN_(this.b8)) {
            this.b8 = this.e.bs.bn.a7;
        }
        return new Size(1, this.h.f(a.ak), this.b8);
    };
    CalloutLayerView.prototype.cj = function (a, b) {
        if (this.b2 == null) {
            this.b2 = new CalloutView(this);
        }
        this.b2.f = a;
        this.b2.v = b;
        var c = this.b2.s();
        this.b2.f = null;
        return c;
    };
    CalloutLayerView.prototype.b4 = function () {
        return new DOMExecutionContext(this.e.bs.bn.i.af);
    };
    CalloutLayerView.prototype.cc = function (a, b, c) {
        a.d = b;
        a.e = c;
    };
    CalloutLayerView.prototype.cd = function (a, b, c) {
        a.i(this.e, b, c);
    };
    CalloutLayerView.prototype.ce = function (a) {
        if (a != null) {
            return this.ci(a.k2);
        }
        return this.e.k2;
    };
    CalloutLayerView.prototype.ch = function () {
        return ((function () {
            var $ret = new Brush();
            $ret.color = Color.u(255, 255, 255, 255);
            return $ret;
        })());
    };
    CalloutLayerView.prototype.cg = function () {
        return ((function () {
            var $ret = new Brush();
            $ret.color = Color.u(255, 255, 255, 255);
            return $ret;
        })());
    };
    CalloutLayerView.prototype.cf = function (a) {
        if (a != null) {
            return this.ci(a.k2);
        }
        return this.e.k2;
    };
    CalloutLayerView.$t = markType(CalloutLayerView, 'CalloutLayerView', AnnotationLayerView.$);
    return CalloutLayerView;
}(AnnotationLayerView));
export { CalloutLayerView };
/**
 * @hidden
 */
var CalloutView = /** @class */ (function (_super) {
    tslib_1.__extends(CalloutView, _super);
    function CalloutView(a) {
        var _this = _super.call(this) || this;
        _this.a = null;
        _this.w = 0;
        _this.j = null;
        _this.r = null;
        _this.o = { $type: Point_$type, x: 0, y: 0 };
        _this.q = null;
        _this.u = new Thickness(0, 0);
        _this.t = null;
        _this.d = 0;
        _this.e = 0;
        _this.b = null;
        _this.q = new Line();
        _this.q.ab = false;
        _this.q._stroke = ((function () {
            var $ret = new Brush();
            $ret.color = Color.u(255, 0, 0, 0);
            return $ret;
        })());
        _this.q.ac = 2;
        _this.j = new TextBlock();
        _this.r = new Rectangle();
        _this.r.ac = 0;
        _this.a = a;
        _this.w = 0;
        return _this;
    }
    CalloutView.prototype.h = function (a) {
        a.z();
        a.af(this.d, this.e);
        var b = new TranslateTransform();
        b.j = this.t.c;
        b.k = this.t.e;
        this.r.j = b;
        a.r(this.q);
        a.z();
        this.r.width = this.j.m + this.u.c + this.u.d;
        this.r.height = this.j.l + this.u.e + this.u.b;
        a.j(this.r.j);
        a.v(this.r);
        this.j.o = this.u.e;
        this.j.n = this.u.c;
        a.w(this.j);
        a.y();
        a.y();
    };
    Object.defineProperty(CalloutView.prototype, "p", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            if (this.o.x != a.x || this.o.y != a.y) {
                var b = this.s();
                var c = b.height;
                var d = b.width;
                this.o = a;
                var e = CalloutViewHelper.b(this.p, d, c, this.q);
                this.t = e;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutView.prototype, "n", {
        get: function () {
            return this.j.am;
        },
        set: function (a) {
            this.j.am = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutView.prototype, "m", {
        get: function () {
            return this.r._stroke;
        },
        set: function (a) {
            this.r._stroke = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutView.prototype, "l", {
        get: function () {
            return this.q._stroke;
        },
        set: function (a) {
            this.q._stroke = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutView.prototype, "k", {
        get: function () {
            return this.r._fill;
        },
        set: function (a) {
            this.r._fill = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutView.prototype, "c", {
        get: function () {
            return this.r.ac;
        },
        set: function (a) {
            this.r.ac = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutView.prototype, "v", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutView.prototype, "f", {
        get: function () {
            return this.j.ak;
        },
        set: function (a) {
            if (a == null) {
                this.j.ak = "";
            }
            else {
                this.j.ak = a.toString();
            }
            var b = this.s();
            var c = b.height;
            var d = b.width;
            var e = CalloutViewHelper.b(this.p, d, c, this.q);
            this.t = e;
        },
        enumerable: true,
        configurable: true
    });
    CalloutView.prototype.s = function () {
        var a = this.a.ck(this.j);
        this.j.m = a.width;
        this.j.l = a.height;
        return new Size(1, a.width + this.u.c + this.u.d, a.height + this.u.e + this.u.b);
    };
    CalloutView.prototype.i = function (a, b, c) {
        if (this.b == null) {
            this.b = new DataContext();
        }
        this.b.series = a;
        this.b.item = c;
        if (b.c != null) {
            b.c.u = this.b;
        }
    };
    CalloutView.prototype.g = function (a, b, c) {
        a.a2.exportCalloutData(this.d, this.e, this.t.c, this.t.e, c, this.r, this.j, b, this.q, "callout", ["Main", "Callout"]);
    };
    CalloutView.$t = markType(CalloutView, 'CalloutView');
    return CalloutView;
}(Base));
export { CalloutView };
