// igniteui-react-charts/ES5/CategoryBucketCalculator_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, typeCast, Type, markType, Array_$type, Number_$type, Point_$type, enumGetBox, runOn, fromEnum } from "igniteui-react-core/ES5/type";
import { IBucketizer_$type } from "./IBucketizer";
import { ArgumentNullException } from "igniteui-react-core/ES5/ArgumentNullException";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { ISortingAxis_$type } from "./ISortingAxis";
import { ScalerParams } from "./ScalerParams";
import { Series } from "./Series";
import { truncate, intDivide, isNaN_, isInfinity } from "igniteui-react-core/ES5/number";
import { FramePreparer } from "./FramePreparer";
import { IIsCategoryBased_$type } from "./IIsCategoryBased";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { ISupportsMarkers_$type } from "./ISupportsMarkers";
import { IProvidesViewport_$type } from "./IProvidesViewport";
import { ISupportsErrorBars_$type } from "./ISupportsErrorBars";
import { CategoryFrame } from "./CategoryFrame";
import { PreparationParams } from "./PreparationParams";
import { IDateTimeAxis_$type } from "./IDateTimeAxis";
import { NumericAxisBase } from "./NumericAxisBase";
import { IHasCategoryTrendline_$type } from "./IHasCategoryTrendline";
import { IHasSingleValueCategory_$type } from "./IHasSingleValueCategory";
import { DefaultCategoryTrendlineHost } from "./DefaultCategoryTrendlineHost";
import { DefaultSingleValueProvider } from "./DefaultSingleValueProvider";
import { SingleValuesHolder } from "./SingleValuesHolder";
import { MarkerSeries } from "./MarkerSeries";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { ICategorySeriesInteractionManager_$type } from "./ICategorySeriesInteractionManager";
import { DVContainer } from "igniteui-react-core/ES5/DVContainer";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { HighlightingInfo } from "./HighlightingInfo";
import { CategoryTransitionSourceFramePreparer } from "./CategoryTransitionSourceFramePreparer";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { CategorySeriesRenderManager } from "./CategorySeriesRenderManager";
import { Pool$1 } from "igniteui-react-core/ES5/Pool$1";
import { Clipper } from "igniteui-react-core/ES5/Clipper";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { round10N } from "igniteui-react-core/ES5/numberExtended";
/**
 * @hidden
 */
var CategoryBucketCalculator = /** @class */ (function (_super) {
    tslib_1.__extends(CategoryBucketCalculator, _super);
    function CategoryBucketCalculator(a) {
        var _this = _super.call(this) || this;
        _this.c = null;
        _this.e = 0;
        _this.f = 0;
        _this.d = 0;
        if (a == null) {
            throw new ArgumentNullException(0, "view");
        }
        _this.c = a;
        _this.e = -1;
        _this.d = 0;
        return _this;
    }
    CategoryBucketCalculator.prototype.getBucket = function (a) {
        return null;
    };
    CategoryBucketCalculator.prototype.getBucketWithoutUnknowns = function (a) {
        return this.getBucket(a);
    };
    CategoryBucketCalculator.prototype.getErrorBucket = function (a, b) {
        return NaN;
    };
    CategoryBucketCalculator.prototype.g = function (a) {
        var b = this.c.bs;
        var c = this.c.br;
        var d = this.c.cp;
        var e = typeCast(CategoryAxisBase.$, d.fetchXAxis());
        var f = d.bz;
        if (b.isEmpty || c.isEmpty || e == null || f == null || f.count == 0) {
            this.d = 0;
            return;
        }
        var g = d.getEffectiveViewport1(this.c);
        var h = d.lp(this.c);
        var i = typeCast(ISortingAxis_$type, e);
        if (i == null || i.i$b == null) {
            var j = new ScalerParams(0, b, c, e.b8, g);
            var k = Math.floor(e.dz(h.left, j, 0));
            var l = Math.ceil(e.dz(h.right, j, 0));
            if (e.b8) {
                l = Math.ceil(e.dz(h.left, j, 0));
                k = Math.floor(e.dz(h.right, j, 0));
            }
            var m = Math.floor((l - k + 1) * a / h.width);
            this.d = truncate(Math.max(1, m));
            this.e = truncate(Math.max(0, Math.floor(k / this.d) - 1));
            this.f = truncate(Math.ceil(l / this.d));
        }
        else {
            this.e = i.i$e(b, c, g, h);
            this.f = i.i$g(b, c, g, h);
            this.d = 1;
        }
    };
    CategoryBucketCalculator.prototype.getBucketInfo = function (a, b, c, d) {
        a = this.e;
        b = this.f;
        c = this.d;
        d = this.c.cp.eu;
        return {
            p0: a,
            p1: b,
            p2: c,
            p3: d
        };
    };
    CategoryBucketCalculator.prototype.cacheValues = function () {
    };
    CategoryBucketCalculator.prototype.unCacheValues = function () {
    };
    CategoryBucketCalculator.$t = markType(CategoryBucketCalculator, 'CategoryBucketCalculator', Base.$, [IBucketizer_$type]);
    return CategoryBucketCalculator;
}(Base));
export { CategoryBucketCalculator };
/**
 * @hidden
 */
var CategoryFramePreparerBase = /** @class */ (function (_super) {
    tslib_1.__extends(CategoryFramePreparerBase, _super);
    function CategoryFramePreparerBase(a) {
        var _rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            _rest[_i - 1] = arguments[_i];
        }
        var _this = this;
        a = (a == void 0) ? 0 : a;
        switch (a) {
            case 0:
                {
                    var c = _rest[0];
                    var d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                    {
                        var f = d[0];
                        var g = d[1];
                        var h = d[2];
                        var i = d[3];
                        var j = d[4];
                        _this = _super.call(this, 1, g, h, i) || this;
                        _this.g = null;
                        _this.h = null;
                        _this.m = false;
                        _this.o = new List$1(Array_$type, 0);
                        _this.p = new List$1(Number_$type, 0);
                        _this.n = new List$1(Number_$type, 0);
                        _this.h = f;
                        _this.g = j;
                    }
                }
                break;
            case 1:
                {
                    var c = _rest[0];
                    var d = _rest[1];
                    var e = _rest[2];
                    var f = _rest[3];
                    var g = _rest[4];
                    _this = _super.call(this, 1, d, e, f) || this;
                    _this.g = null;
                    _this.h = null;
                    _this.m = false;
                    _this.o = new List$1(Array_$type, 0);
                    _this.p = new List$1(Number_$type, 0);
                    _this.n = new List$1(Number_$type, 0);
                    _this.h = c;
                    _this.g = g;
                }
                break;
        }
        return _this;
    }
    CategoryFramePreparerBase.prototype.i = function (a) {
        var b = new PreparationParams();
        b.f = this.h.useHighMarkerFidelity;
        b.b = this.h.scaler;
        b.c = this.h.yScaler;
        b.d = typeCast(ISortingAxis_$type, b.b);
        b.a = typeCast(CategoryFrame.$, a);
        if (b.a == null || b.b == null || b.c == null) {
            return null;
        }
        var c;
        var d;
        var e;
        var f;
        var g = this.g.getBucketInfo(c, d, e, f);
        c = g.p0;
        d = g.p1;
        e = g.p2;
        f = g.p3;
        b.i = c;
        b.j = d;
        b.h = e;
        b.g = f;
        if (b.j < b.i) {
            return null;
        }
        var h;
        var i;
        var j;
        var k = this.a.getViewInfo(i, h);
        i = k.p0;
        h = k.p1;
        j = this.a.getEffectiveViewport();
        b.m = h;
        b.l = i;
        b.k = j;
        if (Rect.l_op_Equality(b.m, Rect.empty) || Rect.l_op_Equality(b.l, Rect.empty)) {
            return null;
        }
        if (this.h != null && (typeCast(Series.$, this.h) !== null && this.h.cs) && this.g != null) {
            b.e = true;
        }
        return b;
    };
    CategoryFramePreparerBase.prototype.r = function (a, b, c, d) {
        return a.getSeriesOffset(b, c, d, this.h.currentMode2Index, this.h.currentCategoryMode);
    };
    CategoryFramePreparerBase.prototype.s = function (a) {
        return this.r(a.b, a.m, a.l, a.k);
    };
    CategoryFramePreparerBase.prototype.d = function (a, b) {
        var c = this.i(a);
        if (c == null || this.g == null) {
            return;
        }
        c.a.f.clear();
        c.a.k.clear();
        c.a.m.clear();
        c.a.p.clear();
        c.a.i.clear();
        c.a.h.clear();
        c.a.g.clear();
        var d = this.c.shouldDisplayMarkers;
        var e = false;
        var f = this.s(c);
        var g = this.j(c);
        if (c.d != null && c.d.i$b != null && c.d.i$b.count != g.e) {
            return;
        }
        if (c.d != null && c.d.i$a) {
            c.d.i$h();
        }
        this.g.cacheValues();
        if (this.m && c.d != null && c.d.i$b.count > 0) {
            var h = c.d.i$b._inner[c.i];
            this.z(g, h, true, c.e);
            while ((isNaN_(g.c) || isNaN_(g.d)) && c.i > 0) {
                c.i = c.i - 1;
                h = c.d.i$b._inner[c.i];
                this.z(g, h, true, c.e);
            }
            var i = c.d.i$b._inner[c.j];
            this.z(g, i, true, c.e);
            while ((isNaN_(g.c) || isNaN_(g.d)) && c.j < g.e - 1) {
                c.j = c.j + 1;
                i = c.d.i$b._inner[c.j];
                this.z(g, i, true, c.e);
            }
        }
        if (this.m && c.d == null) {
            var j = c.i;
            var k = this.g.getBucket(j);
            while ((isNaN_(k[1]) || isNaN_(k[2])) && c.i > 0) {
                c.i = c.i - 1;
                j = c.i;
                k = this.g.getBucket(j);
            }
            var l = c.j;
            k = this.g.getBucket(l);
            while ((isNaN_(k[1]) || isNaN_(k[2])) && c.j < (intDivide(g.e, c.h))) {
                c.j = c.j + 1;
                l = c.j;
                k = this.g.getBucket(l);
            }
        }
        this.w(c, g, f);
        var m = this.t(c, g, f, d, e);
        this.c.updateMarkerCount(m);
        this.v(typeCast(CategoryFrame.$, a), b);
        this.g.unCacheValues();
        return;
    };
    CategoryFramePreparerBase.prototype.v = function (a, b) {
    };
    Object.defineProperty(CategoryFramePreparerBase.prototype, "q", {
        get: function () {
            return this.o;
        },
        enumerable: true,
        configurable: true
    });
    CategoryFramePreparerBase.prototype.t = function (a, b, c, d, e) {
        var _this = this;
        var f = 0;
        var g = false;
        var h = null;
        var i = false;
        var j = a.d != null;
        var k = a.m;
        var l = a.l;
        var m = typeCast(NumericAxisBase.$, a.c) !== null && a.c.jt;
        var n = a.f;
        var o = this.h.mayContainUnknowns;
        var p = this.o;
        var q = this.n;
        var r = this.p;
        p.clear();
        q.clear();
        r.clear();
        var s = this.g;
        var t = this.h.provideCollisionDetector();
        var u = 0;
        var v = a.k;
        var w = new ScalerParams(0, k, l, a.b.j$a, v);
        var x = ((function () {
            var $ret = new ScalerParams(0, k, l, a.c.j$a, v);
            $ret.b = typeCast(Series.$, _this.h) !== null ? _this.h.eh() : NaN;
            return $ret;
        })());
        if (j) {
            u = a.b.j$c(2, w) - a.b.j$c(1, w);
            i = this.k();
        }
        var y = a.i;
        var z = a.j;
        var aa = a.a.f;
        var ab = typeCast(IDateTimeAxis_$type, a.b);
        if (ab != null) {
            ab.StartTimeCursor();
        }
        var _loop_1 = function (ac) {
            var ad = void 0;
            if (!j) {
                if (!o) {
                    ad = s.getBucketWithoutUnknowns(ac);
                }
                else {
                    ad = s.getBucket(ac);
                }
            }
            else {
                ad = ((function () { var ae = _this.f(a, ac, b, u, g, h, c); ac = ae.p1; g = ae.p4; h = ae.p5; return ae.ret; })());
            }
            var ae = !m || (m && ad[1] > 0);
            var af = ad[0];
            if (!j || !isNaN_(ad[0])) {
                aa.add(ad);
                if (g && i) {
                    if (h != null) {
                        aa.add(h);
                    }
                }
                if (d && ae) {
                    var ag = ac * a.h;
                    var ah = ag;
                    if (j && a.d.i$b != null && ag >= 0 && ag < a.d.i$b.count) {
                        ag = a.d.i$b._inner[ag];
                    }
                    var ai = new Array(ad.length);
                    ai[0] = ad[0];
                    ai[1] = ad[1];
                    ai[2] = ad[2];
                    if (n && ag < b.e) {
                        ai = new Array(ad.length);
                        ai[0] = af;
                        this_1.z(b, ag, true, a.e);
                        ai[1] = b.c;
                        ai[2] = b.d;
                    }
                    p.add(ai);
                    q.add(a.a.f.count - 1);
                    r.add(ag);
                }
            }
            out_ac_1 = ac;
        };
        var this_1 = this, out_ac_1;
        for (var ac = y; ac <= z; ++ac) {
            _loop_1(ac);
            ac = out_ac_1;
        }
        if (ab != null) {
            ab.ResetTimeCursor();
        }
        this.y(a, c, j, w, x, e, d);
        if (d) {
            var aj = p.count;
            var ak = void 0;
            for (var al = 0; al < aj; al++) {
                ak = p._inner[al];
                if (this.l(a.a, ak, t, Math.min(r._inner[al], b.e - 1), f, q._inner[al])) {
                    ++f;
                }
            }
        }
        return f;
    };
    CategoryFramePreparerBase.prototype.k = function () {
        if (typeCast(Series.$, this.h) !== null) {
            return this.h.ck;
        }
        return false;
    };
    CategoryFramePreparerBase.prototype.f = function (a, b, c, d, e, f, g) {
        var h;
        f = null;
        e = false;
        var i = typeCast(CategorySeries.$, this.h);
        var j = a.l;
        var k = a.m;
        var l = a.k;
        var m = new ScalerParams(0, k, j, a.b.j$a, l);
        var n = new ScalerParams(0, k, j, a.c.j$a, l);
        var o = a.e;
        if (a.d == null || a.d.i$b == null || a.d.i$b.count == 0) {
            return {
                ret: [NaN, NaN, NaN],
                p1: b,
                p4: e,
                p5: f
            };
        }
        if (i != null && i.bz != null && i.bz.count < a.d.i$b.count) {
            return {
                ret: [NaN, NaN, NaN],
                p1: b,
                p4: e,
                p5: f
            };
        }
        var p = a.d.i$b._inner[b];
        var q = a.d.i$d(p);
        var r = q;
        this.z(c, p, false, o);
        while (b < a.j) {
            p = a.d.i$b._inner[b + 1];
            r = a.d.i$d(p);
            this.z(c, p, true, o);
            if (r - q >= d || isNaN_(c.c) || isNaN_(c.d)) {
                if (e) {
                    var s = a.d.i$b._inner[b];
                    this.z(c, s, true, o);
                    f = [a.b.j$b(q, m), c.c, c.d];
                }
                break;
            }
            if (!e && this.k()) {
                var t = a.d.i$b._inner[b];
                this.z(c, t, true, o);
                var u = [a.b.j$b(q, m), c.c, c.d];
                if (!isNaN_(u[0])) {
                    if (!isNaN_(u[1]) && !isNaN_(u[2])) {
                        if (a.d != null) {
                            u[0] = u[0] + g;
                        }
                        a.a.f.add(u);
                        e = true;
                    }
                    else {
                        break;
                    }
                }
            }
            b++;
            this.u(c, p, o);
        }
        var v = NaN;
        if (!isNaN_(q)) {
            v = a.b.j$b(q, m);
        }
        h = this.e(v, c);
        return {
            ret: h,
            p1: b,
            p4: e,
            p5: f
        };
    };
    CategoryFramePreparerBase.prototype.y = function (a, b, c, d, e, f, g) {
        var h = a.h;
        var i = a.a.f.count;
        var j = a.a.f;
        var k = this.o;
        var l = k.count;
        if (f) {
            var m = a.c.j$b(a.c.kd, e);
            a.c.j$e(a.a.k, 0, a.a.k.count, e);
            var n = a.a.k;
            var o = n.count;
            for (var p = 0; p < n.count; p++) {
                n._inner[p] = Math.abs(n._inner[p] - m);
            }
        }
    };
    CategoryFramePreparerBase.prototype.w = function (a, b, c) {
    };
    CategoryFramePreparerBase.$t = markType(CategoryFramePreparerBase, 'CategoryFramePreparerBase', FramePreparer.$);
    return CategoryFramePreparerBase;
}(FramePreparer));
export { CategoryFramePreparerBase };
/**
 * @hidden
 */
var CategoryFramePreparer = /** @class */ (function (_super) {
    tslib_1.__extends(CategoryFramePreparer, _super);
    function CategoryFramePreparer(a) {
        var _rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            _rest[_i - 1] = arguments[_i];
        }
        var _this = this;
        a = (a == void 0) ? 0 : a;
        switch (a) {
            case 0:
                {
                    var c = _rest[0];
                    var d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                    {
                        var f = d[0];
                        var g = d[1];
                        var h = d[2];
                        var i = d[3];
                        var j = d[4];
                        _this = _super.call(this, 1, f, g, h, i, j) || this;
                        _this.aa = null;
                        _this.ab = null;
                        _this.aa = new DefaultCategoryTrendlineHost();
                        if (typeCast(IHasCategoryTrendline_$type, f) !== null) {
                            _this.aa = typeCast(IHasCategoryTrendline_$type, f);
                        }
                        _this.ab = new DefaultSingleValueProvider();
                        if (typeCast(IHasSingleValueCategory_$type, f) !== null) {
                            _this.ab = typeCast(IHasSingleValueCategory_$type, f);
                        }
                    }
                }
                break;
            case 1:
                {
                    var c = _rest[0];
                    var d = _rest[1];
                    var e = _rest[2];
                    var f = _rest[3];
                    var g = _rest[4];
                    _this = _super.call(this, 1, c, d, e, f, g) || this;
                    _this.aa = null;
                    _this.ab = null;
                    _this.aa = new DefaultCategoryTrendlineHost();
                    if (typeCast(IHasCategoryTrendline_$type, c) !== null) {
                        _this.aa = typeCast(IHasCategoryTrendline_$type, c);
                    }
                    _this.ab = new DefaultSingleValueProvider();
                    if (typeCast(IHasSingleValueCategory_$type, c) !== null) {
                        _this.ab = typeCast(IHasSingleValueCategory_$type, c);
                    }
                }
                break;
        }
        return _this;
    }
    CategoryFramePreparer.prototype.l = function (a, b, c, d, e, f) {
        var g = b[0];
        var h = b[1];
        var i = new Rect(0, g - 5, h - 5, 11, 11);
        if (!isNaN_(g) && !isNaN_(h) && !isInfinity(g) && !isInfinity(h) && c.tryAdd(i)) {
            a.m.add({ $type: Point_$type, x: g, y: h });
            this.c.updateMarkerTemplate(e, d, f);
            return true;
        }
        return false;
    };
    CategoryFramePreparer.prototype.j = function (a) {
        var b = new SingleValuesHolder();
        var c = this.ab.valueColumn;
        b.f = c;
        return b;
    };
    CategoryFramePreparer.prototype.x = function (a, b, c, d, e, f) {
        if (d) {
            b[0] = b[0] + c;
        }
        else {
            b[0] = a.b.j$b(b[0], e) + c;
        }
        b[1] = a.c.j$b(b[1], f);
        if (a.h > 1 || d) {
            b[2] = a.c.j$b(b[2], f);
        }
        else {
            b[2] = b[1];
        }
    };
    CategoryFramePreparer.prototype.y = function (a, b, c, d, e, f, g) {
        _super.prototype.y.call(this, a, b, c, d, e, f, g);
        var h = a.h;
        var i = a.a.f.count;
        var j = a.a.f;
        var k = this.q;
        var l = k.count;
        var m = new List$1(Number_$type, 0);
        if (!c) {
            m.add(0);
        }
        var n = new List$1(Number_$type, 0);
        n.add(1);
        if (c || h > 1) {
            n.add(2);
        }
        a.b.j$d(j, m, 0, i, d);
        a.c.j$d(j, n, 0, i, e);
        for (var o = 0; o < i; o++) {
            j._inner[o][0] = j._inner[o][0] + b;
        }
        if (h == 1 && !c) {
            for (var p = 0; p < i; p++) {
                j._inner[p][2] = j._inner[p][1];
            }
        }
        if (g) {
            a.b.j$d(k, m, 0, l, d);
            a.c.j$d(k, n, 0, l, e);
            for (var q = 0; q < l; q++) {
                k._inner[q][0] = k._inner[q][0] + b;
            }
            if (h == 1 && !c) {
                for (var r = 0; r < l; r++) {
                    k._inner[r][2] = k._inner[r][1];
                }
            }
        }
    };
    CategoryFramePreparer.prototype.w = function (a, b, c) {
        if (this.aa.trendLineType == 0 || this.aa.trendlinePreparer == null || this.aa.trendLinePeriod < 1) {
            return;
        }
        this.aa.trendlinePreparer.prepareLine(this.aa, a, b, c);
    };
    CategoryFramePreparer.prototype.z = function (a, b, c, d) {
        var e = a;
        var f = e.f;
        var g = f.item(b);
        var h = g;
        if (d) {
            var i = this.g.getBucket(b);
            g = i[1];
            h = i[1];
        }
        if (c) {
            e.c = g;
            e.d = h;
        }
        else {
            e.a = g;
            e.b = h;
        }
    };
    CategoryFramePreparer.prototype.u = function (a, b, c) {
        var d = a;
        var e = d.f;
        if (b < e.count) {
            var f = void 0;
            if (c) {
                var g = this.g.getBucket(b);
                f = g[1];
            }
            else {
                f = e.item(b);
            }
            d.a = Math.min(d.a, f);
            d.b = Math.max(d.b, f);
        }
    };
    CategoryFramePreparer.prototype.e = function (a, b) {
        var c = b;
        return [a, c.a, c.b];
    };
    CategoryFramePreparer.$t = markType(CategoryFramePreparer, 'CategoryFramePreparer', CategoryFramePreparerBase.$);
    return CategoryFramePreparer;
}(CategoryFramePreparerBase));
export { CategoryFramePreparer };
/**
 * @hidden
 */
var CategorySeries = /** @class */ (function (_super) {
    tslib_1.__extends(CategorySeries, _super);
    function CategorySeries() {
        var _this = _super.call(this) || this;
        _this.na = null;
        _this.nn = null;
        _this.nr = false;
        _this.nq = false;
        _this.assigningCategoryStyle = null;
        _this.assigningCategoryMarkerStyle = null;
        _this.nl = null;
        _this.nm = null;
        _this.m7 = new CategoryFrame(3);
        _this.m9 = new CategoryFrame(3);
        _this.m6 = new CategoryFrame(3);
        _this.m8 = new CategoryFrame(3);
        _this.nd = null;
        _this.nh = null;
        _this.nd = new CategorySeriesRenderManager();
        _this.nh = new CategoryTransitionSourceFramePreparer();
        _this.nr = false;
        _this.nq = false;
        _this.ab = CategorySeries.$;
        return _this;
    }
    CategorySeries.prototype.bf = function () {
        var a = new CategorySeriesView(this);
        return a;
    };
    Object.defineProperty(CategorySeries.prototype, "no", {
        get: function () {
            var _this = this;
            if (this.nn == null) {
                this.nn = DVContainer.instance.createInstance(ICategorySeriesInteractionManager_$type, function () { return _this.b2.n; });
            }
            return this.nn;
        },
        enumerable: true,
        configurable: true
    });
    CategorySeries.prototype.ch = function (a, b, c) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var d = this.getEffectiveViewport1(this.bi);
        var e = this.fetchXAxis();
        var f = new ScalerParams(1, this.bs.mq, this.lx, e.b8);
        f.c = d;
        var g = this.fetchYAxis();
        var h = new ScalerParams(1, this.bs.mq, this.lx, g.b8);
        h.c = d;
        var i = this.bs.mq;
        var j = this.lx;
        var k = this.currentMode2Index;
        var l = 0;
        if (e.b5 && e.b3) {
            l = e.getSeriesOffset(i, j, d, k, e.categoryMode);
        }
        a.clear();
        this.lz.cc(function (m) {
            if (m._visibility == 0) {
                var n = new SeriesPointOfInterest();
                var o = m.n;
                var p = m.o;
                if (o < _this.lx.left || o > _this.lx.right || p < _this.lx.top || p > _this.lx.bottom) {
                    return;
                }
                if (e.b2 && !e.b5 && e.categoryMode != 0) {
                    n.b = m.ak;
                }
                else {
                    n.b = e.dy(o, f);
                }
                if (g.b2 && !g.b5 && g.categoryMode != 0) {
                    n.c = m.ak;
                }
                else {
                    n.c = g.dy(p, h);
                }
                n.a = m.content;
                if (b) {
                    n.d = _this.n9(n.c, c);
                }
                else {
                    n.d = e.fi(m.ak);
                }
                a.add(n);
            }
        });
        return true;
    };
    CategorySeries.prototype.n9 = function (a, b) {
        if (b >= 0) {
            return round10N(a, b).toString();
        }
        else {
            return a.toString();
        }
    };
    CategorySeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.ne = a;
    };
    CategorySeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    Object.defineProperty(CategorySeries.prototype, "categoryAxis", {
        get: function () {
            return this.m5();
        },
        enumerable: true,
        configurable: true
    });
    CategorySeries.prototype.m5 = function () {
        return this.fetchXAxis();
    };
    CategorySeries.prototype.np = function () {
        return this.fetchYAxis();
    };
    CategorySeries.prototype.get_cl = function () {
        return true;
    };
    CategorySeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.ne.cn.g(this.eu);
    };
    CategorySeries.prototype.ny = function () {
        return (this.assigningCategoryStyle != null && this.nr) || this.cu;
    };
    CategorySeries.prototype.nz = function () {
        return (this.assigningCategoryMarkerStyle != null && this.nq) || this.cu;
    };
    CategorySeries.prototype.om = function (a) {
        if (this.assigningCategoryStyle != null && this.nr) {
            this.assigningCategoryStyle(this, a);
        }
    };
    CategorySeries.prototype.ol = function (a) {
        if (this.assigningCategoryMarkerStyle != null && this.nq) {
            this.assigningCategoryMarkerStyle(this, a);
        }
    };
    CategorySeries.prototype.mz = function (a, b) {
        return null;
    };
    CategorySeries.prototype.hv = function (a, b) {
        _super.prototype.hv.call(this, a, b);
        if (a) {
            this.og(b);
        }
    };
    CategorySeries.prototype.a8 = function () {
        var a = _super.prototype.a8.call(this);
        return a;
    };
    CategorySeries.prototype.ia = function () {
        _super.prototype.ia.call(this);
        var a = this.fetchXAxis();
        if (a != null) {
            a.g5(false);
        }
        var b = this.fetchYAxis();
        if (b != null) {
            b.g5(false);
        }
    };
    CategorySeries.prototype.js = function (a, b) {
        _super.prototype.js.call(this, a, b);
        this.ne.cn.g(this.eu);
        this.jc(false);
    };
    CategorySeries.prototype.jr = function (a, b) {
        if (this.dm && this.ez < 0.05 && (this.bs == null || !this.bs.c5())) {
            this.dn = true;
        }
        _super.prototype.jr.call(this, a, b);
        this.ne.cn.g(this.eu);
        this.jc(false);
    };
    CategorySeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                var e = this.fetchYAxis();
                var f = this.fetchXAxis();
                if (c != null && d == null) {
                    this.hz(f);
                    this.hz(e);
                }
                if (c == null && d != null) {
                    this.i3(f);
                    this.i3(e);
                }
                this.ne.cn.g(this.eu);
                this.jc(false);
                break;
            case "SyncLink":
                if (this.bn != null && this.bs != null) {
                    this.ne.cn.g(this.eu);
                }
                this.jc(false);
                break;
            case "TransitionProgress":
                this.m9.c(this.ez, this.m7, this.m6);
                if (this.cc(this.bi)) {
                    return;
                }
                if (this.ez == 1) {
                    this.m6.v();
                    this.on(this.m6, this.ne);
                    if (this.dm) {
                        this.dm = false;
                        this.hw();
                        this.bt.r = truncate(this.eq());
                        this.bt.b = this.bu;
                    }
                }
                else {
                    this.on(this.m9, this.ne);
                }
                if (this.bs != null) {
                    this.bs.jj(this);
                }
                break;
            case "ErrorBarSettings":
                this.ij();
                break;
            case CategorySeries.$$p[4]:
                this.jc(false);
                break;
            case CategorySeries.$$p[1]:
                this.nr = this.nt();
                this.jc(false);
                break;
            case CategorySeries.$$p[0]:
                this.nq = this.ns();
                this.jc(false);
                break;
        }
    };
    CategorySeries.prototype.nt = function () {
        return this.nw;
    };
    CategorySeries.prototype.ns = function () {
        return this.nv;
    };
    CategorySeries.prototype.n4 = function (a) {
        return this.ed(a, this.fetchXAxis());
    };
    CategorySeries.prototype.get_c1 = function () {
        return typeCast(ISortingAxis_$type, this.m5()) !== null;
    };
    CategorySeries.prototype.ap = function () {
        return this.m5();
    };
    CategorySeries.prototype.fa = function (a) {
        return this.no.b$d(this, a);
    };
    CategorySeries.prototype.fp = function (a) {
        return this.no.b$f(this, a);
    };
    CategorySeries.prototype.e8 = function (a) {
        var b = truncate(Math.round(this.eb(a)));
        return b;
    };
    CategorySeries.prototype.eb = function (a) {
        return this.no.b$c(this, a);
    };
    CategorySeries.prototype.get_nu = function () {
        return false;
    };
    Object.defineProperty(CategorySeries.prototype, "nu", {
        get: function () {
            return this.get_nu();
        },
        enumerable: true,
        configurable: true
    });
    CategorySeries.prototype.au = function (a, b) {
        var c = this.nk(a, b, this.nm);
        this.nm = c;
        return c;
    };
    CategorySeries.prototype.get_cv = function () {
        return true;
    };
    CategorySeries.prototype.at = function (a, b) {
        if (this.nu) {
            var c = this.nj(a, b, this.nl);
            this.nl = c;
            return c;
        }
        else {
            var d = this.ni(a, b, this.nl);
            this.nl = d;
            return d;
        }
    };
    CategorySeries.prototype.n3 = function () {
        return this.ne.cq.a;
    };
    CategorySeries.prototype.nk = function (a, b, c) {
        if (!this.l2()) {
            return null;
        }
        var d = this.m6;
        if (this.ca()) {
            d = this.m9;
        }
        var e = this.n3();
        return this.no.b$b(this, e, d.f, a, b, c);
    };
    CategorySeries.prototype.nj = function (a, b, c) {
        var d = this.ne.cn.e;
        var e = this.ne.cn.f;
        var f = this.ne.cn.d;
        var g = this.m6;
        if (this.ca()) {
            g = this.m9;
        }
        return this.no.b$a(this, g.f, d, e, f, a, b, c);
    };
    CategorySeries.prototype.ni = function (a, b, c) {
        var d = new HighlightingInfo();
        d.b = this;
        d.f = 0;
        d.e = this.bz.count - 1;
        if (c != null && c.f == d.f && c.e == d.e) {
            return c;
        }
        return d;
    };
    CategorySeries.prototype.oi = function (a, b, c, d, e, f) {
        var g = this.b9;
        this.nd.ab(a, b, c, d, e, f, g);
        var h = this.nd.b;
        var i = null;
        var j = this.bz;
        if (g && this.bs != null) {
            i = this.bs.az.getHighlightingInfo(this, j, d, h.m, h.l, true);
            h.b = i;
        }
        h.f = f;
        this.ol(h);
        this.nd.z(i, f, g);
    };
    CategorySeries.prototype.oj = function (a, b, c, d, e, f) {
        var g = this.b9;
        this.nd.ac(a, b, c, d, e, f, g);
        var h = this.nd.c;
        var i = null;
        var j = this.bz;
        if (g && this.bs != null) {
            i = this.bs.az.getHighlightingInfo(this, j, d, h.m, h.l, false);
            h.b = i;
        }
        h.f = f;
        this.om(h);
        this.nd.aa(i, f, g);
    };
    CategorySeries.prototype.dw = function (a, b, c) {
        var d = _super.prototype.dw.call(this, a, b, c);
        d = d && this.n2(this.fetchXAxis());
        d = d && this.n2(this.fetchYAxis());
        var e = c;
        if (!c.l() || b.isEmpty || a.isEmpty || this.bz == null) {
            d = false;
        }
        if (!d) {
            e.cn.d = 0;
        }
        return d;
    };
    CategorySeries.prototype.n2 = function (a) {
        if (a == null || a.bf == null) {
            return false;
        }
        var b = typeCast(CategoryAxisBase.$, a);
        if (b != null) {
            if (b.itemsSource == null) {
                return false;
            }
            if (b.jx < 1) {
                return false;
            }
        }
        else {
            var c = typeCast(NumericAxisBase.$, a);
            if (c != null) {
                return c.j3 != c.j2;
            }
        }
        return true;
    };
    CategorySeries.prototype.dc = function () {
        return this.nx;
    };
    CategorySeries.prototype.nf = function () {
        return 12;
    };
    CategorySeries.prototype.hw = function () {
        _super.prototype.hw.call(this);
        this.m7.v();
        this.m6.v();
        this.m9.v();
    };
    CategorySeries.prototype.jd = function (a) {
        _super.prototype.jd.call(this, a);
        this.ne.cn.g(this.eu);
        if (this.cc(this.bi)) {
            this.m6.y();
            this.m9.y();
            this.m7.y();
            return;
        }
        if (this.de) {
            if (this.ca()) {
                if (this.bt.g) {
                    this.bt.t();
                }
                this.on(this.m9, this.ne);
            }
            else {
                this.on(this.m6, this.ne);
            }
            return;
        }
        if (this.da(a)) {
            var b = this.m7;
            if (this.ca()) {
                if (this.bt.g) {
                    this.bt.t();
                }
                this.m7 = this.m9;
                this.m9 = b;
            }
            else {
                this.m7 = this.m6;
                this.m6 = b;
            }
            this.ok(this.m6, this.ne);
            if (this.dp) {
                this.bt.w();
                this.bt.r = truncate(this.er());
                this.bt.b = this.bv != null ? this.bv : this.bu;
                this.nh.p(this.m7, this.m6, this.isVertical, this.fetchXAxis(), this.fetchYAxis(), this.ng, this.nf(), this.bp, this.bo(), runOn(this, this.n5), this.ne.br);
            }
            this.hs();
            this.jk();
            if (this.dp) {
                this.dn = false;
                this.dm = true;
            }
        }
        else {
            this.ok(this.m6, this.ne);
            this.on(this.m6, this.ne);
        }
    };
    CategorySeries.prototype.bo = function () {
        return 3;
    };
    CategorySeries.prototype.n5 = function () {
        if (this.isVertical) {
            return this.lx.left;
        }
        return this.lx.bottom;
    };
    CategorySeries.prototype.n8 = function () {
        var e_1, _a;
        var a = 0;
        var b = this.fetchXAxis();
        try {
            for (var _b = tslib_1.__values(fromEnum(this.bs.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                if (c == this) {
                    return a;
                }
                var d = typeCast(CategorySeries.$, c);
                if (d != null) {
                    var e = typeCast(CategoryAxisBase.$, d.fetchXAxis());
                    if (e == b && d.preferredCategoryMode(e) == 2) {
                        a++;
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return -1;
    };
    Object.defineProperty(CategorySeries.prototype, "GetErrorBarsXAxis", {
        get: function () {
            return this.fetchXAxis();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CategorySeries.prototype, "GetErrorBarsYAxis", {
        get: function () {
            return this.fetchYAxis();
        },
        enumerable: true,
        configurable: true
    });
    CategorySeries.prototype.oh = function (a, b) {
        this.nd.w(this, this.nz(), runOn(this, this.mz), this.n6(a), this.n7(a));
        var c = this.nd.b;
        var d = c != null;
        var e = this.m5();
        if (e == null) {
            return;
        }
        var f = this.getEffectiveViewport1(a);
        var g = new ScalerParams(0, a.bs, a.br, e.b8, f);
        var h = this.bz == null ? 0 : this.bz.count;
        var i = a;
        var j = b.f;
        var k = i.cn.e;
        var l = i.cn.f;
        var m = i.cn.d;
        var n = k * m;
        for (var o = 0; o < a.cq.count; o++) {
            var p = a.cq.item(o);
            var q = p.content;
            if (d) {
                var r = p.al;
                this.oi(j, r, h, e, g, a.n);
            }
            this.nd.ad(p, q);
        }
    };
    CategorySeries.prototype.on = function (a, b) {
        this.lo = a.aa;
        b.ar();
    };
    CategorySeries.prototype.ok = function (a, b) {
        a.u();
    };
    CategorySeries.prototype.n6 = function (a) {
        return a.cn.d;
    };
    CategorySeries.prototype.n7 = function (a) {
        return a.cn.e;
    };
    CategorySeries.prototype.og = function (a) {
        var b = a;
        b.cq.count = 0;
    };
    CategorySeries.prototype.je = function (a, b) {
        _super.prototype.je.call(this, a, b);
        if (!this.dl) {
            this.bi.az(b);
            return;
        }
        var c = typeCast(CategorySeriesView.$, this.bh);
        c.cn.g(this.eu);
        this.bi.az(b);
        if (this.cc(this.bh)) {
            return;
        }
        this.oo();
        this.dl = false;
    };
    CategorySeries.prototype.oo = function () {
        var a = typeCast(CategorySeriesView.$, this.bh);
        if (!this.df) {
            this.ok(this.m8, a);
        }
        this.df = false;
        this.on(this.m8, a);
    };
    CategorySeries.prototype.nb = function (a) {
        var b = typeCast(CategorySeriesView.$, a);
        if (b != null && b.n) {
            if (b.m) {
                return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), b, this, b.cn);
            }
            else {
                return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), this.bs.bn.m, this, b.cn);
            }
        }
        else {
            return this.na;
        }
    };
    CategorySeries.prototype.am = function (a) {
        if (this.fetchXAxis() == a) {
            return this.fetchYAxis();
        }
        if (this.fetchYAxis() == a) {
            return this.fetchXAxis();
        }
        return null;
    };
    CategorySeries.prototype.ag = function () {
        return [this.fetchXAxis(), this.fetchYAxis()];
    };
    CategorySeries.m0 = function (a, b, c, d, e) {
        var f = null;
        if (b > -1 && !d.isEmpty && !c.isEmpty) {
            var g = a._inner[0][0] < c.left - 2000 ? c.left - 10 : NaN;
            var h = c.bottom + 10;
            var i = a._inner[b][0] > c.right + 2000 ? c.right + 10 : NaN;
            var j = c.top - 10;
            f = new Clipper(1, g, h, i, j, e);
        }
        return f;
    };
    CategorySeries.$t = markType(CategorySeries, 'CategorySeries', MarkerSeries.$, [IHasCategoryModePreference_$type, ISupportsErrorBars_$type]);
    return CategorySeries;
}(MarkerSeries));
export { CategorySeries };
Type.dep(DependencyProperty, PropertyMetadata, CategorySeries, 'raisePropertyChanged', ['IsCustomCategoryMarkerStyleAllowed:nv:op', [0, false], 'IsCustomCategoryStyleAllowed:nw:oq', [0, false], 'IsTransitionInEnabled:nx:or', [0, false], 'TransitionInMode:ng:os', [CategoryTransitionInMode_$type, enumGetBox(CategoryTransitionInMode_$type, 0)], 'UseHighMarkerFidelity:n1:ot', [0, false]]);
/**
 * @hidden
 */
var CategorySeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(CategorySeriesView, _super);
    function CategorySeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cp = null;
        _this.cs = 0;
        _this.cn = null;
        _this.cq = null;
        _this.cs = -1;
        _this.cp = a;
        _this.cn = _this.co();
        _this.cq = ((function () {
            var $ret = new Pool$1(Marker.$);
            $ret.create = runOn(_this, _this.bx);
            $ret.activate = runOn(_this, _this.cg);
            $ret.disactivate = runOn(_this, _this.ci);
            $ret.destroy = runOn(_this, _this.ch);
            return $ret;
        })());
        return _this;
    }
    CategorySeriesView.prototype.cr = function (a) {
        if (this.cs != a.t) {
            return true;
        }
        return false;
    };
    CategorySeriesView.prototype.cu = function (a) {
        this.cs = a.t;
    };
    CategorySeriesView.prototype.updateMarkerTemplate = function (a, b, c) {
        if (!this.by.l4) {
            var d = this.cq.item(a);
            var e = d.content;
            e.item = this.e.bz.item(b);
            d.ak = b;
            d.al = c;
        }
    };
    CategorySeriesView.prototype.co = function () {
        return new CategoryBucketCalculator(this);
    };
    Object.defineProperty(CategorySeriesView.prototype, "shouldDisplayMarkers", {
        get: function () {
            return this.cp.l2();
        },
        enumerable: true,
        configurable: true
    });
    CategorySeriesView.prototype.updateMarkerCount = function (a) {
        this.cq.count = a;
    };
    CategorySeriesView.prototype.cc = function (a) {
        this.cq.c(a);
    };
    CategorySeriesView.prototype.ct = function () {
    };
    CategorySeriesView.$t = markType(CategorySeriesView, 'CategorySeriesView', MarkerSeriesView.$, [ISupportsMarkers_$type]);
    return CategorySeriesView;
}(MarkerSeriesView));
export { CategorySeriesView };
