// igniteui-react-charts/ES5/CategoryXAxis_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { Base, Number_$type, typeCast, Array_$type, markType, TypeRegistrar } from "igniteui-react-core/ES5/type";
import { HorizontalAxisLabelPanel } from "./HorizontalAxisLabelPanel";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { MathUtil } from "igniteui-react-core/ES5/MathUtil";
import { ScalerParams } from "./ScalerParams";
import { NumericScaler } from "./NumericScaler";
import { LinearCategorySnapper } from "./LinearCategorySnapper";
import { NumericYAxis } from "./NumericYAxis";
import { LabelPosition } from "./LabelPosition";
import { AxisRangeChangedEventArgs } from "./AxisRangeChangedEventArgs";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { truncate, isNaN_, isInfinity } from "igniteui-react-core/ES5/number";
import { CategoryAxisBaseView } from "./CategoryAxisBaseView";
/**
 * @hidden
 */
var CategoryXAxis = /** @class */ (function (_super) {
    tslib_1.__extends(CategoryXAxis, _super);
    function CategoryXAxis() {
        var _this = _super.call(this) || this;
        _this.k8 = 1;
        _this.k7 = 1;
        _this.ky = 0;
        _this.kz = 0;
        _this.jl = new List$1(Number_$type, 0);
        _this.ab = CategoryXAxis.$;
        return _this;
    }
    CategoryXAxis.prototype.a1 = function () {
        return new CategoryXAxisView(this);
    };
    CategoryXAxis.prototype.gr = function (a) {
        _super.prototype.gr.call(this, a);
        this.kw = a;
    };
    Object.defineProperty(CategoryXAxis.prototype, "la", {
        get: function () {
            return this.k8;
        },
        set: function (a) {
            this.k8 = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CategoryXAxis.prototype, "k9", {
        get: function () {
            return this.k7;
        },
        set: function (a) {
            this.k7 = a;
        },
        enumerable: true,
        configurable: true
    });
    CategoryXAxis.prototype.as = function () {
        if (this.co) {
            return (TypeRegistrar.create('HorizontalSmartAxisLabelPanel'));
        }
        else {
            return new HorizontalAxisLabelPanel();
        }
    };
    CategoryXAxis.prototype.getCategorySize = function (a, b, c) {
        var d = this.iy(b, c);
        return d.width / (this.jx * a.width);
    };
    CategoryXAxis.prototype.jq = function (a, b, c) {
        var d = !isNaN_(this.jn) ? MathUtil.b(this.jn, 0, 1) : 0;
        var e = 0;
        if (!isNaN_(this.ju)) {
            e = Math.min(this.ju, 1);
        }
        else {
            e = 0;
        }
        ;
        var f = 1 - 0.5 * d;
        var g = this.j0 == 0 ? 1 : this.j0;
        var h = this.getCategorySize(a, b, c) * f / (g - (g - 1) * e);
        return h;
    };
    CategoryXAxis.prototype.getGroupCenter = function (a, b, c, d) {
        var e = 0.5;
        if (this.j0 > 1) {
            var f = !isNaN_(this.jn) ? MathUtil.b(this.jn, 0, 1) : 0;
            var g = 0;
            if (!isNaN_(this.ju)) {
                g = Math.min(this.ju, 1);
            }
            var h = 1 - 0.5 * f;
            var i = h / (this.j0 - (this.j0 - 1) * g);
            var j = (h - i) / (this.j0 - 1);
            e = 0.25 * f + 0.5 * i + a * j;
        }
        return this.getCategorySize(b, c, d) * e;
    };
    CategoryXAxis.prototype.di = function () {
        return this.k5;
    };
    CategoryXAxis.prototype.lf = function (a) {
        var b = this.bf != null ? this.bf.mq : Rect.empty;
        b = b.copy();
        var c = this.i0;
        var d = this.ix();
        var e = new Rect(0, 0, 0, 1, 1);
        var f = new ScalerParams(0, e, e, this.b8, d);
        var g = !b.isEmpty && !c.isEmpty && this.je != null ? this.je.indexOf(a) : -1;
        var h = g > -1 ? this.dv(g, f) : NaN;
        if (!isNaN_(h) && this.bf.c0) {
            if (!isNaN_(h)) {
                if (h < b.left + 0.1 * b.width) {
                    h = h + 0.4 * b.width;
                    b.x = h - 0.5 * b.width;
                }
                if (h > b.right - 0.1 * b.width) {
                    h = h - 0.4 * b.width;
                    b.x = h - 0.5 * b.width;
                }
            }
            this.bf.km(b, false, true);
        }
    };
    CategoryXAxis.prototype.dv = function (a, b) {
        if (!b.c.isEmpty) {
            var c = this.k4(a, NumericScaler.ac, b.c, this.categoryMode);
            c = (c - b.d.left) / b.d.width;
            c = b.d.left + b.d.width * (c - b.e.left) / b.e.width;
            return c;
        }
        return this.k4(a, b.e, b.d, this.categoryMode);
    };
    CategoryXAxis.prototype.kv = function (a) {
        if (typeCast(List$1.$.specialize(Array_$type), a) !== null) {
            return a.asArray();
        }
        return null;
    };
    CategoryXAxis.prototype.gl = function (a, b, c, d, e) {
        var f;
        var g = e.e;
        var h = e.d;
        var i = e.c;
        var j = e.a;
        var k = !i.isEmpty;
        var l = this.categoryMode == 0 ? this.jx - 1 : this.jx;
        if (l < 0) {
            l = 0;
        }
        var m = i.left;
        var n = i.width;
        var o = g.left;
        var p = g.width;
        var q = h.left;
        var r = h.width;
        var s = 0;
        var t = 1;
        var u = this.kv(a);
        var v = false;
        if (u != null) {
            v = true;
        }
        var w;
        for (var x = 0; x < b.count; x++) {
            w = b.item(x);
            for (var y = c; y < d; y++) {
                if (v) {
                    f = u[y];
                }
                else {
                    f = a.item(y);
                }
                if (k) {
                    var z = l >= 1 ? (f[w]) / (l) : l == 0 ? 0.5 : NaN;
                    if (j) {
                        z = 1 - z;
                    }
                    z = m + n * (z - s) / t;
                    var aa = (z - (o * r)) / p;
                    f[w] = aa;
                }
                else {
                    var ab = l >= 1 ? (f[w]) / (l) : l == 0 ? 0.5 : NaN;
                    if (j) {
                        ab = 1 - ab;
                    }
                    ab = q + r * (ab - o) / p;
                    f[w] = ab;
                }
            }
        }
    };
    CategoryXAxis.prototype.dw = function (a, b, c) {
        if (!b.c.isEmpty) {
            var d = this.k4(a, NumericScaler.ac, b.c, c);
            d = (d - b.d.left) / b.d.width;
            d = b.d.left + b.d.width * (d - b.e.left) / b.e.width;
            return d;
        }
        return this.k4(a, b.e, b.d, c);
    };
    CategoryXAxis.prototype.k4 = function (a, b, c, d) {
        var e = d == 0 ? this.jx - 1 : this.jx;
        if (e < 0) {
            e = 0;
        }
        var f = e >= 1 ? (a) / (e) : e == 0 ? 0.5 : NaN;
        if (this.b9) {
            f = 1 - f;
        }
        return c.left + c.width * (f - b.left) / b.width;
    };
    CategoryXAxis.prototype.dy = function (a, b) {
        if (!b.c.isEmpty) {
            var c = b.e.left + b.e.width * (a - b.d.left) / b.d.width;
            c = (c * b.d.width) + b.d.left;
            c = this.getUnscaledValue3(c, NumericScaler.ac, b.c, this.categoryMode);
            return c;
        }
        return this.getUnscaledValue3(a, b.e, b.d, this.categoryMode);
    };
    CategoryXAxis.prototype.dz = function (a, b, c) {
        if (!b.c.isEmpty) {
            var d = b.e.left + b.e.width * (a - b.d.left) / b.d.width;
            d = (d * b.d.width) + b.d.left;
            d = this.getUnscaledValue3(d, NumericScaler.ac, b.c, c);
            return d;
        }
        return this.getUnscaledValue3(a, b.e, b.d, c);
    };
    CategoryXAxis.prototype.getUnscaledValue3 = function (a, b, c, d) {
        var e = b.left + (a - c.left) * b.width / c.width;
        if (this.b9) {
            e = 1 - e;
        }
        var f = d == 0 ? this.jx - 1 : this.jx;
        if (f < 0) {
            f = 0;
        }
        return e * f;
    };
    CategoryXAxis.prototype.lg = function (a, b) {
        if (a == NaN || b == NaN || this.je == null || this.je.count == 0) {
            return;
        }
        var c = a / this.je.count;
        var d = (b + 1) / this.je.count;
        this.bf.m4 = new Rect(0, c, this.bf.m4.y, d - c, this.bf.m4.height);
    };
    CategoryXAxis.prototype.g6 = function (a) {
        _super.prototype.g6.call(this, a);
        var b = this.bf != null ? this.bf.mq : Rect.empty;
        var c = this.i0;
        var d = this.iv();
        var e = this.iu();
        var f = new ScalerParams(0, b, c, this.b8, d);
        var g = this.a2.ay();
        var h = this.a2.a1();
        var i = this.a2.az();
        var j = this.a2.a0();
        var k = this.a2.e();
        var l = this.a2.f();
        var m = this.a2.g();
        var n = this.je;
        var o = this.kx();
        this.hd();
        this.gh(g);
        this.gh(h);
        this.gh(i);
        this.gh(j);
        this.cu.clear();
        this.cv.clear();
        this.jl.clear();
        this.a2.at(this, b, c);
        if (b.isEmpty || c.isEmpty) {
            this.bo.count = 0;
            this.bn.count = 0;
            this.bm.count = 0;
        }
        if (this.bo.count == 0 && this.bn.count == 0) {
            this.a2.y();
        }
        if (this.aw != null) {
            this.aw.registerAxis(this);
        }
        if (this.itemsSource == null || o == null || n.count == 0) {
            this.bo.count = 0;
            this.bn.count = 0;
            this.bm.count = 0;
            this.a2.y();
            return;
        }
        if (!b.isEmpty && !c.isEmpty) {
            var p = this.dy(e.left, f);
            var q = this.dy(e.right, f);
            if (this.b8) {
                p = Math.ceil(p);
                q = Math.floor(q);
            }
            else {
                p = Math.floor(p);
                q = Math.ceil(q);
            }
            var r = c.bottom;
            var s = r - c.top;
            if (this.ag != null) {
                var t = typeCast(NumericYAxis.$, this.ag);
                if (t != null) {
                    var u = this.g9(r, s, b, c, d, c.top, c.bottom);
                    r = u.p0;
                    s = u.p1;
                }
            }
            this.gp(g, r, c, k, true);
            this.a2.ao(s);
            var v = Math.min(p, q);
            var w = Math.max(p, q);
            var x = new LinearCategorySnapper(1, v, w, e.width, this.k5, this.categoryMode, this.d8());
            var y = Math.floor((v - 0) / x.g);
            var z = Math.ceil((w - 0) / x.g);
            this.k0 = x.g;
            if (!isNaN_(y) && !isNaN_(z)) {
                var aa = truncate(y);
                var ab = truncate(z);
                var ac = this.k3(0 + aa * x.g, f);
                var ad = this.dv(x.g, f);
                var ae = Math.abs(ad - this.dv(0, f));
                this.a2.ap(ad);
                var af = truncate(Math.ceil(e.right));
                var ag = truncate(Math.floor(e.left));
                var ah = this.k6;
                var ai = ah > 0 && ah < x.g;
                var aj = c.right;
                var ak = this.jl;
                var al = x.g;
                var am = truncate(al);
                var an = this.j0;
                var ao = this.b8;
                var ap = this.categoryMode;
                for (var aq = aa; aq <= ab; ++aq) {
                    var ar = this.k3(0 + (aq + 1) * x.g, f);
                    if (ac <= aj) {
                        if (aq % 2 == 0) {
                            this.hf(h, ac, ar, c);
                        }
                        this.he(i, ac, c, l, false);
                        ak.add(ac);
                    }
                    if (this.categoryMode != 0 && this.j0 != 0 && this.ci) {
                        if (ai) {
                            var as = ae * this.k6;
                            if (this.b8) {
                                var at = ac;
                                while (at - as > ar) {
                                    at -= as;
                                    if (Math.abs(at - ar) < 0.0001) {
                                        break;
                                    }
                                    this.he(j, at, c, m, false);
                                }
                            }
                            else {
                                var au = ac;
                                while (au + as < ar) {
                                    au += as;
                                    if (Math.abs(au - ar) < 0.0001) {
                                        break;
                                    }
                                    this.he(j, au, c, m, false);
                                }
                            }
                            this.k1 = this.k6;
                        }
                        else {
                            for (var av = 0; av < truncate(x.g); av++) {
                                for (var aw = 0; aw < this.j0; aw++) {
                                    var ax = this.getGroupCenter(aw, b, c, d);
                                    if (this.b8) {
                                        ax = -ax;
                                    }
                                    var ay = this.dv(av + aq * x.g, f) + ax;
                                    this.he(j, ay, c, m, false);
                                }
                            }
                            this.k1 = 0.5;
                        }
                    }
                    var az = this.k2(aq * al, ac, ap == 0, f);
                    var a0 = truncate(Math.round(az));
                    if (a0 >= ag && a0 <= af) {
                        var a1 = 0;
                        if (al >= 1) {
                            a1 = aq * truncate(Math.floor(al));
                        }
                        else {
                            if ((aq * al) * 2 % 2 == 0) {
                                a1 = truncate(Math.floor(aq * al));
                            }
                            else {
                                a1 = -1;
                            }
                        }
                        if (o != null && a1 < n.count && a1 >= 0) {
                            var a2 = o.getItem(a1);
                            var a3 = this.e5(a2);
                            if (!isNaN_(az) && !isInfinity(az) && a3 != null) {
                                if ((typeof a3 === 'string') && Base.equalsStatic(a3, "")) {
                                }
                                else {
                                    this.cu.add1(a3);
                                    this.cv.add(new LabelPosition(az));
                                }
                            }
                        }
                    }
                    ac = ar;
                }
            }
            if ((this.aw == null || this.aw.visibility == 0) && this.ag != null) {
                if (this.aw != null && (this.aw.location == 4 || this.aw.location == 5)) {
                    this.bf.ht();
                }
            }
            this.a2.au(this.cu, this.cv);
            this.kl();
        }
    };
    CategoryXAxis.prototype.cm = function () {
        if (this.cf && this.cn) {
            this.a9 = this.a8();
        }
        if (this.je == null) {
            return false;
        }
        var a = this.je.count;
        if (a != this.k9) {
            var b = new AxisRangeChangedEventArgs(0, 1, 1, this.k9, a);
            this.k9 = a;
            this.g0(b);
            return true;
        }
        return false;
    };
    Object.defineProperty(CategoryXAxis.prototype, "k5", {
        get: function () {
            return this.c(CategoryXAxis.lh);
        },
        set: function (a) {
            this.h(CategoryXAxis.lh, a);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CategoryXAxis.prototype, "k0", {
        get: function () {
            return this.ky;
        },
        set: function (a) {
            if (this.k0 != a) {
                var b = this.ky;
                this.ky = a;
                this.gz("ActualInterval", b, this.k0);
            }
        },
        enumerable: true,
        configurable: true
    });
    CategoryXAxis.prototype.d6 = function () {
        return this.k0;
    };
    Object.defineProperty(CategoryXAxis.prototype, "k6", {
        get: function () {
            return this.c(CategoryXAxis.li);
        },
        set: function (a) {
            this.h(CategoryXAxis.li, a);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CategoryXAxis.prototype, "k1", {
        get: function () {
            return this.kz;
        },
        set: function (a) {
            if (this.k1 != a) {
                var b = this.kz;
                this.kz = a;
                this.gz("ActualMinorInterval", b, this.k1);
            }
        },
        enumerable: true,
        configurable: true
    });
    CategoryXAxis.prototype.jh = function (a) {
        if (a == null) {
            return false;
        }
        var b = this.bb();
        if (b == null) {
            return false;
        }
        return b.synchronizeHorizontally;
    };
    CategoryXAxis.prototype.get_az = function () {
        return 0;
    };
    Object.defineProperty(CategoryXAxis.prototype, "az", {
        get: function () {
            return this.get_az();
        },
        enumerable: true,
        configurable: true
    });
    CategoryXAxis.prototype.kx = function () {
        return this.je;
    };
    CategoryXAxis.prototype.k3 = function (a, b) {
        return this.dv(a, b);
    };
    CategoryXAxis.prototype.k2 = function (a, b, c, d) {
        if (!c) {
            var e = this.k3(a + 1, d);
            return (b + e) / 2;
        }
        return b;
    };
    Object.defineProperty(CategoryXAxis.prototype, "isVertical", {
        get: function () {
            return this.ce;
        },
        enumerable: true,
        configurable: true
    });
    CategoryXAxis.prototype.j$c = function (a, b) {
        return this.dy.apply(this, arguments);
    };
    CategoryXAxis.prototype.j$b = function (a, b) {
        return this.dv.apply(this, arguments);
    };
    CategoryXAxis.prototype.j$d = function (a, b, c, d, e) {
        this.gl.apply(this, arguments);
    };
    CategoryXAxis.prototype.j$e = function (a, b, c, d) {
        this.gm.apply(this, arguments);
    };
    CategoryXAxis.prototype.j$f = function (a, b, c, d) {
        this.gn.apply(this, arguments);
    };
    Object.defineProperty(CategoryXAxis.prototype, "j$a", {
        get: function () {
            return this.b8;
        },
        enumerable: true,
        configurable: true
    });
    CategoryXAxis.$t = markType(CategoryXAxis, 'CategoryXAxis', CategoryAxisBase.$);
    CategoryXAxis.lh = DependencyProperty.i("Interval", Number_$type, CategoryXAxis.$, new PropertyMetadata(2, NaN, function (a, b) {
        a.gz("Interval", b.oldValue, b.newValue);
        a.g5(false);
    }));
    CategoryXAxis.li = DependencyProperty.i("MinorInterval", Number_$type, CategoryXAxis.$, new PropertyMetadata(2, NaN, function (a, b) {
        (typeCast(CategoryXAxis.$, a)).gz("MinorInterval", b.oldValue, b.newValue);
        (typeCast(CategoryXAxis.$, a)).g5(false);
    }));
    return CategoryXAxis;
}(CategoryAxisBase));
export { CategoryXAxis };
/**
 * @hidden
 */
var CategoryXAxisView = /** @class */ (function (_super) {
    tslib_1.__extends(CategoryXAxisView, _super);
    function CategoryXAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bg = null;
        _this.bg = a;
        return _this;
    }
    CategoryXAxisView.$t = markType(CategoryXAxisView, 'CategoryXAxisView', CategoryAxisBaseView.$);
    return CategoryXAxisView;
}(CategoryAxisBaseView));
export { CategoryXAxisView };
