// igniteui-react-charts/ES5/ChaikinVolatilityIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ChaikinVolatilityIndicatorStrategy } from "./ChaikinVolatilityIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var ChaikinVolatilityIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(ChaikinVolatilityIndicator, _super);
    function ChaikinVolatilityIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChaikinVolatilityIndicator.prototype.get_qb = function () {
        return new ChaikinVolatilityIndicatorStrategy();
    };
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    ChaikinVolatilityIndicator.prototype.get_qo = function () {
        return ChaikinVolatilityIndicator.$;
    };
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(ChaikinVolatilityIndicator.qx));
        },
        set: function (a) {
            this.h(ChaikinVolatilityIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    ChaikinVolatilityIndicator.prototype.qh = function () {
        return this.qw;
    };
    ChaikinVolatilityIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    ChaikinVolatilityIndicator.$t = markType(ChaikinVolatilityIndicator, 'ChaikinVolatilityIndicator', StrategyBasedIndicator.$);
    ChaikinVolatilityIndicator.qx = StrategyBasedIndicator.qt(10, ChaikinVolatilityIndicator.$);
    return ChaikinVolatilityIndicator;
}(StrategyBasedIndicator));
export { ChaikinVolatilityIndicator };
