// igniteui-react-charts/ES5/CommodityChannelIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CommodityChannelIndexIndicatorStrategy } from "./CommodityChannelIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var CommodityChannelIndexIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(CommodityChannelIndexIndicator, _super);
    function CommodityChannelIndexIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CommodityChannelIndexIndicator.prototype.get_qb = function () {
        return new CommodityChannelIndexIndicatorStrategy();
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.get_qo = function () {
        return CommodityChannelIndexIndicator.$;
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(CommodityChannelIndexIndicator.qx));
        },
        set: function (a) {
            this.h(CommodityChannelIndexIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.qh = function () {
        return this.qw;
    };
    CommodityChannelIndexIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    CommodityChannelIndexIndicator.$t = markType(CommodityChannelIndexIndicator, 'CommodityChannelIndexIndicator', StrategyBasedIndicator.$);
    CommodityChannelIndexIndicator.qx = StrategyBasedIndicator.qt(20, CommodityChannelIndexIndicator.$);
    return CommodityChannelIndexIndicator;
}(StrategyBasedIndicator));
export { CommodityChannelIndexIndicator };
