// igniteui-react-charts/ES5/CustomIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CustomIndicatorStrategy } from "./CustomIndicatorStrategy";
import { String_$type, fromEnum, markType } from "igniteui-react-core/ES5/type";
import { FinancialEventArgs } from "./FinancialEventArgs";
import { List$1 } from "igniteui-react-core/ES5/List$1";
/**
 * @hidden
 */
var CustomIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(CustomIndicator, _super);
    function CustomIndicator() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.qx = null;
        _this.qw = null;
        return _this;
    }
    CustomIndicator.prototype.get_qb = function () {
        return new CustomIndicatorStrategy();
    };
    Object.defineProperty(CustomIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    CustomIndicator.prototype.get_qo = function () {
        return CustomIndicator.$;
    };
    Object.defineProperty(CustomIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    CustomIndicator.prototype.pn = function (a, b) {
        var e_1, _a;
        var c = this.mj(a, b);
        var d = this.mk(c);
        var e = new List$1(String_$type, 0);
        if (this.qw != null) {
            var f = new FinancialEventArgs(c.calculateFrom, c.calculateCount, c, d, this);
            this.qw(this, f);
            if (f.basedOn != null) {
                try {
                    for (var _b = tslib_1.__values(fromEnum(f.basedOn)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var g = _c.value;
                        e.add(g);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        }
        return e;
    };
    Object.defineProperty(CustomIndicator.prototype, "indicator", {
        get: function () {
            return this.qx;
        },
        set: function (a) {
            this.qx = a;
            this.pv();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CustomIndicator.prototype, "basedOnColumns", {
        get: function () {
            return this.qw;
        },
        set: function (a) {
            this.qw = a;
        },
        enumerable: true,
        configurable: true
    });
    CustomIndicator.prototype.pk = function (a, b) {
        _super.prototype.pk.call(this, a, b);
        if (this.qx != null) {
            var c = this.mj(a, b);
            if (b == 0) {
                return false;
            }
            if (!this.ng(this.pn(a, b))) {
                return false;
            }
            this.qx(this, new FinancialEventArgs(a, b, c, this.mk(c), this));
            if (this.qd(c) && this.ms != null) {
                this.ms.ck();
            }
            return true;
        }
        return false;
    };
    CustomIndicator.$t = markType(CustomIndicator, 'CustomIndicator', StrategyBasedIndicator.$);
    return CustomIndicator;
}(StrategyBasedIndicator));
export { CustomIndicator };
