// igniteui-react-charts/ES5/DetrendedPriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { DetrendedPriceOscillatorIndicatorStrategy } from "./DetrendedPriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var DetrendedPriceOscillatorIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(DetrendedPriceOscillatorIndicator, _super);
    function DetrendedPriceOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DetrendedPriceOscillatorIndicator.prototype.get_qb = function () {
        return new DetrendedPriceOscillatorIndicatorStrategy();
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.get_qo = function () {
        return DetrendedPriceOscillatorIndicator.$;
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(DetrendedPriceOscillatorIndicator.qx));
        },
        set: function (a) {
            this.h(DetrendedPriceOscillatorIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.qh = function () {
        return this.qw;
    };
    DetrendedPriceOscillatorIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    DetrendedPriceOscillatorIndicator.$t = markType(DetrendedPriceOscillatorIndicator, 'DetrendedPriceOscillatorIndicator', StrategyBasedIndicator.$);
    DetrendedPriceOscillatorIndicator.qx = StrategyBasedIndicator.qt(20, DetrendedPriceOscillatorIndicator.$);
    return DetrendedPriceOscillatorIndicator;
}(StrategyBasedIndicator));
export { DetrendedPriceOscillatorIndicator };
