// igniteui-react-charts/ES5/FastStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FastStochasticOscillatorIndicatorStrategy } from "./FastStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
import { Brush } from "igniteui-react-core/ES5/Brush";
/**
 * @hidden
 */
var FastStochasticOscillatorIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(FastStochasticOscillatorIndicator, _super);
    function FastStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    FastStochasticOscillatorIndicator.prototype.get_qb = function () {
        return new FastStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    FastStochasticOscillatorIndicator.prototype.get_qo = function () {
        return FastStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(FastStochasticOscillatorIndicator.qx));
        },
        set: function (a) {
            this.h(FastStochasticOscillatorIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    FastStochasticOscillatorIndicator.prototype.qh = function () {
        return this.qw;
    };
    FastStochasticOscillatorIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    FastStochasticOscillatorIndicator.$t = markType(FastStochasticOscillatorIndicator, 'FastStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FastStochasticOscillatorIndicator.qx = StrategyBasedIndicator.qt(14, FastStochasticOscillatorIndicator.$);
    return FastStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { FastStochasticOscillatorIndicator };
