// igniteui-react-charts/ES5/FinancialPriceSeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, markType, getInstanceType } from "igniteui-react-core/ES5/type";
import { IFinancialPriceSeriesProxy_$type } from "./IFinancialPriceSeriesProxy";
import { stringEndsWith } from "igniteui-react-core/ES5/string";
/**
 * @hidden
 */
var FinancialPriceSeriesProxy = /** @class */ (function (_super) {
    tslib_1.__extends(FinancialPriceSeriesProxy, _super);
    function FinancialPriceSeriesProxy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FinancialPriceSeriesProxy.prototype.getCloseMemberPath = function (a) {
        return a.ob;
    };
    FinancialPriceSeriesProxy.prototype.getHighMemberPath = function (a) {
        return a.oe;
    };
    FinancialPriceSeriesProxy.prototype.getLowMemberPath = function (a) {
        return a.oj;
    };
    FinancialPriceSeriesProxy.prototype.getOpenMemberPath = function (a) {
        return a.on;
    };
    FinancialPriceSeriesProxy.prototype.getVolumeMemberPath = function (a) {
        return a.or;
    };
    FinancialPriceSeriesProxy.prototype.matchesType = function (a) {
        if (!a.cr) {
            return false;
        }
        if (stringEndsWith(getInstanceType(a).typeName, "Indicator") || stringEndsWith(getInstanceType(a).typeName, "Overlay")) {
            return false;
        }
        return true;
    };
    FinancialPriceSeriesProxy.prototype.setCategoryMode = function (a, b) {
        a.me = b;
    };
    FinancialPriceSeriesProxy.prototype.setCloseMemberPath = function (a, b) {
        a.ob = b;
    };
    FinancialPriceSeriesProxy.prototype.setDisplayType = function (a, b) {
        a.pj = b;
    };
    FinancialPriceSeriesProxy.prototype.setHighMemberPath = function (a, b) {
        a.oe = b;
    };
    FinancialPriceSeriesProxy.prototype.setLowMemberPath = function (a, b) {
        a.oj = b;
    };
    FinancialPriceSeriesProxy.prototype.setOpenMemberPath = function (a, b) {
        a.on = b;
    };
    FinancialPriceSeriesProxy.prototype.setVolumeMemberPath = function (a, b) {
        a.or = b;
    };
    FinancialPriceSeriesProxy.prototype.setXAxis = function (a, b) {
        a.l7 = b;
    };
    FinancialPriceSeriesProxy.prototype.setYAxis = function (a, b) {
        a.ms = b;
    };
    FinancialPriceSeriesProxy.$t = markType(FinancialPriceSeriesProxy, 'FinancialPriceSeriesProxy', Base.$, [IFinancialPriceSeriesProxy_$type]);
    return FinancialPriceSeriesProxy;
}(Base));
export { FinancialPriceSeriesProxy };
