// igniteui-react-charts/ES5/FinancialOverlay.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { FinancialSeries } from "./FinancialSeries";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { ISortingAxis_$type } from "./ISortingAxis";
import { Type, typeCast, markType } from "igniteui-react-core/ES5/type";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { ScalerParams } from "./ScalerParams";
import { AxisRange } from "./AxisRange";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { isNaN_ } from "igniteui-react-core/ES5/number";
/**
 * @hidden
 */
var FinancialOverlay = /** @class */ (function (_super) {
    tslib_1.__extends(FinancialOverlay, _super);
    function FinancialOverlay() {
        var _this = _super.call(this) || this;
        _this.ph = false;
        _this.ph = false;
        return _this;
    }
    FinancialOverlay.prototype.hy = function (a, b, c, d) {
        _super.prototype.hy.call(this, a, b, c, d);
        if (this.l7 != null && typeCast(ISortingAxis_$type, this.l7) !== null) {
            this.l7.i$i();
        }
        this.mn.bu.e(this.eu);
        this.pi();
        if (this.ms != null) {
            this.ms.ck();
        }
        this.jc(true);
    };
    FinancialOverlay.prototype.get_dv = function () {
        return true;
    };
    FinancialOverlay.prototype.iy = function (a, b, c, d) {
        switch (b) {
            case "FastItemsSource":
                this.ph = false;
                break;
            case FinancialOverlay.$$p[0]:
                this.ph = false;
                this.jc(false);
                this.ij();
                break;
            case "XAxis":
                this.ph = false;
                break;
        }
        _super.prototype.iy.call(this, a, b, c, d);
    };
    FinancialOverlay.prototype.bd = function () {
        if (this.be == 0) {
            return 1;
        }
        else {
            return _super.prototype.bd.call(this);
        }
    };
    FinancialOverlay.prototype.dh = function (a, b) {
        if (this.dg(a, b)) {
            return true;
        }
        return false;
    };
    FinancialOverlay.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.jc(false);
    };
    FinancialOverlay.prototype.c9 = function (a) {
        var b = this.bz != null ? this.bz.indexOf(a) : -1;
        var c = this.bi.bs;
        c = c.copy();
        var d = this.bi.br;
        var e = new Rect(0, 0, 0, 1, 1);
        var f = this.lr(this.bi);
        var g = new ScalerParams(0, e, e, this.l7.b8, f);
        var h = new ScalerParams(0, e, e, this.ms.b8, f);
        h.b = this.eh();
        if (b >= 0 && Rect.l_op_Inequality(c, null) && Rect.l_op_Inequality(d, null)) {
            if (this.l7 != null) {
                var i = this.l7.dv(b, g);
                if (i < c.left + 0.1 * c.width) {
                    i = i + 0.4 * c.width;
                }
                if (i > c.right - 0.1 * c.width) {
                    i = i - 0.4 * c.width;
                }
                c.x = i - 0.5 * c.width;
            }
            if (this.ms != null && this.mz != null && this.mx != null && b < this.mz.count && b < this.mx.count) {
                var j = this.ms.dv(this.mz.item(b), h);
                var k = this.ms.dv(this.mx.item(b), h);
                if (!isNaN_(j) && !isNaN_(k)) {
                    var l = Math.abs(j - k);
                    if (c.height < l) {
                        c.height = l;
                        c.y = Math.min(j, k);
                    }
                    else {
                        if (j < c.top + 0.1 * c.height) {
                            j = j + 0.4 * c.height;
                        }
                        if (j > c.bottom - 0.1 * c.height) {
                            j = j - 0.4 * c.height;
                        }
                        c.y = j - 0.5 * c.height;
                    }
                }
            }
            this.bn.ap(this.bs, c, true);
        }
        return b >= 0;
    };
    FinancialOverlay.prototype.an = function (a) {
        if (a != null && a == this.ms && this.mz != null && this.mx != null) {
            return new AxisRange(this.mz.minimum, this.mx.maximum);
        }
        return null;
    };
    FinancialOverlay.$t = markType(FinancialOverlay, 'FinancialOverlay', FinancialSeries.$);
    return FinancialOverlay;
}(FinancialSeries));
export { FinancialOverlay };
Type.dep(DependencyProperty, PropertyMetadata, FinancialOverlay, 'raisePropertyChanged', ['IgnoreFirst:pj:pl', [1, 0]]);
