// igniteui-react-charts/ES5/FullStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FullStochasticOscillatorIndicatorStrategy } from "./FullStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
import { Brush } from "igniteui-react-core/ES5/Brush";
/**
 * @hidden
 */
var FullStochasticOscillatorIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(FullStochasticOscillatorIndicator, _super);
    function FullStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    FullStochasticOscillatorIndicator.prototype.get_qb = function () {
        return new FullStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.get_qo = function () {
        return FullStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.q1));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.q1, a);
        },
        enumerable: true,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.qh = function () {
        return this.qw;
    };
    FullStochasticOscillatorIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "smoothingPeriod", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.q2));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.q2, a);
        },
        enumerable: true,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.qj = function () {
        return this.smoothingPeriod;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "qy", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.q3));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.q3, a);
        },
        enumerable: true,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.qg = function () {
        return this.qy;
    };
    FullStochasticOscillatorIndicator.prototype.pr = function () {
        return this.qy;
    };
    FullStochasticOscillatorIndicator.$t = markType(FullStochasticOscillatorIndicator, 'FullStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FullStochasticOscillatorIndicator.q1 = StrategyBasedIndicator.qt(14, FullStochasticOscillatorIndicator.$);
    FullStochasticOscillatorIndicator.q2 = StrategyBasedIndicator.qu(3, FullStochasticOscillatorIndicator.$, "SmoothingPeriod");
    FullStochasticOscillatorIndicator.q3 = StrategyBasedIndicator.qu(3, FullStochasticOscillatorIndicator.$, "TriggerPeriod");
    return FullStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { FullStochasticOscillatorIndicator };
