// igniteui-react-charts/ES5/HighDensityScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Series } from "./Series";
import { NumericXAxis } from "./NumericXAxis";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { NumericYAxis } from "./NumericYAxis";
import { Color } from "igniteui-react-core/ES5/Color";
import { ScatterFrame } from "./ScatterFrame";
import { KDTree2D } from "./KDTree2D";
import { Type, typeGetValue, runOn, delegateRemove, typeCast, delegateCombine, markType, Nullable$1 } from "igniteui-react-core/ES5/type";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { IFastItemsSource_$type } from "igniteui-react-core/ES5/IFastItemsSource";
import { SeriesView } from "./SeriesView";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { KDPointData } from "./KDPointData";
import { ScalerParams } from "./ScalerParams";
import { ProgressiveLoadStatusEventArgs } from "./ProgressiveLoadStatusEventArgs";
import { SearchArgs } from "./SearchArgs";
import { KDTreeNode2D } from "./KDTreeNode2D";
import { AxisRange } from "./AxisRange";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { truncate, isNaN_, isInfinity, logBase, intDivide } from "igniteui-react-core/ES5/number";
import { boxArray$1, arrayShallowClone } from "igniteui-react-core/ES5/array";
import { dateNow } from "igniteui-react-core/ES5/date";
import { RenderingContext } from "igniteui-react-core/ES5/RenderingContext";
import { CanvasViewRenderer } from "igniteui-react-core/ES5/CanvasViewRenderer";
import { Rectangle } from "igniteui-react-core/ES5/Rectangle";
/**
 * @hidden
 */
var HighDensityScatterSeries = /** @class */ (function (_super) {
    tslib_1.__extends(HighDensityScatterSeries, _super);
    function HighDensityScatterSeries() {
        var _this = _super.call(this) || this;
        _this.mc = null;
        _this.me = null;
        _this.nk = 0;
        _this.mg = false;
        _this.m0 = 0;
        _this.mz = 0;
        _this.my = 0;
        _this.mx = 255;
        _this.mw = 0;
        _this.mv = 0;
        _this.m3 = 255;
        _this.m2 = 0;
        _this.m1 = 0;
        _this.l4 = null;
        _this.l9 = null;
        _this.ma = null;
        _this.l5 = null;
        _this.l7 = null;
        _this.mi = false;
        _this.nl = 0;
        _this.nn = 0;
        _this.nh = 0;
        _this.ng = 0;
        _this.progressiveLoadStatusChanged = null;
        _this.mn = null;
        _this.mt = 0;
        _this.ms = 0;
        _this.mu = 0;
        _this.m6 = 0;
        _this.m4 = 0;
        _this.m5 = 0;
        _this.m7 = 0;
        _this.m9 = 0;
        _this.nb = 0;
        _this.na = 0;
        _this.m8 = 0;
        _this.mp = 0;
        _this.mq = 0;
        _this.mr = 0;
        _this.mo = 0;
        _this.nj = 0;
        _this.ni = 0;
        _this.l1 = null;
        _this.nm = 0;
        _this.mh = false;
        _this._itemIndexes = null;
        _this.l2 = null;
        _this.lz = null;
        _this.mb = null;
        _this.ab = HighDensityScatterSeries.$;
        _this.c7 = false;
        return _this;
    }
    HighDensityScatterSeries.prototype.get_c4 = function () {
        return true;
    };
    Object.defineProperty(HighDensityScatterSeries.prototype, "md", {
        get: function () {
            return this.mc;
        },
        set: function (a) {
            if (this.mc != a) {
                var b = this.md;
                this.mc = a;
                this.raisePropertyChanged("XColumn", b, this.md);
            }
        },
        enumerable: true,
        configurable: true
    });
    HighDensityScatterSeries.prototype.fs = function (a, b) {
        var c = this.gu(b);
        if (c == HighDensityScatterSeries.$$p[8]) {
            return this.ft(a, b, this.n2);
        }
        if (c == HighDensityScatterSeries.$$p[10]) {
            return this.ft(a, b, this.n6);
        }
        return _super.prototype.fs.call(this, a, b);
    };
    Object.defineProperty(HighDensityScatterSeries.prototype, "mf", {
        get: function () {
            return this.me;
        },
        set: function (a) {
            if (this.me != a) {
                var b = this.mf;
                this.me = a;
                this.raisePropertyChanged("YColumn", b, this.mf);
            }
        },
        enumerable: true,
        configurable: true
    });
    HighDensityScatterSeries.prototype.ok = function () {
        if (this.l4 == null) {
            return;
        }
        this.l4.s();
        var a = this.l4;
        a.progressiveThunkCompleted = delegateRemove(a.progressiveThunkCompleted, runOn(this, this.n8));
        this.l4 = null;
    };
    HighDensityScatterSeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                if (this.bs != null) {
                    var e = this.getEffectiveViewport1(this.bi);
                    this.mh = !e.isEmpty;
                }
                break;
            case "FastItemsSource":
                this.ok();
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    c.deregisterColumn(this.md);
                    c.deregisterColumn(this.mf);
                    this.md = null;
                    this.mf = null;
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    this.md = this.bw(this.n2);
                    this.mf = this.bw(this.n6);
                }
                if ((this.l8 != null && !this.l8.ck()) || (this.l6 != null && !this.l6.ck())) {
                    this.jc(false);
                }
                this.ik();
                break;
            case HighDensityScatterSeries.$$p[7]:
                var f = this.l5;
                this.l5 = this.l6;
                this.ok();
                if (f != null) {
                    f.rangeChanged = delegateRemove(f.rangeChanged, runOn(this, this.ob));
                }
                if (this.l5 != null) {
                    var g = this.l5;
                    g.rangeChanged = delegateCombine(g.rangeChanged, runOn(this, this.ob));
                }
                if (c != null) {
                    c.bv(this);
                }
                if (d != null) {
                    d.ch(this);
                }
                if ((this.l6 != null && !this.l6.ck()) || (d == null && c != null)) {
                    this.jc(false);
                }
                break;
            case HighDensityScatterSeries.$$p[9]:
                var h = this.l7;
                this.l7 = this.l8;
                this.ok();
                if (h != null) {
                    h.rangeChanged = delegateRemove(h.rangeChanged, runOn(this, this.ob));
                }
                if (this.l7 != null) {
                    var i = this.l7;
                    i.rangeChanged = delegateCombine(i.rangeChanged, runOn(this, this.ob));
                }
                if (c != null) {
                    c.bv(this);
                }
                if (d != null) {
                    d.ch(this);
                }
                if ((this.l8 != null && !this.l8.ck()) || (d == null && c != null)) {
                    this.jc(false);
                }
                break;
            case "MouseOverEnabled":
                this.mi = this.c7;
                this.jc(false);
                break;
            case HighDensityScatterSeries.$$p[8]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.md);
                    this.md = this.bw(this.n2);
                }
                break;
            case "XColumn":
                this.ok();
                if (this.l6 != null && !this.l6.ck()) {
                    this.jc(false);
                }
                this.ik();
                break;
            case HighDensityScatterSeries.$$p[10]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.mf);
                    this.mf = this.bw(this.n6);
                }
                break;
            case "YColumn":
                this.ok();
                if (this.l8 != null && !this.l8.ck()) {
                    this.jc(false);
                }
                this.ik();
                break;
            case HighDensityScatterSeries.$$p[6]:
                this.ok();
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[3]:
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[1]:
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[0]:
            case HighDensityScatterSeries.$$p[2]:
                if (!this.l3.by(this.oz, this.oy)) {
                    return;
                }
                this.m0 = this.oz.o;
                this.mz = this.oz.n;
                this.my = this.oz.m;
                this.mx = this.oy.o;
                this.mw = this.oy.n;
                this.mv = this.oy.m;
                this.m3 = this.mx - this.m0;
                this.m2 = this.mw - this.mz;
                this.m1 = this.mv - this.my;
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[4]:
                this.nk = Math.max(typeGetValue(d) - 1, 0);
                this.mg = this.nk > 0;
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[5]:
                this.jc(false);
                this.ij();
                break;
        }
    };
    HighDensityScatterSeries.prototype.mk = function (a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    HighDensityScatterSeries.prototype.mj = function (a) {
        if (typeCast(NumericXAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    HighDensityScatterSeries.prototype.ob = function (a, b) {
        this.ok();
        this.jc(false);
    };
    HighDensityScatterSeries.prototype.dw = function (a, b, c) {
        var d = _super.prototype.dw.call(this, a, b, c);
        if (this.l8 == null || this.l6 == null || this.l6.bf == null || this.l8.bf == null || this.mf == null || this.md == null || this.mf.count < 1 || this.md.count < 1 || this.mf.count != this.md.count || this.l6.j3 == this.l6.j2 || this.l8.j3 == this.l8.j2 || this.lx.isEmpty) {
            d = false;
        }
        return d;
    };
    Object.defineProperty(HighDensityScatterSeries.prototype, "nr", {
        get: function () {
            return this.nl;
        },
        set: function (a) {
            var b = this.nl;
            this.nl = a;
            this.raisePropertyChanged("ProgressiveStatus", b, this.nl);
        },
        enumerable: true,
        configurable: true
    });
    HighDensityScatterSeries.prototype.jd = function (a) {
        _super.prototype.jd.call(this, a);
        this.od(a, this.bi);
    };
    HighDensityScatterSeries.prototype.od = function (a, b) {
        if (this.cc(b)) {
            return;
        }
        if (this.l4 == null && !this.mm) {
            var c = new List$1(KDPointData.$, 2, this.md.count);
            var d = this.l6;
            var e = this.l8;
            var f = new Rect(0, 0, 0, 1, 1);
            var g = new Rect(0, 0, 0, 1, 1);
            var h = new ScalerParams(1, f, g, d.b8);
            var i = new ScalerParams(1, f, g, e.b8);
            var j_1;
            var k_1;
            var _loop_1 = function (l) {
                j_1 = d.dv(this_1.md.item(l), h);
                k_1 = e.dv(this_1.mf.item(l), i);
                if (!isNaN_(j_1) && !isNaN_(k_1) && !isInfinity(j_1) && !isInfinity(k_1)) {
                    c.add(((function () {
                        var $ret = new KDPointData();
                        $ret.a = j_1;
                        $ret.b = k_1;
                        $ret.c = l;
                        return $ret;
                    })()));
                }
            };
            var this_1 = this;
            for (var l = 0; l < this.md.count; l++) {
                _loop_1(l);
            }
            if (this.ml) {
                this.ng = 1;
                this.nh = truncate(logBase(c.count, 2)) + 3;
                this.nr = truncate(((this.ng / this.nh) * 100));
                if (this.progressiveLoadStatusChanged != null) {
                    this.progressiveLoadStatusChanged(this, new ProgressiveLoadStatusEventArgs(this.nr));
                }
                this.l4 = KDTree2D.b(c.toArray(), 1);
                var m = this.l4;
                m.progressiveThunkCompleted = delegateCombine(m.progressiveThunkCompleted, runOn(this, this.n8));
                if (!this.l4.f(this.e4)) {
                    var n = this.l4;
                    n.progressiveThunkCompleted = delegateRemove(n.progressiveThunkCompleted, runOn(this, this.n8));
                }
                this.ij();
            }
            else {
                var o = dateNow();
                this.l4 = new KDTree2D(0, c.toArray(), 1);
                var p = dateNow();
            }
        }
        if (this.ml && !this.mm) {
            this.oe(b);
        }
        else {
            this.of(b);
        }
    };
    HighDensityScatterSeries.prototype.hy = function (a, b, c, d) {
        _super.prototype.hy.call(this, a, b, c, d);
        this.ok();
        var e = false;
        if (this.l6 != null && !this.l6.ck()) {
            e = true;
        }
        if (this.l8 != null && !this.l8.ck()) {
            e = true;
        }
        if (e) {
            this.jc(false);
        }
        this.ik();
    };
    HighDensityScatterSeries.prototype.oa = function (a) {
        var b = a;
        var c = this.nj * this.ni;
        if (this.mi) {
            if (this._itemIndexes == null || this._itemIndexes.length != c) {
                this._itemIndexes = new Array(c);
                for (var d = 0; d < c; d++) {
                    this._itemIndexes[d] = 0;
                }
            }
            else {
                for (var e = 0; e < c; e++) {
                    this._itemIndexes[e] = 0;
                }
            }
        }
    };
    HighDensityScatterSeries.prototype.of = function (a) {
        var b;
        var c;
        var d = null;
        if (a.n) {
            d = this._itemIndexes;
        }
        var e = a.getViewInfo(c, b);
        c = e.p0;
        b = e.p1;
        var f = a;
        var g = this.getEffectiveViewport1(a);
        this.l9 = new ScalerParams(0, b, c, this.l5.b8, g);
        this.ma = new ScalerParams(0, b, c, this.l7.b8, g);
        this.n9(a);
        this.oa(a);
        var h = this.nj * this.ni;
        if (this.l2 == null || this.l2.length != h) {
            this.l2 = new Array(h);
            this.lz = new Array(h);
        }
        var i = this.l2.length;
        var j = this.l2;
        var k = this.lz;
        for (var l = 0; l < i; l++) {
            j[l] = 0;
            k[l] = 0;
        }
        this.nn = truncate(Math.round(this.eu));
        this.l1 = f.bv();
        var m = this.l1.length;
        var n = this.l1;
        for (var o = 0; o < m; o++) {
            n[o] = 0;
        }
        var p = this.ne;
        var q = this.nd;
        if (isNaN_(p) || isInfinity(p)) {
            p = 0;
        }
        if (isNaN_(q) || isInfinity(q)) {
            q = 50;
        }
        this.mt = Math.min(p, q);
        this.ms = Math.max(p, q);
        if (this.mt < 0) {
            this.mt = 0;
        }
        if (this.ms < 0) {
            this.ms = 0;
        }
        this.mu = this.ms - this.mt;
        if (this.mm) {
            this.oc(a);
        }
        else {
            this.om(a);
        }
        f.b6();
        if (a.n) {
            this._itemIndexes = d;
            this.l3.b7();
        }
    };
    HighDensityScatterSeries.prototype.oe = function (a) {
        if (this.l4 == null) {
            return;
        }
        this.of(a);
    };
    HighDensityScatterSeries.prototype.n8 = function (a, b) {
        var _this = this;
        this.l3.b5(function () {
            if (_this.ng < _this.nh - 1) {
                _this.ng++;
            }
            _this.nr = truncate(((_this.ng / _this.nh) * 100));
            if (_this.progressiveLoadStatusChanged != null) {
                _this.progressiveLoadStatusChanged(_this, new ProgressiveLoadStatusEventArgs(_this.nr));
            }
            _this.oe(_this.bi);
            _this.ij();
            if (_this.l4 != null) {
                if (!_this.l4.f(_this.e4)) {
                    var c = _this.l4;
                    c.progressiveThunkCompleted = delegateRemove(c.progressiveThunkCompleted, runOn(_this, _this.n8));
                    _this.nr = 100;
                    if (_this.progressiveLoadStatusChanged != null) {
                        _this.progressiveLoadStatusChanged(_this, new ProgressiveLoadStatusEventArgs(_this.nr));
                    }
                }
            }
        });
    };
    HighDensityScatterSeries.prototype.oc = function (a) {
        var b = arrayShallowClone(this.md.asArray());
        this.l6.gm(boxArray$1(b), 0, b.length, this.l9);
        var c = arrayShallowClone(this.mf.asArray());
        this.nm = 0;
        var d = this.nk;
        var e = this.mg;
        var f = this.l2.length;
        this.l8.gm(boxArray$1(c), 0, c.length, this.ma);
        var g = 255 << 24 | truncate(this.m0) << 16 | truncate(this.mz) << 8 | truncate(this.my);
        for (var h = 0; h < b.length; h++) {
            var i = truncate(b[h]);
            var j = truncate(c[h]);
            i = truncate(Math.floor(i));
            j = truncate(Math.floor(j));
            if (i < 0 || i >= this.nj || j < 0 || j >= this.ni) {
                continue;
            }
            if (e) {
                var k = i - d;
                var l = i + d;
                l = l > (this.nj - 1) ? (this.nj - 1) : l;
                k = k < 0 ? 0 : k;
                var m = j + d;
                var n = j - d;
                for (var o = k; o <= l; o++) {
                    for (var p = n; p <= m; p++) {
                        var q = (p * this.nj) + o;
                        if (q < 0 || q > f - 1) {
                            continue;
                        }
                        this.l2[q] = this.l2[q] + 1;
                        this.lz[q] = 1;
                        this.nm++;
                        if (this.mi) {
                            this._itemIndexes[q] = h + 1;
                        }
                    }
                }
            }
            else {
                var r = (j * this.nj) + i;
                this.l2[r] = this.l2[r] + 1;
                this.lz[r] = 1;
                this.nm++;
                if (this.mi) {
                    this._itemIndexes[r] = h + 1;
                }
            }
        }
        this.og();
    };
    HighDensityScatterSeries.prototype.om = function (a) {
        var b;
        var c;
        var d = a.getViewInfo(b, c);
        b = d.p0;
        c = d.p1;
        if (this.l4 == null) {
            return;
        }
        if (this.mn == null) {
            this.mn = new List$1(KDTreeNode2D.$, 2, truncate(Math.round(b.width * b.height)));
        }
        else {
            this.mn.clear();
        }
        var e = c.left;
        var f = c.right;
        var g = c.top;
        var h = c.bottom;
        this.na = c.top;
        this.m8 = c.height;
        this.m9 = c.left;
        this.nb = c.width;
        this.m6 = b.top;
        this.m4 = b.height;
        this.m5 = b.left;
        this.m7 = b.width;
        var i = this.getEffectiveViewport1(a);
        this.mp = (i.left - this.m5) / this.m7;
        this.mq = (i.top - this.m6) / this.m4;
        var j = (i.right - this.m5) / this.m7;
        this.mr = j - this.mp;
        var k = (i.bottom - this.m6) / this.m4;
        this.mo = k - this.mq;
        if (this.mh) {
            e = (e - this.mp) / this.mr;
            f = (f - this.mp) / this.mr;
            g = (g - this.mq) / this.mo;
            h = (h - this.mq) / this.mo;
        }
        var l = this.eu / b.width * c.width;
        var m = this.eu / b.height * c.height;
        var n = Math.min(l, m);
        var o = ((function () {
            var $ret = new SearchArgs();
            $ret.c = e;
            $ret.a = f;
            $ret.d = g;
            $ret.b = h;
            $ret.e = l;
            $ret.f = m;
            $ret.g = 0x7FFFFFFF;
            return $ret;
        })());
        this.l4.t(this.mn, o, 0, 1, 0, 1);
        var p;
        this.nm = 0;
        for (var q = 0; q < this.mn.count; q++) {
            p = this.mn._inner[q];
            this.oh(p);
        }
        this.og();
    };
    HighDensityScatterSeries.prototype.oh = function (a) {
        if (a.g) {
            return;
        }
        var b = a.e != null && a.e.a;
        var c = a.a == null ? 0 : a.a.length;
        var d = (a.h - c) + 1;
        if (b && d > 0) {
        }
        else {
            d = 1;
        }
        this.oj(a.b, d, b, a.e);
        if (c > 0 && !b) {
            var e = void 0;
            for (var f = 0; f < c; f++) {
                e = a.a[f];
                this.oj(e, d, false, a.e);
            }
        }
        if (a.e != null) {
            a.e.a = false;
        }
    };
    HighDensityScatterSeries.prototype.c9 = function (a) {
        var b = this.bs != null ? this.bs.mq : Rect.empty;
        var c = this.bs != null ? this.bs.m3 : Rect.empty;
        var d = new Rect(0, 0, 0, 1, 1);
        var e = this.lr(this.bi);
        var f = new ScalerParams(0, d, d, this.l6.b8, e);
        var g = new ScalerParams(0, d, d, this.l8.b8, e);
        var h = !b.isEmpty && !c.isEmpty && this.bz != null ? this.bz.indexOf(a) : -1;
        var i = this.l6 != null && this.md != null && h < this.md.count ? this.l6.dv(this.md.item(h), f) : NaN;
        var j = this.l8 != null && this.mf != null && h < this.mf.count ? this.l8.dv(this.mf.item(h), g) : NaN;
        if (!isNaN_(i)) {
            if (i < b.left + 0.1 * b.width) {
                i = i + 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
            if (i > b.right - 0.1 * b.width) {
                i = i - 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
        }
        if (!isNaN_(j)) {
            if (j < b.top + 0.1 * b.height) {
                j = j + 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
            if (j > b.bottom - 0.1 * b.height) {
                j = j - 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
        }
        if (this.bn != null) {
            this.bn.ap(this.bs, b, true);
        }
        return h >= 0;
    };
    HighDensityScatterSeries.prototype.oj = function (a, b, c, d) {
        var e = this.np(b);
        var f = a.c;
        var g = this.nk;
        if (c) {
            var h = d.d;
            var i = d.b;
            var j = d.e;
            var k = d.c;
            if (this.mh) {
                h = this.mp + this.mr * h;
                i = this.mp + this.mr * i;
                j = this.mq + this.mo * j;
                k = this.mq + this.mo * k;
            }
            var l = truncate((this.m5 + this.m7 * (h - this.m9) / this.nb));
            var m = truncate((this.m5 + this.m7 * (i - this.m9) / this.nb));
            var n = truncate((this.m6 + this.m4 * (j - this.na) / this.m8));
            var o = truncate((this.m6 + this.m4 * (k - this.na) / this.m8));
            l = truncate(Math.floor(l));
            m = truncate(Math.floor(m));
            n = truncate(Math.floor(n));
            o = truncate(Math.floor(o));
            if (this.mg) {
                l -= g;
                m += g;
                n -= g;
                o += g;
            }
            var p = ((m - l) + 1) * ((o - n) + 1);
            var q = b / p;
            var r = q;
            for (var s = l; s <= m; s++) {
                for (var t = n; t <= o; t++) {
                    this.oi(f, s, t, q, r);
                }
            }
        }
        else {
            var u = a.a;
            var v = a.b;
            if (this.mh) {
                u = this.mp + this.mr * u;
                v = this.mq + this.mo * v;
            }
            var w = truncate((this.m5 + this.m7 * (u - this.m9) / this.nb));
            var x = truncate((this.m6 + this.m4 * (v - this.na) / this.m8));
            w = truncate(Math.floor(w));
            x = truncate(Math.floor(x));
            if (this.mg) {
                var y = w - g;
                var z = w + g;
                var aa = x + g;
                var ab = x - g;
                var ac = ((z - y) + 1) * ((aa - ab) + 1);
                var ad = b / ac;
                var ae = ad;
                for (var af = y; af <= z; af++) {
                    for (var ag = ab; ag <= aa; ag++) {
                        this.oi(f, af, ag, ad, ae);
                    }
                }
            }
            else {
                this.oi(f, w, x, b, 1);
            }
        }
    };
    HighDensityScatterSeries.prototype.no = function (a, b) {
        return truncate((Math.floor(255 * b))) << 24 | truncate(Math.floor(b * (this.m0 + a * this.m3))) << 16 | truncate(Math.floor(b * (this.mz + a * this.m2))) << 8 | truncate(Math.floor(b * (this.my + a * this.m1)));
    };
    HighDensityScatterSeries.prototype.np = function (a) {
        return 255 << 24 | truncate(Math.floor((this.m0 + a * this.m3))) << 16 | truncate(Math.floor((this.mz + a * this.m2))) << 8 | truncate(Math.floor((this.my + a * this.m1)));
    };
    HighDensityScatterSeries.prototype.nc = function (a) {
        return (a >> 24 & 255) / 255;
    };
    HighDensityScatterSeries.prototype.oi = function (a, b, c, d, e) {
        if (b < 0 || b >= this.nj || c < 0 || c >= this.ni) {
            return;
        }
        var f = (c * this.nj) + b;
        this.l2[f] = this.l2[f] + d;
        this.lz[f] = this.lz[f] + e;
        if (this.mi) {
            this._itemIndexes[f] = a + 1;
        }
        this.nm += truncate(d);
    };
    HighDensityScatterSeries.prototype.og = function () {
        var a = 0;
        var b = 0;
        var c = this.mu;
        var d = this.mt;
        var e = this.ms;
        var f = this.l2.length;
        var g = this.l2;
        var h = this.lz;
        var i = this.l1;
        var j = NaN;
        var k = NaN;
        var l = 0;
        var m = 0;
        for (var n = 0; n < f; n++) {
            a = g[n];
            if (a != 0) {
                b = h[n];
                if (a >= e) {
                    a = 1;
                }
                else if (a <= d) {
                    a = 0;
                }
                else {
                    a = (a - d) / (c);
                }
                if (b > 1) {
                    b = 1;
                }
                if (b < 0.2) {
                    b = 0.2;
                }
                if (j != a || k != b) {
                    l = this.no(a, b);
                    j = a;
                    k = b;
                }
                i[m] = l >> 16 & 255;
                i[m + 1] = l >> 8 & 255;
                i[m + 2] = l & 255;
                i[m + 3] = l >> 24 & 255;
            }
            m += 4;
        }
    };
    HighDensityScatterSeries.prototype.n9 = function (a) {
        var b = a;
        b.b3();
    };
    HighDensityScatterSeries.prototype.bf = function () {
        return new HighDensityScatterSeriesView(this);
    };
    HighDensityScatterSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.l3 = a;
    };
    HighDensityScatterSeries.prototype.an = function (a) {
        if (a != null && a == this.l6 && this.md != null) {
            return new AxisRange(this.md.minimum, this.md.maximum);
        }
        if (a != null && a == this.l8 && this.mf != null) {
            return new AxisRange(this.mf.minimum, this.mf.maximum);
        }
        return null;
    };
    HighDensityScatterSeries.prototype.jr = function (a, b) {
        if (this.bs != null) {
            var c = this.getEffectiveViewport1(this.bi);
            this.mh = !c.isEmpty;
        }
        this.jc(false);
    };
    HighDensityScatterSeries.prototype.js = function (a, b) {
        this.jc(false);
    };
    HighDensityScatterSeries.prototype.fp = function (a) {
        if (!this.mi || this._itemIndexes == null || this.bs == null || this.bz == null || this._itemIndexes.length != (this.nj * this.ni)) {
            return null;
        }
        var b = this.bs.mq;
        var c = (a.x - b.left) / b.width;
        var d = (a.y - b.top) / b.height;
        var e = truncate(Math.round(this.lx.left + (this.lx.width * c)));
        var f = truncate(Math.round(this.lx.top + (this.lx.height * d)));
        var g = this.ns(e, f);
        if (g < 0 || g > this._itemIndexes.length - 1) {
            return null;
        }
        var h = this._itemIndexes[g] - 1;
        if (h < 0 || h > this.bz.count) {
            return null;
        }
        return this.bz.item(h);
    };
    HighDensityScatterSeries.prototype.ns = function (a, b) {
        var c = (this.nj * b) + a;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        var d = 1;
        c = (this.nj * (b - d)) + a;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b + d)) + a;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b)) + a - 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b)) + a + 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b - d)) + a - 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b + d)) + a + 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b - d)) + a + 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b + d)) + a - 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        return 0;
    };
    HighDensityScatterSeries.prototype.ol = function (a, b, c) {
        this.l1 = a;
        this.nj = b;
        this.ni = c;
    };
    HighDensityScatterSeries.prototype.get_dq = function () {
        return true;
    };
    HighDensityScatterSeries.prototype.set_dq = function (a) {
        _super.prototype.set_dq.call(this, a);
    };
    HighDensityScatterSeries.prototype.hv = function (a, b) {
        _super.prototype.hv.call(this, a, b);
        this.l3.b4();
        if (this.bh != null) {
            this.bh.b4();
        }
        this.ij();
    };
    HighDensityScatterSeries.prototype.je = function (a, b) {
        _super.prototype.je.call(this, a, b);
        if (!this.dl) {
            this.bi.az(b);
            return;
        }
        this.bi.az(b);
        if (this.cc(this.bh)) {
            return;
        }
        var c = typeCast(HighDensityScatterSeriesView.$, this.bh);
        var d = new ScatterFrame();
        this.od(false, c);
        this.dl = false;
    };
    HighDensityScatterSeries.prototype.jb = function (a, b, c, d, e) {
        _super.prototype.jb.call(this, a, b, c, d, e);
        var f = this.a0.a$h.item(d);
        var g = f;
        f.ay(c);
        if (this.cc(g)) {
            return;
        }
        if (this.mb == null) {
            this.mb = new ScatterFrame();
        }
        this.od(false, f);
    };
    HighDensityScatterSeries.prototype.ht = function () {
        _super.prototype.ht.call(this);
        this.l6 = null;
        this.l8 = null;
    };
    HighDensityScatterSeries.$t = markType(HighDensityScatterSeries, 'HighDensityScatterSeries', Series.$);
    return HighDensityScatterSeries;
}(Series));
export { HighDensityScatterSeries };
Type.dep(DependencyProperty, PropertyMetadata, HighDensityScatterSeries, 'raisePropertyChanged', ['HeatMaximumColor:oy:on', [Color.$, null], 'HeatMaximum:nd:oo', [1, 50], 'HeatMinimumColor:oz:op', [Color.$, null], 'HeatMinimum:ne:oq', [1, 0], 'PointExtent:nq:or', [1, 1], 'ProgressiveLoad:ml:os', [0, true], 'UseBruteForce:mm:ot', [0, false], 'XAxis:l6:ou', [NumericXAxis.$, null], 'XMemberPath:n2:ov', [2, null], 'YAxis:l8:ow', [NumericYAxis.$, null], 'YMemberPath:n6:ox', [2, null]]);
/**
 * @hidden
 */
var HighDensityScatterSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(HighDensityScatterSeriesView, _super);
    function HighDensityScatterSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.bw = null;
        _this.bu = null;
        _this.b2 = 0;
        _this.b1 = 0;
        _this.bz = null;
        _this.b0 = null;
        _this.bx = null;
        _this.bw = a;
        return _this;
    }
    HighDensityScatterSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        if (!this.n) {
            this.e.eu = 4;
            this.bw.oz = Color.u(255, 0, 0, 0);
            this.bw.oy = Color.u(255, 255, 0, 0);
        }
    };
    HighDensityScatterSeriesView.prototype.b7 = function () {
        this.bw.ol(this.bu, this.b2, this.b1);
    };
    HighDensityScatterSeriesView.prototype.b3 = function () {
        if (this.bz == null) {
            this.bz = this.e.e4.createElement("canvas");
            var a = this.e.e4.get2DCanvasContext(this.bz);
            this.bx = new RenderingContext(new CanvasViewRenderer(), a);
        }
        var b;
        if (this.m) {
            b = this.br;
        }
        else {
            b = this.e.bs.mv();
        }
        var c = truncate(Math.round(b.left));
        var d = truncate(Math.round(b.top));
        var e = truncate(Math.round(b.width));
        var f = truncate(Math.round(b.height));
        var g = this.b2;
        var h = this.b1;
        this.b2 = e;
        this.b1 = f;
        if (g != this.b2 || h != this.b1 || this.bu == null) {
            this.bz.setAttribute("width", e.toString());
            this.bz.setAttribute("height", f.toString());
            this.b0 = this.bx.g().getImageData(0, 0, e, f);
            var imageDataArray_ = this.b0.data;
            this.bu = imageDataArray_;
        }
        this.bw.ol(this.bu, this.b2, this.b1);
    };
    HighDensityScatterSeriesView.prototype.bv = function () {
        return this.bu;
    };
    HighDensityScatterSeriesView.prototype.b6 = function () {
        var a = truncate(Math.round(this.br.left));
        var b = truncate(Math.round(this.br.left));
        var c = truncate(Math.round(this.br.width));
        var d = truncate(Math.round(this.br.height));
        this.bx.g().putImageData(this.b0, 0, 0);
        this.an();
    };
    HighDensityScatterSeriesView.prototype.b5 = function (a) {
        if (this.e.e4 != null) {
            this.e.e4.setTimeout(a, 0);
        }
        else {
            window.setTimeout(a, 0);
        }
    };
    HighDensityScatterSeriesView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        if (b) {
            if (!this.bw.c7) {
                return;
            }
            var c = new Rectangle();
            c.n = this.br.left;
            c.o = this.br.top;
            c.width = this.br.width;
            c.height = this.br.height;
            c._fill = this.bl();
            a.v(c);
            return;
        }
        if (this.bz == null) {
            return;
        }
        var d = truncate(Math.round(this.br.left));
        var e = truncate(Math.round(this.br.top));
        var f = truncate(Math.round(this.br.width));
        var g = truncate(Math.round(this.br.height));
        this.h.n(this.bz.getNativeElement(), this.e._opacity, d, e, f, g, d, e, f, g);
    };
    HighDensityScatterSeriesView.prototype.by = function (a, b) {
        return Color.e(Nullable$1.toNullable(Color.$, a), Nullable$1.toNullable(Color.$, null)) && Color.e(Nullable$1.toNullable(Color.$, b), Nullable$1.toNullable(Color.$, null));
    };
    HighDensityScatterSeriesView.prototype.b4 = function () {
        if (this.bu != null) {
            for (var a = 0; a < this.bu.length; a++) {
                this.bu[a] = 0;
            }
            this.b6();
        }
    };
    HighDensityScatterSeriesView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        if (this.bu == null) {
            this.e.a2.exportSeriesPixelsData(a, null, this.b2, this.b1);
            return;
        }
        var b = new Array(intDivide(this.bu.length, 4));
        var c;
        for (var d = 0; d < intDivide(this.bu.length, 4); d++) {
            c = d * 4;
            b[d] = this.bu[c + 3] << 24 | this.bu[c] << 16 | this.bu[c + 1] << 8 | this.bu[c + 2];
        }
        this.e.a2.exportSeriesPixelsData(a, b, this.b2, this.b1);
    };
    HighDensityScatterSeriesView.$t = markType(HighDensityScatterSeriesView, 'HighDensityScatterSeriesView', SeriesView.$);
    return HighDensityScatterSeriesView;
}(SeriesView));
export { HighDensityScatterSeriesView };
