// igniteui-react-charts/ES5/igr-scatter-triangulation-series.js
import * as tslib_1 from "tslib";
import { delegateCombine, delegateRemove } from "igniteui-react-core/ES5/type";
import { IgrNumericXAxis } from "./igr-numeric-x-axis";
import { IgrNumericYAxis } from "./igr-numeric-y-axis";
import { TriangulationStatusEventArgs } from "igniteui-react-core/ES5/igr-triangulation-status-event-args";
import { IgrSeries } from "./igr-series";
/**
 * Base class for series which triangulate XY data prior to rendering.
*/
var IgrScatterTriangulationSeries = /** @class */ (function (_super) {
    tslib_1.__extends(IgrScatterTriangulationSeries, _super);
    function IgrScatterTriangulationSeries(props) {
        var _this = _super.call(this, props) || this;
        _this._xAxisName = null;
        _this._yAxisName = null;
        _this._triangulationStatusChanged = null;
        _this._triangulationStatusChanged_wrapped = null;
        return _this;
    }
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "xMemberPath", {
        /**
         * The name of the property from which to extract the X-coordinate for each item in the ItemsSource.
        */
        get: function () {
            return this.i.mx;
        },
        set: function (v) {
            this.i.mx = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "yMemberPath", {
        /**
         * The name of the property from which to extract the Y-coordinate for each item in the ItemsSource.
        */
        get: function () {
            return this.i.m1;
        },
        set: function (v) {
            this.i.m1 = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "xAxis", {
        /**
         * The X-Axis for this series.
        */
        get: function () {
            if (this.i.lz == null) {
                return null;
            }
            if (!this.i.lz.externalObject) {
                var e = IgrNumericXAxis._createFromInternal(this.i.lz);
                if (e) {
                    e._implementation = this.i.lz;
                }
                this.i.lz.externalObject = e;
            }
            return this.i.lz.externalObject;
        },
        set: function (v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.lz = null : this.i.lz = v.i;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "xAxisName", {
        /**
        * Gets or sets the name to use to resolve xAxis from markup.
        */
        get: function () {
            return this._xAxisName;
        },
        set: function (v) {
            this._xAxisName = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "yAxis", {
        /**
         * The Y-Axis for this series.
        */
        get: function () {
            if (this.i.l0 == null) {
                return null;
            }
            if (!this.i.l0.externalObject) {
                var e = IgrNumericYAxis._createFromInternal(this.i.l0);
                if (e) {
                    e._implementation = this.i.l0;
                }
                this.i.l0.externalObject = e;
            }
            return this.i.l0.externalObject;
        },
        set: function (v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.l0 = null : this.i.l0 = v.i;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "yAxisName", {
        /**
        * Gets or sets the name to use to resolve yAxis from markup.
        */
        get: function () {
            return this._yAxisName;
        },
        set: function (v) {
            this._yAxisName = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "trianglesSource", {
        /**
         * The source of triangulation data.
         * This property is optional.  If it is left as null, the triangulation will be created based on the items in the ItemsSource.  Triangulation is a demanding operation, so the runtime performance will be better when specifying a TriangulationSource, especially when a large number of data items are present.
        */
        get: function () {
            return this.i.trianglesSource;
        },
        set: function (v) {
            this.i.trianglesSource = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "triangleVertexMemberPath1", {
        /**
         * The name of the property of the TrianglesSource items which, for each triangle, contains the index of the first vertex point in the ItemsSource.
        */
        get: function () {
            return this.i.mp;
        },
        set: function (v) {
            this.i.mp = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "triangleVertexMemberPath2", {
        /**
         * The name of the property of the TrianglesSource items which, for each triangle, contains the index of the second vertex point in the ItemsSource.
        */
        get: function () {
            return this.i.mr;
        },
        set: function (v) {
            this.i.mr = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "triangleVertexMemberPath3", {
        /**
         * The name of the property of the TrianglesSource items which, for each triangle, contains the index of the third vertex point in the ItemsSource.
        */
        get: function () {
            return this.i.mt;
        },
        set: function (v) {
            this.i.mt = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "isShape", {
        /**
         * Gets whether or not this series is a shape series
        */
        get: function () {
            return this.i.c5;
        },
        enumerable: true,
        configurable: true
    });
    IgrScatterTriangulationSeries.prototype.bindAxes = function (axes) {
        _super.prototype.bindAxes.call(this, axes);
        for (var i = 0; i < axes.length; i++) {
            if (this.xAxisName && this.xAxisName.length > 0 &&
                axes[i].name == this.xAxisName) {
                this.xAxis = axes[i];
            }
        }
        for (var i = 0; i < axes.length; i++) {
            if (this.yAxisName && this.yAxisName.length > 0 &&
                axes[i].name == this.yAxisName) {
                this.yAxis = axes[i];
            }
        }
    };
    IgrScatterTriangulationSeries.prototype._styling = function (container, component, parent) {
        _super.prototype._styling.call(this, container, component, parent);
        this._inStyling = true;
        if (this.xAxis && this.xAxis._styling) {
            this.xAxis._styling(container, component, this);
        }
        if (this.yAxis && this.yAxis._styling) {
            this.yAxis._styling(container, component, this);
        }
        this._inStyling = false;
    };
    IgrScatterTriangulationSeries.prototype.getItemValue = function (item, memberPathName) {
        var iv = this.i.fs(item, memberPathName);
        return (iv);
    };
    /**
     * Determine if object can be used as YAxis
    
    * @param axis  * The object to check
    */
    IgrScatterTriangulationSeries.prototype.canUseAsYAxis = function (axis) {
        var iv = this.i.mh(axis);
        return (iv);
    };
    /**
     * Determine if object can be used as XAxis
    
    * @param axis  * The object to check
    */
    IgrScatterTriangulationSeries.prototype.canUseAsXAxis = function (axis) {
        var iv = this.i.mg(axis);
        return (iv);
    };
    Object.defineProperty(IgrScatterTriangulationSeries.prototype, "triangulationStatusChanged", {
        /**
         * Raised when the status of an ongoing Triangulation has changed.
        */
        get: function () {
            return this._triangulationStatusChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._triangulationStatusChanged_wrapped !== null) {
                this.i.triangulationStatusChanged = delegateRemove(this.i.triangulationStatusChanged, this._triangulationStatusChanged_wrapped);
                this._triangulationStatusChanged_wrapped = null;
                this._triangulationStatusChanged = null;
            }
            this._triangulationStatusChanged = ev;
            this._triangulationStatusChanged_wrapped = function (o, e) {
                var outerArgs = new TriangulationStatusEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeTriangulationStatusChanged) {
                    _this.beforeTriangulationStatusChanged(_this, outerArgs);
                }
                if (_this._triangulationStatusChanged) {
                    _this._triangulationStatusChanged(_this, outerArgs);
                }
            };
            this.i.triangulationStatusChanged = delegateCombine(this.i.triangulationStatusChanged, this._triangulationStatusChanged_wrapped);
        },
        enumerable: true,
        configurable: true
    });
    return IgrScatterTriangulationSeries;
}(IgrSeries));
export { IgrScatterTriangulationSeries };
