// igniteui-react-charts/ES5/igr-axis-annotation.js
import { AxisAnnotation as AxisAnnotation_internal } from "./AxisAnnotation";
import { brushToString, stringToBrush } from "igniteui-react-core/ES5/componentUtil";
/**
 * Represents an axis annotation
*/
var AxisAnnotation = /** @class */ (function () {
    function AxisAnnotation() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    AxisAnnotation.prototype.createImplementation = function () {
        return new AxisAnnotation_internal();
    };
    Object.defineProperty(AxisAnnotation.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: true,
        configurable: true
    });
    AxisAnnotation.prototype.onImplementationCreated = function () {
    };
    AxisAnnotation.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    };
    Object.defineProperty(AxisAnnotation.prototype, "value", {
        /**
         * Gets or sets the value at which this annotation is displayed on the axis.
        */
        get: function () {
            return this.i.value;
        },
        set: function (v) {
            this.i.value = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "text", {
        /**
         * Gets or sets the text that should be displayed for the annotation. Leave null to display the value automatically.
        */
        get: function () {
            return this.i.text;
        },
        set: function (v) {
            this.i.text = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "formatLabel", {
        /**
         * Sets or gets a function which takes an object that produces a formatted label for the value of this annotation.
        */
        get: function () {
            return this.i.formatLabel;
        },
        set: function (v) {
            this.i.formatLabel = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "textColor", {
        /**
         * Gets or sets the color to use for the text in the annotation.
        */
        get: function () {
            return brushToString(this.i.textColor);
        },
        set: function (v) {
            this.i.textColor = stringToBrush(v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "background", {
        /**
         * Gets or sets the background to use for tha axis annotation box.
        */
        get: function () {
            return brushToString(this.i.background);
        },
        set: function (v) {
            this.i.background = stringToBrush(v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "outline", {
        /**
         * Gets or sets the Outline to use for tha axis annotation box.
        */
        get: function () {
            return brushToString(this.i.outline);
        },
        set: function (v) {
            this.i.outline = stringToBrush(v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "strokeThickness", {
        /**
         * Gets or sets the border stroke thickness to use for tha axis annotation box.
        */
        get: function () {
            return this.i.strokeThickness;
        },
        set: function (v) {
            this.i.strokeThickness = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "backgroundPaddingLeft", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.backgroundPaddingLeft;
        },
        set: function (v) {
            this.i.backgroundPaddingLeft = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "backgroundPaddingTop", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.backgroundPaddingTop;
        },
        set: function (v) {
            this.i.backgroundPaddingTop = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "backgroundPaddingRight", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.backgroundPaddingRight;
        },
        set: function (v) {
            this.i.backgroundPaddingRight = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AxisAnnotation.prototype, "backgroundPaddingBottom", {
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.backgroundPaddingBottom;
        },
        set: function (v) {
            this.i.backgroundPaddingBottom = +v;
        },
        enumerable: true,
        configurable: true
    });
    AxisAnnotation.prototype.resetCachedExtent = function () {
        this.i.resetCachedExtent();
    };
    AxisAnnotation.prototype.resolveLabelValue = function () {
        var iv = this.i.resolveLabelValue();
        return (iv);
    };
    return AxisAnnotation;
}());
export { AxisAnnotation };
