// igniteui-react-charts/ES5/LegendBaseViewManager_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, typeCast, runOn, fromEnum, markType, INotifyPropertyChanged_$type, PropertyChangedEventArgs } from "igniteui-react-core/ES5/type";
import { DomWrapper_$type } from "igniteui-react-core/ES5/dom";
import { Dictionary$2 } from "igniteui-react-core/ES5/Dictionary$2";
import { RenderingContext } from "igniteui-react-core/ES5/RenderingContext";
import { ContentControl } from "igniteui-react-core/ES5/ContentControl";
import { ILegendContext_$type } from "igniteui-react-core/ES5/ILegendContext";
import { DataTemplatePassInfo } from "igniteui-react-core/ES5/DataTemplatePassInfo";
import { DataTemplateMeasureInfo } from "igniteui-react-core/ES5/DataTemplateMeasureInfo";
import { DataTemplateRenderInfo } from "igniteui-react-core/ES5/DataTemplateRenderInfo";
import { CanvasViewRenderer } from "igniteui-react-core/ES5/CanvasViewRenderer";
import { Size } from "igniteui-react-core/ES5/Size";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { FontUtil } from "igniteui-react-core/ES5/FontUtil";
import { LegendItemHelper } from "igniteui-react-core/ES5/LegendItemHelper";
import { isNaN_ } from "igniteui-react-core/ES5/number";
import { IChartLegend_$type } from "igniteui-react-core/ES5/IChartLegend";
import { ObservableCollection$1 } from "igniteui-react-core/ES5/ObservableCollection$1";
import { UIElement } from "igniteui-react-core/ES5/UIElement";
import { ILegendSeries_$type } from "igniteui-react-core/ES5/ILegendSeries";
import { PropertyUpdatedEventArgs } from "igniteui-react-core/ES5/PropertyUpdatedEventArgs";
import { LegendMouseButtonEventArgs } from "./LegendMouseButtonEventArgs";
import { LegendMouseEventArgs } from "./LegendMouseEventArgs";
import { LegendVisualData } from "./LegendVisualData";
import { MouseButtonEventArgs } from "igniteui-react-core/ES5/MouseButtonEventArgs";
import { MouseEventArgs } from "igniteui-react-core/ES5/MouseEventArgs";
import { RectData } from "igniteui-react-core/ES5/RectData";
import { AppearanceHelper } from "igniteui-react-core/ES5/AppearanceHelper";
import { LegendItemVisualData } from "./LegendItemVisualData";
import { stringIsNullOrEmpty } from "igniteui-react-core/ES5/string";
/**
 * @hidden
 */
var LegendBaseViewManager = /** @class */ (function (_super) {
    tslib_1.__extends(LegendBaseViewManager, _super);
    function LegendBaseViewManager(a) {
        var _this = _super.call(this) || this;
        _this.a = null;
        _this.g = null;
        _this.m = null;
        _this.h = null;
        _this.d = false;
        _this.f = null;
        _this.n = null;
        _this.c = null;
        _this.a = a;
        _this.f = new Dictionary$2(Base.$, DomWrapper_$type, 0);
        _this.q();
        return _this;
    }
    LegendBaseViewManager.prototype.q = function () {
        var _this = this;
        if (this.g == null) {
            this.g = function (a) {
                var b = LegendItemHelper.b;
                if (window.devicePixelRatio != undefined && (isNaN_(b) || window.devicePixelRatio != b)) {
                    LegendItemHelper.b = window.devicePixelRatio;
                    _this.a.aj();
                }
            };
            window.addEventListener("resize", this.g, false);
        }
        this.g(null);
    };
    LegendBaseViewManager.prototype.t = function (a) {
        if (a == null) {
            if (this.g != null) {
                window.removeEventListener("resize", this.g, false);
                this.g = null;
            }
            this.m = null;
            this.h = null;
            return;
        }
        this.h = a;
        this.m = this.j();
        this.m.addClass(this.a.s);
        this.h.append(this.m);
    };
    LegendBaseViewManager.prototype.j = function () {
        return this.h.createElement("table");
    };
    LegendBaseViewManager.prototype.i = function () {
        var a = this.h.createElement("tr");
        a.addClass(this.a.t);
        var b = this.a.k;
        if (b) {
            a.setStyleProperty("display", "block");
            a.setStyleProperty("float", "left");
            a.setStyleProperty("margin", "0px");
            a.setStyleProperty("margin-right", "8px");
        }
        return a;
    };
    LegendBaseViewManager.prototype.k = function (a) {
        return this.l(a, this.m);
    };
    LegendBaseViewManager.prototype.l = function (a, b) {
        var c = typeCast(ContentControl.$, a);
        var d = this.i();
        if (c != null) {
            var e = typeCast(ILegendContext_$type, c.content);
            if (e != null && c.ah != null) {
                var f = new DataTemplatePassInfo();
                f.context = d;
                f.passID = "LegendItem";
                var g = new DataTemplateMeasureInfo();
                g.passInfo = f;
                g.width = NaN;
                g.height = NaN;
                g.data = e;
                g.context = d;
                var h = new DataTemplateRenderInfo();
                h.passInfo = f;
                if (c.ah.measure != null) {
                    c.ah.measure(g);
                }
                h.context = this.h.getSubRenderer(d);
                h.availableWidth = g.width;
                h.availableHeight = g.height;
                h.data = e;
                h.xPosition = 0;
                h.yPosition = 0;
                c.ah.render(h);
                var i = d.findByClass(".ui-legend-item-badge");
                for (var j = 0; j < i.length; j++) {
                    i[j].removeClass("ui-legend-item-badge").addClass(this.a.r);
                }
                i = d.findByClass(".ui-legend-item-text");
                for (var k = 0; k < i.length; k++) {
                    if (this.a.k) {
                        i[k].setStyleProperty("white-space", "nowrap");
                    }
                    i[k].removeClass("ui-legend-item-text").addClass(this.a.u);
                }
            }
        }
        return d;
    };
    LegendBaseViewManager.prototype.p = function (a, b) {
        var c = this.l(a, this.m);
        c.listen("mouseup", runOn(this.a, this.a.ae));
        c.listen("mousedown", runOn(this.a, this.a.ab));
        c.listen("mouseleave", runOn(this.a, this.a.ac));
        c.listen("mousemove", runOn(this.a, this.a.ad));
        this.f.addItem(a, c);
        if ((this.m.getChildCount() == 0) || (b >= this.m.getChildCount())) {
            this.m.append(c);
        }
        else {
            var d = this.m.getChildAt(b);
            d.before(c);
        }
    };
    LegendBaseViewManager.prototype.v = function (a) {
    };
    LegendBaseViewManager.prototype.w = function (a) {
        var _this = this;
        var b;
        if (((function () { var c = _this.f.tryGetValue(a, b); b = c.p1; return c.ret; })())) {
            b.unlistenAll();
            b.remove();
            this.f.removeItem(a);
        }
    };
    LegendBaseViewManager.prototype.o = function (visual_) {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(fromEnum(this.f)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var pair_ = _c.value;
                if (pair_.value.getNativeElement() === visual_) {
                    return pair_.key;
                }
                ;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return null;
    };
    LegendBaseViewManager.prototype.e = function (a) {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(fromEnum(this.f.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                var c = typeCast(ContentControl.$, b);
                if (c != null && c.content != null) {
                    var d = typeCast(ILegendContext_$type, c.content);
                    if (d != null && d.itemLabel == a.itemLabel && d.legendLabel == a.legendLabel && d.series == a.series && d.itemBrush == a.itemBrush) {
                        return true;
                    }
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return false;
    };
    LegendBaseViewManager.prototype.b = function (a) {
        var b = this.h.rootWrapper.width();
        var c = this.h.rootWrapper.height();
        this.m.remove();
        var d = this.h.createElement("div");
        d.setStyleProperty("position", "relative");
        this.n = this.h.createElement("<canvas style=\"position : absolute; top : 0; left : 0\" />");
        this.n.setStyleProperty("position", "absolute");
        this.n.setStyleProperty("top", "0px");
        this.n.setStyleProperty("left", "0px");
        this.h.append(d);
        d.append(this.n);
        this.c = new RenderingContext(new CanvasViewRenderer(), this.h.get2DCanvasContext(this.n));
        this.n.setAttribute("width", b.toString());
        this.n.setAttribute("height", c.toString());
        return this.c;
    };
    LegendBaseViewManager.prototype.aa = function () {
        return new Size(1, this.h.rootWrapper.width(), Math.max(10, this.h.rootWrapper.height() - 15));
    };
    LegendBaseViewManager.prototype.x = function (a, b, c, d) {
        var e_3, _a;
        var e = a.g();
        e.beginPath();
        e.moveTo(b.aj._inner[0].x, b.aj._inner[0].y);
        for (var f = 1; f < b.aj.count; f++) {
            e.lineTo(b.aj._inner[f].x, b.aj._inner[f].y);
        }
        e.lineTo(b.aj._inner[0].x, b.aj._inner[0].y);
        var g = e.createLinearGradient(d.left, d.top, d.left, d.top + d.height);
        try {
            for (var _b = tslib_1.__values(fromEnum(c.a)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var h = _c.value;
                g.addColorStop(h.a, h.b._fill);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        e.fillStyle = g;
        e.fill();
    };
    LegendBaseViewManager.prototype.r = function () {
    };
    LegendBaseViewManager.prototype.s = function () {
    };
    LegendBaseViewManager.prototype.u = function (a) {
        var b = this.m == null ? 0 : this.m.getChildCount();
        for (var c = 0; c < b; c++) {
            var d = this.m.getChildAt(c);
            d.setStyleProperty("display", a ? "block" : "");
            d.setStyleProperty("float", a ? "left" : "");
            d.setStyleProperty("margin", a ? "0px" : "");
            d.setStyleProperty("margin-right", a ? "8px" : "");
            d.getChildAt(1).setStyleProperty("white-space", a ? "nowrap" : "");
        }
    };
    LegendBaseViewManager.prototype.y = function (a) {
        this.m.setStyleProperty("color", a);
    };
    LegendBaseViewManager.prototype.z = function (a) {
        var b = FontUtil.getFontInfoFromString(this.h, a);
        this.m.setStyleProperty("font-family", b.n);
        this.m.setStyleProperty("font-size", b.d + "px");
        this.m.setStyleProperty("font-style", b.q);
    };
    LegendBaseViewManager.$t = markType(LegendBaseViewManager, 'LegendBaseViewManager');
    return LegendBaseViewManager;
}(Base));
export { LegendBaseViewManager };
/**
 * @hidden
 */
var LegendBase = /** @class */ (function (_super) {
    tslib_1.__extends(LegendBase, _super);
    function LegendBase() {
        var _this = _super.call(this) || this;
        _this.ak = null;
        _this.at = null;
        _this._seriesOwner = null;
        _this._chartOwner = null;
        _this.propertyChanged = null;
        _this.propertyUpdated = null;
        _this.legendItemMouseLeftButtonDown = null;
        _this.legendItemMouseLeftButtonUp = null;
        _this.legendItemMouseEnter = null;
        _this.legendItemMouseLeave = null;
        _this.legendItemMouseMove = null;
        var a = _this.aj();
        _this.a6(a);
        a.ai();
        _this.at = new ObservableCollection$1(UIElement.$, 0);
        _this.aw = "ig-chart-legend-items-list";
        _this.ax = "ig-chart-legend-item";
        _this.av = "ig-chart-legend-item-badge";
        _this.ay = "ig-chart-legend-item-text";
        return _this;
    }
    LegendBase.prototype.aj = function () {
        return new LegendBaseView(this);
    };
    LegendBase.prototype.a6 = function (a) {
        this.ak = a;
    };
    LegendBase.prototype.addChildInOrder = function (a, b) {
    };
    LegendBase.prototype.a0 = function (a, b) {
    };
    Object.defineProperty(LegendBase.prototype, "seriesOwner", {
        get: function () {
            return this._seriesOwner;
        },
        set: function (a) {
            this._seriesOwner = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LegendBase.prototype, "chartOwner", {
        get: function () {
            return this._chartOwner;
        },
        set: function (a) {
            this._chartOwner = a;
        },
        enumerable: true,
        configurable: true
    });
    LegendBase.prototype.get_isItemwise = function () {
        return false;
    };
    Object.defineProperty(LegendBase.prototype, "isItemwise", {
        get: function () {
            return this.get_isItemwise();
        },
        enumerable: true,
        configurable: true
    });
    LegendBase.prototype.get_isScale = function () {
        return false;
    };
    Object.defineProperty(LegendBase.prototype, "isScale", {
        get: function () {
            return this.get_isScale();
        },
        enumerable: true,
        configurable: true
    });
    LegendBase.prototype.clearLegendItemsForSeries = function (a) {
        var e_4, _a, e_5, _b;
        if (a == null || this.at == null || this.at.count == 0) {
            return;
        }
        var b = new ObservableCollection$1(UIElement.$, 0);
        try {
            for (var _c = tslib_1.__values(fromEnum(this.at)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var c = _d.value;
                var d = typeCast(ContentControl.$, c);
                if (d != null && d.content != null) {
                    var e = typeCast(ILegendContext_$type, d.content);
                    if (e != null && e.series == a) {
                        b.add(c);
                    }
                }
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_4) throw e_4.error; }
        }
        try {
            for (var _e = tslib_1.__values(fromEnum(b)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var f = _f.value;
                this.at.remove(f);
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_5) throw e_5.error; }
        }
    };
    LegendBase.prototype.containsChild = function (a) {
        return this.at.contains(a);
    };
    LegendBase.prototype.removeChild = function (a) {
        this.at.remove(a);
    };
    Object.defineProperty(LegendBase.prototype, "ai", {
        get: function () {
            if (this.seriesOwner != null) {
                return this.seriesOwner;
            }
            else {
                return this.chartOwner;
            }
        },
        enumerable: true,
        configurable: true
    });
    LegendBase.prototype.a7 = function (a, b, c) {
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
        if (this.propertyUpdated != null) {
            this.propertyUpdated(this, new PropertyUpdatedEventArgs(a, b, c));
        }
    };
    LegendBase.prototype.a3 = function (a) {
        if (this.legendItemMouseLeftButtonDown != null) {
            this.legendItemMouseLeftButtonDown(this, a);
        }
    };
    LegendBase.prototype.a4 = function (a) {
        if (this.legendItemMouseLeftButtonUp != null) {
            this.legendItemMouseLeftButtonUp(this, a);
        }
    };
    LegendBase.prototype.a1 = function (a) {
        if (this.legendItemMouseEnter != null) {
            this.legendItemMouseEnter(this, a);
        }
    };
    LegendBase.prototype.a2 = function (a) {
        if (this.legendItemMouseLeave != null) {
            this.legendItemMouseLeave(this, a);
        }
    };
    LegendBase.prototype.a5 = function (a) {
        if (this.legendItemMouseMove != null) {
            this.legendItemMouseMove(this, a);
        }
    };
    LegendBase.prototype.provideContainer = function (a) {
        this.ak.ah(a);
    };
    Object.defineProperty(LegendBase.prototype, "aw", {
        get: function () {
            return this.ak.s;
        },
        set: function (a) {
            this.ak.s = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LegendBase.prototype, "ax", {
        get: function () {
            return this.ak.t;
        },
        set: function (a) {
            this.ak.t = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LegendBase.prototype, "av", {
        get: function () {
            return this.ak.r;
        },
        set: function (a) {
            this.ak.r = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LegendBase.prototype, "ay", {
        get: function () {
            return this.ak.u;
        },
        set: function (a) {
            this.ak.u = a;
        },
        enumerable: true,
        configurable: true
    });
    LegendBase.prototype.exportVisualData = function () {
        return this.ak.f();
    };
    LegendBase.prototype.exportSerializedVisualData = function () {
        return this.exportVisualData().serialize();
    };
    LegendBase.prototype.get_isFinancial = function () {
        return false;
    };
    Object.defineProperty(LegendBase.prototype, "isFinancial", {
        get: function () {
            return this.get_isFinancial();
        },
        enumerable: true,
        configurable: true
    });
    LegendBase.$t = markType(LegendBase, 'LegendBase', ContentControl.$, [INotifyPropertyChanged_$type, IChartLegend_$type]);
    return LegendBase;
}(ContentControl));
export { LegendBase };
/**
 * @hidden
 */
var LegendBaseView = /** @class */ (function (_super) {
    tslib_1.__extends(LegendBaseView, _super);
    function LegendBaseView(a) {
        var _this = _super.call(this) || this;
        _this.c = null;
        _this.b = null;
        _this.s = null;
        _this.t = null;
        _this.r = null;
        _this.u = null;
        _this.l = false;
        _this.i = false;
        _this.p = null;
        _this.q = null;
        _this.b = a;
        _this.c = new LegendBaseViewManager(_this);
        return _this;
    }
    LegendBaseView.prototype.ai = function () {
    };
    LegendBaseView.prototype.get_j = function () {
        return false;
    };
    Object.defineProperty(LegendBaseView.prototype, "j", {
        get: function () {
            return this.get_j();
        },
        enumerable: true,
        configurable: true
    });
    LegendBaseView.prototype.d = function (a) {
        var b;
        var c;
        var d;
        var e = this.aa(a, b, c, d);
        b = e.p1;
        c = e.p2;
        d = e.p3;
        var f = new MouseButtonEventArgs();
        var g = new LegendMouseButtonEventArgs(b, c, d, f, a);
        return g;
    };
    LegendBaseView.prototype.e = function (a) {
        var b;
        var c;
        var d;
        var e = this.aa(a, b, c, d);
        b = e.p1;
        c = e.p2;
        d = e.p3;
        var f = new MouseEventArgs();
        var g = new LegendMouseEventArgs(b, c, d, f, a);
        return g;
    };
    LegendBaseView.prototype.aa = function (a, b, c, d) {
        b = this.b.ai;
        c = null;
        d = null;
        if (a != null) {
            var e = typeCast(ContentControl.$, a);
            if (e != null && e.content != null && typeCast(ILegendContext_$type, e.content) !== null) {
                var f = typeCast(ILegendContext_$type, e.content);
                c = typeCast(ILegendSeries_$type, f.series);
                if (c != null) {
                    b = c.container;
                }
                d = f.item;
            }
        }
        return {
            p1: b,
            p2: c,
            p3: d
        };
    };
    LegendBaseView.prototype.z = function () {
    };
    LegendBaseView.prototype.m = function () {
        return true;
    };
    LegendBaseView.prototype.y = function (a) {
    };
    LegendBaseView.prototype.ak = function (a) {
    };
    LegendBaseView.prototype.ah = function (a) {
        this.c.t(a);
    };
    LegendBaseView.prototype.al = function (a) {
        this.c.w(a);
    };
    LegendBaseView.prototype.x = function (a) {
        var b = this.b.at.indexOf(a);
        this.c.p(a, b);
    };
    LegendBaseView.prototype.af = function () {
        this.c.r();
    };
    LegendBaseView.prototype.ag = function () {
        this.c.s();
    };
    Object.defineProperty(LegendBaseView.prototype, "k", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.c.u(this.i);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LegendBaseView.prototype, "v", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.c.y(this.p);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LegendBaseView.prototype, "w", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.c.z(this.q);
        },
        enumerable: true,
        configurable: true
    });
    LegendBaseView.prototype.aj = function () {
    };
    LegendBaseView.prototype.n = function (e_) {
        var ev_ = e_.originalEvent;
        return ev_.currentTarget;
    };
    LegendBaseView.prototype.o = function (e_) {
        return this.c.o(this.n(e_));
    };
    LegendBaseView.prototype.ad = function (a) {
        if (!this.l) {
            this.b.a1(this.e(this.o(a)));
        }
        this.l = true;
    };
    LegendBaseView.prototype.ac = function (a) {
        if (this.l) {
            this.b.a2(this.e(this.o(a)));
        }
        this.l = false;
    };
    LegendBaseView.prototype.ab = function (a) {
        this.b.a3(this.d(this.o(a)));
    };
    LegendBaseView.prototype.ae = function (a) {
        this.b.a4(this.d(this.o(a)));
    };
    LegendBaseView.prototype.h = function (a) {
        var b = a.querySelectorAll("span *");
        var c = Rect.empty;
        for (var e = 0; e < b.length; e++) {
            var d = b[e];
            if (stringIsNullOrEmpty(d.getText())) {
                continue;
            }
            var f = new Rect(0, d.getOffset().left, d.getOffset().top, d.width(), d.height());
            f.union(c);
            c = f;
        }
        return RectData.b(c);
    };
    LegendBaseView.prototype.g = function (a) {
        var b = a.querySelectorAll("canvas");
        var c = b.length > 0 ? b[0] : null;
        return c != null ? new RectData(c.getOffset().left, c.getOffset().top, c.width(), c.height()) : null;
    };
    LegendBaseView.prototype.a = function (a) {
        if (this.b == null || this.b.at == null || this.b.at.count <= a) {
            return null;
        }
        var b = this.b.at._inner[a];
        if (b == null) {
            return null;
        }
        var c = b.content;
        if (c == null) {
            return null;
        }
        var d = c.actualItemBrush != null ? c.actualItemBrush : c.itemBrush;
        return AppearanceHelper.b(d);
    };
    LegendBaseView.prototype.f = function () {
        var a = new LegendVisualData();
        var b = this.c.m;
        var c = b.getChildCount();
        for (var d = 0; d < c; d++) {
            var e = new LegendItemVisualData();
            var f = b.getChildAt(d);
            e.label = f.getText();
            e.label = e.label != null ? e.label.trim() : null;
            e.bounds = new RectData(f.getOffset().left, f.getOffset().top, f.width(), f.height());
            e.labelBounds = this.h(f);
            e.badgeBounds = this.g(f);
            a.items.add(e);
        }
        for (var g = 0; g < this.b.at.count; g++) {
            a.items._inner[g].appearance.fill = this.a(g);
        }
        return a;
    };
    LegendBaseView.$t = markType(LegendBaseView, 'LegendBaseView');
    return LegendBaseView;
}(Base));
export { LegendBaseView };
