// igniteui-react-charts/ES5/MoneyFlowIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MoneyFlowIndexIndicatorStrategy } from "./MoneyFlowIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var MoneyFlowIndexIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(MoneyFlowIndexIndicator, _super);
    function MoneyFlowIndexIndicator() {
        var _this = _super.call(this) || this;
        _this.pj = 1;
        return _this;
    }
    MoneyFlowIndexIndicator.prototype.get_qb = function () {
        return new MoneyFlowIndexIndicatorStrategy();
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.get_qo = function () {
        return MoneyFlowIndexIndicator.$;
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(MoneyFlowIndexIndicator.qx));
        },
        set: function (a) {
            this.h(MoneyFlowIndexIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.qh = function () {
        return this.qw;
    };
    MoneyFlowIndexIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    MoneyFlowIndexIndicator.$t = markType(MoneyFlowIndexIndicator, 'MoneyFlowIndexIndicator', StrategyBasedIndicator.$);
    MoneyFlowIndexIndicator.qx = StrategyBasedIndicator.qt(14, MoneyFlowIndexIndicator.$);
    return MoneyFlowIndexIndicator;
}(StrategyBasedIndicator));
export { MoneyFlowIndexIndicator };
