// igniteui-react-charts/ES5/PointSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
import { markType } from "igniteui-react-core/ES5/type";
import { AnchoredCategorySeriesView } from "./AnchoredCategorySeriesView";
import { Color } from "igniteui-react-core/ES5/Color";
/**
 * @hidden
 */
var PointSeries = /** @class */ (function (_super) {
    tslib_1.__extends(PointSeries, _super);
    function PointSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PointSeries.$;
        return _this;
    }
    PointSeries.prototype.bf = function () {
        return new PointSeriesView(this);
    };
    PointSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.qg = a;
    };
    PointSeries.prototype.dh = function (a, b) {
        var c = this.lk(this.ln(a), true, false);
        if (this.l3(a, b)) {
            return true;
        }
        return false;
    };
    PointSeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
    };
    PointSeries.prototype.get_cy = function () {
        return false;
    };
    Object.defineProperty(PointSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.n1;
        },
        enumerable: true,
        configurable: true
    });
    PointSeries.$t = markType(PointSeries, 'PointSeries', HorizontalAnchoredCategorySeries.$);
    return PointSeries;
}(HorizontalAnchoredCategorySeries));
export { PointSeries };
/**
 * @hidden
 */
var PointSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(PointSeriesView, _super);
    function PointSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c6 = null;
        _this.c6 = a;
        return _this;
    }
    PointSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        if (!this.n) {
            this.by.l0 = 2;
            this.e.bj = 4;
        }
    };
    PointSeriesView.prototype.ac = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 3;
        this.e.ew = 2;
        this.e.ex = 2;
        this.e.du = false;
    };
    PointSeriesView.$t = markType(PointSeriesView, 'PointSeriesView', AnchoredCategorySeriesView.$);
    return PointSeriesView;
}(AnchoredCategorySeriesView));
export { PointSeriesView };
