// igniteui-react-charts/ES5/PolarSplineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { Type, markType } from "igniteui-react-core/ES5/type";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Path } from "igniteui-react-core/ES5/Path";
import { PathGeometry } from "igniteui-react-core/ES5/PathGeometry";
import { PathFigure } from "igniteui-react-core/ES5/PathFigure";
import { Extensions } from "igniteui-react-core/ES5/Extensions";
import { Numeric } from "igniteui-react-core/ES5/Numeric";
import { Color } from "igniteui-react-core/ES5/Color";
/**
 * @hidden
 */
var PolarSplineSeries = /** @class */ (function (_super) {
    tslib_1.__extends(PolarSplineSeries, _super);
    function PolarSplineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarSplineSeries.$;
        return _this;
    }
    PolarSplineSeries.prototype.bf = function () {
        return new PolarSplineSeriesView(this);
    };
    PolarSplineSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.pc = a;
    };
    PolarSplineSeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case PolarSplineSeries.$$p[0]:
                this.jc(false);
                break;
        }
    };
    PolarSplineSeries.prototype.pa = function (a) {
        var b = a;
        b.c0();
    };
    PolarSplineSeries.prototype.pb = function (a, b) {
        this.pa(b);
        var c = b;
        c.c1(a.m, this.pd);
    };
    PolarSplineSeries.$t = markType(PolarSplineSeries, 'PolarSplineSeries', PolarLineSeriesBase.$);
    return PolarSplineSeries;
}(PolarLineSeriesBase));
export { PolarSplineSeries };
Type.dep(DependencyProperty, PropertyMetadata, PolarSplineSeries, 'raisePropertyChanged', ['Stiffness:pd:pf', [1, 0.5]]);
/**
 * @hidden
 */
var PolarSplineSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(PolarSplineSeriesView, _super);
    function PolarSplineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cz = null;
        _this.c2 = ((function () {
            var $ret = new Path();
            $ret.aj = new PathGeometry();
            return $ret;
        })());
        _this.cz = a;
        return _this;
    }
    PolarSplineSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        var a = new PathGeometry();
        a.b.add(((function () {
            var $ret = new PathFigure();
            return $ret;
        })()));
        this.c2.aj = a;
    };
    PolarSplineSeriesView.prototype.c0 = function () {
        Extensions.e(this.c2.aj);
        this.an();
    };
    PolarSplineSeriesView.prototype.c1 = function (a, b) {
        this.c2.aj.b = Numeric.h(a.count, function (c) { return a._inner[c].x; }, function (c) { return a._inner[c].y; }, b);
        this.an();
    };
    PolarSplineSeriesView.prototype.a8 = function () {
        _super.prototype.a8.call(this);
        this.c2._stroke = this.e.k2;
        this.c2.ac = this.e.ey;
        this.c2.ai = this.e.la;
        this.c2.ad = this.e.lb;
    };
    PolarSplineSeriesView.prototype.a9 = function () {
        _super.prototype.a9.call(this);
        var a = this.bl();
        this.c2._stroke = a;
        this.c2.ac = this.e.ey + 3;
    };
    PolarSplineSeriesView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        a.s(this.c2);
    };
    PolarSplineSeriesView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        this.e.a2.exportPathData(a, this.c2, "polarShape", ["Main"]);
    };
    PolarSplineSeriesView.prototype.ac = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 3;
        this.e.ew = 1;
        this.e.ex = 4;
        this.e.du = false;
    };
    PolarSplineSeriesView.$t = markType(PolarSplineSeriesView, 'PolarSplineSeriesView', PolarLineSeriesBaseView.$);
    return PolarSplineSeriesView;
}(PolarLineSeriesBaseView));
export { PolarSplineSeriesView };
