// igniteui-react-charts/ES5/PolarSplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { Type, markType } from "igniteui-react-core/ES5/type";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Path } from "igniteui-react-core/ES5/Path";
import { PathGeometry } from "igniteui-react-core/ES5/PathGeometry";
import { Extensions } from "igniteui-react-core/ES5/Extensions";
import { Numeric } from "igniteui-react-core/ES5/Numeric";
import { Color } from "igniteui-react-core/ES5/Color";
/**
 * @hidden
 */
var PolarSplineAreaSeries = /** @class */ (function (_super) {
    tslib_1.__extends(PolarSplineAreaSeries, _super);
    function PolarSplineAreaSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarSplineAreaSeries.$;
        return _this;
    }
    PolarSplineAreaSeries.prototype.bf = function () {
        return new PolarSplineAreaSeriesView(this);
    };
    PolarSplineAreaSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.pc = a;
    };
    PolarSplineAreaSeries.prototype.pa = function (a) {
        var b = a;
        b.c0();
    };
    PolarSplineAreaSeries.prototype.pb = function (a, b) {
        this.pa(b);
        if (a.m.count < 1) {
            return;
        }
        var c = b;
        c.c3._opacity = this.d6;
        c.c1(a.m, this.pd);
    };
    PolarSplineAreaSeries.prototype.get_o7 = function () {
        return true;
    };
    PolarSplineAreaSeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case PolarSplineAreaSeries.$$p[0]:
                this.jc(false);
                break;
        }
    };
    PolarSplineAreaSeries.$t = markType(PolarSplineAreaSeries, 'PolarSplineAreaSeries', PolarLineSeriesBase.$);
    return PolarSplineAreaSeries;
}(PolarLineSeriesBase));
export { PolarSplineAreaSeries };
Type.dep(DependencyProperty, PropertyMetadata, PolarSplineAreaSeries, 'raisePropertyChanged', ['Stiffness:pd:pf', [1, 0.5]]);
/**
 * @hidden
 */
var PolarSplineAreaSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(PolarSplineAreaSeriesView, _super);
    function PolarSplineAreaSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cz = null;
        _this.c4 = ((function () {
            var $ret = new Path();
            $ret.aj = new PathGeometry();
            return $ret;
        })());
        _this.c2 = ((function () {
            var $ret = new Path();
            $ret.aj = new PathGeometry();
            return $ret;
        })());
        _this.cz = a;
        return _this;
    }
    PolarSplineAreaSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        if (!this.n) {
            this.e.bj = 1;
        }
    };
    Object.defineProperty(PolarSplineAreaSeriesView.prototype, "c3", {
        get: function () {
            return this.c2;
        },
        enumerable: true,
        configurable: true
    });
    PolarSplineAreaSeriesView.prototype.c0 = function () {
        Extensions.e(this.c2.aj);
        Extensions.e(this.c4.aj);
        this.an();
    };
    PolarSplineAreaSeriesView.prototype.c1 = function (a, b) {
        this.c4.aj.b = Numeric.h(a.count + 1, function (c) { return c < a.count ? a._inner[c].x : a._inner[c - a.count].x; }, function (c) { return c < a.count ? a._inner[c].y : a._inner[c - a.count].y; }, b);
        this.c2.aj.b = Numeric.h(a.count + 1, function (c) { return c < a.count ? a._inner[c].x : a._inner[c - a.count].x; }, function (c) { return c < a.count ? a._inner[c].y : a._inner[c - a.count].y; }, b);
        this.an();
    };
    PolarSplineAreaSeriesView.prototype.a8 = function () {
        _super.prototype.a8.call(this);
        this.c2._fill = this.e.k2;
        this.c2._opacity = this.e.d6;
        this.c4._stroke = this.e.k3;
        this.c4.ac = this.e.ey;
        this.c4.ai = this.e.la;
        this.c4.ad = this.e.lb;
    };
    PolarSplineAreaSeriesView.prototype.a9 = function () {
        _super.prototype.a9.call(this);
        var a = this.bl();
        this.c2._fill = a;
        this.c2._opacity = 1;
        this.c4._stroke = a;
        this.c4.ac = this.e.ey + 3;
    };
    PolarSplineAreaSeriesView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        if (a.d) {
            a.s(this.c2);
            a.s(this.c4);
        }
    };
    PolarSplineAreaSeriesView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        this.e.a2.exportPathData(a, this.c4, "polarShape", ["Main"]);
        this.e.a2.exportPathData(a, this.c2, "fillShape", ["Fill"]);
    };
    PolarSplineAreaSeriesView.prototype.ac = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 5;
        this.e.ew = 5;
        this.e.ex = 5;
    };
    PolarSplineAreaSeriesView.$t = markType(PolarSplineAreaSeriesView, 'PolarSplineAreaSeriesView', PolarLineSeriesBaseView.$);
    return PolarSplineAreaSeriesView;
}(PolarLineSeriesBaseView));
export { PolarSplineAreaSeriesView };
