// igniteui-react-charts/ES5/PolarLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { UnknownValuePlotting_$type } from "igniteui-react-core/ES5/UnknownValuePlotting";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { Type, enumGetBox, markType } from "igniteui-react-core/ES5/type";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Path } from "igniteui-react-core/ES5/Path";
/**
 * @hidden
 */
var PolarLineSeries = /** @class */ (function (_super) {
    tslib_1.__extends(PolarLineSeries, _super);
    function PolarLineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarLineSeries.$;
        return _this;
    }
    PolarLineSeries.prototype.bf = function () {
        return new PolarLineSeriesView(this);
    };
    PolarLineSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.pc = a;
    };
    PolarLineSeries.prototype.pa = function (a) {
        var b = a;
        b.c0();
    };
    PolarLineSeries.prototype.pb = function (a, b) {
        var c = b;
        c.c1(a.m, this.eu);
    };
    PolarLineSeries.prototype.o2 = function () {
        return this.pd;
    };
    PolarLineSeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case PolarLineSeries.$$p[0]:
                this.jc(false);
                this.ij();
                break;
        }
    };
    PolarLineSeries.$t = markType(PolarLineSeries, 'PolarLineSeries', PolarLineSeriesBase.$);
    return PolarLineSeries;
}(PolarLineSeriesBase));
export { PolarLineSeries };
Type.dep(DependencyProperty, PropertyMetadata, PolarLineSeries, 'raisePropertyChanged', ['UnknownValuePlotting:pd:pf', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
/**
 * @hidden
 */
var PolarLineSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(PolarLineSeriesView, _super);
    function PolarLineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cz = null;
        _this.c2 = new Path();
        _this.cz = a;
        return _this;
    }
    PolarLineSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
    };
    PolarLineSeriesView.prototype.c0 = function () {
        this.c2.aj = null;
        this.an();
    };
    PolarLineSeriesView.prototype.c1 = function (a, b) {
        this.cz.oz.d(this.c2, a, b);
        this.an();
    };
    PolarLineSeriesView.prototype.a8 = function () {
        _super.prototype.a8.call(this);
        this.c2._stroke = this.e.k2;
        this.c2.ac = this.e.ey;
        this.c2.ai = this.e.la;
        this.c2.ad = this.e.lb;
    };
    PolarLineSeriesView.prototype.a9 = function () {
        _super.prototype.a9.call(this);
        var a = this.bl();
        this.c2._stroke = a;
        this.c2.ac = this.e.ey + 3;
    };
    PolarLineSeriesView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        a.s(this.c2);
    };
    PolarLineSeriesView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        this.e.a2.exportPathData(a, this.c2, "polarShape", ["Main"]);
    };
    PolarLineSeriesView.$t = markType(PolarLineSeriesView, 'PolarLineSeriesView', PolarLineSeriesBaseView.$);
    return PolarLineSeriesView;
}(PolarLineSeriesBaseView));
export { PolarLineSeriesView };
