// igniteui-react-charts/ES5/PolarAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { UnknownValuePlotting_$type } from "igniteui-react-core/ES5/UnknownValuePlotting";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { Type, enumGetBox, markType, Point_$type } from "igniteui-react-core/ES5/type";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Path } from "igniteui-react-core/ES5/Path";
import { Color } from "igniteui-react-core/ES5/Color";
/**
 * @hidden
 */
var PolarAreaSeries = /** @class */ (function (_super) {
    tslib_1.__extends(PolarAreaSeries, _super);
    function PolarAreaSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarAreaSeries.$;
        return _this;
    }
    PolarAreaSeries.prototype.bf = function () {
        return new PolarAreaSeriesView(this);
    };
    PolarAreaSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.pc = a;
    };
    PolarAreaSeries.prototype.pa = function (a) {
        var b = a;
        b.c0();
    };
    PolarAreaSeries.prototype.pb = function (a, b) {
        if (a.m.count > 0) {
            var c = a.m._inner[0];
            var d = a.m._inner[a.m.count - 1];
            var e = c.x == d.x && c.y == d.y;
            if (!e) {
                a.m.add(a.m._inner[0]);
            }
        }
        var f = b;
        f.c3._opacity = this.d6;
        f.c1(a.m, this.eu);
    };
    PolarAreaSeries.prototype.get_o7 = function () {
        return true;
    };
    PolarAreaSeries.prototype.o2 = function () {
        return this.pd;
    };
    PolarAreaSeries.$t = markType(PolarAreaSeries, 'PolarAreaSeries', PolarLineSeriesBase.$);
    return PolarAreaSeries;
}(PolarLineSeriesBase));
export { PolarAreaSeries };
Type.dep(DependencyProperty, PropertyMetadata, PolarAreaSeries, 'raisePropertyChanged', ['UnknownValuePlotting:pd:pf', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
/**
 * @hidden
 */
var PolarAreaSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(PolarAreaSeriesView, _super);
    function PolarAreaSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cz = null;
        _this.c4 = new Path();
        _this.c2 = new Path();
        _this.cz = a;
        return _this;
    }
    PolarAreaSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        if (!this.n) {
            this.e.bj = 1;
        }
    };
    Object.defineProperty(PolarAreaSeriesView.prototype, "c3", {
        get: function () {
            return this.c2;
        },
        enumerable: true,
        configurable: true
    });
    PolarAreaSeriesView.prototype.c0 = function () {
        this.c2.aj = null;
        this.c4.aj = null;
        this.an();
    };
    PolarAreaSeriesView.prototype.c1 = function (a, b) {
        var c = new List$1(Point_$type, 1, a);
        this.cz.oz.d(this.c4, c, b);
        this.cz.oz.d(this.c2, c, b);
        this.an();
    };
    PolarAreaSeriesView.prototype.a8 = function () {
        _super.prototype.a8.call(this);
        this.c4._stroke = this.e.k3;
        this.c4.ac = this.e.ey;
        this.c4.ai = this.e.la;
        this.c4.ad = this.e.lb;
        this.c2._fill = this.e.k2;
        this.c2._opacity = this.e.d6;
    };
    PolarAreaSeriesView.prototype.a9 = function () {
        _super.prototype.a9.call(this);
        var a = this.bl();
        this.c4._stroke = a;
        this.c4.ac = this.e.ey + 3;
        this.c2._fill = a;
        this.c2._opacity = 1;
    };
    PolarAreaSeriesView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        if (a.d) {
            a.s(this.c2);
            a.s(this.c4);
        }
    };
    PolarAreaSeriesView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        this.e.a2.exportPathData(a, this.c4, "polarShape", ["Main"]);
        this.e.a2.exportPathData(a, this.c2, "fillShape", ["Fill"]);
    };
    PolarAreaSeriesView.prototype.ac = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 5;
        this.e.ew = 5;
        this.e.ex = 5;
    };
    PolarAreaSeriesView.$t = markType(PolarAreaSeriesView, 'PolarAreaSeriesView', PolarLineSeriesBaseView.$);
    return PolarAreaSeriesView;
}(PolarLineSeriesBaseView));
export { PolarAreaSeriesView };
