// igniteui-react-charts/ES5/RateOfChangeAndMomentumIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RateOfChangeAndMomentumIndicatorStrategy } from "./RateOfChangeAndMomentumIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var RateOfChangeAndMomentumIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(RateOfChangeAndMomentumIndicator, _super);
    function RateOfChangeAndMomentumIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RateOfChangeAndMomentumIndicator.prototype.get_qb = function () {
        return new RateOfChangeAndMomentumIndicatorStrategy();
    };
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    RateOfChangeAndMomentumIndicator.prototype.get_qo = function () {
        return RateOfChangeAndMomentumIndicator.$;
    };
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(RateOfChangeAndMomentumIndicator.qx));
        },
        set: function (a) {
            this.h(RateOfChangeAndMomentumIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    RateOfChangeAndMomentumIndicator.prototype.qh = function () {
        return this.qw;
    };
    RateOfChangeAndMomentumIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    RateOfChangeAndMomentumIndicator.$t = markType(RateOfChangeAndMomentumIndicator, 'RateOfChangeAndMomentumIndicator', StrategyBasedIndicator.$);
    RateOfChangeAndMomentumIndicator.qx = StrategyBasedIndicator.qt(14, RateOfChangeAndMomentumIndicator.$);
    return RateOfChangeAndMomentumIndicator;
}(StrategyBasedIndicator));
export { RateOfChangeAndMomentumIndicator };
