// igniteui-react-charts/ES5/RelativeStrengthIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RelativeStrengthIndexIndicatorStrategy } from "./RelativeStrengthIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var RelativeStrengthIndexIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(RelativeStrengthIndexIndicator, _super);
    function RelativeStrengthIndexIndicator() {
        var _this = _super.call(this) || this;
        _this.pj = 1;
        return _this;
    }
    RelativeStrengthIndexIndicator.prototype.get_qb = function () {
        return new RelativeStrengthIndexIndicatorStrategy();
    };
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    RelativeStrengthIndexIndicator.prototype.get_qo = function () {
        return RelativeStrengthIndexIndicator.$;
    };
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(RelativeStrengthIndexIndicator.qx));
        },
        set: function (a) {
            this.h(RelativeStrengthIndexIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    RelativeStrengthIndexIndicator.prototype.qh = function () {
        return this.qw;
    };
    RelativeStrengthIndexIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    RelativeStrengthIndexIndicator.$t = markType(RelativeStrengthIndexIndicator, 'RelativeStrengthIndexIndicator', StrategyBasedIndicator.$);
    RelativeStrengthIndexIndicator.qx = StrategyBasedIndicator.qt(14, RelativeStrengthIndexIndicator.$);
    return RelativeStrengthIndexIndicator;
}(StrategyBasedIndicator));
export { RelativeStrengthIndexIndicator };
