// igniteui-react-charts/ES5/ScatterPolygonSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { ShapeSeriesBase } from "./ShapeSeriesBase";
import { ShapeAxisInfoCache } from "./ShapeAxisInfoCache";
import { StyleSelector } from "./StyleSelector";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { Style } from "igniteui-react-core/ES5/Style";
import { MarkerType_$type } from "./MarkerType";
import { DataTemplate } from "igniteui-react-core/ES5/DataTemplate";
import { Brush } from "igniteui-react-core/ES5/Brush";
import { CollisionAvoidanceType_$type } from "./CollisionAvoidanceType";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { ScalerParams } from "./ScalerParams";
import { Type, Base, enumGetBox, typeCast, fromEnum, markType, PointUtil, runOn } from "igniteui-react-core/ES5/type";
import { Dictionary$2 } from "igniteui-react-core/ES5/Dictionary$2";
import { OwnedPoint } from "./OwnedPoint";
import { ShapeSeriesViewBase } from "./ShapeSeriesViewBase";
import { MarkerSeries } from "./MarkerSeries";
import { DataContext } from "igniteui-react-core/ES5/DataContext";
import { Marker } from "./Marker";
import { HashPool$2 } from "igniteui-react-core/ES5/HashPool$2";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { RectUtil } from "igniteui-react-core/ES5/RectUtil";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { NumericMarkerManager } from "./NumericMarkerManager";
import { DataTemplateMeasureInfo } from "igniteui-react-core/ES5/DataTemplateMeasureInfo";
import { Size } from "igniteui-react-core/ES5/Size";
import { DataTemplatePassInfo } from "igniteui-react-core/ES5/DataTemplatePassInfo";
import { DataTemplateRenderInfo } from "igniteui-react-core/ES5/DataTemplateRenderInfo";
import { MarkerVisualData } from "./MarkerVisualData";
import { isNaN_, isInfinity } from "igniteui-react-core/ES5/number";
import { stringIsNullOrEmpty } from "igniteui-react-core/ES5/string";
/**
 * @hidden
 */
var ScatterPolygonSeries = /** @class */ (function (_super) {
    tslib_1.__extends(ScatterPolygonSeries, _super);
    function ScatterPolygonSeries() {
        var _this = _super.call(this) || this;
        _this.ng = null;
        _this.nd = new ScalerParams(1, ScatterPolygonSeries.oc, ScatterPolygonSeries.oc, false);
        _this.ne = new ScalerParams(1, ScatterPolygonSeries.oc, ScatterPolygonSeries.oc, false);
        _this.ab = ScatterPolygonSeries.$;
        return _this;
    }
    ScatterPolygonSeries.prototype.mg = function () {
        return this.ni();
    };
    ScatterPolygonSeries.prototype.my = function (a) {
        var _this = this;
        var b = a;
        b.cn(this.oe);
        b.co(this.nh);
        _super.prototype.my.call(this, a);
        if (this.cc(a)) {
            return;
        }
        var c, d;
        var e = a.getViewInfo(c, d);
        c = e.p0;
        d = e.p1;
        var f = this.getEffectiveViewport1(a);
        this.nd = new ScalerParams(1, d, c, this.l5.b8);
        this.nd.c = f;
        this.ne = new ScalerParams(1, d, c, this.l7.b8);
        this.ne.c = f;
        if (c.width < 1 || c.height < 1) {
            return;
        }
        this.ng = ((function () {
            var $ret = new ShapeAxisInfoCache();
            $ret.a = _this.l5;
            $ret.b = _this.l7;
            $ret.e = _this.l5.b8;
            $ret.f = _this.l7.b8;
            $ret.d = _this.bz;
            $ret.c = _this.ma;
            return $ret;
        })());
        if (this.ni()) {
            var g = new Dictionary$2(Base.$, OwnedPoint.$, 0);
            b.c4.l(g, 400, d, c, this.eu);
            b.c4.j(g, false);
        }
    };
    ScatterPolygonSeries.prototype.ni = function () {
        return this.nv != null && ((this.nc != 1 && this.nc != 0) || this.nw != null);
    };
    ScatterPolygonSeries.prototype.bf = function () {
        return new ScatterPolygonSeriesView(this);
    };
    ScatterPolygonSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.nf = a;
    };
    ScatterPolygonSeries.prototype.get_hasMarkers = function () {
        return true;
    };
    ScatterPolygonSeries.prototype.k5 = function () {
        return this.n8;
    };
    ScatterPolygonSeries.prototype.k6 = function () {
        return this.n9;
    };
    ScatterPolygonSeries.prototype.jy = function () {
        return this.nv;
    };
    ScatterPolygonSeries.prototype.jn = function () {
        _super.prototype.jn.call(this);
        if (this.nw != null) {
            this.nf.dc();
        }
        else {
            var a = MarkerSeries.l1(this, this.nc);
            var b = MarkerSeries.l9(a);
            if (b == null) {
                this.nv = MarkerSeries.mk;
            }
            else {
                this.nf.c9(b);
            }
        }
        if (this.oa != null) {
            this.nf.da();
        }
        else {
            this.n8 = this.bs == null ? null : this.bs.ma(this.fh);
        }
        if (this.ob != null) {
            this.nf.db();
        }
        else {
            this.n9 = this.bs == null ? null : this.bs.mb(this.fh);
        }
    };
    ScatterPolygonSeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case ScatterPolygonSeries.$$p[3]:
            case ScatterPolygonSeries.$$p[8]:
            case ScatterPolygonSeries.$$p[5]:
            case ScatterPolygonSeries.$$p[7]:
                this.jn();
                this.jc(false);
                break;
            case ScatterPolygonSeries.$$p[2]:
                if (c == MarkerSeries.mk || d == MarkerSeries.mk || (c == null || d != null)) {
                    this.nf.de();
                    this.jc(false);
                }
                this.jc(false);
                this.ij();
                break;
            case ScatterPolygonSeries.$$p[9]:
                this.nf.cn(this.oe);
                this.jc(false);
                this.ij();
                break;
            case ScatterPolygonSeries.$$p[10]:
                this.nf.co(this.nh);
                this.jc(false);
                this.ij();
                break;
            case ScatterPolygonSeries.$$p[4]:
                this.jc(false);
                this.ij();
                break;
        }
    };
    ScatterPolygonSeries.prototype.ar = function (a, b) {
        var c = this.nf.c1(a);
        var d = null;
        if (c != null) {
            d = c.content;
        }
        if (d != null) {
            return d;
        }
        return _super.prototype.ar.call(this, a, b);
    };
    ScatterPolygonSeries.prototype.nu = function (a, b) {
        var e_1, _a, e_2, _b;
        var c = new List$1(Base.$, 0);
        try {
            for (var _c = tslib_1.__values(fromEnum(b.c)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var d = _d.value;
                if (!a.containsKey(d)) {
                    c.add1(d);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        try {
            for (var _e = tslib_1.__values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var e = _f.value;
                b.g(e);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    ScatterPolygonSeries.prototype.ch = function (a, b, c) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var d = this.l5;
        var e = new ScalerParams(1, this.bs.mq, this.lx, d.b8);
        e.c = this.getEffectiveViewport1(this.bi);
        var f = this.l7;
        var g = new ScalerParams(1, this.bs.mq, this.lx, f.b8);
        g.c = this.getEffectiveViewport1(this.bi);
        a.clear();
        this.nf.dd(function (h) {
            if (h._visibility == 0) {
                var i = new SeriesPointOfInterest();
                var j = h.n;
                var k = h.o;
                if (j < _this.lx.left || j > _this.lx.right || k < _this.lx.top || k > _this.lx.bottom) {
                    return;
                }
                i.b = d.dy(j, e);
                i.c = f.dy(k, g);
                i.a = h.content;
                a.add(i);
            }
        });
        return true;
    };
    ScatterPolygonSeries.prototype.na = function (a, b, c, d, e) {
        var _this = this;
        if (c == null || c.length != this.ng.d.count) {
            c = new Array(this.ng.d.count);
            for (var f = 0; f < this.ng.d.count; f++) {
                c[f] = PointUtil.create();
            }
        }
        var g = ((function () {
            var $ret = new ScalerParams(1, d, e, _this.l5.b8);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var h = ((function () {
            var $ret = new ScalerParams(1, d, e, _this.l7.b8);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var i = this.ng.a.dy(g.d.left, g);
        var j = this.ng.a.dy(g.d.right, g);
        var k = this.ng.b.dy(h.d.bottom, h);
        var l = this.ng.b.dy(h.d.top, h);
        if (this.ng.e) {
            var m = i;
            i = j;
            j = m;
        }
        if (this.ng.f) {
            var n = k;
            k = l;
            l = n;
        }
        var o = this.ng;
        var p = o.a;
        var q = o.b;
        var r;
        var s;
        var t = g.d.left;
        var u = g.d.right;
        var v = h.d.top;
        var w = h.d.bottom;
        for (var x = 0; x < this.ng.d.count; x++) {
            r = 0;
            s = 0;
            var y = null;
            var z = a.b3.tryGetValue(x, y);
            y = z.p1;
            if (y == null) {
                c[x].x = NaN;
                c[x].y = NaN;
            }
            else {
                var aa = y.c;
                var ab = RectUtil.l(aa);
                r = ab.x;
                s = ab.y;
                if (r >= t && r <= u && s >= v && s <= w) {
                    c[x].x = r;
                    c[x].y = s;
                }
                else {
                    c[x].x = NaN;
                    c[x].y = NaN;
                }
            }
        }
        return c;
    };
    ScatterPolygonSeries.prototype.m9 = function (a, b) {
        var e_3, _a;
        if (b == null || b.length != a.e) {
            b = new Array(a.e);
        }
        var c = 0;
        var d = this.bz;
        try {
            for (var _b = tslib_1.__values(fromEnum(a.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                b[c] = d.indexOf(e);
                c++;
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return b;
    };
    ScatterPolygonSeries.$t = markType(ScatterPolygonSeries, 'ScatterPolygonSeries', ShapeSeriesBase.$);
    ScatterPolygonSeries.oc = new Rect(0, 0, 0, 1, 1);
    return ScatterPolygonSeries;
}(ShapeSeriesBase));
export { ScatterPolygonSeries };
Type.dep(DependencyProperty, PropertyMetadata, ScatterPolygonSeries, 'raisePropertyChanged', ['ActualMarkerBrush:n8:nx', [Brush.$], 'ActualMarkerOutline:n9:ny', [Brush.$], 'ActualMarkerTemplate:nv:nz', [DataTemplate.$, null], 'MarkerBrush:oa:n0', [Brush.$], 'MarkerCollisionAvoidance:nb:n1', [CollisionAvoidanceType_$type, enumGetBox(CollisionAvoidanceType_$type, 0)], 'MarkerOutline:ob:n2', [Brush.$], 'MarkerStyle:od:n3', [Style.$], 'MarkerTemplate:nw:n4', [DataTemplate.$, null], 'MarkerType:nc:n5', [MarkerType_$type, enumGetBox(MarkerType_$type, 1)], 'ShapeStyle:oe:n6', [Style.$, null], 'ShapeStyleSelector:nh:n7', [StyleSelector.$, null]]);
/**
 * @hidden
 */
var ScatterPolygonSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(ScatterPolygonSeriesView, _super);
    function ScatterPolygonSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c0 = null;
        _this.cz = null;
        _this.c5 = null;
        _this.c3 = null;
        _this.c7 = false;
        _this.c6 = null;
        _this.c8 = null;
        _this.dl = null;
        _this.c5 = a;
        _this.c6 = new HashPool$2(Base.$, Marker.$);
        _this.c8 = new List$1(Marker.$, 0);
        _this.df(_this.c6);
        _this.c3 = new NumericMarkerManager(1, function (b) { return _this.c6.item(b); }, function (b) { return _this.c5.ng.d.item(b); }, runOn(_this, _this.dj), runOn(_this, _this.cy), runOn(_this, _this.cx), function () { return _this.c5.nb; });
        _this.c3.q = runOn(_this, _this.dm);
        return _this;
    }
    ScatterPolygonSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        if (!this.n) {
            this.e.bj = 8;
        }
    };
    Object.defineProperty(ScatterPolygonSeriesView.prototype, "c4", {
        get: function () {
            return this.c3;
        },
        enumerable: true,
        configurable: true
    });
    ScatterPolygonSeriesView.prototype.dj = function (a) {
        this.c5.nu(a, this.c6);
    };
    ScatterPolygonSeriesView.prototype.cy = function () {
        this.c0 = this.c5.na(this, this.c6, this.c0, this.bs, this.br);
        return this.c0;
    };
    ScatterPolygonSeriesView.prototype.cx = function () {
        this.cz = this.c5.m9(this.c6, this.cz);
        return this.cz;
    };
    ScatterPolygonSeriesView.prototype.cu = function (a, b) {
        return this.cv(a, b, true);
    };
    ScatterPolygonSeriesView.prototype.cc = function (a, b) {
        _super.prototype.cc.call(this, a, b);
        ShapeSeriesViewBase.cb(this, a, b);
    };
    ScatterPolygonSeriesView.prototype.ca = function (a, b) {
        var c = a;
        if (c == null) {
            return;
        }
        c.aj = b;
    };
    ScatterPolygonSeriesView.prototype.dd = function (a) {
        this.c6.f(a);
    };
    ScatterPolygonSeriesView.prototype.df = function (a) {
        a.create = runOn(this, this.c2);
        a.destroy = runOn(this, this.dh);
        a.activate = runOn(this, this.dg);
        a.disactivate = runOn(this, this.di);
    };
    ScatterPolygonSeriesView.prototype.c2 = function () {
        var _this = this;
        var a = new Marker();
        if (!this.c7) {
            a.content = ((function () {
                var $ret = new DataContext();
                $ret.series = _this.e;
                return $ret;
            })());
        }
        a.ah = this.e.nv;
        this.c8.add(a);
        return a;
    };
    ScatterPolygonSeriesView.prototype.dh = function (a) {
        a.content = null;
        this.c8.remove(a);
    };
    ScatterPolygonSeriesView.prototype.dg = function (a) {
        a._visibility = 0;
    };
    ScatterPolygonSeriesView.prototype.di = function (a) {
        if (a.content != null) {
            (typeCast(DataContext.$, a.content)).item = null;
        }
        a._visibility = 1;
    };
    ScatterPolygonSeriesView.prototype.dk = function (a) {
        this.c7 = a;
    };
    ScatterPolygonSeriesView.prototype.be = function (a, b) {
        _super.prototype.be.call(this, a, b);
        var c = a;
        var d = c.content;
        if (d != null) {
            if (this.c5.n8 != null) {
                d.actualItemBrush = this.c5.n8;
            }
            else {
                d.actualItemBrush = this.by.k2;
            }
            if (this.c5.n8 != null) {
                d.outline = this.c5.n9;
            }
            else {
                d.outline = this.by.k3;
            }
            d.thickness = 0.5;
        }
    };
    ScatterPolygonSeriesView.prototype.bf = function (a, b) {
        _super.prototype.bf.call(this, a, b);
        var c = a;
        var d = this.bm(b);
        var e = c.content;
        if (e != null) {
            e.actualItemBrush = d;
            e.outline = d;
            e.thickness = 1 + 3;
        }
    };
    ScatterPolygonSeriesView.prototype.a2 = function (a, b) {
        _super.prototype.a2.call(this, a, b);
        var c = new DataTemplatePassInfo();
        c.isHitTestRender = b;
        c.context = a.g();
        c.viewportTop = this.br.top;
        c.viewportLeft = this.br.left;
        c.viewportWidth = this.br.width;
        c.viewportHeight = this.br.height;
        c.passID = "Markers";
        var d = new DataTemplateRenderInfo();
        d.passInfo = c;
        d.isHitTestRender = b;
        var e = new DataTemplateMeasureInfo();
        e.passInfo = c;
        var f = false;
        var g = a.g();
        e.context = g;
        d.context = g;
        var h = 0;
        var i = 0;
        if (this.c5.nv != null && this.c5.nv.passStarting != null) {
            this.c5.nv.passStarting(c);
        }
        for (var j = 0; j < this.c8.count; j++) {
            var k = this.c8._inner[j];
            if (k._visibility == 1) {
                continue;
            }
            this.bd(k, j, b);
            if (!f) {
                e.width = k.width;
                e.height = k.height;
                e.renderOffsetX = 0;
                e.renderOffsetY = 0;
                e.renderContext = a;
                var l = k.ah;
                if (l.measure != null) {
                    e.data = k.content;
                    l.measure(e);
                    f = e.isConstant;
                    if (f) {
                        h = e.width;
                        i = e.height;
                    }
                }
                d.availableWidth = e.width;
                d.availableHeight = e.height;
                d.renderOffsetX = e.renderOffsetX;
                d.renderOffsetY = e.renderOffsetY;
                d.renderContext = a;
            }
            else {
                d.availableWidth = h;
                d.availableHeight = i;
            }
            if (!isNaN_(k.width) && !isInfinity(k.width)) {
                d.availableWidth = k.width;
            }
            if (!isNaN_(k.height) && !isInfinity(k.height)) {
                d.availableHeight = k.height;
            }
            a.p(d, k);
            k.m = d.availableWidth;
            k.l = d.availableHeight;
            k.ai = d.renderOffsetX;
            k.aj = d.renderOffsetY;
        }
        if (this.c5.nv != null && this.c5.nv.passCompleted != null) {
            this.c5.nv.passCompleted(c);
        }
    };
    ScatterPolygonSeriesView.prototype.dc = function () {
        this.c5.nv = null;
        this.c5.nv = this.c5.nw;
    };
    ScatterPolygonSeriesView.prototype.da = function () {
        this.c5.n8 = null;
        this.c5.n8 = this.c5.oa;
    };
    ScatterPolygonSeriesView.prototype.db = function () {
        this.c5.n9 = null;
        this.c5.n9 = this.c5.ob;
    };
    ScatterPolygonSeriesView.prototype.c9 = function (a) {
        switch (a) {
            case "CircleMarkerTemplate":
                this.c5.nv = this.c5.bs.kn;
                break;
            case "TriangleMarkerTemplate":
                this.c5.nv = this.c5.bs.kw;
                break;
            case "PyramidMarkerTemplate":
                this.c5.nv = this.c5.bs.kt;
                break;
            case "SquareMarkerTemplate":
                this.c5.nv = this.c5.bs.ku;
                break;
            case "DiamondMarkerTemplate":
                this.c5.nv = this.c5.bs.ko;
                break;
            case "PentagonMarkerTemplate":
                this.c5.nv = this.c5.bs.kr;
                break;
            case "HexagonMarkerTemplate":
                this.c5.nv = this.c5.bs.kp;
                break;
            case "TetragramMarkerTemplate":
                this.c5.nv = this.c5.bs.kr;
                break;
            case "PentagramMarkerTemplate":
                this.c5.nv = this.c5.bs.ks;
                break;
            case "HexagramMarkerTemplate":
                this.c5.nv = this.c5.bs.kq;
                break;
        }
    };
    ScatterPolygonSeriesView.prototype.de = function () {
        var e_4, _a;
        try {
            for (var _b = tslib_1.__values(fromEnum(this.c8)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var a = _c.value;
                a.ah = this.c5.nv;
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_4) throw e_4.error; }
        }
        this.an();
    };
    ScatterPolygonSeriesView.prototype.dm = function (a) {
        if (this.dl == null) {
            this.dl = new DataTemplateMeasureInfo();
            this.dl.context = this.h.g();
        }
        this.dl.width = a.width;
        this.dl.height = a.height;
        this.dl.data = a.content;
        var b = a.ah;
        if (b.measure != null) {
            b.measure(this.dl);
        }
        return new Size(1, this.dl.width, this.dl.height);
    };
    ScatterPolygonSeriesView.prototype.c1 = function (a) {
        var b;
        var c;
        var d;
        var e;
        for (var f = this.c8.count - 1; f >= 0; f--) {
            var g = this.c8._inner[f];
            if (g._visibility == 1 || g._opacity == 0) {
                continue;
            }
            b = (g.m / 2) + 3;
            c = (g.l / 2) + 3;
            d = g.ai;
            e = g.aj;
            if ((g.n + d) - b <= a.x && (g.n + d) + b >= a.x && (g.o + e) - c <= a.y && (g.o + e) + c >= a.y) {
                return g;
            }
        }
        return null;
    };
    ScatterPolygonSeriesView.prototype.ai = function (a) {
        var e_5, _a;
        var b = a;
        _super.prototype.ai.call(this, b);
        try {
            for (var _b = tslib_1.__values(fromEnum(this.c6.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = this.c6.item(c);
                if (d._visibility != 1) {
                    var e = typeCast(ScatterPolygonSeries.$, this.e);
                    var f = this.e.a2.getMarkerVisualData(e, d, e.nv);
                    b.markerShapes.add(typeCast(MarkerVisualData.$, f));
                }
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_5) throw e_5.error; }
        }
    };
    ScatterPolygonSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.a0.a$q(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.c5.fx + "</span><br/>";
        a += "</div>";
        return a;
    };
    ScatterPolygonSeriesView.$t = markType(ScatterPolygonSeriesView, 'ScatterPolygonSeriesView', ShapeSeriesViewBase.$);
    return ScatterPolygonSeriesView;
}(ShapeSeriesViewBase));
export { ScatterPolygonSeriesView };
