// igniteui-react-charts/ES5/ScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { ScatterBase } from "./ScatterBase";
import { markType } from "igniteui-react-core/ES5/type";
import { ScatterFrame } from "./ScatterFrame";
import { ScatterBaseView } from "./ScatterBaseView";
import { Color } from "igniteui-react-core/ES5/Color";
/**
 * @hidden
 */
var ScatterSeries = /** @class */ (function (_super) {
    tslib_1.__extends(ScatterSeries, _super);
    function ScatterSeries() {
        var _this = _super.call(this) || this;
        _this.ab = ScatterSeries.$;
        _this.m8 = new ScatterFrame();
        _this.na = new ScatterFrame();
        _this.m7 = new ScatterFrame();
        return _this;
    }
    ScatterSeries.prototype.bf = function () {
        return new ScatterSeriesView(this);
    };
    ScatterSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.ow = a;
    };
    ScatterSeries.prototype.get_cy = function () {
        return false;
    };
    ScatterSeries.$t = markType(ScatterSeries, 'ScatterSeries', ScatterBase.$);
    return ScatterSeries;
}(ScatterBase));
export { ScatterSeries };
/**
 * @hidden
 */
var ScatterSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(ScatterSeriesView, _super);
    function ScatterSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.da = null;
        _this.da = a;
        return _this;
    }
    ScatterSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        if (!this.n) {
            this.by.l0 = 2;
            this.e.bj = 4;
        }
    };
    ScatterSeriesView.prototype.ac = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 3;
        this.e.ew = 2;
        this.e.ex = 2;
        this.e.du = false;
    };
    ScatterSeriesView.$t = markType(ScatterSeriesView, 'ScatterSeriesView', ScatterBaseView.$);
    return ScatterSeriesView;
}(ScatterBaseView));
export { ScatterSeriesView };
