// igniteui-react-charts/ES5/ShapeSeriesBase_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Series } from "./Series";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { IEnumerable$1_$type, Type, typeCast, Point_$type, fromEnum, Array_$type, Number_$type, markType, runOn } from "igniteui-react-core/ES5/type";
import { NumericXAxis } from "./NumericXAxis";
import { NumericYAxis } from "./NumericYAxis";
import { List$1 } from "igniteui-react-core/ES5/List$1";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { IFastItemsSource_$type } from "igniteui-react-core/ES5/IFastItemsSource";
import { PointCollectionUtil } from "igniteui-react-core/ES5/PointCollectionUtil";
import { FlattenedShape } from "./FlattenedShape";
import { SeriesView } from "./SeriesView";
import { Clipper } from "igniteui-react-core/ES5/Clipper";
import { ScalerParams } from "./ScalerParams";
import { Dictionary$2 } from "igniteui-react-core/ES5/Dictionary$2";
import { Tuple$2 } from "igniteui-react-core/ES5/Tuple$2";
import { PolySimplification } from "igniteui-react-core/ES5/PolySimplification";
import { DefaultFlattener } from "./DefaultFlattener";
import { DataContext } from "igniteui-react-core/ES5/DataContext";
import { FrameworkElement } from "igniteui-react-core/ES5/FrameworkElement";
import { AxisRange } from "./AxisRange";
import { StyleShapeEventArgs } from "./StyleShapeEventArgs";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { boxArray$1 } from "igniteui-react-core/ES5/array";
import { InvalidOperationException } from "igniteui-react-core/ES5/InvalidOperationException";
import { FastItemObjectColumn } from "igniteui-react-core/ES5/FastItemObjectColumn";
import { Path } from "igniteui-react-core/ES5/Path";
import { Pool$1 } from "igniteui-react-core/ES5/Pool$1";
import { ShapeHitRegion } from "./ShapeHitRegion";
import { PointCollection } from "igniteui-react-core/ES5/PointCollection";
import { PathGeometry } from "igniteui-react-core/ES5/PathGeometry";
import { PolyLineSegment } from "igniteui-react-core/ES5/PolyLineSegment";
import { PathFigure } from "igniteui-react-core/ES5/PathFigure";
import { GeometryUtil } from "igniteui-react-core/ES5/GeometryUtil";
import { PolygonUtil } from "igniteui-react-core/ES5/PolygonUtil";
import { RectUtil } from "igniteui-react-core/ES5/RectUtil";
import { truncate } from "igniteui-react-core/ES5/number";
/**
 * @hidden
 */
var ShapeSeriesBase = /** @class */ (function (_super) {
    tslib_1.__extends(ShapeSeriesBase, _super);
    function ShapeSeriesBase() {
        var _this = _super.call(this) || this;
        _this.l9 = null;
        _this.l4 = null;
        _this.l6 = null;
        _this.ml = 0;
        _this.mh = null;
        _this.l0 = null;
        _this.l2 = null;
        _this.l1 = null;
        _this.lz = null;
        _this.m6 = null;
        _this.mn = 0;
        _this.styleShape = null;
        FastItemObjectColumn.m();
        _this.mn = _this.mm;
        _this.m6 = Rect.empty;
        return _this;
    }
    ShapeSeriesBase.prototype.ht = function () {
        _super.prototype.ht.call(this);
        this.l5 = null;
        this.l7 = null;
    };
    ShapeSeriesBase.prototype.ag = function () {
        return [this.l5, this.l7];
    };
    ShapeSeriesBase.prototype.fs = function (a, b) {
        var c = this.gu(b);
        if (c == ShapeSeriesBase.$$p[1]) {
            return this.ft(a, b, this.ms);
        }
        return null;
    };
    Object.defineProperty(ShapeSeriesBase.prototype, "ma", {
        get: function () {
            return this.l9;
        },
        set: function (a) {
            var b = this.ma != a;
            if (b) {
                var c = this.ma;
                this.l9 = a;
                this.raisePropertyChanged("ShapeColumn", c, this.ma);
            }
        },
        enumerable: true,
        configurable: true
    });
    ShapeSeriesBase.prototype.md = function (array_) {
        var a = (Array.isArray(array_));
        return a;
    };
    ShapeSeriesBase.prototype.convertToListList = function (a) {
        if (a == null) {
            return null;
        }
        var b = new List$1(List$1.$.specialize(Point_$type), 0);
        var c;
        var curr_;
        var d;
        var e;
        for (var f = 0; f < a.length; f++) {
            c = a[f];
            var g = new List$1(Point_$type, 0);
            for (var h = 0; h < c.length; h++) {
                curr_ = c[h];
                d = (curr_.x);
                e = (curr_.y);
                g.add({ $type: Point_$type, x: d, y: e });
            }
            b.add(g);
        }
        return b;
    };
    ShapeSeriesBase.prototype.by = function (a) {
        if (this.bz != null && this.md(this.itemsSource) && a != null && a.split('!').length != 2) {
            if (this.coercionMethods == null) {
                this.coercionMethods = {};
            }
            var methods_ = this.coercionMethods;
            methods_.convertToListList = this.convertToListList;
            a += "!convertToListList";
        }
        return _super.prototype.by.call(this, a);
    };
    ShapeSeriesBase.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        var e = false;
        var f = false;
        switch (b) {
            case "FastItemsSource":
                var g = typeCast(IFastItemsSource_$type, c);
                if (g != null) {
                    g.deregisterColumn(this.ma);
                    this.ma = null;
                }
                var h = typeCast(IFastItemsSource_$type, d);
                if (h != null) {
                    this.ma = this.by(this.ms);
                }
                e = this.l5 != null && !this.l5.ck();
                f = this.l7 != null && !this.l7.ck();
                if (e || f) {
                    this.jc(false);
                }
                break;
            case ShapeSeriesBase.$$p[1]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.ma);
                    this.ma = this.by(this.ms);
                }
                break;
            case "ShapeColumn":
                this.mz();
                e = this.l5 != null && !this.l5.ck();
                f = this.l7 != null && !this.l7.ck();
                if (e || f) {
                    this.jc(false);
                }
                this.ij();
                break;
            case ShapeSeriesBase.$$p[2]:
            case ShapeSeriesBase.$$p[3]:
                this.l4 = this.l5;
                this.l6 = this.l7;
                var i = c;
                var j = d;
                this.m0(i);
                this.mx(j);
                if ((j != null && !j.ck()) || (j == null && i != null)) {
                    this.jc(false);
                }
                this.ij();
                break;
            case "Resolution":
                this.ml = this.eu;
                this.ij();
                break;
            case "ActualBrush":
                this.jc(false);
                break;
            case ShapeSeriesBase.$$p[0]:
                this.mn = this.mm;
                this.jc(false);
                this.ij();
                break;
            case "Thickness":
                this.jc(false);
                this.ij();
                break;
        }
    };
    ShapeSeriesBase.prototype.mc = function (a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    ShapeSeriesBase.prototype.mb = function (a) {
        if (typeCast(NumericXAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    ShapeSeriesBase.prototype.mz = function () {
        this.mh = null;
        if (this.ma == null) {
            return;
        }
        this.mh = new List$1(Rect.$, 2, this.ma.count);
        for (var a = 0; a < this.ma.count; a++) {
            this.mh.add(PointCollectionUtil.j(this.ma.item(a)));
        }
    };
    ShapeSeriesBase.prototype.m7 = function (a) {
        if (typeCast(IEnumerable$1_$type.specialize(IEnumerable$1_$type.specialize(Point_$type)), a) !== null) {
            return PointCollectionUtil.h(a);
        }
        else if (typeCast(IEnumerable$1_$type.specialize(Point_$type), a) !== null) {
            return PointCollectionUtil.g(a);
        }
        return Rect.empty;
    };
    ShapeSeriesBase.prototype.mj = function (a, b, c, d, e) {
        var e_1, _a;
        return this.mk(a, b, c, d, e);
        if (typeCast(List$1.$.specialize(List$1.$.specialize(Point_$type)), c) !== null) {
            return this.mk(a, b, c, d, e);
        }
        else if (typeCast(IEnumerable$1_$type.specialize(IEnumerable$1_$type.specialize(Point_$type)), c) !== null) {
            var f = c;
            var g = new List$1(FlattenedShape.$, 0);
            var h = null;
            try {
                for (var _b = tslib_1.__values(fromEnum(f)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var i = _c.value;
                    h = this.l3(a, b, PointCollectionUtil.b(i), d, e);
                    if (h == null) {
                        continue;
                    }
                    g.add(h);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return g;
        }
        else if (typeCast(IEnumerable$1_$type.specialize(Point_$type), c) !== null) {
            var j = c;
            var k = this.l3(a, b, PointCollectionUtil.b(j), d, e);
            var l = new List$1(FlattenedShape.$, 0);
            if (k != null) {
                l.add(k);
            }
            return l;
        }
        else {
            throw new InvalidOperationException(1, "Unrecognized data in ShapeColumn");
        }
    };
    ShapeSeriesBase.prototype.get_me = function () {
        return true;
    };
    Object.defineProperty(ShapeSeriesBase.prototype, "me", {
        get: function () {
            return this.get_me();
        },
        enumerable: true,
        configurable: true
    });
    ShapeSeriesBase.prototype.mg = function () {
        return false;
    };
    ShapeSeriesBase.prototype.jd = function (a) {
        this.my(this.bi);
    };
    ShapeSeriesBase.prototype.jb = function (a, b, c, d, e) {
        _super.prototype.jb.call(this, a, b, c, d, e);
        var f = this.a0.a$h.item(d);
        var g = f;
        f.ay(c);
        if (this.cc(f)) {
            return;
        }
        this.my(f);
    };
    ShapeSeriesBase.prototype.my = function (a) {
        if (this.cc(a)) {
            return;
        }
        var b;
        var c;
        var d = a;
        var e = a.getViewInfo(c, b);
        c = e.p0;
        b = e.p1;
        if (c.width < 1 || c.height < 1 || this.ma == null) {
            return;
        }
        var f = this.l5;
        var g = this.l7;
        var h = this.getEffectiveViewport1(a);
        var i = new ScalerParams(1, b, c, f.b8);
        i.c = h;
        var j = new ScalerParams(1, b, c, g.b8);
        j.c = h;
        d.cj();
        var k = new Rect(0, 0, 0, 0, 0);
        var l;
        var m;
        var n;
        var o;
        var p = this.ma.count;
        if (this.l2 == null || this.l2.length != p) {
            this.l2 = new Array(p);
            this.l0 = new Array(p);
            this.l1 = new Array(p);
            this.lz = new Array(p);
        }
        var q = this.l2;
        var r = this.l0;
        var s = this.l1;
        var t = this.lz;
        var u;
        for (var v = 0; v < p; v++) {
            u = this.mh._inner[v];
            q[v] = u.top;
            r[v] = u.left;
            s[v] = u.right;
            t[v] = u.bottom;
        }
        this.l4.gm(boxArray$1(r), 0, r.length, i);
        this.l4.gm(boxArray$1(s), 0, s.length, i);
        this.l6.gm(boxArray$1(q), 0, q.length, j);
        this.l6.gm(boxArray$1(t), 0, t.length, j);
        d.ck(r, q, s, t);
        var w = this.ma;
        var x = this.bz;
        this.m1(d);
        var y = c.left;
        var z = c.right;
        var aa = c.top;
        var ab = c.bottom;
        var ac = null;
        if (this.m6.isEmpty) {
            ac = null;
        }
        else {
            ac = new Clipper(1, this.m6.left, this.m6.bottom, this.m6.right, this.m6.top, this.me);
        }
        var ad = this.mg();
        if (ad) {
            d.b3.clear();
        }
        var ae;
        for (var af = 0; af < p; af++) {
            n = r[af];
            l = q[af];
            m = t[af];
            o = s[af];
            k.x = Math.min(n, o);
            k.y = Math.min(l, m);
            k.width = Math.max(n, o) - k.x;
            k.height = Math.max(l, m) - k.y;
            if (k.intersectsWith(c)) {
                if (k.x < y || k.y < aa || k.x > z || k.y > ab) {
                    ae = ac;
                }
                else {
                    ae = null;
                }
                var ag = this.mj(a, ae, w.item(af), i, j);
                if (ad) {
                    var ah = 0;
                    var ai = null;
                    for (var aj = 0; aj < ag.count; aj++) {
                        var ak = ag._inner[aj];
                        var al = ak.b.width * ak.b.height;
                        if (al > ah) {
                            ah = al;
                            ai = ak;
                        }
                    }
                    d.b3.addItem(af, ai);
                }
                d.b8(af, x.item(af), k, ag);
            }
        }
        d.ci();
    };
    ShapeSeriesBase.prototype.ia = function () {
        _super.prototype.ia.call(this);
        if (this.l5 != null) {
            this.l5.g4();
        }
        if (this.l7 != null) {
            this.l7.g4();
        }
    };
    ShapeSeriesBase.prototype.js = function (a, b) {
        this.jc(false);
    };
    ShapeSeriesBase.prototype.jr = function (a, b) {
        this.jc(false);
    };
    ShapeSeriesBase.prototype.hy = function (a, b, c, d) {
        _super.prototype.hy.call(this, a, b, c, d);
        switch (a) {
            case 3:
                if (d == this.ms) {
                    this.mh._inner[b] = this.m7(this.ma.item(b));
                }
                break;
            case 1:
                for (var e = b; e < b + c; e++) {
                    this.mh.insert(e, this.m7(this.ma.item(e)));
                }
                break;
            case 0:
                this.mh.v(b, c);
                break;
            case 2:
                for (var f = b; f < b + c; f++) {
                    this.mh._inner[f] = this.m7(this.ma.item(f));
                }
                break;
            case 4:
                this.mz();
                break;
        }
        this.jc(false);
    };
    ShapeSeriesBase.prototype.dw = function (a, b, c) {
        var d = true;
        if (!_super.prototype.dw.call(this, a, b, c) || (this.mf && this.ma == null) || b.isEmpty || a.isEmpty || this.l5 == null || this.l7 == null || this.l5.bf == null || this.l7.bf == null || this.l5.j3 == this.l5.j2 || this.l7.j3 == this.l7.j2 || this.ma == null || this.ma.count == 0 || this.mh == null || this.mh.count == 0 || this.mh.count != this.ma.count) {
            d = false;
        }
        return d;
    };
    ShapeSeriesBase.prototype.get_mf = function () {
        return true;
    };
    Object.defineProperty(ShapeSeriesBase.prototype, "mf", {
        get: function () {
            return this.get_mf();
        },
        enumerable: true,
        configurable: true
    });
    ShapeSeriesBase.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.l8 = typeCast(ShapeSeriesViewBase.$, a);
    };
    ShapeSeriesBase.prototype.mk = function (a, b, c, d, e) {
        var f = new List$1(FlattenedShape.$, 0);
        var g;
        var h = c.count;
        for (var i = 0; i < h; i++) {
            g = c._inner[i];
            var j = this.l3(a, b, g, d, e);
            if (j == null) {
                continue;
            }
            f.add(j);
        }
        return f;
    };
    ShapeSeriesBase.prototype.l3 = function (a, b, c, d, e) {
        var f = this.mw(c, d, e);
        var g = f.c;
        var h = f.d;
        var i = g.length;
        var j = 1.7976931348623157E+308;
        var k = 1.7976931348623157E+308;
        var l = -1.7976931348623157E+308;
        var m = -1.7976931348623157E+308;
        var n;
        var o;
        for (var p = 0; p < i; p++) {
            n = g[p];
            o = h[p];
            j = n < j ? n : j;
            k = o < k ? o : k;
            l = n > l ? n : l;
            m = o > m ? o : m;
        }
        var q = !(j > this.m6.right || l < this.m6.left || k > this.m6.bottom || m < this.m6.top);
        if (!q) {
            return null;
        }
        var r = b != null && (j < this.m6.left || k < this.m6.top || l > this.m6.right || m > this.m6.bottom);
        var s = new Rect(0, j, k, l - j, m - k);
        if (!a.b2(s)) {
            return null;
        }
        j = j < this.m6.left ? this.m6.left : j;
        l = l > this.m6.right ? this.m6.right : l;
        k = k < this.m6.top ? this.m6.top : k;
        m = m > this.m6.bottom ? this.m6.bottom : m;
        var t = PolySimplification.b(f, this.eu);
        var u = new DefaultFlattener().c(f.c, f.d, t, this.ml);
        var v = u.count;
        if (r) {
            var w = new List$1(Point_$type, 0);
            b.i = w;
            for (var x = 0; x < v; x++) {
                b.j(u._inner[x]);
            }
            b.i = null;
            u = w;
        }
        if (u == null || u.count == 0) {
            return null;
        }
        var y = new FlattenedShape();
        y.a = u;
        y.b = new Rect(0, j, k, l - j, m - k);
        y.c = s;
        return y;
    };
    ShapeSeriesBase.prototype.mw = function (a, b, c) {
        var d = this.l4;
        var e = this.l6;
        var f = a.count;
        var g = new Array(f);
        var h = new Array(f);
        for (var i = 0; i < f; i++) {
            g[i] = a._inner[i].x;
            h[i] = a._inner[i].y;
        }
        d.gm(boxArray$1(g), 0, g.length, b);
        e.gm(boxArray$1(h), 0, h.length, c);
        return new Tuple$2(Array_$type, Array_$type, g, h);
    };
    ShapeSeriesBase.prototype.m1 = function (a) {
        a.cp();
    };
    ShapeSeriesBase.prototype.hv = function (a, b) {
        _super.prototype.hv.call(this, a, b);
        var c = b;
        c.cd();
    };
    ShapeSeriesBase.prototype.mx = function (a) {
        if (a != null) {
            a.ch(this);
        }
    };
    ShapeSeriesBase.prototype.m0 = function (a) {
        if (a != null) {
            a.bv(this);
        }
    };
    ShapeSeriesBase.prototype.ar = function (a, b) {
        var c = this.l8.cr(a, b);
        var d = null;
        if (c != null) {
            d = c.dataContext;
        }
        if (d != null) {
            return d;
        }
        return _super.prototype.ar.call(this, a, b);
    };
    ShapeSeriesBase.prototype.mp = function (a) {
        return this.bz.item(a);
    };
    ShapeSeriesBase.prototype.get_dq = function () {
        return true;
    };
    ShapeSeriesBase.prototype.set_dq = function (a) {
        _super.prototype.set_dq.call(this, a);
    };
    ShapeSeriesBase.prototype.an = function (a) {
        var e_2, _a, e_3, _b;
        if (typeCast(NumericXAxis.$, a) !== null && this.mh != null) {
            var b = 1.7976931348623157E+308;
            var c = -1.7976931348623157E+308;
            try {
                for (var _c = tslib_1.__values(fromEnum(this.mh)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var d = _d.value;
                    b = Math.min(b, d.left);
                    c = Math.max(c, d.right);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return new AxisRange(b, c);
        }
        if (typeCast(NumericYAxis.$, a) !== null && this.mh != null) {
            var e = 1.7976931348623157E+308;
            var f = -1.7976931348623157E+308;
            try {
                for (var _e = tslib_1.__values(fromEnum(this.mh)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var g = _f.value;
                    e = Math.min(e, g.top);
                    f = Math.max(f, g.bottom);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                }
                finally { if (e_3) throw e_3.error; }
            }
            return new AxisRange(e, f);
        }
        return null;
    };
    ShapeSeriesBase.prototype.get_c5 = function () {
        return true;
    };
    ShapeSeriesBase.prototype.m8 = function (a, b, c) {
        if (this.styleShape != null) {
            var d = new StyleShapeEventArgs(a, b, c);
            this.styleShape(this, d);
            return d.d();
        }
        return null;
    };
    ShapeSeriesBase.$t = markType(ShapeSeriesBase, 'ShapeSeriesBase', Series.$);
    return ShapeSeriesBase;
}(Series));
export { ShapeSeriesBase };
Type.dep(DependencyProperty, PropertyMetadata, ShapeSeriesBase, 'raisePropertyChanged', ['ShapeFilterResolution:mm:m2', [1, 2], 'ShapeMemberPath:ms:m3', [2, null], 'XAxis:l5:m4', [NumericXAxis.$, null], 'YAxis:l7:m5', [NumericYAxis.$, null]]);
/**
 * @hidden
 */
var ShapeSeriesViewBase = /** @class */ (function (_super) {
    tslib_1.__extends(ShapeSeriesViewBase, _super);
    function ShapeSeriesViewBase(a) {
        var _this = _super.call(this, a) || this;
        _this.b3 = null;
        _this.by = null;
        _this.b4 = new List$1(ShapeHitRegion.$, 0);
        _this.b7 = 0;
        _this.b6 = null;
        _this.b1 = null;
        _this.bz = null;
        _this.cw = null;
        _this.bv = null;
        _this.bx = null;
        _this.bw = null;
        _this.bu = null;
        _this.by = a;
        _this.b1 = _this.b0();
        _this.b6 = new List$1(Path.$, 0);
        _this.b3 = new Dictionary$2(Number_$type, FlattenedShape.$, 0);
        return _this;
    }
    ShapeSeriesViewBase.prototype.b0 = function () {
        var _this = this;
        return ((function () {
            var $ret = new Pool$1(FrameworkElement.$);
            $ret.create = runOn(_this, _this.cq);
            $ret.activate = runOn(_this, _this.cf);
            $ret.disactivate = runOn(_this, _this.cg);
            $ret.destroy = runOn(_this, _this.ch);
            return $ret;
        })());
    };
    ShapeSeriesViewBase.prototype.b2 = function (a) {
        return a.width >= this.by.mn && a.height >= this.by.mn;
    };
    ShapeSeriesViewBase.b5 = function (a) {
        var b = new List$1(PointCollection.$, 0);
        var c;
        var d = a.count;
        for (var e = 0; e < d; e++) {
            c = a._inner[e];
            var f = new PointCollection(1, c.a);
            b.add(f);
        }
        return b;
    };
    ShapeSeriesViewBase.prototype.b8 = function (a, b, c, d) {
        if (d != null && d.count > 0) {
            var e = ShapeSeriesViewBase.b5(d);
            c.intersect(this.br);
            if (!this.b2(c)) {
            }
            else {
                var f = this.cu(a, e);
                if (f != null) {
                    var g = this.ct(this.b7, b);
                    for (var h = 0; h < d.count; h++) {
                        var i = new ShapeHitRegion();
                        i.d = d._inner[h].b;
                        i.a = d._inner[h].a;
                        i.b = a;
                        i.c = g;
                        this.b4.add(i);
                    }
                    this.ce(g);
                    this.ca(g, f);
                    this.cc(g, b);
                    this.b7 = this.b7 + 1;
                }
            }
        }
    };
    ShapeSeriesViewBase.prototype.ct = function (a, b) {
        var c = this.b1.item(this.b7);
        if (c.dataContext != null) {
            c.dataContext.item = b;
        }
        return c;
    };
    ShapeSeriesViewBase.cb = function (a, b, c) {
        if (b._fill == null) {
            b._fill = a.e.k2;
        }
        if (b._stroke == null) {
            b._stroke = a.e.k3;
        }
        if (b.y == null) {
            b.ac = a.e.ey;
        }
    };
    ShapeSeriesViewBase.prototype.cc = function (a, b) {
        if (a == null) {
            return;
        }
        if (this.bz != null) {
            a.y = this.bz.selectStyle(b, a);
        }
        a.y = this.by.m8(b, a, a.y);
        if (a.y == null) {
            a.y = this.cw;
        }
    };
    ShapeSeriesViewBase.prototype.ce = function (a) {
    };
    ShapeSeriesViewBase.prototype.b9 = function (a, b) {
        var c = truncate(Math.floor(a.left));
        var d = truncate(Math.ceil(a.right));
        var e = truncate(Math.floor(a.top));
        var f = truncate(Math.ceil(a.bottom));
        for (var g = c; g < d; g++) {
            for (var h = e; h < f; h++) {
                this.cm(g, h, b);
            }
        }
    };
    ShapeSeriesViewBase.prototype.cm = function (a, b, c) {
    };
    ShapeSeriesViewBase.prototype.cj = function () {
        this.b7 = 0;
    };
    ShapeSeriesViewBase.prototype.ci = function () {
        this.b1.count = this.b7;
        this.an();
    };
    ShapeSeriesViewBase.prototype.cl = function (a, b) {
    };
    ShapeSeriesViewBase.prototype.cd = function () {
        this.b1.count = 0;
        this.an();
    };
    ShapeSeriesViewBase.prototype.cq = function () {
        var a = this.by;
        var b = ((function () {
            var $ret = new Path();
            $ret.dataContext = ((function () {
                var $ret = new DataContext();
                $ret.series = a;
                return $ret;
            })());
            return $ret;
        })());
        this.b6.add(b);
        return b;
    };
    ShapeSeriesViewBase.prototype.cf = function (a) {
        a._visibility = 0;
    };
    ShapeSeriesViewBase.prototype.cg = function (a) {
        a.dataContext.item = null;
        a._visibility = 1;
    };
    ShapeSeriesViewBase.prototype.ch = function (a) {
        this.b6.remove(a);
    };
    ShapeSeriesViewBase.prototype.cv = function (a, b, c) {
        if (b == null) {
            return null;
        }
        var d = new PathGeometry();
        var e;
        var f = b.count;
        for (var g = 0; g < f; g++) {
            e = b._inner[g];
            if (e.count < 1) {
                continue;
            }
            var h = new PolyLineSegment();
            h._points = e;
            var i = new PathFigure();
            i._isFilled = c;
            i._isClosed = c;
            i._startPoint = e._inner[0];
            i._segments.add(h);
            d.b.add(i);
        }
        return d;
    };
    ShapeSeriesViewBase.prototype.bb = function (a, b) {
        _super.prototype.bb.call(this, a, b);
        var c = a;
        c._fill = this.e.k2;
        c._stroke = this.e.k3;
        c.ac = this.e.ey;
        if (c.y != null) {
            this.h.i(c, c.y);
        }
    };
    ShapeSeriesViewBase.prototype.bc = function (a, b) {
        _super.prototype.bc.call(this, a, b);
        var c = a;
        var d = this.bm(b);
        c._fill = d;
        c._stroke = d;
        c.ac = this.e.ey + 3;
    };
    ShapeSeriesViewBase.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        for (var c = 0; c < this.b6.count; c++) {
            var d = this.b6._inner[c];
            if (d._visibility != 1) {
                this.ba(d, c, b);
                a.s(d);
            }
        }
    };
    ShapeSeriesViewBase.prototype.co = function (a) {
        this.bz = a;
    };
    ShapeSeriesViewBase.prototype.cn = function (a) {
        this.cw = a;
    };
    ShapeSeriesViewBase.prototype.cr = function (a, b) {
        return this.cs(a, b, false);
    };
    ShapeSeriesViewBase.prototype.cs = function (a, b, c) {
        if (this.b4.count == 0) {
            return null;
        }
        var d = a.x;
        var e = a.y;
        var f = this.b4.count;
        var g = this.b4;
        var h = null;
        for (var i = 0; i < f; i++) {
            h = g._inner[i];
            if (h.a == null || h.a.count == 0) {
                continue;
            }
            if (d >= h.d.left && d <= h.d.right && e >= h.d.top && e <= h.d.bottom) {
                if (c) {
                    var j = this.by.ey / 2 + this.by.ee(b);
                    if (GeometryUtil.b(a, h.a._inner[0], h.a, j)) {
                        return h.c;
                    }
                }
                else {
                    if (PolygonUtil.a(h.a, a)) {
                        return h.c;
                    }
                }
            }
        }
        return null;
    };
    ShapeSeriesViewBase.prototype.ck = function (a, b, c, d) {
        this.bv = a;
        this.bx = b;
        this.bw = c;
        this.bu = d;
        this.b4.clear();
    };
    ShapeSeriesViewBase.prototype.cp = function () {
        var _this = this;
        var a = this.br;
        var b = ((function () {
            var $ret = new Path();
            $ret.y = _this.cw;
            return $ret;
        })());
        b.ac = 1;
        if (this.h != null) {
            this.h.i(b, b.y);
        }
        this.by.m6 = RectUtil.r(a, b.ac);
    };
    ShapeSeriesViewBase.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        if (this.b1.a == null) {
            return;
        }
        for (var b = 0; b < this.b1.a.count; b++) {
            var c = typeCast(Path.$, this.b1.a._inner[b]);
            if (c._visibility != 1) {
                this.e.a2.exportPathData(a, c, "shape" + b, ["Main "]);
            }
        }
    };
    ShapeSeriesViewBase.$t = markType(ShapeSeriesViewBase, 'ShapeSeriesViewBase', SeriesView.$);
    return ShapeSeriesViewBase;
}(SeriesView));
export { ShapeSeriesViewBase };
