// igniteui-react-charts/ES5/SplineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { SplineSeriesBase } from "./SplineSeriesBase";
import { CategoryLineRasterizer } from "./CategoryLineRasterizer";
import { typeCast, runOn, markType } from "igniteui-react-core/ES5/type";
import { Path } from "igniteui-react-core/ES5/Path";
import { ScalerParams } from "./ScalerParams";
import { SplineSeriesBaseView } from "./SplineSeriesBaseView";
import { ISortingAxis_$type } from "./ISortingAxis";
/**
 * @hidden
 */
var SplineSeries = /** @class */ (function (_super) {
    tslib_1.__extends(SplineSeries, _super);
    function SplineSeries() {
        var _this = _super.call(this) || this;
        _this.qq = null;
        _this.qq = new CategoryLineRasterizer();
        _this.ab = SplineSeries.$;
        return _this;
    }
    SplineSeries.prototype.bf = function () {
        return new SplineSeriesView(this);
    };
    SplineSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.qr = a;
    };
    SplineSeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    SplineSeries.prototype.hv = function (a, b) {
        _super.prototype.hv.call(this, a, b);
        var c = b;
        c.c8();
    };
    SplineSeries.prototype.on = function (a, b) {
        _super.prototype.on.call(this, a, b);
        var c = b;
        var d = c.cn.d;
        var e = typeCast(SplineSeriesView.$, b);
        var f = a.f;
        var g = this.getEffectiveViewport1(b);
        this.nd.x(this, this.ny(), this.px, runOn(this, this.mz), this.n6(b), this.n7(b));
        var h = false;
        var i = this.nd.c;
        if (i != null) {
            h = true;
        }
        if (h) {
            var j = new ScalerParams(0, b.bs, b.br, this.px.b8, g);
            this.oj(f, -1, this.valueColumn.count, this.px, j, b.n);
        }
        var k = e.de;
        var l = e.df;
        var m = e.dd;
        this.nd.ae(k, true, false, true, false);
        this.nd.ae(l, true, false, true, false);
        this.nd.ae(m, false, true, false, false);
        m._opacity = 0.75 * this.nd.i;
        if (b.cr(a)) {
            e.c9(f.count, f, true, 1, this.ov(f, f.count - 1, b.br, b.bs), d, this.eu);
            b.cu(a);
        }
    };
    Object.defineProperty(SplineSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.n1;
        },
        enumerable: true,
        configurable: true
    });
    SplineSeries.$t = markType(SplineSeries, 'SplineSeries', SplineSeriesBase.$);
    return SplineSeries;
}(SplineSeriesBase));
export { SplineSeries };
/**
 * @hidden
 */
var SplineSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(SplineSeriesView, _super);
    function SplineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c7 = null;
        _this.dh = new Path();
        _this.dg = new Path();
        _this.di = new Path();
        _this.dc = new Path();
        _this.db = new Path();
        _this.da = new Path();
        _this.c7 = a;
        return _this;
    }
    Object.defineProperty(SplineSeriesView.prototype, "de", {
        get: function () {
            return this.dh;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplineSeriesView.prototype, "df", {
        get: function () {
            return this.di;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplineSeriesView.prototype, "dd", {
        get: function () {
            return this.dg;
        },
        enumerable: true,
        configurable: true
    });
    SplineSeriesView.prototype.c8 = function () {
        this.dg.aj = null;
        this.dh.aj = null;
        this.di.aj = null;
        this.an();
    };
    SplineSeriesView.prototype.c9 = function (a, b, c, d, e, f, g) {
        this.c7.qq.a = typeCast(ISortingAxis_$type, this.cp.fetchXAxis()) !== null ? true : false;
        this.c7.qq.m(this.dh, this.dg, this.di, a, b, c, d, e, f, g);
        this.an();
    };
    SplineSeriesView.prototype.a9 = function () {
        _super.prototype.a9.call(this);
        this.db.aj = this.dh.aj;
        this.dc.aj = this.di.aj;
        this.da.aj = this.dg.aj;
        var a = this.bl();
        this.db._stroke = a;
        this.db.ac = this.e.ey + 3;
        this.dc._stroke = a;
        this.dc.ac = this.e.ey + 3;
        this.da._fill = a;
        this.da._opacity = 1;
    };
    SplineSeriesView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        if (a.d) {
            if (b) {
                a.s(this.da);
                a.s(this.db);
                a.s(this.dc);
            }
            else {
                a.s(this.dg);
                a.s(this.dh);
                a.s(this.di);
            }
        }
    };
    SplineSeriesView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        this.e.a2.exportPathData(a, this.dh, "lowerShape", ["Lower", "Main"]);
        this.e.a2.exportPathData(a, this.di, "upperShape", ["Upper"]);
        this.e.a2.exportPathData(a, this.dg, "translucentShape", ["Translucent"]);
    };
    SplineSeriesView.$t = markType(SplineSeriesView, 'SplineSeriesView', SplineSeriesBaseView.$);
    return SplineSeriesView;
}(SplineSeriesBaseView));
export { SplineSeriesView };
