// igniteui-react-charts/ES5/SplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { SplineSeriesBase } from "./SplineSeriesBase";
import { CategoryLineRasterizer } from "./CategoryLineRasterizer";
import { typeCast, runOn, markType } from "igniteui-react-core/ES5/type";
import { Path } from "igniteui-react-core/ES5/Path";
import { ScalerParams } from "./ScalerParams";
import { SplineSeriesBaseView } from "./SplineSeriesBaseView";
import { ISortingAxis_$type } from "./ISortingAxis";
import { Color } from "igniteui-react-core/ES5/Color";
/**
 * @hidden
 */
var SplineAreaSeries = /** @class */ (function (_super) {
    tslib_1.__extends(SplineAreaSeries, _super);
    function SplineAreaSeries() {
        var _this = _super.call(this) || this;
        _this.qq = null;
        _this.qq = new CategoryLineRasterizer();
        _this.ab = SplineAreaSeries.$;
        return _this;
    }
    SplineAreaSeries.prototype.bf = function () {
        return new SplineAreaSeriesView(this);
    };
    SplineAreaSeries.prototype.nf = function () {
        return 1;
    };
    SplineAreaSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.qr = a;
    };
    SplineAreaSeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    SplineAreaSeries.prototype.hv = function (a, b) {
        _super.prototype.hv.call(this, a, b);
        var c = b;
        c.c8();
    };
    SplineAreaSeries.prototype.dh = function (a, b) {
        var c = this.lk(this.ln(a), true, false);
        if (this.p7(a, c, b)) {
            return true;
        }
        if (this.dg(a, b)) {
            return true;
        }
        if (this.l3(a, b)) {
            return true;
        }
        return false;
    };
    SplineAreaSeries.prototype.on = function (a, b) {
        var _this = this;
        _super.prototype.on.call(this, a, b);
        var c = this.ne.cn.d;
        var d = typeCast(SplineAreaSeriesView.$, b);
        var e = a.f;
        var f = this.getEffectiveViewport1(b);
        this.nd.x(this, this.ny(), this.px, runOn(this, this.mz), this.n6(b), this.n7(b));
        var g = false;
        var h = this.nd.c;
        if (h != null) {
            g = true;
        }
        if (g) {
            var i = new ScalerParams(0, b.bs, b.br, this.px.b8, f);
            this.oj(e, -1, this.valueColumn.count, this.px, i, b.n);
        }
        var j = d.df;
        var k = d.dg;
        var l = d.di;
        var m = d.dk;
        this.nd.ae(j, true, false, true, true);
        this.nd.ae(k, true, false, true, true);
        this.nd.ae(l, false, true, false, false);
        this.nd.ae(m, false, true, false, false);
        l._opacity = this.nd.i * this.d6;
        m._opacity = 0.5 * this.nd.i * this.d6;
        if (b.cr(a)) {
            d.da(a.f.count, e, true, c, this.eu, function (n, o, p, q, r) { return _this.ph(n, a.f.count, b); }, 1, a.f.count == 1 ? null : this.ow(e, e.count - 1, b.br, b.bs, true));
            b.cu(a);
        }
        var n = this.fetchYAxis();
    };
    Object.defineProperty(SplineAreaSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.n1;
        },
        enumerable: true,
        configurable: true
    });
    SplineAreaSeries.$t = markType(SplineAreaSeries, 'SplineAreaSeries', SplineSeriesBase.$);
    return SplineAreaSeries;
}(SplineSeriesBase));
export { SplineAreaSeries };
/**
 * @hidden
 */
var SplineAreaSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(SplineAreaSeriesView, _super);
    function SplineAreaSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c7 = null;
        _this.dh = new Path();
        _this.dl = new Path();
        _this.dj = new Path();
        _this.dm = new Path();
        _this.db = new Path();
        _this.dd = new Path();
        _this.dc = new Path();
        _this.de = new Path();
        _this.c7 = a;
        return _this;
    }
    SplineAreaSeriesView.prototype.ap = function () {
        _super.prototype.ap.call(this);
        if (!this.n) {
            this.e.bj = 1;
        }
    };
    Object.defineProperty(SplineAreaSeriesView.prototype, "df", {
        get: function () {
            return this.dl;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplineAreaSeriesView.prototype, "dg", {
        get: function () {
            return this.dm;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplineAreaSeriesView.prototype, "di", {
        get: function () {
            return this.dh;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplineAreaSeriesView.prototype, "dk", {
        get: function () {
            return this.dj;
        },
        enumerable: true,
        configurable: true
    });
    SplineAreaSeriesView.prototype.c8 = function () {
        this.dh.aj = null;
        this.dj.aj = null;
        this.dl.aj = null;
        this.dm.aj = null;
        this.an();
    };
    SplineAreaSeriesView.prototype.c9 = function (a, b, c, d, e, f, g) {
        this.da(a, b, c, d, e, f, g, null);
    };
    SplineAreaSeriesView.prototype.da = function (a, b, c, d, e, f, g, h) {
        this.c7.qq.a = typeCast(ISortingAxis_$type, this.cp.fetchXAxis()) !== null ? true : false;
        this.c7.qq.j(this.dh, this.dl, this.dj, this.dm, a, b, c, d, e, f, g, h);
        this.an();
    };
    SplineAreaSeriesView.prototype.a9 = function () {
        _super.prototype.a9.call(this);
        this.db.aj = this.dh.aj;
        this.dd.aj = this.dl.aj;
        this.dc.aj = this.dj.aj;
        this.de.aj = this.dm.aj;
        var a = this.bl();
        this.db._fill = a;
        this.db._opacity = 1;
        this.dc._fill = a;
        this.dc._opacity = 1;
        this.dd._stroke = a;
        this.dd.ac = this.e.ey + 3;
        this.de._stroke = a;
        this.de.ac = this.e.ey + 3;
    };
    SplineAreaSeriesView.prototype.a3 = function (a, b) {
        _super.prototype.a3.call(this, a, b);
        if (b) {
            a.s(this.db);
            a.s(this.dc);
            a.s(this.dd);
            a.s(this.de);
        }
        else {
            a.s(this.dh);
            a.s(this.dj);
            a.s(this.dl);
            a.s(this.dm);
        }
    };
    SplineAreaSeriesView.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        this.e.a2.exportPathData(a, this.dl, "lowerShape", ["Lower"]);
        this.e.a2.exportPathData(a, this.dm, "upperShape", ["Upper", "Main"]);
        this.e.a2.exportPathData(a, this.dj, "translucentShape", ["Translucent"]);
        this.e.a2.exportPathData(a, this.dh, "fillShape", ["Fill"]);
    };
    SplineAreaSeriesView.prototype.ac = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 5;
        this.e.ew = 1;
        this.e.ex = -3;
    };
    SplineAreaSeriesView.$t = markType(SplineAreaSeriesView, 'SplineAreaSeriesView', SplineSeriesBaseView.$);
    return SplineAreaSeriesView;
}(SplineSeriesBaseView));
export { SplineAreaSeriesView };
