// igniteui-react-charts/ES5/Stacked100BarSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StackedBarSeries } from "./StackedBarSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core/ES5/type";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedBarSeriesView } from "./StackedBarSeriesView";
import { Stacked100BarBucketCalculator } from "./Stacked100BarBucketCalculator";
/**
 * @hidden
 */
var Stacked100BarSeries = /** @class */ (function (_super) {
    tslib_1.__extends(Stacked100BarSeries, _super);
    function Stacked100BarSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100BarSeries.$;
        return _this;
    }
    Stacked100BarSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.jc(false);
    };
    Stacked100BarSeries.prototype.bf = function () {
        return new Stacked100BarSeriesView(this);
    };
    Stacked100BarSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.p8 = a;
    };
    Stacked100BarSeries.prototype.ox = function () {
        return this.p8;
    };
    Stacked100BarSeries.prototype.pk = function () {
        _super.prototype.pk.call(this);
        Stacked100DataPreparer.a(this);
    };
    Object.defineProperty(Stacked100BarSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.n1;
        },
        enumerable: true,
        configurable: true
    });
    Stacked100BarSeries.$t = markType(Stacked100BarSeries, 'Stacked100BarSeries', StackedBarSeries.$, [IStacked100Series_$type]);
    return Stacked100BarSeries;
}(StackedBarSeries));
export { Stacked100BarSeries };
/**
 * @hidden
 */
var Stacked100BarSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(Stacked100BarSeriesView, _super);
    function Stacked100BarSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c2 = null;
        _this.c2 = a;
        return _this;
    }
    Stacked100BarSeriesView.prototype.co = function () {
        return new Stacked100BarBucketCalculator(this);
    };
    Stacked100BarSeriesView.$t = markType(Stacked100BarSeriesView, 'Stacked100BarSeriesView', StackedBarSeriesView.$);
    return Stacked100BarSeriesView;
}(StackedBarSeriesView));
export { Stacked100BarSeriesView };
