// igniteui-react-charts/ES5/Stacked100ColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StackedColumnSeries } from "./StackedColumnSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core/ES5/type";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedColumnSeriesView } from "./StackedColumnSeriesView";
import { Stacked100ColumnBucketCalculator } from "./Stacked100ColumnBucketCalculator";
/**
 * @hidden
 */
var Stacked100ColumnSeries = /** @class */ (function (_super) {
    tslib_1.__extends(Stacked100ColumnSeries, _super);
    function Stacked100ColumnSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100ColumnSeries.$;
        return _this;
    }
    Stacked100ColumnSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.jc(false);
    };
    Stacked100ColumnSeries.prototype.bf = function () {
        return new Stacked100ColumnSeriesView(this);
    };
    Stacked100ColumnSeries.prototype.it = function (a) {
        _super.prototype.it.call(this, a);
        this.p8 = a;
    };
    Stacked100ColumnSeries.prototype.ox = function () {
        return this.p8;
    };
    Stacked100ColumnSeries.prototype.pk = function () {
        _super.prototype.pk.call(this);
        Stacked100DataPreparer.a(this);
    };
    Object.defineProperty(Stacked100ColumnSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.n1;
        },
        enumerable: true,
        configurable: true
    });
    Stacked100ColumnSeries.$t = markType(Stacked100ColumnSeries, 'Stacked100ColumnSeries', StackedColumnSeries.$, [IStacked100Series_$type]);
    return Stacked100ColumnSeries;
}(StackedColumnSeries));
export { Stacked100ColumnSeries };
/**
 * @hidden
 */
var Stacked100ColumnSeriesView = /** @class */ (function (_super) {
    tslib_1.__extends(Stacked100ColumnSeriesView, _super);
    function Stacked100ColumnSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c1 = null;
        _this.c1 = a;
        return _this;
    }
    Stacked100ColumnSeriesView.prototype.co = function () {
        return new Stacked100ColumnBucketCalculator(this);
    };
    Stacked100ColumnSeriesView.$t = markType(Stacked100ColumnSeriesView, 'Stacked100ColumnSeriesView', StackedColumnSeriesView.$);
    return Stacked100ColumnSeriesView;
}(StackedColumnSeriesView));
export { Stacked100ColumnSeriesView };
