// igniteui-react-charts/ES5/StandardDeviationIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StandardDeviationIndicatorStrategy } from "./StandardDeviationIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var StandardDeviationIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(StandardDeviationIndicator, _super);
    function StandardDeviationIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StandardDeviationIndicator.prototype.get_qb = function () {
        return new StandardDeviationIndicatorStrategy();
    };
    Object.defineProperty(StandardDeviationIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    StandardDeviationIndicator.prototype.get_qo = function () {
        return StandardDeviationIndicator.$;
    };
    Object.defineProperty(StandardDeviationIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StandardDeviationIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(StandardDeviationIndicator.qx));
        },
        set: function (a) {
            this.h(StandardDeviationIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    StandardDeviationIndicator.prototype.qh = function () {
        return this.qw;
    };
    StandardDeviationIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    StandardDeviationIndicator.$t = markType(StandardDeviationIndicator, 'StandardDeviationIndicator', StrategyBasedIndicator.$);
    StandardDeviationIndicator.qx = StrategyBasedIndicator.qt(20, StandardDeviationIndicator.$);
    return StandardDeviationIndicator;
}(StrategyBasedIndicator));
export { StandardDeviationIndicator };
