// igniteui-react-charts/ES5/StochRSIIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StochRSIIndicatorStrategy } from "./StochRSIIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var StochRSIIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(StochRSIIndicator, _super);
    function StochRSIIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StochRSIIndicator.prototype.get_qb = function () {
        return new StochRSIIndicatorStrategy();
    };
    Object.defineProperty(StochRSIIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    StochRSIIndicator.prototype.get_qo = function () {
        return StochRSIIndicator.$;
    };
    Object.defineProperty(StochRSIIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StochRSIIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(StochRSIIndicator.qx));
        },
        set: function (a) {
            this.h(StochRSIIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    StochRSIIndicator.prototype.qh = function () {
        return this.qw;
    };
    StochRSIIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    StochRSIIndicator.$t = markType(StochRSIIndicator, 'StochRSIIndicator', StrategyBasedIndicator.$);
    StochRSIIndicator.qx = StrategyBasedIndicator.qt(14, StochRSIIndicator.$);
    return StochRSIIndicator;
}(StrategyBasedIndicator));
export { StochRSIIndicator };
