// igniteui-react-charts/ES5/TRIXIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { TRIXIndicatorStrategy } from "./TRIXIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
import { Brush } from "igniteui-react-core/ES5/Brush";
/**
 * @hidden
 */
var TRIXIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(TRIXIndicator, _super);
    function TRIXIndicator() {
        var _this = _super.call(this) || this;
        _this.pj = 0;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 9;
        _this.trendLinePeriod = 9;
        return _this;
    }
    TRIXIndicator.prototype.get_qb = function () {
        return new TRIXIndicatorStrategy();
    };
    Object.defineProperty(TRIXIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    TRIXIndicator.prototype.get_qo = function () {
        return TRIXIndicator.$;
    };
    Object.defineProperty(TRIXIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TRIXIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(TRIXIndicator.qx));
        },
        set: function (a) {
            this.h(TRIXIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    TRIXIndicator.prototype.qh = function () {
        return this.qw;
    };
    TRIXIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    TRIXIndicator.$t = markType(TRIXIndicator, 'TRIXIndicator', StrategyBasedIndicator.$);
    TRIXIndicator.qx = StrategyBasedIndicator.qt(15, TRIXIndicator.$);
    return TRIXIndicator;
}(StrategyBasedIndicator));
export { TRIXIndicator };
