// igniteui-react-charts/ES5/ValueBrushScale.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { BrushScale } from "./BrushScale";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { Type, typeGetValue, markType } from "igniteui-react-core/ES5/type";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { isNaN_, isInfinity, logBase } from "igniteui-react-core/ES5/number";
/**
 * @hidden
 */
var ValueBrushScale = /** @class */ (function (_super) {
    tslib_1.__extends(ValueBrushScale, _super);
    function ValueBrushScale() {
        var _this = _super.call(this) || this;
        _this.w = false;
        _this.aa = 10;
        return _this;
    }
    ValueBrushScale.prototype.getBrushByIndex = function (a, b) {
        if (b == null || this.brushes == null || this.brushes.count == 0 || a < 0 || a >= b.count) {
            return null;
        }
        if (b.count == 0) {
            return this.brushes.item(0);
        }
        var c = isNaN_(this.minimumValue) || isInfinity(this.minimumValue) ? b.minimum : this.minimumValue;
        var d = isNaN_(this.maximumValue) || isInfinity(this.maximumValue) ? b.maximum : this.maximumValue;
        var e = b.item(a);
        if (c == d) {
            return e == c ? this.brushes.item(0) : null;
        }
        return this.an(c, d, e);
    };
    ValueBrushScale.prototype.getBrushByValue = function (a, b) {
        if (b == null || this.brushes == null || this.brushes.count == 0) {
            return null;
        }
        if (b.count <= 1) {
            return this.brushes.item(0);
        }
        var c = isNaN_(this.minimumValue) || isInfinity(this.minimumValue) ? b.minimum : this.minimumValue;
        var d = isNaN_(this.maximumValue) || isInfinity(this.maximumValue) ? b.maximum : this.maximumValue;
        if (a < c) {
            return null;
        }
        if (a > d) {
            return null;
        }
        return this.an(c, d, a);
    };
    ValueBrushScale.prototype.an = function (a, b, c) {
        if (this.w && this.aa > 0) {
            var d = logBase(a, this.aa);
            var e = logBase(b, this.aa);
            var f = logBase(c, this.aa);
            return this.am(d, e, f);
        }
        return this.am(a, b, c);
    };
    ValueBrushScale.prototype.am = function (a, b, c) {
        if (c < a || c > b) {
            return null;
        }
        var d = (c - a) / (b - a);
        var e = d * (this.brushes.count - 1);
        return this.v(e);
    };
    ValueBrushScale.prototype.get_isReady = function () {
        return this.brushes != null && this.brushes.count > 0;
    };
    ValueBrushScale.prototype.q = function (a, b, c, d) {
        switch (b) {
            case ValueBrushScale.$$p[1]:
                this.aa = typeGetValue(d);
                break;
            case ValueBrushScale.$$p[0]:
                this.w = d;
                break;
        }
        _super.prototype.q.call(this, a, b, c, d);
    };
    ValueBrushScale.$t = markType(ValueBrushScale, 'ValueBrushScale', BrushScale.$);
    return ValueBrushScale;
}(BrushScale));
export { ValueBrushScale };
Type.dep(DependencyProperty, PropertyMetadata, ValueBrushScale, 'r', ['IsLogarithmic', [0, false], 'LogarithmBase', [1, 10], 'MaximumValue', [1, NaN], 'MinimumValue', [1, NaN]]);
