// igniteui-react-charts/ES5/VerticalAnchoredCategorySeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { AnchoredCategorySeries } from "./AnchoredCategorySeries";
import { NumericXAxis } from "./NumericXAxis";
import { DependencyProperty } from "igniteui-react-core/ES5/DependencyProperty";
import { CategoryYAxis } from "./CategoryYAxis";
import { IVerticalAnchoredCategorySeriesInteractionManager_$type } from "./IVerticalAnchoredCategorySeriesInteractionManager";
import { DVContainer } from "igniteui-react-core/ES5/DVContainer";
import { Type, runOn, typeCast, EnumUtil, markType } from "igniteui-react-core/ES5/type";
import { Axis } from "./Axis";
import { ScalerParams } from "./ScalerParams";
import { Rect } from "igniteui-react-core/ES5/Rect";
import { TrendLineType_$type } from "igniteui-react-core/ES5/TrendLineType";
import { CategoryTrendLineManagerFactory } from "./CategoryTrendLineManagerFactory";
import { PropertyMetadata } from "igniteui-react-core/ES5/PropertyMetadata";
import { truncate } from "igniteui-react-core/ES5/number";
/**
 * @hidden
 */
var VerticalAnchoredCategorySeries = /** @class */ (function (_super) {
    tslib_1.__extends(VerticalAnchoredCategorySeries, _super);
    function VerticalAnchoredCategorySeries() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.py = null;
        return _this;
    }
    VerticalAnchoredCategorySeries.prototype.ht = function () {
        _super.prototype.ht.call(this);
        this.xAxis = null;
        this.yAxis = null;
    };
    Object.defineProperty(VerticalAnchoredCategorySeries.prototype, "pz", {
        get: function () {
            var _this = this;
            if (this.py == null) {
                this.py = DVContainer.instance.createInstance(IVerticalAnchoredCategorySeriesInteractionManager_$type, function () { return _this.b2.n; });
            }
            return this.py;
        },
        enumerable: true,
        configurable: true
    });
    VerticalAnchoredCategorySeries.prototype.get_isVertical = function () {
        return true;
    };
    VerticalAnchoredCategorySeries.prototype.m5 = function () {
        return this.yAxis;
    };
    VerticalAnchoredCategorySeries.prototype.np = function () {
        return this.xAxis;
    };
    VerticalAnchoredCategorySeries.prototype.getOffsetValue = function () {
        return this.na.r(this.m5(), this.bi.bs, this.bi.br, this.getEffectiveViewport1(this.bi));
    };
    VerticalAnchoredCategorySeries.prototype.getCategoryWidth = function () {
        return this.yAxis.getCategorySize(this.bi.bs, this.bi.br, this.getEffectiveViewport1(this.bi));
    };
    VerticalAnchoredCategorySeries.prototype.fc = function (a, b) {
        return this.fd(a, b, this.m5(), runOn(this, this.n4), this.valueColumn);
    };
    VerticalAnchoredCategorySeries.prototype.fe = function (a, b) {
        return this.ff(a, b, this.m5(), runOn(this, this.n4), this.valueColumn);
    };
    VerticalAnchoredCategorySeries.prototype.d9 = function (a, b, c, d, e) {
        if (this.valueColumn == null) {
            return Number.POSITIVE_INFINITY;
        }
        return this.ea(a, b, this.yAxis, d, e, this.valueColumn.count, runOn(this, this.n4));
    };
    VerticalAnchoredCategorySeries.prototype.bd = function () {
        if (this.be == 0) {
            return 1;
        }
        else {
            return _super.prototype.bd.call(this);
        }
    };
    VerticalAnchoredCategorySeries.prototype.lv = function (a) {
        if (!this.hasMarkers || !this.l2()) {
            return Rect.empty;
        }
        return this.pz.e$b(this, this.m6.m, a, runOn(this, this.p7));
    };
    VerticalAnchoredCategorySeries.prototype.p7 = function (a) {
        return this.ox.c5(a);
    };
    VerticalAnchoredCategorySeries.prototype.dh = function (a, b) {
        if (this.dg(a, b)) {
            return true;
        }
        if (this.l3(a, b)) {
            return true;
        }
        return false;
    };
    VerticalAnchoredCategorySeries.prototype.en = function (a, b, c) {
        if (this.bs == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.bi);
        var e = this.m5();
        var f = new ScalerParams(0, this.bs.mq, this.bi.br, e.b8, d);
        var g = this.na.r(e, this.bs.mq, this.bi.br, d);
        return this.ep(this.valueColumn, a, e, f, g, runOn(this, this.n4), b, c);
    };
    VerticalAnchoredCategorySeries.prototype.lk = function (a, b, c) {
        var d = this.m5();
        return this.lm(a, b, c, this.na.r(d, this.bi.bs, this.bi.br, this.getEffectiveViewport1(this.bi)), this.np(), d, null, null, null);
    };
    VerticalAnchoredCategorySeries.prototype.fetchXAxis = function () {
        return this.xAxis;
    };
    VerticalAnchoredCategorySeries.prototype.fetchYAxis = function () {
        return this.yAxis;
    };
    VerticalAnchoredCategorySeries.prototype.pf = function (a) {
        this.xAxis = typeCast(NumericXAxis.$, a);
    };
    VerticalAnchoredCategorySeries.prototype.pg = function (a) {
        this.yAxis = typeCast(CategoryYAxis.$, a);
    };
    VerticalAnchoredCategorySeries.prototype.n0 = function () {
        return this.xAxis != null && this.xAxis.ck();
    };
    VerticalAnchoredCategorySeries.prototype.mz = function (a, b) {
        return this.ou(a, b, this.yAxis);
    };
    VerticalAnchoredCategorySeries.prototype.eb = function (a) {
        return this.pz.e$a(this, a);
    };
    VerticalAnchoredCategorySeries.prototype.n4 = function (a) {
        return this.ed(a, this.m5());
    };
    VerticalAnchoredCategorySeries.prototype.e8 = function (a) {
        return truncate(Math.round(this.eb(a)));
    };
    VerticalAnchoredCategorySeries.prototype.fp = function (a) {
        var b = this.e8(a);
        return b >= 0 && this.bz != null && b < this.bz.count ? this.bz.item(b) : null;
    };
    VerticalAnchoredCategorySeries.prototype.iy = function (a, b, c, d) {
        _super.prototype.iy.call(this, a, b, c, d);
        switch (b) {
            case VerticalAnchoredCategorySeries.$$p[0]:
                if (c != d) {
                    this.hz(typeCast(Axis.$, c));
                    this.i3(typeCast(Axis.$, d));
                    this.ne.cn.g(this.eu);
                    this.jc(false);
                    this.ij();
                }
                break;
            case VerticalAnchoredCategorySeries.$$p[1]:
                if (c != d) {
                    this.hz(typeCast(Axis.$, c));
                    this.i3(typeCast(Axis.$, d));
                    this.ne.cn.g(this.eu);
                    this.n0();
                    this.jc(false);
                    this.ij();
                }
                break;
            case "TrendLineType":
                var e = EnumUtil.getEnumValue(TrendLineType_$type, d);
                if (e != 0) {
                    this.ox.cy = CategoryTrendLineManagerFactory.a(this.ox.cy, this.yAxis, this.jw, this, this.b2);
                }
                break;
        }
    };
    VerticalAnchoredCategorySeries.prototype.p2 = function (a) {
        if (typeCast(CategoryYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    VerticalAnchoredCategorySeries.prototype.p1 = function (a) {
        if (typeCast(NumericXAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    Object.defineProperty(VerticalAnchoredCategorySeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.n1;
        },
        enumerable: true,
        configurable: true
    });
    VerticalAnchoredCategorySeries.$t = markType(VerticalAnchoredCategorySeries, 'VerticalAnchoredCategorySeries', AnchoredCategorySeries.$);
    return VerticalAnchoredCategorySeries;
}(AnchoredCategorySeries));
export { VerticalAnchoredCategorySeries };
Type.dep(DependencyProperty, PropertyMetadata, VerticalAnchoredCategorySeries, 'raisePropertyChanged', ['XAxis::p5', [NumericXAxis.$, null], 'YAxis::p6', [CategoryYAxis.$, null]]);
