// igniteui-react-charts/ES5/WilliamsPercentRIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { WilliamsPercentRIndicatorStrategy } from "./WilliamsPercentRIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES5/type";
/**
 * @hidden
 */
var WilliamsPercentRIndicator = /** @class */ (function (_super) {
    tslib_1.__extends(WilliamsPercentRIndicator, _super);
    function WilliamsPercentRIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    WilliamsPercentRIndicator.prototype.get_qb = function () {
        return new WilliamsPercentRIndicatorStrategy();
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "qb", {
        get: function () {
            return this.get_qb();
        },
        enumerable: true,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.get_qo = function () {
        return WilliamsPercentRIndicator.$;
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "qo", {
        get: function () {
            return this.get_qo();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "qw", {
        get: function () {
            return typeGetValue(this.c(WilliamsPercentRIndicator.qx));
        },
        set: function (a) {
            this.h(WilliamsPercentRIndicator.qx, a);
        },
        enumerable: true,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.qh = function () {
        return this.qw;
    };
    WilliamsPercentRIndicator.prototype.qq = function (a) {
        this.qw = a;
    };
    WilliamsPercentRIndicator.$t = markType(WilliamsPercentRIndicator, 'WilliamsPercentRIndicator', StrategyBasedIndicator.$);
    WilliamsPercentRIndicator.qx = StrategyBasedIndicator.qt(14, WilliamsPercentRIndicator.$);
    return WilliamsPercentRIndicator;
}(StrategyBasedIndicator));
export { WilliamsPercentRIndicator };
