// igniteui-react-charts/ES2015/AbsoluteVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { AbsoluteVolumeOscillatorIndicatorStrategy } from "./AbsoluteVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class AbsoluteVolumeOscillatorIndicator extends StrategyBasedIndicator {
    get_qb() {
        return new AbsoluteVolumeOscillatorIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return AbsoluteVolumeOscillatorIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get shortPeriod() {
        return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.qx));
    }
    set shortPeriod(a) {
        this.h(AbsoluteVolumeOscillatorIndicator.qx, a);
    }
    qj() {
        return this.shortPeriod;
    }
    get longPeriod() {
        return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.qw));
    }
    set longPeriod(a) {
        this.h(AbsoluteVolumeOscillatorIndicator.qw, a);
    }
    qg() {
        return this.longPeriod;
    }
}
AbsoluteVolumeOscillatorIndicator.$t = markType(AbsoluteVolumeOscillatorIndicator, 'AbsoluteVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
AbsoluteVolumeOscillatorIndicator.qx = StrategyBasedIndicator.qv(10, AbsoluteVolumeOscillatorIndicator.$);
AbsoluteVolumeOscillatorIndicator.qw = StrategyBasedIndicator.qr(30, AbsoluteVolumeOscillatorIndicator.$);
