// igniteui-react-charts/ES2015/BollingerBandWidthIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { BollingerBandWidthIndicatorStrategy } from "./BollingerBandWidthIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class BollingerBandWidthIndicator extends StrategyBasedIndicator {
    get_qb() {
        return new BollingerBandWidthIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return BollingerBandWidthIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qx() {
        return typeGetValue(this.c(BollingerBandWidthIndicator.q0));
    }
    set qx(a) {
        this.h(BollingerBandWidthIndicator.q0, a);
    }
    qh() {
        return this.qx;
    }
    get qw() {
        return this.c(BollingerBandWidthIndicator.qz);
    }
    set qw(a) {
        this.h(BollingerBandWidthIndicator.qz, a);
    }
    qf() {
        return this.qw;
    }
}
BollingerBandWidthIndicator.$t = markType(BollingerBandWidthIndicator, 'BollingerBandWidthIndicator', StrategyBasedIndicator.$);
BollingerBandWidthIndicator.q0 = StrategyBasedIndicator.qt(20, BollingerBandWidthIndicator.$);
BollingerBandWidthIndicator.qz = StrategyBasedIndicator.qs(2, BollingerBandWidthIndicator.$);
