// igniteui-react-charts/ES2015/BollingerBandsOverlay_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialOverlay } from "./FinancialOverlay";
import { DependencyProperty } from "igniteui-react-core/ES2015/DependencyProperty";
import { List$1 } from "igniteui-react-core/ES2015/List$1";
import { Type, Number_$type, typeCast, runOn, markType, getEnumerator, Point_$type, fromEnum } from "igniteui-react-core/ES2015/type";
import { CategoryFrame } from "./CategoryFrame";
import { Series } from "./Series";
import { SafeEnumerable } from "./SafeEnumerable";
import { TrendCalculators } from "igniteui-react-core/ES2015/TrendCalculators";
import { AxisRange } from "./AxisRange";
import { FinancialSeriesView } from "./FinancialSeriesView";
import { ScalerParams } from "./ScalerParams";
import { ISortingAxis_$type } from "./ISortingAxis";
import { FinancialBucketCalculator } from "./FinancialBucketCalculator";
import { Polyline } from "igniteui-react-core/ES2015/Polyline";
import { Polygon } from "igniteui-react-core/ES2015/Polygon";
import { PropertyMetadata } from "igniteui-react-core/ES2015/PropertyMetadata";
import { isNaN_ } from "igniteui-react-core/ES2015/number";
import { Flattener } from "igniteui-react-core/ES2015/Flattener";
/**
 * @hidden
 */
export class BollingerBandsOverlay extends FinancialOverlay {
    constructor() {
        super();
        this.pp = 0;
        this.pq = 0;
        this.pn = new List$1(Number_$type, 0);
        this.po = new List$1(Number_$type, 0);
        this.ab = BollingerBandsOverlay.$;
        this.ma = new CategoryFrame(4);
        this.mc = new CategoryFrame(4);
        this.l9 = new CategoryFrame(4);
    }
    iy(a, b, c, d) {
        if (this.nq().contains(b)) {
            this.ph = false;
        }
        super.iy(a, b, c, d);
        switch (b) {
            case BollingerBandsOverlay.$$p[1]:
                this.ph = false;
                this.jc(false);
                break;
            case BollingerBandsOverlay.$$p[0]:
                this.jc(false);
                break;
        }
    }
    pi() {
        this.pn.clear();
        this.po.clear();
        let a = getEnumerator(TrendCalculators.e(new SafeEnumerable(this.np), this.ps));
        let b = getEnumerator(TrendCalculators.f(new SafeEnumerable(this.np), this.ps));
        this.pq = 1.7976931348623157E+308;
        this.pp = -1.7976931348623157E+308;
        let c = true;
        let d = true;
        let e = this.pr;
        while (c || d) {
            if (a.moveNext()) {
                this.pn.add(a.current);
            }
            else {
                c = false;
            }
            if (b.moveNext()) {
                this.po.add(b.current);
            }
            else {
                d = false;
            }
            if (c && d) {
                this.pq = Math.min(this.pq, a.current - b.current * e);
                this.pp = Math.max(this.pp, a.current + b.current * e);
            }
        }
        return true;
    }
    an(a) {
        if (this.pn == null || this.po == null || a == null || this.bz == null || a != this.ms) {
            return null;
        }
        let b = new AxisRange(this.pq, this.pp);
        return b;
    }
    oy(a, b) {
        super.oy(a, b);
        let c = b.bs;
        let d = b.br;
        let e = this.getEffectiveViewport1(b);
        let f = this.l7;
        let g = this.ms;
        let h = new ScalerParams(0, c, d, f.b8, e);
        let i = new ScalerParams(0, c, d, g.b8, e);
        a.f.clear();
        a.m.clear();
        a.p.clear();
        let j = typeCast(ISortingAxis_$type, this.l7);
        if (j != null && j.i$b.count != this.bz.count) {
            return;
        }
        let k = 0;
        k = this.ny(c, d, e);
        if (!this.ph) {
            this.ph = this.pi();
            if (this.ms != null) {
                this.ms.cl(true);
            }
        }
        i.b = this.eh();
        let l = this.l7.dy(2, h) - this.l7.dy(1, h);
        for (let m = b.bu.c; m <= b.bu.d; ++m) {
            let n;
            if (j == null) {
                n = b.bu.getBucket(m);
            }
            else {
                let o = j.i$b._inner[m];
                let p = j.i$d(o);
                let q = this.pn._inner[m];
                let r = this.po._inner[m];
                let s = q;
                let t = r;
                let u = p;
                let v = 1;
                while (m < b.bu.d) {
                    o = j.i$b._inner[m + 1];
                    u = j.i$d(o);
                    if (u - p > l) {
                        break;
                    }
                    m++;
                    s += this.pn._inner[m];
                    t += this.po._inner[m];
                    v++;
                }
                s /= v;
                t /= v;
                let w = s - t * this.pr;
                let x = s;
                let y = s + t * this.pr;
                let z = NaN;
                if (!isNaN_(p)) {
                    z = this.l7.dv(p, h);
                }
                n = [z, w, x, y];
            }
            let aa = Math.max(1, l);
            if (!isNaN_(n[0]) && m * aa >= this.pj) {
                if (this.l7 != null && typeCast(ISortingAxis_$type, this.l7) !== null) {
                    n[0] = n[0] + k;
                }
                else {
                    n[0] = f.dv(n[0], h) + k;
                }
                n[1] = g.dv(n[1], i);
                n[2] = g.dv(n[2], i);
                n[3] = g.dv(n[3], i);
                a.f.add(n);
            }
        }
    }
    hv(a, b) {
        super.hv(a, b);
        let c = b;
        if (c != null) {
            c.b1();
        }
    }
    en(a, b, c) {
        if (this.pn == null) {
            return super.en(a, b, c);
        }
        let d = this.getEffectiveViewport1(this.bi);
        let e = this.ny(this.bi.bs, this.bi.br, d);
        let f = new ScalerParams(0, this.bi.bs, this.bi.br, this.l7.b8, d);
        f.c = this.bs.m3;
        return this.ep(this.pn, a, this.l7, f, e, runOn(this, this.nx), b, c);
    }
    fe(a, b) {
        if (this.pn == null) {
            return super.fe(a, b);
        }
        return this.ff(a, b, this.l7, runOn(this, this.nx), this.pn);
    }
    fc(a, b) {
        if (this.pn == null) {
            return super.fc(a, b);
        }
        return this.fd(a, b, this.l7, runOn(this, this.nx), this.pn);
    }
    dw(a, b, c) {
        let d = super.dw(a, b, c);
        if (!this.ng(this.nq())) {
            d = false;
        }
        return d;
    }
    o0(a, b) {
        super.o0(a, b);
        let c = typeCast(BollingerBandsOverlayView.$, b);
        let d = this.getEffectiveViewport1(b);
        if (b.bx(a)) {
            c.b1();
            let e = a.f.count;
            let f = (g) => a.f._inner[g][0];
            let g = (h) => a.f._inner[e - 1 - h][0];
            let h = (i) => a.f._inner[i][1];
            let i = (j) => a.f._inner[j][2];
            let j = (k) => a.f._inner[e - 1 - k][3];
            c.b2(e, f, g, h, i, j);
            b.bz(a);
        }
        this.mf.x(this, this.nf(), this.l7, runOn(this, this.lz), this.n4(b), this.n6(b));
        let k = false;
        let l = this.mf.c;
        let m = a.f;
        let n = this.bz.count;
        if (l != null) {
            k = true;
        }
        if (k) {
            let o = new ScalerParams(0, b.bs, b.br, this.l7.b8, d);
            this.ox(m, -1, n, this.l7, o, b.n);
        }
        let p = c.b9;
        let q = c.ca;
        let r = c.cb;
        let s = c.b4;
        this.mf.ae(q, true, false, true, true);
        this.mf.ao = null;
        this.mf.an = null;
        this.mf.ae(p, true, false, false, true);
        this.mf.ae(r, true, false, false, true);
        this.mf.ae(s, false, true, false, false);
        s._opacity = this.mf.i * this.d6;
    }
    bf() {
        return new BollingerBandsOverlayView(this);
    }
    it(a) {
        super.it(a);
        this.pm = typeCast(BollingerBandsOverlayView.$, a);
    }
    n7() {
        return 3;
    }
    eh() {
        return Series.ei(this.pn, null);
    }
}
BollingerBandsOverlay.$t = markType(BollingerBandsOverlay, 'BollingerBandsOverlay', FinancialOverlay.$);
Type.dep(DependencyProperty, PropertyMetadata, BollingerBandsOverlay, 'raisePropertyChanged', ['Multiplier:pr:pv', [1, 2], 'Period:ps:pw', [1, 14]]);
/**
 * @hidden
 */
export class BollingerBandsBucketCalculator extends FinancialBucketCalculator {
    constructor(a) {
        super(a);
        this.f = null;
        this.f = a;
    }
    getBucket(a) {
        let b = a * this.b;
        let c = Math.min(b + this.b - 1, this.a.bw.bz.count - 1);
        if (b <= c) {
            let d = this.f.b0.pr;
            let e = 0;
            let f = 0;
            let g = 0;
            for (let h = b; h <= c; ++h) {
                if (!isNaN_(this.f.b0.pn._inner[h]) && !isNaN_(this.f.b0.po._inner[h])) {
                    e += this.f.b0.pn._inner[h];
                    f += this.f.b0.po._inner[h];
                    ++g;
                }
            }
            if (g > 0) {
                e = e / g;
                f = f / g;
                return [0.5 * (b + c), e - f * d, e, e + f * d];
            }
        }
        return [NaN, NaN, NaN, NaN];
    }
}
BollingerBandsBucketCalculator.$t = markType(BollingerBandsBucketCalculator, 'BollingerBandsBucketCalculator', FinancialBucketCalculator.$);
/**
 * @hidden
 */
export class BollingerBandsOverlayView extends FinancialSeriesView {
    constructor(a) {
        super(a);
        this.b0 = null;
        this.b5 = new Polygon();
        this.cc = new Polyline();
        this.cd = new Polyline();
        this.ce = new Polyline();
        this.b3 = new Polygon();
        this.b6 = new Polyline();
        this.b7 = new Polyline();
        this.b8 = new Polyline();
        this.b0 = a;
    }
    ap() {
        super.ap();
        if (!this.n) {
            this.e.bj = 1;
        }
    }
    bv() {
        return new BollingerBandsBucketCalculator(this);
    }
    b2(a, b, c, d, e, f) {
        for (let g of fromEnum(Flattener.d(a, b, d, this.e.eu))) {
            this.b5.aj.add({ $type: Point_$type, x: b(g), y: d(g) });
            this.cc.aj.add({ $type: Point_$type, x: b(g), y: d(g) });
        }
        for (let h of fromEnum(Flattener.d(a, b, e, this.e.eu))) {
            this.cd.aj.add({ $type: Point_$type, x: b(h), y: e(h) });
        }
        for (let i of fromEnum(Flattener.d(a, c, f, this.e.eu))) {
            this.b5.aj.add({ $type: Point_$type, x: c(i), y: f(i) });
            this.ce.aj.add({ $type: Point_$type, x: c(i), y: f(i) });
        }
        this.an();
    }
    b1() {
        this.b5.aj.clear();
        this.cc.aj.clear();
        this.cd.aj.clear();
        this.ce.aj.clear();
        this.an();
    }
    get b4() {
        return this.b5;
    }
    get b9() {
        return this.cc;
    }
    get ca() {
        return this.cd;
    }
    get cb() {
        return this.ce;
    }
    a9() {
        super.a9();
        this.b3.aj = this.b5.aj;
        this.b6.aj = this.cc.aj;
        this.b7.aj = this.cd.aj;
        this.b8.aj = this.ce.aj;
        let a = this.bl();
        this.b3._fill = a;
        this.b3._opacity = 1;
        this.b6._stroke = a;
        this.b6.ac = this.e.ey + 3;
        this.b7._stroke = a;
        this.b7.ac = this.e.ey + 3;
        this.b8._stroke = a;
        this.b8.ac = this.e.ey + 3;
    }
    a3(a, b) {
        super.a3(a, b);
        if (a.d) {
            if (b) {
                a.t(this.b3);
                a.u(this.b6);
                a.u(this.b7);
                a.u(this.b8);
            }
            else {
                a.t(this.b5);
                a.u(this.cc);
                a.u(this.cd);
                a.u(this.ce);
            }
        }
    }
    ai(a) {
        super.ai(a);
        this.e.a2.exportPolygonData(a, this.b5, "FillShape", ["Fill"]);
        this.e.a2.exportPolylineData(a, this.cc, "BottomLine", ["Lower", "Main"]);
        this.e.a2.exportPolylineData(a, this.cd, "CentralLine", ["Central"]);
        this.e.a2.exportPolylineData(a, this.ce, "TopLine", ["Upper"]);
    }
}
BollingerBandsOverlayView.$t = markType(BollingerBandsOverlayView, 'BollingerBandsOverlayView', FinancialSeriesView.$);
