// igniteui-react-charts/ES2015/CategoryItemHighlightLayer_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayer } from "./AnnotationLayer";
import { Series } from "./Series";
import { DependencyProperty } from "igniteui-react-core/ES2015/DependencyProperty";
import { CategoryItemHighlightType_$type } from "./CategoryItemHighlightType";
import { DataTemplate } from "igniteui-react-core/ES2015/DataTemplate";
import { CategoryItemHighlightLayerFrame } from "./CategoryItemHighlightLayerFrame";
import { Type, Base, enumGetBox, fromEnum, typeCast, Point_$type, markType, runOn } from "igniteui-react-core/ES2015/type";
import { Dictionary$2 } from "igniteui-react-core/ES2015/Dictionary$2";
import { ItemHighlightFrameRect } from "./ItemHighlightFrameRect";
import { List$1 } from "igniteui-react-core/ES2015/List$1";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { Rect } from "igniteui-react-core/ES2015/Rect";
import { MarkerSeries } from "./MarkerSeries";
import { DataContext } from "igniteui-react-core/ES2015/DataContext";
import { Pool$1 } from "igniteui-react-core/ES2015/Pool$1";
import { Rectangle } from "igniteui-react-core/ES2015/Rectangle";
import { ContentControl } from "igniteui-react-core/ES2015/ContentControl";
import { PropertyMetadata } from "igniteui-react-core/ES2015/PropertyMetadata";
import { isNaN_, isInfinity, truncate } from "igniteui-react-core/ES2015/number";
import { AnnotationLayerView } from "./AnnotationLayerView";
import { Marker } from "./Marker";
import { BrushUtil } from "igniteui-react-core/ES2015/BrushUtil";
import { Color } from "igniteui-react-core/ES2015/Color";
import { DataTemplatePassInfo } from "igniteui-react-core/ES2015/DataTemplatePassInfo";
import { DataTemplateRenderInfo } from "igniteui-react-core/ES2015/DataTemplateRenderInfo";
import { DataTemplateMeasureInfo } from "igniteui-react-core/ES2015/DataTemplateMeasureInfo";
/**
 * @hidden
 */
export class CategoryItemHighlightLayer extends AnnotationLayer {
    constructor() {
        super();
        this.m8 = false;
        this.m7 = true;
        this.nb = 10;
        let a = new CategoryItemHighlightLayerFrame();
        let b = new CategoryItemHighlightLayerFrame();
        let c = new CategoryItemHighlightLayerFrame();
        this.l1 = a;
        this.l0 = b;
        this.l2 = c;
        this.ab = CategoryItemHighlightLayer.$;
    }
    bf() {
        return new CategoryItemHighlightLayerView(this);
    }
    it(a) {
        super.it(a);
        this.m3 = a;
    }
    mk(a, b) {
        if (this.m6 == null || this.m6 == a) {
            if (!this.aq.c) {
                this.jc(b);
            }
        }
    }
    l7(a) {
        if (!super.l7(a)) {
            return false;
        }
        let b = this.mf(a);
        if (!b.c) {
            return false;
        }
        return true;
    }
    mm(a, b) {
        super.mm(a, b);
        let c = a;
        let d = this.na;
        c.g.clear();
        if (isNaN_(this.m0.x) && isNaN_(this.m0.y)) {
            return;
        }
        if (this.l7(this.m6)) {
            this.nl(this.m6, c, d);
        }
        else {
            for (let e of fromEnum(this.bs.series)) {
                if (this.l7(e)) {
                    this.nl(e, c, d);
                }
            }
        }
    }
    iy(a, b, c, d) {
        super.iy(a, b, c, d);
        switch (b) {
            case CategoryItemHighlightLayer.$$p[4]:
                this.jc(true);
                break;
            case CategoryItemHighlightLayer.$$p[5]:
                this.m8 = d;
                this.jc(true);
                break;
            case CategoryItemHighlightLayer.$$p[1]:
                this.jc(true);
                break;
            case CategoryItemHighlightLayer.$$p[0]:
                this.nb = this.nc;
                this.jc(true);
                break;
            case CategoryItemHighlightLayer.$$p[3]:
                this.m7 = d;
                this.jc(true);
                break;
        }
    }
    nl(a, b, c) {
        if (this.bi == null || this.bi.bs.isEmpty || this.bi.br.isEmpty) {
            return;
        }
        if (!a.cl && !a.cr) {
            return;
        }
        let d = a;
        if (d.categoryAxis == null) {
            return;
        }
        let e = this.m0.x;
        e = this.bi.br.left + this.bi.br.width * (e - this.bi.bs.left) / this.bi.bs.width;
        let f = this.m0.y;
        f = this.bi.br.top + this.bi.br.height * (f - this.bi.bs.top) / this.bi.bs.height;
        let g = this.m9;
        let h = true;
        let i = typeCast(IHasCategoryModePreference_$type, a);
        let j = 0;
        if (i != null) {
            j = i.preferredCategoryMode(d.categoryAxis);
            if (j == 2) {
                h = false;
            }
        }
        let k = true;
        if (!h || this.m4 == 2) {
            k = false;
        }
        if (this.m4 == 1) {
            k = true;
        }
        let l = this.mf(a);
        if (!l.c) {
            return;
        }
        if (k) {
            let m = a.lk(l.d, this.na, g);
            if (j != 0 && !d.categoryAxis.cc) {
                let n = m.x;
                if (a.cr) {
                    n -= a.getOffsetValue();
                    n += a.getCategoryWidth() * 0.5;
                    m = { $type: Point_$type, x: n, y: m.y };
                }
                else if (a.cl) {
                    n -= a.getOffsetValue();
                    n += a.getCategoryWidth() * 0.5;
                    m = { $type: Point_$type, x: n, y: m.y };
                }
            }
            if (!isNaN_(m.x) && !isNaN_(m.y)) {
                m = this.ln(m);
            }
            this.nj(m, b, a);
        }
        else {
            let o = a.lk(l.d, this.na, g);
            let p;
            let q = this.getEffectiveViewport1(this.bi);
            if (a.isVertical) {
                let r = o.y;
                let s = d.categoryAxis;
                let t = this.nb;
                if (j != 0) {
                    t = s.jq(this.bi.bs, this.bi.br, q);
                }
                if (isNaN_(t) || isInfinity(t)) {
                    return;
                }
                p = new Rect(0, this.lx.left, r - t / 2, this.lx.width, t);
            }
            else {
                let u = o.x;
                let v = d.categoryAxis;
                let w = this.nb;
                if (j != 0) {
                    w = v.jq(this.bi.bs, this.bi.br, q);
                }
                if (isNaN_(w) || isInfinity(w)) {
                    return;
                }
                p = new Rect(0, u - w / 2, this.lx.top, w, this.lx.height);
            }
            p.intersect(this.bi.br);
            this.nk(p, b, a);
        }
    }
    nj(a, b, c) {
        let d = new ItemHighlightFrameRect();
        d.b = a.x;
        d.e = a.y;
        d.i = this.my();
        d.j = this.mz();
        if (d.i == null) {
            if (c.hasMarkers) {
                d.i = this.m3.cf(c.k5());
            }
            else {
                d.i = this.m3.cf(c.k2);
            }
        }
        if (d.j == null) {
            if (c.hasMarkers) {
                d.j = this.m3.cf(c.k6());
            }
            else {
                d.j = this.m3.cf(c.k3);
            }
        }
        d.d = this.ey;
        d.f = Base.getHashCodeStatic(c);
        d.g = this.m5(c, a);
        let e = (typeCast(MarkerSeries.$, c) !== null) ? c.mh : null;
        if (e == null || e == MarkerSeries.mk) {
            let f = MarkerSeries.l1(c, 2);
            switch (f) {
                case 3:
                case 13:
                    e = this.bs.kn;
                    break;
                case 4:
                    e = this.bs.kw;
                    break;
                case 5:
                    e = this.bs.kt;
                    break;
                case 6:
                    e = this.bs.ku;
                    break;
                case 7:
                    e = this.bs.ko;
                    break;
                case 8:
                    e = this.bs.kr;
                    break;
                case 9:
                    e = this.bs.kp;
                    break;
                case 10:
                    e = this.bs.kv;
                    break;
                case 11:
                    e = this.bs.ks;
                    break;
                case 12:
                    e = this.bs.kq;
                    break;
            }
        }
        d.h = this.no != null ? this.no : e;
        if (!isNaN_(d.b) && !isNaN_(d.e)) {
            b.g.addItem(d.f, d);
        }
    }
    m5(a, b) {
        let c = a.fp(b);
        let d = a.eb(b);
        if (this.na && d != Math.floor(d)) {
            c = null;
        }
        return ((() => {
            let $ret = new DataContext();
            $ret.series = a;
            $ret.item = c;
            return $ret;
        })());
    }
    nk(a, b, c) {
        let d = new ItemHighlightFrameRect();
        d.b = a.left;
        d.e = a.top;
        d.c = a.right;
        d.a = a.bottom;
        d.i = this.my();
        d.j = this.mz();
        if (d.i == null) {
            if (c != null) {
                d.i = this.m3.cg(c.k2);
            }
        }
        if (d.j == null) {
            if (c != null) {
                d.j = this.m3.cg(c.k3);
            }
        }
        d.d = this.ey;
        d.f = Base.getHashCodeStatic(c);
        if (!a.isEmpty) {
            b.g.addItem(d.f, d);
        }
    }
    mp(a, b) {
        super.mp(a, b);
        let c = this.m3.b3;
        let d = this.m3.b2;
        let e = 0;
        let f = 0;
        let g = a;
        for (let h of fromEnum(g.g.values)) {
            if (h.h != null) {
                let i = d.item(f);
                this.nm(h, i);
                f++;
            }
            else {
                let j = c.item(e);
                this.nn(h, j);
                e++;
            }
        }
        c.count = e;
        d.count = f;
    }
    nm(a, b) {
        if (isNaN_(a.b) || isNaN_(a.e) || isNaN_(a.d) || a.g == null) {
            b._visibility = 1;
            return;
        }
        let c = { $type: Point_$type, x: a.b, y: a.e };
        let d = a.g;
        d.itemBrush = a.i;
        let e = d.series.lk(c, this.ca() || this.m8, this.m7);
        if (isNaN_(e.x) || isNaN_(e.y)) {
            b._visibility = 1;
            return;
        }
        this.m3.b9(b, e.x, e.y);
        b.ah = a.h;
        b.content = a.g;
    }
    nn(a, b) {
        if (isNaN_(a.b) || isNaN_(a.e) || isNaN_(a.a) || isNaN_(a.c) || isNaN_(a.d)) {
            b._visibility = 1;
            return;
        }
        this.m3.ca(b, a.b, a.e, a.c - a.b, a.a - a.e);
        b._fill = a.i;
        b._stroke = a.j;
        b.ac = a.d;
        if (this.la != null) {
            b.ai = this.la;
        }
        else {
            b.ai = null;
        }
        b.ad = this.lb;
    }
    c8() {
        return true;
    }
}
CategoryItemHighlightLayer.$t = markType(CategoryItemHighlightLayer, 'CategoryItemHighlightLayer', AnnotationLayer.$);
Type.dep(DependencyProperty, PropertyMetadata, CategoryItemHighlightLayer, 'raisePropertyChanged', ['BandHighlightWidth:nc:np', [1, 10], 'HighlightType:m4:nq', [CategoryItemHighlightType_$type, enumGetBox(CategoryItemHighlightType_$type, 0)], 'MarkerTemplate:no:nr', [DataTemplate.$, null], 'SkipUnknownValues:m9:ns', [0, true], 'TargetSeries:m6:nt', [Series.$, null], 'UseInterpolation:na:nu', [0, false]]);
/**
 * @hidden
 */
export class CategoryItemHighlightLayerView extends AnnotationLayerView {
    constructor(a) {
        super(a);
        this.b1 = null;
        this.b3 = null;
        this.b2 = null;
        this.b4 = null;
        this.b5 = null;
        this.b1 = a;
        this.b4 = new List$1(ContentControl.$, 0);
        this.b5 = new List$1(Rectangle.$, 0);
    }
    ap() {
        super.ap();
        this.b3 = ((() => {
            let $ret = new Pool$1(Rectangle.$);
            $ret.create = runOn(this, this.ch);
            $ret.activate = runOn(this, this.cb);
            $ret.disactivate = runOn(this, this.cd);
            $ret.destroy = runOn(this, this.cc);
            return $ret;
        })());
        this.b2 = ((() => {
            let $ret = new Pool$1(ContentControl.$);
            $ret.create = runOn(this, this.ce);
            $ret.activate = runOn(this, this.b6);
            $ret.disactivate = runOn(this, this.b8);
            $ret.destroy = runOn(this, this.b7);
            return $ret;
        })());
    }
    ch() {
        let a = new Rectangle();
        a.ab = false;
        this.b5.add(a);
        return a;
    }
    cb(a) {
        a._visibility = 0;
    }
    cd(a) {
        a._visibility = 1;
    }
    cc(a) {
        this.b5.remove(a);
    }
    ce() {
        let a = new Marker();
        this.b4.add(a);
        return a;
    }
    b6(a) {
        a._visibility = 0;
    }
    b8(a) {
        a._visibility = 1;
    }
    b7(a) {
        this.b4.remove(a);
    }
    cf(a) {
        if (a == null) {
            return a;
        }
        return BrushUtil.h(a, 0.3);
    }
    cg(a) {
        if (a != null) {
            let b = BrushUtil.h(a, 0.3);
            if (b.isGradient) {
                let c = b;
                if (c.gradientStops != null) {
                    for (let d = 0; d < c.gradientStops.length; d++) {
                        let e = c.gradientStops[d];
                        e.color = Color.u(truncate(Math.round((e.color.l) * 0.3)), e.color.o, e.color.n, e.color.m);
                        e.color = e.color;
                    }
                }
            }
            else {
                b.color = Color.u(truncate(Math.round((b.color.l & 255) * 0.3)), b.color.o, b.color.n, b.color.m);
                b.color = b.color;
            }
            return b;
        }
        else {
            return null;
        }
    }
    ca(a, b, c, d, e) {
        a._visibility = 0;
        a.n = b;
        a.o = c;
        a.width = d;
        a.height = e;
    }
    b9(a, b, c) {
        a.n = b;
        a.o = c;
    }
    be(a, b) {
        super.be(a, b);
        let c = a;
        let d = c.content;
        if (d != null) {
            if (d.itemBrush != null) {
                d.actualItemBrush = d.itemBrush;
            }
            let e = d.series;
            if (e.hasMarkers) {
                d.outline = e.k6();
            }
            d.thickness = 0.5;
        }
    }
    a3(a, b) {
        super.a3(a, b);
        if (a.d && !b) {
            for (let c = 0; c < this.b5.count; c++) {
                let d = this.b5._inner[c];
                if (d._visibility == 0) {
                    a.v(d);
                }
            }
            if (this.b4.count > 0) {
                let e = new DataTemplatePassInfo();
                e.isHitTestRender = b;
                e.context = a.g();
                e.viewportTop = this.br.top;
                e.viewportLeft = this.br.left;
                e.viewportWidth = this.br.width;
                e.viewportHeight = this.br.height;
                e.passID = "ItemMarkers";
                let f = new DataTemplateRenderInfo();
                f.isHitTestRender = b;
                f.passInfo = e;
                let g = new DataTemplateMeasureInfo();
                g.passInfo = e;
                let h = false;
                let i = a.g();
                g.context = i;
                f.context = i;
                g.renderContext = a;
                f.renderContext = a;
                let j = 0;
                let k = 0;
                let l = new Dictionary$2(DataTemplate.$, DataTemplate.$, 0);
                for (let m = 0; m < this.b4.count; m++) {
                    let n = this.b4._inner[m];
                    let o = n.ah;
                    if (!l.containsKey(o)) {
                        l.addItem(o, o);
                        if (o != null && o.passStarting != null) {
                            o.passStarting(e);
                        }
                    }
                    if (n._visibility == 1) {
                        continue;
                    }
                    this.bd(n, m, b);
                    if (!h) {
                        g.data = n.content;
                        g.width = n.width;
                        g.height = n.height;
                        if (o.measure != null) {
                            g.data = n.content;
                            o.measure(g);
                            h = g.isConstant;
                            if (h) {
                                j = g.width;
                                k = g.height;
                            }
                        }
                        f.availableWidth = g.width;
                        f.availableHeight = g.height;
                    }
                    else {
                        f.availableWidth = j;
                        f.availableHeight = k;
                    }
                    if (!isNaN_(n.width) && !isInfinity(n.width)) {
                        f.availableWidth = n.width;
                    }
                    if (!isNaN_(n.height) && !isInfinity(n.height)) {
                        f.availableHeight = n.height;
                    }
                    a.p(f, n);
                    n.m = f.availableWidth;
                    n.l = f.availableHeight;
                }
                for (let p of fromEnum(l.keys)) {
                    if (p != null && p.passCompleted != null) {
                        p.passCompleted(e);
                    }
                }
            }
        }
    }
    ai(a) {
        super.ai(a);
        for (let b = 0; b < this.b5.count; b++) {
            this.e.a2.exportRectangleData(a, this.b5._inner[b], "catItemRect", ["Main", "CategoryItem"]);
        }
        for (let c = 0; c < this.b4.count; c++) {
            this.e.a2.exportMarkerVisualDataFromMarker(this.e, a, this.b4._inner[c]);
        }
    }
}
CategoryItemHighlightLayerView.$t = markType(CategoryItemHighlightLayerView, 'CategoryItemHighlightLayerView', AnnotationLayerView.$);
