// igniteui-react-charts/ES2015/ChaikinVolatilityIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ChaikinVolatilityIndicatorStrategy } from "./ChaikinVolatilityIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class ChaikinVolatilityIndicator extends StrategyBasedIndicator {
    get_qb() {
        return new ChaikinVolatilityIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return ChaikinVolatilityIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(ChaikinVolatilityIndicator.qx));
    }
    set qw(a) {
        this.h(ChaikinVolatilityIndicator.qx, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
}
ChaikinVolatilityIndicator.$t = markType(ChaikinVolatilityIndicator, 'ChaikinVolatilityIndicator', StrategyBasedIndicator.$);
ChaikinVolatilityIndicator.qx = StrategyBasedIndicator.qt(10, ChaikinVolatilityIndicator.$);
