// igniteui-react-charts/ES2015/ChaikinOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { ChaikinOscillatorIndicatorStrategy } from "./ChaikinOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class ChaikinOscillatorIndicator extends StrategyBasedIndicator {
    get_qb() {
        return new ChaikinOscillatorIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return ChaikinOscillatorIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get shortPeriod() {
        return typeGetValue(this.c(ChaikinOscillatorIndicator.qx));
    }
    set shortPeriod(a) {
        this.h(ChaikinOscillatorIndicator.qx, a);
    }
    qj() {
        return this.shortPeriod;
    }
    get longPeriod() {
        return typeGetValue(this.c(ChaikinOscillatorIndicator.qw));
    }
    set longPeriod(a) {
        this.h(ChaikinOscillatorIndicator.qw, a);
    }
    qg() {
        return this.longPeriod;
    }
}
ChaikinOscillatorIndicator.$t = markType(ChaikinOscillatorIndicator, 'ChaikinOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
ChaikinOscillatorIndicator.qx = StrategyBasedIndicator.qv(3, ChaikinOscillatorIndicator.$);
ChaikinOscillatorIndicator.qw = StrategyBasedIndicator.qr(10, ChaikinOscillatorIndicator.$);
