// igniteui-react-charts/ES2015/FinancialOverlay.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialSeries } from "./FinancialSeries";
import { DependencyProperty } from "igniteui-react-core/ES2015/DependencyProperty";
import { ISortingAxis_$type } from "./ISortingAxis";
import { Type, typeCast, markType } from "igniteui-react-core/ES2015/type";
import { Rect } from "igniteui-react-core/ES2015/Rect";
import { ScalerParams } from "./ScalerParams";
import { AxisRange } from "./AxisRange";
import { PropertyMetadata } from "igniteui-react-core/ES2015/PropertyMetadata";
import { isNaN_ } from "igniteui-react-core/ES2015/number";
/**
 * @hidden
 */
export class FinancialOverlay extends FinancialSeries {
    constructor() {
        super();
        this.ph = false;
        this.ph = false;
    }
    hy(a, b, c, d) {
        super.hy(a, b, c, d);
        if (this.l7 != null && typeCast(ISortingAxis_$type, this.l7) !== null) {
            this.l7.i$i();
        }
        this.mn.bu.e(this.eu);
        this.pi();
        if (this.ms != null) {
            this.ms.ck();
        }
        this.jc(true);
    }
    get_dv() {
        return true;
    }
    iy(a, b, c, d) {
        switch (b) {
            case "FastItemsSource":
                this.ph = false;
                break;
            case FinancialOverlay.$$p[0]:
                this.ph = false;
                this.jc(false);
                this.ij();
                break;
            case "XAxis":
                this.ph = false;
                break;
        }
        super.iy(a, b, c, d);
    }
    bd() {
        if (this.be == 0) {
            return 1;
        }
        else {
            return super.bd();
        }
    }
    dh(a, b) {
        if (this.dg(a, b)) {
            return true;
        }
        return false;
    }
    ac() {
        super.ac();
        this.jc(false);
    }
    c9(a) {
        let b = this.bz != null ? this.bz.indexOf(a) : -1;
        let c = this.bi.bs;
        c = c.copy();
        let d = this.bi.br;
        let e = new Rect(0, 0, 0, 1, 1);
        let f = this.lr(this.bi);
        let g = new ScalerParams(0, e, e, this.l7.b8, f);
        let h = new ScalerParams(0, e, e, this.ms.b8, f);
        h.b = this.eh();
        if (b >= 0 && Rect.l_op_Inequality(c, null) && Rect.l_op_Inequality(d, null)) {
            if (this.l7 != null) {
                let i = this.l7.dv(b, g);
                if (i < c.left + 0.1 * c.width) {
                    i = i + 0.4 * c.width;
                }
                if (i > c.right - 0.1 * c.width) {
                    i = i - 0.4 * c.width;
                }
                c.x = i - 0.5 * c.width;
            }
            if (this.ms != null && this.mz != null && this.mx != null && b < this.mz.count && b < this.mx.count) {
                let j = this.ms.dv(this.mz.item(b), h);
                let k = this.ms.dv(this.mx.item(b), h);
                if (!isNaN_(j) && !isNaN_(k)) {
                    let l = Math.abs(j - k);
                    if (c.height < l) {
                        c.height = l;
                        c.y = Math.min(j, k);
                    }
                    else {
                        if (j < c.top + 0.1 * c.height) {
                            j = j + 0.4 * c.height;
                        }
                        if (j > c.bottom - 0.1 * c.height) {
                            j = j - 0.4 * c.height;
                        }
                        c.y = j - 0.5 * c.height;
                    }
                }
            }
            this.bn.ap(this.bs, c, true);
        }
        return b >= 0;
    }
    an(a) {
        if (a != null && a == this.ms && this.mz != null && this.mx != null) {
            return new AxisRange(this.mz.minimum, this.mx.maximum);
        }
        return null;
    }
}
FinancialOverlay.$t = markType(FinancialOverlay, 'FinancialOverlay', FinancialSeries.$);
Type.dep(DependencyProperty, PropertyMetadata, FinancialOverlay, 'raisePropertyChanged', ['IgnoreFirst:pj:pl', [1, 0]]);
