// igniteui-react-charts/ES2015/FinancialChart_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Brush } from "igniteui-react-core/ES2015/Brush";
import { XYChart } from "./XYChart";
import { IndicatorPane } from "./IndicatorPane";
import { Base, BaseError, EventArgs, runOn, delegateRemove, delegateCombine, typeCast, fromEnum, EnumUtil, Number_$type, String_$type, enumGetBox, markType, TypeRegistrar, fromEn } from "igniteui-react-core/ES2015/type";
import { List$1 } from "igniteui-react-core/ES2015/List$1";
import { DataSeriesType_$type } from "igniteui-react-core/ES2015/DataSeriesType";
import { IDataSeriesAdapterRule_$type } from "igniteui-react-core/ES2015/IDataSeriesAdapterRule";
import { SimpleFinancialPriceSeriesRule } from "igniteui-react-core/ES2015/SimpleFinancialPriceSeriesRule";
import { SubCollectionsRule } from "igniteui-react-core/ES2015/SubCollectionsRule";
import { XamDataChart } from "./XamDataChart";
import { FinancialChartVolumeType_$type } from "./FinancialChartVolumeType";
import { FinancialChartXAxisMode_$type } from "./FinancialChartXAxisMode";
import { FinancialChartYAxisMode_$type } from "./FinancialChartYAxisMode";
import { FinancialChartType_$type } from "./FinancialChartType";
import { NumericYAxis } from "./NumericYAxis";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { FinancialChartZoomSliderType_$type } from "./FinancialChartZoomSliderType";
import { BrushCollection } from "igniteui-react-core/ES2015/BrushCollection";
import { FinancialChartRangeSelectorOptionCollection } from "./FinancialChartRangeSelectorOptionCollection";
import { ObservableCollection$1 } from "igniteui-react-core/ES2015/ObservableCollection$1";
import { TemplatedHtmlControl } from "./TemplatedHtmlControl";
import { FinancialChartContext } from "./FinancialChartContext";
import { FinancialChartToolbarContext } from "./FinancialChartToolbarContext";
import { FinancialChartViewModel } from "./FinancialChartViewModel";
import { FinancialChartToolbarViewModel } from "./FinancialChartToolbarViewModel";
import { Series } from "./Series";
import { Rect } from "igniteui-react-core/ES2015/Rect";
import { Axis } from "./Axis";
import { AxisLabelSettings } from "./AxisLabelSettings";
import { MarkerSeries } from "./MarkerSeries";
import { SyncSettings } from "./SyncSettings";
import { SyncManager } from "./SyncManager";
import { FinancialOverlayType_$type } from "./FinancialOverlayType";
import { ArrayExtension } from "igniteui-react-core/ES2015/ArrayExtension";
import { FinancialIndicatorType_$type } from "./FinancialIndicatorType";
import { IndicatorDisplayType_$type } from "./IndicatorDisplayType";
import { ITimeRangeSelectorAxis_$type } from "./ITimeRangeSelectorAxis";
import { FinancialChartVisualData } from "./FinancialChartVisualData";
import { FinancialChartCustomIndicatorArgs } from "./FinancialChartCustomIndicatorArgs";
import { ScalerParams } from "./ScalerParams";
import { stringReplace, stringJoin, stringIsNullOrEmpty } from "igniteui-react-core/ES2015/string";
import { timeSpanFromDays } from "igniteui-react-core/ES2015/timespan";
import { dateMaxValue, dateMinValue } from "igniteui-react-core/ES2015/date";
import { isNaN_, isInfinity } from "igniteui-react-core/ES2015/number";
import { FinancialChartTypePicker } from "./FinancialChartTypePicker";
import { FinancialChartIndicatorMenu } from "./FinancialChartIndicatorMenu";
import { TrendLineType_$type } from "igniteui-react-core/ES2015/TrendLineType";
import { FinancialChartIndicatorMenuContext } from "./FinancialChartIndicatorMenuContext";
import { FinancialChartTypePickerContext } from "./FinancialChartTypePickerContext";
import { FinancialChartRangeSelectorContext } from "./FinancialChartRangeSelectorContext";
import { FinancialChartToolbarVisualData } from "./FinancialChartToolbarVisualData";
import { FinancialChartRangeSelectorViewModel } from "./FinancialChartRangeSelectorViewModel";
import { FinancialChartRangeSelectorView } from "./FinancialChartRangeSelectorView";
/**
 * @hidden
 */
export class FinancialChart extends XYChart {
    constructor() {
        super();
        this.qd = null;
        this.sv = null;
        this.st = null;
        this.sq = null;
        this.ss = null;
        this.su = null;
        this.sr = null;
        this.pj = null;
        this.td = null;
        this.te = null;
        this.ta = null;
        this.tf = null;
        this.tg = null;
        this.tb = null;
        this.tc = null;
        this.q1 = null;
        this.r6 = 40;
        this.p7 = null;
        this.p5 = null;
        this.p9 = null;
        this.qe = null;
        this.to = null;
        this.qx = null;
        this.rl = false;
        this.rk = false;
        this.rq = false;
        this.q7 = false;
        this.s1 = 10;
        this.r8 = NaN;
        this.sa = NaN;
        this.r9 = NaN;
        this.sb = 0;
        this.pp = 0;
        this.ps = 0;
        this.pu = 0;
        this.q4 = true;
        this.pl = 0;
        this.q5 = false;
        this.pz = null;
        this.p1 = null;
        this.rs = null;
        this.q6 = true;
        this.qy = null;
        this.pw = 1;
        this.on = null;
        this.oo = null;
        this.op = null;
        this.oq = null;
        this.os = null;
        this.or = null;
        this.om = null;
        this.ol = null;
        this.r7 = 1;
        this.r5 = 1;
        this.r4 = 1;
        this.qb = null;
        this.sx = -2147483648;
        this.r3 = 2;
        this.s0 = 3;
        this.sy = 10;
        this.sw = 30;
        this.sz = 9;
        this.pg = null;
        this.ph = null;
        this.pc = null;
        this.applyCustomIndicators = null;
        this.w8 = null;
        this.sc = 1;
        this.q3 = false;
        this.p3 = null;
        this.r0 = dateMinValue();
        this.rz = dateMaxValue();
        this.rx = null;
        this.f6();
        this.rx = new ObservableCollection$1(Base.$, 0);
        if (this.xAxisBreaks != null) {
            let a = this.xAxisBreaks;
            a.collectionChanged = delegateCombine(a.collectionChanged, (a, b) => this.wx());
        }
        this.wi();
    }
    g3(a, b, c) {
        super.g3(a, b, c);
        switch (a) {
            case "ToolbarTemplate":
            case "RangeSelectorTemplate":
                this.pk = new FinancialChartToolbar(this, this.tl);
                break;
            case "Toolbar":
                let d = b;
                if (d != null) {
                    d.chartTypeChanged = delegateRemove(d.chartTypeChanged, runOn(this, this.v6));
                    d.trendLineTypeChanged = delegateRemove(d.trendLineTypeChanged, runOn(this, this.v9));
                    d.volumeTypeChanged = delegateRemove(d.volumeTypeChanged, runOn(this, this.wa));
                    d.l(null);
                }
                if (this.pk != null) {
                    this.pk.ac = this.pr;
                    this.pk.ai = this.av;
                    this.pk.y = this.pm;
                    this.pk.al = this.ri;
                    if (this.ss != null) {
                        this.pk.l(this.ss);
                    }
                    let e = this.pk;
                    e.chartTypeChanged = delegateCombine(e.chartTypeChanged, runOn(this, this.v6));
                    ;
                    let f = this.pk;
                    f.trendLineTypeChanged = delegateCombine(f.trendLineTypeChanged, runOn(this, this.v9));
                    let g = this.pk;
                    g.volumeTypeChanged = delegateCombine(g.volumeTypeChanged, runOn(this, this.wa));
                }
                break;
            case "ChartType":
                if (this.pk != null) {
                    this.pk.y = this.pm;
                }
                break;
            case "VolumeType":
                if (this.pk != null) {
                    this.pk.ac = this.pr;
                }
                break;
            case "TrendLineType":
                if (this.pk != null) {
                    this.pk.ai = this.av;
                }
                break;
            case "IsToolbarVisible":
                if (this.pk == null) {
                    break;
                }
                this.pk.al = this.ri;
                this.setGridRows();
                break;
        }
    }
    wa(a, b) {
        this.pr = this.pk.ac;
    }
    v9(a, b) {
        this.av = this.pk.ai;
    }
    v6(a, b) {
        this.pm = this.pk.y;
    }
    wi() {
        this.qd = new IndicatorPane();
        this.p4 = TypeRegistrar.create("FinancialLegend");
    }
    provideGrid(a) {
        this.sv = a;
        this.w3();
        this.setGridRows();
        if (this.pk != null) {
            this.pk.bg(a);
        }
    }
    onDocumentClick(a) {
        if (this.pk != null) {
            this.pk.bl(a);
        }
    }
    getContext() {
        let a = this.pk.v();
        let b = new FinancialChartContext();
        b.vm = new FinancialChartViewModel();
        b.vm.toolbar = a;
        b.container = this.di;
        b.toolbarContainer = this.ss;
        b.legendContainer = this.sr;
        b.indicatorsContainer = this.sq;
        b.volumeContainer = this.st;
        b.zoomContainer = this.su;
        return b;
    }
    provideContainer(a) {
        let b = a;
        if (b != null) {
            let c = "financialChartMainGrid";
            let d = b.createElement("div");
            d.setStyleProperty("width", "100%");
            d.setStyleProperty("height", "100%");
            if (this.pk == null) {
                this.pk = new FinancialChartToolbar(this, this.tl);
            }
            this.pk.w.isIndicatorMenuVisible = this.q9 && this.ra;
            this.ss = b.getSubRenderer(d);
            this.pk.l(this.ss);
            let e = b.createElement("div");
            e.setStyleProperty("width", "100%");
            e.setStyleProperty("height", "100%");
            let f = b.createElement("div");
            f.setStyleProperty("width", "100%");
            f.setStyleProperty("height", "100%");
            this.sr = b.getSubRenderer(f);
            let g = b.createElement("div");
            g.setStyleProperty("width", "100%");
            g.setStyleProperty("height", "100%");
            this.sq = b.getSubRenderer(g);
            let h = b.createElement("div");
            h.setStyleProperty("width", "100%");
            h.setStyleProperty("height", "100%");
            this.st = b.getSubRenderer(h);
            let i = b.createElement("div");
            i.setStyleProperty("width", "100%");
            i.setStyleProperty("height", "100%");
            this.su = b.getSubRenderer(i);
            a = b.getSubRenderer(e);
        }
        else {
            this.ss = null;
            if (this.pk != null) {
                this.pk.l(null);
            }
        }
        super.provideContainer(a);
        if (this.volumeChart != null) {
            this.volumeChart.provideContainer(this.st);
        }
        if (this.qd != null) {
            this.qd.h(this.sq);
        }
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.provideContainer(this.su);
            if (a != null) {
                this.vq();
            }
            else {
                this.q2 = null;
            }
        }
        if (this.p4 != null) {
            this.p4.provideContainer(this.sr);
        }
        this.setGridRows();
    }
    wt(a, b, c) {
        if (a != null) {
            a.provideContainer(null);
        }
        if (b != null) {
            b.provideContainer(c);
        }
        this.setGridRows();
    }
    w7(a, b) {
        this.wt(a, b, this.su);
        if (b == null) {
            return;
        }
        this.vq();
    }
    vq() {
        if (this.su == null) {
            return;
        }
        if (this.q2 != null) {
            this.q2.provideContainer(null);
        }
        this.q2 = TypeRegistrar.create("ZoomSlider");
        this.q2.provideContainer(this.su);
        let a = this.q2;
        a.windowRectChanged = delegateCombine(a.windowRectChanged, (a, b) => {
            if (this.dataChart != null) {
                this.dataChart.m4 = new Rect(0, this.q2.windowRect.left, this.dataChart.mq.top, this.q2.windowRect.width, this.dataChart.mq.height);
            }
        });
        this.vr();
    }
    w5(a, b) {
        this.wt(a, b, this.st);
    }
    setGridRows() {
        if (this.sv == null) {
            return;
        }
        let a = new Array(6);
        a[0] = (this.ri ? this.sh : 0) + "px";
        a[1] = this.re ? "min-content" : "0px";
        a[2] = "auto";
        a[3] = this.indicatorCharts == null || this.indicatorCharts.count == 0 ? "0px" : this.indicatorCharts.count * 150 + "px";
        a[4] = this.volumeChart == null ? "0px" : "100px";
        a[5] = this.zoomSliderChart == null ? "0px" : "100px";
        if (this.q2 != null) {
            if (a[5] == "0px") {
                this.q2.hide();
            }
            else {
                this.q2.show();
            }
        }
        this.sv.setStyleProperty("-ms-grid-rows", stringReplace(stringJoin(" ", ...a), "auto", "1fr"));
        this.sv.setStyleProperty("grid-template-rows", stringJoin(" ", ...a));
        if (this.di == null) {
            return;
        }
        if (this.dataChart != null) {
            this.dataChart.notifyContainerResized();
        }
        if (this.qd != null) {
            this.qd.f();
        }
        if (this.volumeChart != null) {
            this.volumeChart.notifyContainerResized();
        }
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.notifyContainerResized();
        }
        if (this.q2 != null) {
            this.q2.notifySizeChanged();
        }
    }
    gj() {
        super.gj();
        this.ie = 3;
        this.ni = 0;
        this.cu = NaN;
        this.cs = 30;
        this.bq = true;
        this.bw = false;
        if (this.dataChart != null) {
            let a = this.dataChart.bn.i.e();
            this.ov = a.d;
            this.ow = a.e;
            this.ou = a.d;
            this.w9 = this.dataChart.pm;
        }
        let b = new BrushCollection();
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "rgba(116, 70, 185, .3)";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "rgba(159, 179, 40, .3)";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "rgba(46, 156, 166, .3)";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "rgba(63, 81, 181, .3)";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "rgba(154, 154, 154, .3)";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "rgba(67, 156, 71, .3)";
            return $ret;
        })()));
        this.ox = b;
        let c = new BrushCollection();
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "#7446B9";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "#9FB328";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "#2E9CA6";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "#3F51B5";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "#9A9A9A";
            return $ret;
        })()));
        b.add(((() => {
            let $ret = new Brush();
            $ret.fill = "#439C47";
            return $ret;
        })()));
        this.oy = c;
    }
    wp() {
        if (this.qd != null) {
            this.qd.b = null;
            this.qd.b = this.indicatorCharts;
        }
        this.setGridRows();
    }
    wq() {
        this.setGridRows();
    }
    wz() {
        this.wp();
    }
    vp() {
        if (this.q2 == null || this.zoomSliderChart == null || this.dataChart == null || this.dataChart.series.count == 0) {
            return;
        }
        let a = this.di.rootWrapper.width();
        let b = this.dataChart.m3.left + this.dataChart.m3.width;
        let c = this.dataChart.m3.left;
        let d = a - b;
        this.q2.startInset = c - this.q2.trackStartInset;
        this.q2.endInset = d - this.q2.trackEndInset;
        if (this.q2.endInset < 0) {
            let e = this.q2.endInset;
            this.q2.endInset = 0;
            this.dataChart.el += (e * -1);
        }
        if (this.q2.startInset < 0) {
            let f = this.q2.startInset;
            this.q2.startInset = 0;
            this.dataChart.ej += (f * -1);
        }
        this.zoomSliderChart.ej = c;
        this.zoomSliderChart.el = d;
        this.zoomSliderChart.ef = this.q2.barExtent;
    }
    wg() {
        this.vp();
    }
    gg(a) {
        super.gg(a);
        if (a != null) {
            a.ee = 0;
        }
    }
    xa(a) {
        return a.mj();
    }
    xd() {
        return this.sr == null ? null : this.xc(this.sr.rootWrapper);
    }
    xb(a) {
        if (a == null || a.d == null) {
            return null;
        }
        return this.xc(a.d.rootWrapper);
    }
    xc(a) {
        if (a == null) {
            return null;
        }
        let b = a.getOffset();
        return new Rect(0, b.left, b.top, a.width(), a.height());
    }
    sd() {
        if (this.sv == null) {
            return NaN;
        }
        let a = this.sv.getOffset();
        return this.sv.width() + (a == null ? 0 : a.left);
    }
    gr(source_, a, b) {
        super.gr(source_, a, b);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.hy(source_, a, b);
        }
    }
    gs(source_, a, b) {
        super.gs(source_, a, b);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.hz(source_, a, b);
        }
    }
    gq(source_) {
        super.gq(source_);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.hv(source_);
        }
    }
    gu(source_, a, b, c) {
        super.gu(source_, a, b, c);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.h0(source_, a, b, c);
        }
    }
    gm(a, b) {
        super.gm(a, b);
        if (this.rf(a)) {
            a.toolTip = null;
        }
    }
    get pk() {
        return this.pj;
    }
    set pk(a) {
        let b = this.pk;
        if (a != b) {
            this.pj = a;
            this.g1("Toolbar", b, this.pk);
        }
    }
    get tk() {
        return this.td;
    }
    set tk(a) {
        let b = this.tk;
        if (a != b) {
            this.td = a;
            this.g1("RangeSelectorTemplate", b, this.tk);
        }
    }
    get tl() {
        return this.te;
    }
    set tl(a) {
        let b = this.tl;
        if (a != b) {
            this.te = a;
            this.g1("ToolbarTemplate", b, this.tl);
        }
    }
    get th() {
        return this.ta;
    }
    set th(a) {
        let b = this.th;
        if (a != b) {
            this.ta = a;
            this.g1("ChartTypePickerTemplate", b, this.th);
        }
    }
    get tm() {
        return this.tf;
    }
    set tm(a) {
        let b = this.tm;
        if (a != b) {
            this.tf = a;
            this.g1("TrendLineTypePickerTemplate", b, this.tm);
        }
    }
    get tn() {
        return this.tg;
    }
    set tn(a) {
        let b = this.tn;
        if (a != b) {
            this.tg = a;
            this.g1("VolumeTypePickerTemplate", b, this.tn);
        }
    }
    get ti() {
        return this.tb;
    }
    set ti(a) {
        let b = this.ti;
        if (a != b) {
            this.tb = a;
            this.g1("IndicatorMenuTemplate", b, this.ti);
        }
    }
    get tj() {
        return this.tc;
    }
    set tj(a) {
        let b = this.tj;
        if (a != b) {
            this.tc = a;
            this.g1("OverlayPickerTemplate", b, this.tj);
        }
    }
    get q2() {
        return this.q1;
    }
    set q2(a) {
        let b = this.q2;
        if (a != b) {
            this.q1 = a;
            this.g1("ZoomSlider", b, this.q2);
        }
    }
    get_cs() {
        return super.get_cs();
    }
    set_cs(a) {
        super.set_cs(a);
    }
    get_bq() {
        return super.get_bq();
    }
    set_bq(a) {
        super.set_bq(a);
    }
    get sh() {
        return this.r6;
    }
    set sh(a) {
        let b = this.sh;
        if (a != b) {
            this.r6 = a;
            this.g1("ToolbarHeight", b, this.sh);
        }
    }
    h7() {
        return this.xAxis;
    }
    h8() {
        return this.yAxis;
    }
    u(a) {
        return 20;
    }
    get_b5() {
        return ((() => {
            let $ret = new List$1(DataSeriesType_$type, 0);
            $ret.add(20);
            return $ret;
        })());
    }
    get b5() {
        return this.get_b5();
    }
    get_b4() {
        return ((() => {
            let $ret = new List$1(IDataSeriesAdapterRule_$type, 0);
            $ret.add(new SimpleFinancialPriceSeriesRule());
            $ret.add(new SubCollectionsRule());
            return $ret;
        })());
    }
    get b4() {
        return this.get_b4();
    }
    o6(a) {
        return this.o7(a);
    }
    qk(a) {
        return this.ql(a);
    }
    o5(a) {
        let b = this.o6(0);
        b.name = a;
        return b;
    }
    f3() {
        this.xAxis = this.o5("xAxis");
        this.yAxis = this.qk(0);
        this.yAxis.name = "yAxis";
        this.volumeAxis = this.qk(0);
        this.volumeAxis.name = "volumeAxis";
        this.volumeXAxis = this.o5("volumeXAxis");
        this.ru = new List$1(NumericYAxis.$, 0);
        this.rw = new List$1(CategoryAxisBase.$, 0);
        this.indicatorAxis = this.qk(0);
        this.indicatorAxis.name = "indicatorAxis";
        this.indicatorXAxis = this.o5("indicatorXAxis_0");
        this.ru.add(this.indicatorAxis);
        this.rw.add(this.indicatorXAxis);
        if (this.xAxis.aw == null) {
            this.xAxis.aw = new AxisLabelSettings();
        }
        if (this.yAxis.aw == null) {
            this.yAxis.aw = new AxisLabelSettings();
        }
        if (this.volumeAxis.aw == null) {
            this.volumeAxis.aw = new AxisLabelSettings();
        }
        if (this.volumeXAxis.aw == null) {
            this.volumeAxis.aw = new AxisLabelSettings();
        }
        if (this.indicatorAxis.aw == null) {
            this.indicatorAxis.aw = new AxisLabelSettings();
        }
        if (this.indicatorXAxis.aw == null) {
            this.indicatorXAxis.aw = new AxisLabelSettings();
        }
        this.zoomSliderAxis = this.qk(0);
        this.zoomSliderAxis.name = "zoomSliderAxis";
        this.zoomSliderXAxis = this.o5("zoomSliderXAxis");
        if (this.zoomSliderAxis.aw == null) {
            this.zoomSliderAxis.aw = new AxisLabelSettings();
        }
        if (this.zoomSliderXAxis.aw == null) {
            this.zoomSliderXAxis.aw = new AxisLabelSettings();
        }
    }
    f1(a, b) {
        let c = typeCast(MarkerSeries.$, a);
        if (c != null) {
            c.l0 = this.ai(a, b);
        }
    }
    hh(a) {
        if (a.suggestedSecondaryAxis == 2) {
            this.yAxis.js = true;
            this.yAxis.kf = 10;
        }
    }
    hk() {
        this.yAxis.js = this.rn;
        this.yAxis.kf = this.s9;
    }
    po(a) {
        let b;
        if (this.pm == 0) {
            if (a == null || a.index > 0 || stringIsNullOrEmpty(a.getMemberPathFor(12))) {
                b = 4;
            }
            else {
                b = 2;
            }
        }
        else {
            b = this.pm;
        }
        b = this.pn(b);
        return b;
    }
    pn(a) {
        if ((a == 1 || a == 2)) {
            if (TypeRegistrar.isRegistered("FinancialPriceSeries")) {
                return a;
            }
            a = 4;
        }
        if (a == 4) {
            if (TypeRegistrar.isRegistered("LineSeries")) {
                return a;
            }
            a = 3;
        }
        if (a == 3) {
            if (TypeRegistrar.isRegistered("ColumnSeries")) {
                return a;
            }
        }
        return a;
    }
    px(a) {
        if ((a == 2 || a == 3)) {
            if (TypeRegistrar.isRegistered("FinancialPriceSeries")) {
                return a;
            }
            a = 5;
        }
        if (a == 5) {
            if (TypeRegistrar.isRegistered("LineSeries")) {
                return a;
            }
            a = 6;
        }
        if (a == 6) {
            if (TypeRegistrar.isRegistered("AreaSeries")) {
                return a;
            }
            a = 4;
        }
        if (a == 4) {
            if (TypeRegistrar.isRegistered("ColumnSeries")) {
                return a;
            }
        }
        return a;
    }
    pq(a) {
        if (a == 2) {
            if (TypeRegistrar.isRegistered("LineSeries")) {
                return a;
            }
            a = 3;
        }
        if (a == 3) {
            if (TypeRegistrar.isRegistered("AreaSeries")) {
                return a;
            }
            a = 1;
        }
        if (a == 1) {
            if (TypeRegistrar.isRegistered("ColumnSeries")) {
                return a;
            }
        }
        return a;
    }
    tv(a) {
        let b = a.getMemberPathFor(14);
        if (stringIsNullOrEmpty(b)) {
            b = a.getMemberPathFor(11);
        }
        return b;
    }
    o7(axisType_) {
        if ((this.externalObject)) {
            return (this.externalObject.createXAxis(axisType_).i);
        }
        else {
            let a;
            switch (axisType_) {
                case 0:
                    if (TypeRegistrar.isRegistered("OrdinalTimeXAxis")) {
                        a = TypeRegistrar.create("OrdinalTimeXAxis");
                    }
                    else if (TypeRegistrar.isRegistered("TimeXAxis")) {
                        a = TypeRegistrar.create("TimeXAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
                case 1:
                    if (TypeRegistrar.isRegistered("TimeXAxis")) {
                        a = TypeRegistrar.create("TimeXAxis");
                    }
                    else if (TypeRegistrar.isRegistered("OrdinalTimeXAxis")) {
                        a = TypeRegistrar.create("OrdinalTimeXAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
            }
            return null;
        }
    }
    ql(axisType_) {
        if ((this.externalObject)) {
            return (this.externalObject.createYAxis(axisType_).i);
        }
        else {
            let a;
            switch (axisType_) {
                case 0:
                    if (TypeRegistrar.isRegistered("NumericYAxis")) {
                        a = TypeRegistrar.create("NumericYAxis");
                    }
                    else if (TypeRegistrar.isRegistered("PercentChangeYAxis")) {
                        a = TypeRegistrar.create("PercentChangeYAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
                case 1:
                    if (TypeRegistrar.isRegistered("PercentChangeYAxis")) {
                        a = TypeRegistrar.create("PercentChangeYAxis");
                    }
                    else if (TypeRegistrar.isRegistered("NumericYAxis")) {
                        a = TypeRegistrar.create("NumericYAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
            }
            return null;
        }
    }
    qr() {
        return this.ar(1);
    }
    qs() {
        return this.ar(0);
    }
    qq() {
        return this.ar(2);
    }
    qt() {
        return this.ar(20);
    }
    get q8() {
        return (TypeRegistrar.isRegistered("HorizontalAnchoredCategorySeriesProxy"));
    }
    get p8() {
        if (TypeRegistrar.isRegistered("HorizontalAnchoredCategorySeriesProxy")) {
            return TypeRegistrar.create("HorizontalAnchoredCategorySeriesProxy");
        }
        else {
            throw new BaseError(1, "horizontal series module is not loaded, but is required.");
        }
    }
    get rb() {
        return (TypeRegistrar.isRegistered("FinancialPriceSeriesProxy"));
    }
    get p6() {
        if (TypeRegistrar.isRegistered("FinancialPriceSeriesProxy")) {
            return TypeRegistrar.create("FinancialPriceSeriesProxy");
        }
        else {
            throw new BaseError(1, "financial series module is not loaded, but is required.");
        }
    }
    get q9() {
        return (TypeRegistrar.isRegistered("IndicatorProxy"));
    }
    get qa() {
        if (TypeRegistrar.isRegistered("IndicatorProxy")) {
            return TypeRegistrar.create("IndicatorProxy");
        }
        else {
            throw new BaseError(1, "financial series module is not loaded, but is required.");
        }
    }
    get ra() {
        return (TypeRegistrar.isRegistered("OverlayProxy"));
    }
    get qf() {
        if (TypeRegistrar.isRegistered("OverlayProxy")) {
            return TypeRegistrar.create("OverlayProxy");
        }
        else {
            throw new BaseError(1, "overlay series module is not loaded, but is required.");
        }
    }
    aq(a, b) {
        let c = this.tv(a);
        switch (this.po(a)) {
            case 3:
                let d = this.qr();
                this.p8.setValueMemberPath(d, c);
                this.p8.setXAxis(d, this.xAxis);
                this.p8.setYAxis(d, this.yAxis);
                this.p8.setCategoryMode(d, 1);
                return d;
            case 4:
                let e = this.qs();
                this.p8.setValueMemberPath(e, c);
                this.p8.setXAxis(e, this.xAxis);
                this.p8.setYAxis(e, this.yAxis);
                this.p8.setCategoryMode(e, 1);
                return e;
            default:
            case 1:
            case 2:
                let f = this.qt();
                this.p6.setOpenMemberPath(f, a.getMemberPathFor(11));
                this.p6.setHighMemberPath(f, a.getMemberPathFor(12));
                this.p6.setLowMemberPath(f, a.getMemberPathFor(13));
                this.p6.setCloseMemberPath(f, a.getMemberPathFor(14));
                this.p6.setVolumeMemberPath(f, a.getMemberPathFor(15));
                this.p6.setXAxis(f, this.xAxis);
                this.p6.setYAxis(f, this.yAxis);
                this.p6.setDisplayType(f, this.pm == 1 ? 1 : 0);
                this.p6.setCategoryMode(f, 1);
                return f;
        }
    }
    hj() {
        super.hj();
        this.we(this.xAxis);
        this.w1(false);
        if (this.pr != 0 && this.volumeChart != null) {
            this.wk();
        }
        this.we(this.volumeXAxis);
        this.we(this.indicatorXAxis);
        this.wy();
        this.w0();
        if (this.py != 0) {
            this.wl();
            this.we(this.zoomSliderXAxis);
        }
    }
    m6(a, b) {
        super.m6(a, b);
        a(this.volumeXAxis);
        if (this.rw != null) {
            for (let c of fromEnum(this.rw)) {
                a(c);
            }
        }
        let d = true;
        switch (b) {
            case "XAxisMajorStroke":
            case "XAxisMajorStrokeThickness":
                d = false;
                break;
        }
        if (d) {
            a(this.zoomSliderXAxis);
        }
    }
    na(a, b) {
        super.na(a, b);
        if (b) {
            a(this.volumeAxis);
            if (this.ru != null) {
                for (let c of fromEnum(this.ru)) {
                    a(c);
                }
            }
            a(this.zoomSliderAxis);
        }
    }
    m8(a) {
        super.m8(a);
        this.m0(this.volumeXAxis, a);
        this.m0(this.indicatorXAxis, a);
        this.m0(this.zoomSliderXAxis, a);
    }
    nc(a) {
        super.nc(a);
        this.m0(this.volumeAxis, a);
        this.m0(this.indicatorAxis, a);
        this.m0(this.zoomSliderAxis, a);
    }
    nb(a) {
        super.nb(a);
        this.mz(this.volumeAxis, a);
        this.mz(this.indicatorAxis, a);
        this.mz(this.zoomSliderAxis, a);
    }
    m7(a) {
        super.m7(a);
        this.mz(this.volumeXAxis, a);
        this.mz(this.indicatorXAxis, a);
        this.mz(this.zoomSliderXAxis, a);
    }
    gf(a) {
        super.gf(a);
        if (a == null) {
            return;
        }
        if (a == this.zoomSliderChart) {
            a.nr = a.nt = false;
            a.legend = null;
            if (this.dataChart != null) {
                a.m9 = this.dataChart.m9;
                a.nc = this.dataChart.nc;
            }
        }
        else {
            SyncManager.d(a, ((() => {
                let $ret = new SyncSettings();
                $ret.syncChannel = this.up();
                $ret.synchronizeHorizontally = true;
                $ret.synchronizeVertically = false;
                return $ret;
            })()));
            SyncManager.c(a);
        }
        a.bc = 2;
    }
    up() {
        if (this.to == null) {
            FinancialChart.s8++;
            this.to = FinancialChart.s8.toString();
        }
        return this.to;
    }
    v0(a, b) {
        this.vm();
        this.wg();
    }
    vm() {
        if (this.yAxis.bz()) {
            return;
        }
        let a = this.yAxis.at.a5;
        if (this.pr != 0) {
            a = Math.max(a, this.volumeAxis.at.a5);
        }
        for (let b of fromEnum(this.ru)) {
            a = Math.max(a, b.at.a5);
        }
        this.vn(a);
    }
    vn(a) {
        if (this.yAxis.bz()) {
            return;
        }
        this.yAxis.d3 = this.volumeAxis.d3 = a;
        for (let b of fromEnum(this.ru)) {
            b.d3 = a;
        }
        this.zoomSliderAxis.d3 = this.yAxis.d3;
    }
    vs(a) {
        if (this.indicatorCharts == null) {
            return;
        }
        for (let b = 0; b < this.indicatorCharts.count; b++) {
            let c = this.indicatorCharts._inner[b];
            for (let d = 0; d < c.series.count; d++) {
                a(c.series._inner[d], b);
            }
        }
    }
    vv(a) {
        if (this.volumeChart == null) {
            return;
        }
        for (let b = 0; b < this.volumeChart.series.count; b++) {
            a(this.volumeChart.series._inner[b], b);
        }
    }
    vt(a) {
        if (this.dataChart == null) {
            return;
        }
        for (let b = 0; b < this.dataChart.series.count; b++) {
            let c = this.dataChart.series._inner[b];
            if (!this.rf(c)) {
                continue;
            }
            a(c, b);
        }
    }
    vu(a, b) {
        if (this.dataChart == null) {
            return;
        }
        for (let c = 0; c < this.dataChart.series.count; c++) {
            let d = this.dataChart.series._inner[c];
            if (this.rf(d)) {
                continue;
            }
            let e = this.s2(d, c);
            a(d, e);
        }
        if (b) {
            this.vw(a);
        }
    }
    vw(a) {
        if (this.zoomSliderChart == null) {
            return;
        }
        for (let b = 0; b < this.zoomSliderChart.series.count; b++) {
            a(this.zoomSliderChart.series._inner[b], b);
        }
    }
    s2(a, b) {
        if (a.bq != this.dataChart) {
            return b;
        }
        let c = this.v == null || this.v.dataSeries == null ? 0 : this.v.dataSeries.count;
        let d = this.p2 == null ? 0 : this.p2.count;
        return b - (c * d);
    }
    f9(a) {
        super.f9(a);
        this.vv(a);
        this.vs(a);
        this.vw(a);
    }
    get volumeChart() {
        return this.qx;
    }
    set volumeChart(a) {
        let b = this.volumeChart;
        if (a != b) {
            this.qx = a;
            this.g1("VolumeChart", b, this.volumeChart);
        }
    }
    wk() {
        if (this.dataChart == null) {
            return;
        }
        this.volumeChart.series.clear();
        let a = new List$1(Series.$, 0);
        let b = -1;
        for (let c = 0; c < this.dataChart.series.count; c++) {
            let d = this.dataChart.series._inner[c];
            if (this.rf(d) || d.cj) {
                continue;
            }
            b++;
            let e;
            let f = this.pr;
            f = this.pq(f);
            switch (f) {
                case 2:
                    e = this.qs();
                    break;
                case 3:
                    e = this.qq();
                    break;
                default:
                case 1:
                    e = this.qr();
                    break;
            }
            this.p8.setCategoryMode(e, 1);
            e.name = "volumeSeries_" + b;
            e.itemsSource = d.itemsSource;
            this.p8.setXAxis(e, this.volumeXAxis);
            this.p8.setYAxis(e, this.volumeAxis);
            this.p8.setValueMemberPath(e, this.v != null && this.v.dataSeries != null && this.v.dataSeries.count > b ? this.v.dataSeries._inner[b].getMemberPathFor(15) : null);
            e.fx = d.fx + " Volume";
            e.k4 = ArrayExtension.getModulus$1(Brush.$, this.oz, b);
            e.k7 = ArrayExtension.getModulus$1(Brush.$, this.o0, b);
            e.ey = this.si;
            a.add(e);
        }
        let g = this.b1("volume", false);
        if (g != null) {
            a.o(g);
        }
        for (let h = 0; h < a.count; h++) {
            let i = a._inner[h];
            this.volumeChart.series.add(i);
            this.gk(i, h);
        }
    }
    rc(a) {
        if (!this.q8) {
            return false;
        }
        return this.p8.matchesType(a);
    }
    rg(a) {
        if (!this.rb) {
            return false;
        }
        return this.p6.matchesType(a);
    }
    rd(a) {
        if (!this.q9) {
            return false;
        }
        return this.qa.matchesType(a);
    }
    rf(a) {
        if (!this.ra) {
            return false;
        }
        return this.qf.matchesType(a);
    }
    wj() {
        if (this.dataChart == null) {
            return;
        }
        let a = 0;
        let b = this.p0 == null ? 0 : this.p0.count;
        let c = this.customIndicatorNames == null ? 0 : this.customIndicatorNames.count;
        let d = b + c;
        for (let e = 0; e < d; e++) {
            let f = new List$1(Series.$, 0);
            let g = this.indicatorCharts._inner[e];
            let h = e >= b;
            let i = h ? this.customIndicatorNames._inner[e - b] : EnumUtil.getName(FinancialIndicatorType_$type, this.p0._inner[e]);
            let j = -1;
            for (let k = 0; k < this.dataChart.series.count; k++) {
                let l = this.dataChart.series._inner[k];
                if (this.rf(l) || l.cj) {
                    continue;
                }
                j++;
                let m;
                if (h) {
                    let n = this.qu("Custom");
                    this.qa.handleCustomIndicatorEvent(n, runOn(this, this.vy));
                    m = n;
                }
                else {
                    m = this.qw(this.p0._inner[e]);
                }
                if (m == null) {
                    continue;
                }
                m.name = "indicator_" + a++;
                this.qa.setXAxis(m, this.rw._inner[e]);
                this.qa.setYAxis(m, this.ru._inner[e]);
                m.itemsSource = l.itemsSource;
                this.qa.setOpenMemberPath(m, this.v.dataSeries._inner[j].getMemberPathFor(11));
                this.qa.setHighMemberPath(m, this.v.dataSeries._inner[j].getMemberPathFor(12));
                this.qa.setLowMemberPath(m, this.v.dataSeries._inner[j].getMemberPathFor(13));
                this.qa.setCloseMemberPath(m, this.v.dataSeries._inner[j].getMemberPathFor(14));
                this.qa.setVolumeMemberPath(m, this.v.dataSeries._inner[j].getMemberPathFor(15));
                m.fx = l.fx + " " + i;
                m.k4 = ArrayExtension.getModulus$1(Brush.$, this.ot, e);
                m.ey = this.sf;
                if (this.rd(m)) {
                    this.qa.setNegativeBrush(m, ArrayExtension.getModulus$1(Brush.$, this.ou, e));
                    this.qa.setDisplayType(m, ArrayExtension.getModulus$1(IndicatorDisplayType_$type, this.qc, e));
                }
                if (this.s4 != -2147483648 && this.rd(m)) {
                    this.qa.setPeriodOverride(m, this.s4);
                }
                if (this.rd(m)) {
                    this.qa.setSmoothingPeriod(m, this.s7);
                }
                if (this.rd(m)) {
                    this.qa.setSignalPeriod(m, this.s6);
                }
                if (this.rd(m)) {
                    this.qa.setShortPeriod(m, this.s5);
                    this.qa.setLongPeriod(m, this.s3);
                }
                this.gk(m, e);
                f.add(m);
            }
            let o = this.b1(i, false);
            for (let p = 0; p < f.count; p++) {
                let q = f._inner[p];
                g.series.add(q);
            }
            if (o != null) {
                for (let r = 0; r < o.count; r++) {
                    let s = o._inner[r];
                    g.series.add(s);
                    this.gk(s, r);
                }
            }
        }
    }
    qu(indicatorName_) {
        return (this.externalObject.createIndicator(indicatorName_ + 'Indicator').i);
    }
    qw(a) {
        return this.qu(EnumUtil.getName(FinancialIndicatorType_$type, a));
    }
    w1(a) {
        if (this.pk == null) {
            return;
        }
        let b = this.xAxis;
        if (b.actualMinimumValue > b.actualMaximumValue) {
            return;
        }
        this.pk.as = b.actualMinimumValue;
        this.pk.ar = b.actualMaximumValue;
        if (a) {
            this.pk.au = b.visibleMinimum;
            this.pk.at = b.visibleMaximum;
        }
        this.pk.bn();
    }
    dj() {
        let a = ((() => {
            let $ret = new FinancialChartVisualData();
            $ret.mainChart = super.dj();
            $ret.volumeChart = this.volumeChart == null ? null : this.volumeChart.oe();
            $ret.toolbar = this.pk == null ? null : this.pk.aj();
            $ret.zoomChart = this.zoomSliderChart == null ? null : this.zoomSliderChart.oe();
            return $ret;
        })());
        if (this.indicatorCharts != null) {
            a.indicatorCharts = new List$1(Base.$, 0);
            for (let b of fromEnum(this.indicatorCharts)) {
                a.indicatorCharts.add1(b.oe());
            }
        }
        if (this.dataChart != null) {
            a.mainChartLeft = this.xa(this.dataChart).x;
            a.mainChartTop = this.xa(this.dataChart).y;
        }
        if (this.volumeChart != null) {
            a.volumeChartLeft = this.xa(this.volumeChart).x;
            a.volumeChartTop = this.xa(this.volumeChart).y;
        }
        if (this.zoomSliderChart != null) {
            a.zoomChartLeft = this.xa(this.zoomSliderChart).x;
            a.zoomChartTop = this.xa(this.zoomSliderChart).y;
        }
        if (this.pk != null) {
            let c = this.xb(this.pk);
            a.toolbarLeft = c.left;
            a.toolbarTop = c.top;
            a.toolbarWidth = c.width;
            a.toolbarHeight = c.height;
        }
        if (this.indicatorCharts != null) {
            a.indicatorChartLefts = new List$1(Number_$type, 0);
            a.indicatorChartTops = new List$1(Number_$type, 0);
            for (let d of fromEnum(this.indicatorCharts)) {
                a.indicatorChartLefts.add(this.xa(d).x);
                a.indicatorChartTops.add(this.xa(d).y);
            }
        }
        let e = this.xd();
        a.legendLeft = e.left;
        a.legendTop = e.top;
        a.legendWidth = e.width;
        a.legendHeight = e.height;
        a.legend = this.p4 == null ? null : this.p4.exportVisualData();
        return a;
    }
    tt() {
        let a = this.dj();
        a.scaleByViewport();
        return a.serialize();
    }
    qv(overlayName_) {
        return (this.externalObject.createOverlay(overlayName_ + 'Overlay').i);
    }
    w0() {
        if (this.dataChart == null) {
            return;
        }
        let a = new List$1(Series.$, 0);
        for (let b of fromEnum(this.dataChart.series)) {
            if (this.rf(b)) {
                a.add(b);
            }
        }
        for (let c of fromEnum(a)) {
            this.dataChart.series.remove(c);
        }
        if (this.p2 == null) {
            return;
        }
        let d = new List$1(Series.$, 0);
        let e = -1;
        for (let f = 0; f < this.dataChart.series.count; f++) {
            let g = this.dataChart.series._inner[f];
            if (this.rf(g) || g.cj) {
                continue;
            }
            e++;
            for (let h = 0; h < this.p2.count; h++) {
                let i = this.p2._inner[h];
                let j;
                switch (i) {
                    case 1:
                        j = this.qv(EnumUtil.getName(FinancialOverlayType_$type, i));
                        break;
                    default:
                    case 0:
                        j = this.qv(EnumUtil.getName(FinancialOverlayType_$type, i));
                        break;
                }
                j.name = g.name + "_" + EnumUtil.getName(FinancialOverlayType_$type, i);
                this.qf.setXAxis(j, this.xAxis);
                this.qf.setYAxis(j, this.yAxis);
                j.itemsSource = g.itemsSource;
                let k = e * this.p2.count + h;
                j.k4 = ArrayExtension.getModulus$1(Brush.$, this.ox, k);
                j.k7 = ArrayExtension.getModulus$1(Brush.$, this.oy, k);
                j.ey = this.sg;
                j.fx = g.fx + " " + EnumUtil.getName(FinancialOverlayType_$type, i);
                let l = g;
                if (l == null) {
                    continue;
                }
                let m = this.v.dataSeries._inner[e];
                this.qf.setOpenMemberPath(j, m.getMemberPathFor(11));
                this.qf.setHighMemberPath(j, m.getMemberPathFor(12));
                this.qf.setLowMemberPath(j, m.getMemberPathFor(13));
                this.qf.setCloseMemberPath(j, m.getMemberPathFor(14));
                this.qf.setVolumeMemberPath(j, m.getMemberPathFor(15));
                d.add(j);
            }
        }
        let n = new Array(this.dataChart.series.count);
        this.dataChart.series.copyTo(n, 0);
        this.dataChart.skipDataCleanup = true;
        this.dataChart.series.clear();
        this.dataChart.skipDataCleanup = false;
        for (let o = 0; o < n.length; o++) {
            n[o].fh = -1;
        }
        d.o(n);
        for (let p = 0; p < d.count; p++) {
            let q = d._inner[p];
            this.dataChart.series.add(q);
            this.gk(q, p);
        }
    }
    wy() {
        if (this.rl) {
            return;
        }
        this.rl = true;
        if (this.indicatorCharts != null) {
            let a = this.indicatorCharts;
            this.indicatorCharts = null;
            for (let b of fromEnum(a)) {
                this.wf(b);
            }
        }
        if ((this.p0 == null || this.p0.count == 0) && (this.customIndicatorNames == null || this.customIndicatorNames.count == 0)) {
            this.indicatorCharts = null;
            if (this.ru != null) {
                this.ru.clear();
            }
            if (this.rw != null) {
                this.rw.clear();
            }
            this.wp();
            this.wz();
            this.rl = false;
            return;
        }
        if (this.indicatorCharts == null) {
            this.indicatorCharts = new List$1(XamDataChart.$, 0);
        }
        this.indicatorCharts.clear();
        this.ru.clear();
        this.rw.clear();
        let c = this.p0 == null ? 0 : this.p0.count;
        let d = this.customIndicatorNames == null ? 0 : this.customIndicatorNames.count;
        let e = c + d;
        for (let f = 0; f < e; f++) {
            let g = new XamDataChart();
            this.indicatorCharts.add(g);
            this.wd(g);
            let h = f >= c;
            g.gq = h ? this.customIndicatorNames._inner[f - c] : EnumUtil.getName(FinancialIndicatorType_$type, this.p0._inner[f]);
            if (this.indicatorAxis != null) {
                let i;
                let j;
                if (f == 0) {
                    i = this.indicatorAxis;
                    j = this.indicatorXAxis;
                }
                else {
                    i = new NumericYAxis();
                    if (!this.indicatorXAxis.b5) {
                        j = this.o6(0);
                    }
                    else {
                        j = this.o6(1);
                    }
                }
                j.name = "indicatorXAxis_" + f;
                i.name = "indicatorAxis_" + f;
                g.axes.add(i);
                g.axes.add(j);
                this.ru.add(i);
                this.rw.add(j);
                if (this.bh) {
                    this.m3(i);
                }
                if (this.bh) {
                    this.m1(j);
                }
            }
        }
        this.wj();
        this.wp();
        if (this.pk != null) {
            this.pk.ae = this.p0;
        }
        this.vo();
        this.wz();
        this.rl = false;
    }
    gc() {
        super.gc();
        this.m1(this.xAxis);
        this.m1(this.volumeXAxis);
        if (this.rw != null) {
            for (let a of fromEnum(this.rw)) {
                this.m1(a);
            }
        }
        this.m3(this.yAxis);
        this.m3(this.volumeAxis);
        if (this.ru != null) {
            for (let b of fromEnum(this.ru)) {
                this.m3(b);
            }
        }
        this.m1(this.zoomSliderXAxis);
        this.m3(this.zoomSliderAxis);
    }
    get_bv() {
        return !this.rj;
    }
    get rp() {
        return (this.externalObject._dynamicContent);
    }
    w6(a, b) {
        this.ws(a, b);
        this.vo();
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.axes.add(this.zoomSliderAxis);
            this.zoomSliderChart.axes.add(this.zoomSliderXAxis);
        }
        this.wl();
        this.zoomSliderXAxis.itemsSource = this.xAxis.itemsSource;
        this.wq();
        this.w7(a, b);
    }
    wl() {
        if (this.dataChart == null || this.zoomSliderChart == null) {
            return;
        }
        this.zoomSliderChart.series.clear();
        let a = new List$1(Series.$, 0);
        for (let b = 0; b < this.dataChart.series.count; b++) {
            let c = this.dataChart.series._inner[b];
            if (this.rf(c) || c.cj) {
                continue;
            }
            let d;
            let e = this.py;
            e = this.px(e);
            let f = this.s2(c, b);
            if (e == 1) {
                switch (this.po(this.v != null && this.v.dataSeries != null && this.v.dataSeries.count > f ? this.v.dataSeries._inner[f] : null)) {
                    case 1:
                        e = 2;
                        break;
                    case 2:
                        e = 3;
                        break;
                    case 3:
                        e = 4;
                        break;
                    case 4:
                        e = 5;
                        break;
                }
            }
            switch (e) {
                case 2:
                    d = this.qt();
                    this.p6.setDisplayType(d, 1);
                    break;
                case 3:
                    d = this.qt();
                    this.p6.setDisplayType(d, 0);
                    break;
                case 5:
                    d = this.qs();
                    break;
                case 6:
                    d = this.qq();
                    break;
                default:
                case 4:
                    d = this.qr();
                    break;
            }
            let g = this.v != null && this.v.dataSeries != null && this.v.dataSeries.count > f ? this.v.dataSeries._inner[f] : null;
            if (d.cl && !d.cr) {
                this.p8.setCategoryMode(d, 1);
                this.p8.setXAxis(d, this.zoomSliderXAxis);
                this.p8.setYAxis(d, this.zoomSliderAxis);
                this.p8.setValueMemberPath(d, g == null ? null : this.tv(g));
            }
            else if (d.cr) {
                this.p6.setCategoryMode(d, 1);
                this.p6.setXAxis(d, this.zoomSliderXAxis);
                this.p6.setYAxis(d, this.zoomSliderAxis);
                if (g != null) {
                    this.p6.setOpenMemberPath(d, g.getMemberPathFor(11));
                    this.p6.setHighMemberPath(d, g.getMemberPathFor(12));
                    this.p6.setLowMemberPath(d, g.getMemberPathFor(13));
                    this.p6.setCloseMemberPath(d, g.getMemberPathFor(14));
                    this.p6.setVolumeMemberPath(d, g.getMemberPathFor(15));
                }
            }
            d.name = "zoomSliderSeries_" + f;
            d.itemsSource = c.itemsSource;
            d.fx = c.fx;
            a.add(d);
        }
        for (let h = 0; h < a.count; h++) {
            let i = a._inner[h];
            this.zoomSliderChart.series.add(i);
            this.gk(i, h);
        }
    }
    gy(a, b, c) {
        super.gy(a, b, c);
        if (a == null || !b) {
            return;
        }
        let d = true;
        let e = true;
        switch (c) {
            case "Brushes":
            case "Outlines":
                d = e = false;
                break;
        }
        if (this.indicatorCharts != null && d) {
            for (let f of fromEnum(this.indicatorCharts)) {
                a(f);
            }
        }
        if (this.volumeChart != null && e) {
            a(this.volumeChart);
        }
        if (this.zoomSliderChart != null) {
            a(this.zoomSliderChart);
        }
    }
    vy(a, b) {
        let c = b;
        let d = a;
        let e = -1;
        if (this.indicatorCharts != null) {
            e = this.indicatorCharts.indexOf(d.bq);
        }
        e -= this.p0 == null ? 0 : this.p0.count;
        if (this.applyCustomIndicators != null) {
            this.applyCustomIndicators(this, new FinancialChartCustomIndicatorArgs(c, d, e));
        }
    }
    wu(a, b, c) {
        if (a != null) {
            if (b != null) {
                a.axes.remove(b);
            }
            a.axes.add(c);
        }
        if (this.bh) {
            this.m1(c);
        }
    }
    get ro() {
        return TypeRegistrar.isRegistered("IgxZoomSliderComponent");
    }
    vz(a, b) {
        this.wy();
    }
    rh(a, b) {
        if (b <= (0)) {
            return false;
        }
        switch (a) {
            case 5: return true;
            case 4:
            case 3: return b >= timeSpanFromDays(365);
            case 2: return b >= timeSpanFromDays(180);
            case 1: return b >= timeSpanFromDays(90);
            case 0: return b >= timeSpanFromDays(30);
        }
        return false;
    }
    wo() {
        let a = this.pi != null ? this.pi : new FinancialChartRangeSelectorOptionCollection(1, [5, 4, 3, 2, 1, 0]);
        let b = new FinancialChartRangeSelectorOptionCollection(0);
        let c = this.xAxis;
        let d = +(c.actualMaximumValue) - +(c.actualMinimumValue);
        for (let e of fromEnum(a)) {
            if (this.rh(e, d)) {
                b.add(e);
            }
        }
        this.ph = b;
        if (this.pk != null) {
            this.pk.u = this.ph;
        }
    }
    v3(a, b) {
        this.wo();
    }
    v8(a, b) {
        this.p2 = this.pk.ag;
    }
    v7(a, b) {
        this.p0 = this.pk.ae;
    }
    v1(a, b) {
        this.wy();
    }
    v2(a, b) {
        this.w0();
    }
    wc(a, b) {
        let c = a;
        if (c.bd == null) {
            return;
        }
        if (this.rj) {
            this.wn(c.bd);
        }
    }
    wf(a) {
        SyncManager.e(a);
        a.axes.clear();
        a.series.clear();
        a.gridAreaRectChanged = delegateRemove(a.gridAreaRectChanged, runOn(this, this.v0));
        if (this.volumeChart == null && this.indicatorCharts == null) {
            this.yAxis.d3 = this.volumeAxis.d3 = this.indicatorAxis.d3 = 0;
            this.xAxis.aw.visibility = this.oj;
        }
    }
    wd(a) {
        let b = a != this.zoomSliderChart;
        if (b) {
            if (this.dataChart != null) {
                a.m4 = new Rect(0, this.dataChart.mq.x, a.m4.y, this.dataChart.mq.width, a.m4.height);
            }
            let c = ((() => {
                let $ret = new SyncSettings();
                $ret.syncChannel = "one";
                $ret.synchronizeHorizontally = true;
                $ret.synchronizeVertically = false;
                return $ret;
            })());
            SyncManager.d(a, c);
            SyncManager.c(a);
        }
        a.gridAreaRectChanged = delegateCombine(a.gridAreaRectChanged, runOn(this, this.v0));
        this.ge(a);
    }
    vo() {
        if (this.indicatorCharts != null || this.volumeChart != null) {
            this.mz(this.xAxis, (a) => a.visibility = 1);
        }
        for (let a = 0; a < this.rw.count; a++) {
            let b = this.rw._inner[a];
            let c = a < this.rw.count - 1 || this.volumeChart != null;
            this.mz(b, (d) => d.visibility = c ? 1 : 0);
        }
    }
    ws(a, b) {
        if (a != null) {
            this.wf(a);
        }
        if (b != null) {
            this.wd(b);
        }
    }
    w4(a, b) {
        this.ws(a, b);
        this.vo();
        if (this.volumeChart != null) {
            if (this.volumeAxis != null) {
                this.volumeChart.axes.add(this.volumeAxis);
                this.volumeChart.axes.add(this.volumeXAxis);
            }
            this.wk();
            this.volumeXAxis.itemsSource = this.xAxis.itemsSource;
        }
        this.w5(a, b);
    }
    v4(a, b) {
        this.w2();
    }
    w2() {
        if (this.pk == null || +(this.pk.au) == +(dateMaxValue()) || +(this.pk.at) == +(dateMinValue()) || this.dataChart == null) {
            return;
        }
        let a = this.xAxis;
        let b = a.getWindowRectForSelection(this.pk.au, this.pk.at);
        if (b.width == 0 || isNaN_(b.width) || isInfinity(b.width)) {
            return;
        }
        this.rk = true;
        this.dataChart.m4 = new Rect(0, b.left, this.dataChart.m4.top, b.width, this.dataChart.m4.height);
        if (this.q2 != null) {
            this.q2.windowRect = this.dataChart.mq;
        }
        if (this.rj) {
            this.wn(this.dataChart);
        }
        this.rk = false;
    }
    wb(a, b) {
        this.wo();
        this.w1(!this.rk);
    }
    vx(a, b) {
        if (!this.rk) {
            if (this.rj) {
                this.wn(a);
            }
            let c = false;
            let d = this.g6(() => this.w1(true), c);
            c = d.p1;
            if (this.q2 != null) {
                this.q2.windowRect = this.dataChart.mq;
            }
        }
    }
    ww(a, b, c, d) {
        if (a == null || b == null) {
            return;
        }
        if (c != -1) {
            this.rk = true;
            a.m4 = new Rect(0, d, a.mq.top, c, a.mq.height);
            this.rk = false;
        }
        let e = 1.7976931348623157E+308;
        let f = -1.7976931348623157E+308;
        let g = new Rect(0, 0, 0, 1, 1);
        let h = new ScalerParams(0, g, g, this.yAxis.b8, a.mu);
        for (let i of fromEnum(a.series)) {
            h.b = i.eh();
            let j = i.ao();
            let k = j == null || j.c == 1.7976931348623157E+308 || isNaN_(j.c);
            if (k) {
                continue;
            }
            let l = b.dv(j.c, h);
            let m = b.dv(j.b, h);
            f = Math.max(f, l);
            e = Math.min(e, m);
        }
        let n = a.mq.width;
        if (c != -1) {
            n = c;
        }
        let o = a.mq.left;
        if (d != -1) {
            o = d;
        }
        if (e == 1.7976931348623157E+308 || f == -1.7976931348623157E+308 || isNaN_(e) || isNaN_(f)) {
            return;
        }
        this.rk = true;
        let p = new Rect(0, o, e, n, f - e);
        a.m4 = p;
        this.rk = false;
    }
    wv(a) {
        if (a == null) {
            return;
        }
        if (a == this.dataChart) {
            this.ww(this.dataChart, this.yAxis, -1, -1);
            if (this.volumeChart != null) {
                this.ww(this.volumeChart, this.volumeAxis, this.dataChart.mq.width, this.dataChart.mq.left);
            }
        }
        else if (a == this.volumeChart) {
            this.ww(this.volumeChart, this.volumeAxis, -1, -1);
            this.ww(this.dataChart, this.yAxis, this.volumeChart.mq.width, this.volumeChart.mq.left);
        }
    }
    wn(a) {
        this.g5(() => this.wv(a));
    }
    v5(a, b) {
        if (this.rj) {
            this.wn(a);
        }
    }
    w3() {
        if (!this.rp) {
            return;
        }
        this.vr();
        if (this.py == 0) {
            this.zoomSliderChart = null;
        }
        else if (this.zoomSliderChart == null) {
            this.zoomSliderChart = ((() => {
                let $ret = new XamDataChart();
                $ret.name = "ZoomSliderChart";
                return $ret;
            })());
        }
        else {
            this.wl();
        }
    }
    vr() {
        let container_ = this.su.rootWrapper.getNativeElement();
        let containerParent_ = container_.parentElement;
        if (this.q2 != null && containerParent_ != null && !this.rq) {
            this.rq = true;
            if (TypeRegistrar.isRegistered("IgxZoomSliderComponent")) {
                let z_ = this.q2;
                let f_ = this.externalObject;
                let sliderComponent_ = TypeRegistrar.get("IgxZoomSliderComponent");
                let sliderInstance_ = TypeRegistrar.create("IgxZoomSliderComponent");
                sliderInstance_._implementation = z_;
                sliderComponent_._staticStyling(container_, sliderInstance_, f_);
            }
        }
    }
    wx() {
        if (this.pt != 1) {
            return;
        }
        this.m5((a) => {
            let b;
            b = a.breaks;
            b.clear();
            if (this.xAxisBreaks == null) {
                return;
            }
            for (let c of fromEnum(this.xAxisBreaks)) {
                b.add(c);
            }
        });
    }
    ho(a, b, c) {
        super.ho(a, b, c);
        if (this.v.dataSeries.count > 0) {
            if (this.pt == 1 && stringIsNullOrEmpty(this.ea)) {
                this.ac.setXMemberPath(a, c.getMemberPathFor(17));
            }
            if (stringIsNullOrEmpty(this.ec)) {
                this.ac.setYMemberPath(a, this.tv(c));
            }
            if (stringIsNullOrEmpty(this.d6)) {
                this.ac.setLabelMemberPath(a, this.tv(c));
            }
            if (stringIsNullOrEmpty(this.d3)) {
                this.ac.setContentMemberPath(a, this.tv(c));
            }
        }
    }
    m1(a) {
        super.m1(a);
        let b = a;
        this.we(b);
        if (a.bd != null) {
            for (let c of fromEnum(a.bd.series)) {
                if (this.rc(c)) {
                    this.p8.setXAxis(c, b);
                }
                else if (this.rg(c)) {
                    this.p6.setXAxis(c, b);
                }
                else if (this.rd(c)) {
                    this.qa.setXAxis(c, b);
                }
                else if (this.rf(c)) {
                    this.qf.setXAxis(c, b);
                }
            }
        }
        let d = a == this.zoomSliderXAxis;
        if (d) {
            a.ii = this.w9;
            a.d2 = this.sp;
            this.mz(a, (e) => e.visibility = 1);
        }
        let e = a;
        e.minimumValue = this.r2;
        e.maximumValue = this.r1;
    }
    m3(a) {
        super.m3(a);
        if (a.bd != null && a == this.yAxis) {
            for (let b of fromEnum(a.bd.series)) {
                if (this.rc(b)) {
                    this.p8.setYAxis(b, this.yAxis);
                }
                else if (this.rg(b)) {
                    this.p6.setYAxis(b, this.yAxis);
                }
                else if (this.rd(b)) {
                    this.qa.setYAxis(b, this.yAxis);
                }
                else if (this.rf(b)) {
                    this.qf.setYAxis(b, this.yAxis);
                }
            }
        }
        a.jn = this.rm;
        a.kb = this.sn;
        a.ka = this.sm;
        if (a == this.yAxis) {
            a.j7 = this.sl;
        }
        let c = a == this.zoomSliderAxis;
        if (c) {
            this.mz(a, (d) => d.visibility = 1);
        }
    }
    we(a) {
        a.itemsSource = this.dataChart != null && this.dataChart.series.count > 0 ? this.dataChart.series._inner[0].itemsSource : null;
        a.e6 = XYChart.kv(a, this.v.dataSeries);
        if (this.v.dataSeries.count > 0) {
            let b = this.v.dataSeries._inner[0].getMemberPathFor(17);
            a.gv(b);
        }
    }
    gk(a, b) {
        super.gk(a, b);
        if (a == null) {
            return;
        }
        let c = this.dataChart != null && this.dataChart.series.contains(a);
        if (c && a.cz) {
            a.jh(this.ov != null ? ArrayExtension.getModulus$1(Brush.$, this.ov, b) : null, this.ow != null ? ArrayExtension.getModulus$1(Brush.$, this.ow, b) : null);
        }
        let d = this.zoomSliderChart == a.bq;
        let e = this.rd(a);
        if (this.rf(a)) {
            a.k4 = this.ox != null ? ArrayExtension.getModulus$1(Brush.$, this.ox, b) : null;
            a.k7 = this.oy != null ? ArrayExtension.getModulus$1(Brush.$, this.oy, b) : null;
            a.ey = this.sg;
        }
        else if (e) {
            a.k4 = this.ot != null ? ArrayExtension.getModulus$1(Brush.$, this.ot, b) : null;
            a.ey = this.sf;
        }
        else if (!a.cj && (c || d)) {
            a.k4 = this.j != null ? ArrayExtension.getModulus$1(Brush.$, this.j, b) : null;
            a.k7 = this.n != null ? ArrayExtension.getModulus$1(Brush.$, this.n, b) : null;
        }
        let f = this.volumeChart != null && this.volumeChart == a.bq;
        if (f) {
            a.ey = this.si;
        }
        a.legend = d || f || e || !a.isUsableInLegend ? null : this.legend != null ? this.legend : this.re ? this.p4 : null;
    }
    g1(a, b, c) {
        super.g1(a, b, c);
        switch (a) {
            case "YAxisIsLogarithmic":
                this.yAxis.js = this.rn;
                if (this.volumeAxis != null) {
                    this.volumeAxis.js = this.rn;
                }
                if (this.zoomSliderAxis != null) {
                    this.zoomSliderAxis.js = this.rn;
                }
                break;
            case "YAxisLogarithmBase":
                this.yAxis.kf = this.s9;
                if (this.volumeAxis != null) {
                    this.volumeAxis.kf = this.s9;
                }
                if (this.zoomSliderAxis != null) {
                    this.zoomSliderAxis.kf = this.s9;
                }
                break;
            case "VolumeType":
                let d = EnumUtil.getEnumValue(FinancialChartVolumeType_$type, b);
                if (this.pr == 0) {
                    this.volumeChart = null;
                }
                else if (this.volumeChart == null) {
                    this.volumeChart = ((() => {
                        let $ret = new XamDataChart();
                        $ret.name = "VolumeChart";
                        return $ret;
                    })());
                }
                else {
                    this.wk();
                }
                break;
            case "VolumeChart":
                this.w4(typeCast(XamDataChart.$, b), typeCast(XamDataChart.$, c));
                break;
            case "DataChart":
                if (typeCast(XamDataChart.$, b) !== null) {
                    let e = b;
                    e.gridAreaRectChanged = delegateRemove(e.gridAreaRectChanged, runOn(this, this.v0));
                    e.actualWindowRectChanged = delegateRemove(e.actualWindowRectChanged, runOn(this, this.vx));
                    e.seriesDataUpdated = delegateRemove(e.seriesDataUpdated, runOn(this, this.v5));
                }
                if (this.dataChart != null) {
                    let f = this.dataChart;
                    f.gridAreaRectChanged = delegateCombine(f.gridAreaRectChanged, runOn(this, this.v0));
                    let g = this.dataChart;
                    g.actualWindowRectChanged = delegateCombine(g.actualWindowRectChanged, runOn(this, this.vx));
                    let h = this.dataChart;
                    h.seriesDataUpdated = delegateCombine(h.seriesDataUpdated, runOn(this, this.v5));
                }
                this.w3();
                break;
            case "XAxisMode":
                this.xAxis = this.o6(this.pt);
                this.xAxis.name = "xAxis";
                this.volumeXAxis = this.o6(this.pt);
                this.volumeXAxis.name = "volumeXAxis";
                this.indicatorXAxis = this.o6(this.pt);
                this.indicatorXAxis.name = "indicatorXAxis_0";
                this.zoomSliderXAxis = this.o6(this.pt);
                this.zoomSliderXAxis.name = "zoomSliderXAxis";
                this.mz(this.zoomSliderXAxis, (i) => i.visibility = 1);
                this.vo();
                if (this.pt == 1) {
                    this.wx();
                }
                break;
            case "YAxisMode":
                this.yAxis = this.qk(this.pv);
                this.yAxis.name = "yAxis";
                this.zoomSliderAxis = this.qk(this.pv);
                this.zoomSliderAxis.name = "zoomSliderAxis";
                this.volumeAxis = this.qk(this.pv);
                this.volumeAxis.name = "volumeAxis";
                break;
            case "Toolbar":
                let i = typeCast(FinancialChartToolbar.$, b);
                if (i != null) {
                    i.rangeChanged = delegateRemove(i.rangeChanged, runOn(this, this.v4));
                    i.indicatorTypesChanged = delegateRemove(i.indicatorTypesChanged, runOn(this, this.v7));
                    i.overlayTypesChanged = delegateRemove(i.overlayTypesChanged, runOn(this, this.v8));
                }
                if (this.pk != null) {
                    this.pk.ae = this.p0;
                    this.pk.ag = this.p2;
                    this.pk.u = this.ph;
                    this.w1(true);
                    let j = this.pk;
                    j.rangeChanged = delegateCombine(j.rangeChanged, runOn(this, this.v4));
                    let k = this.pk;
                    k.indicatorTypesChanged = delegateCombine(k.indicatorTypesChanged, runOn(this, this.v7));
                    let l = this.pk;
                    l.overlayTypesChanged = delegateCombine(l.overlayTypesChanged, runOn(this, this.v8));
                }
                break;
            case "XAxis":
                let m = typeCast(CategoryAxisBase.$, b);
                if (m != null) {
                    if (this.dataChart != null) {
                        this.dataChart.axes.remove(m);
                    }
                }
                let n = typeCast(ITimeRangeSelectorAxis_$type, b);
                if (n != null) {
                    n.visibleRangeChanged = delegateRemove(n.visibleRangeChanged, runOn(this, this.wb));
                }
                if (this.dataChart != null) {
                    this.dataChart.axes.add(this.xAxis);
                }
                if (this.bh) {
                    this.m1(this.xAxis);
                    this.f7();
                }
                this.w1(false);
                this.w2();
                this.wo();
                let o = this.xAxis;
                o.visibleRangeChanged = delegateCombine(o.visibleRangeChanged, runOn(this, this.wb));
                break;
            case "ChartType":
                this.hj();
                break;
            case "VolumeXAxis":
                this.wu(this.volumeChart, typeCast(Axis.$, b), this.volumeXAxis);
                break;
            case "ZoomSliderXAxis":
                this.wu(this.zoomSliderChart, typeCast(Axis.$, b), this.zoomSliderXAxis);
                break;
            case "IndicatorXAxis":
                this.wy();
                break;
            case "YAxis":
                let p = typeCast(NumericYAxis.$, b);
                if (p != null) {
                    if (this.dataChart != null) {
                        this.dataChart.axes.remove(p);
                    }
                }
                if (this.dataChart != null) {
                    this.dataChart.axes.add(this.yAxis);
                }
                if (this.bh) {
                    this.m3(this.yAxis);
                    this.m3(this.yAxis);
                }
                let q = this.yAxis;
                q.rangeChanged = delegateCombine(q.rangeChanged, runOn(this, this.wc));
                break;
            case "VolumeAxis":
                let r = typeCast(Axis.$, b);
                if (this.volumeChart != null && r != null) {
                    this.volumeChart.axes.remove(r);
                }
                if (this.volumeChart != null && this.volumeAxis != null) {
                    this.volumeChart.axes.add(this.volumeAxis);
                    this.wk();
                }
                if (this.bh) {
                    this.m3(this.volumeAxis);
                }
                let s = this.volumeAxis;
                s.rangeChanged = delegateCombine(s.rangeChanged, runOn(this, this.wc));
                break;
            case "IndicatorAxis":
                if (this.indicatorCharts != null) {
                    for (let t = 0; t < this.indicatorCharts.count; t++) {
                        let u = this.indicatorCharts._inner[t];
                        let v = new List$1(Axis.$, 0);
                        for (let w of fromEnum(u.axes)) {
                            if (typeCast(NumericYAxis.$, w) !== null) {
                                v.add(w);
                            }
                        }
                        for (let x of fromEnum(v)) {
                            u.axes.remove(x);
                        }
                        let y = null;
                        if (t == 0) {
                            y = this.indicatorAxis;
                        }
                        else if (this.pv == 1) {
                            y = this.qk(this.pv);
                        }
                        else if (typeCast(NumericYAxis.$, this.indicatorAxis) !== null) {
                            y = new NumericYAxis();
                        }
                        if (y != null) {
                            u.axes.add(y);
                        }
                        if (this.bh) {
                            this.m3(y);
                        }
                    }
                }
                break;
            case "IsWindowSyncedToVisibleRange":
                if (this.dataChart != null) {
                    this.dataChart.nt = this.bw && this.bv;
                }
                if (this.volumeChart != null) {
                    this.volumeChart.nt = this.bw && this.bv;
                }
                if (this.rj) {
                    this.wv(this.dataChart);
                }
                else {
                    this.volumeChart.oc = 1;
                }
                break;
            case "IndicatorTypes":
                let z = typeCast(ObservableCollection$1.$.specialize(FinancialIndicatorType_$type), b);
                if (z != null) {
                    z.collectionChanged = delegateRemove(z.collectionChanged, runOn(this, this.v1));
                }
                let aa = typeCast(ObservableCollection$1.$.specialize(FinancialIndicatorType_$type), c);
                if (aa != null) {
                    aa.collectionChanged = delegateCombine(aa.collectionChanged, runOn(this, this.v1));
                }
                this.wy();
                break;
            case "OverlayTypes":
                let ab = typeCast(ObservableCollection$1.$.specialize(FinancialOverlayType_$type), b);
                if (ab != null) {
                    ab.collectionChanged = delegateRemove(ab.collectionChanged, runOn(this, this.v2));
                }
                let ac = typeCast(ObservableCollection$1.$.specialize(FinancialOverlayType_$type), c);
                if (ac != null) {
                    ac.collectionChanged = delegateCombine(ac.collectionChanged, runOn(this, this.v2));
                }
                this.w0();
                break;
            case "YAxisAbbreviateLargeNumbers":
                this.m9((ad) => ad.jn = this.rm);
                break;
            case "YAxisMinimumValue":
                this.na((ad) => ad.kb = this.sn, false);
                break;
            case "YAxisMaximumValue":
                this.na((ad) => ad.ka = this.sm, false);
                break;
            case "YAxisInterval":
                this.na((ad) => ad.j7 = this.sl, false);
                break;
            case "ZoomSliderType":
                this.w3();
                break;
            case "ZoomSliderChart":
                if (!this.ro) {
                    return;
                }
                this.w6(typeCast(XamDataChart.$, b), typeCast(XamDataChart.$, c));
                break;
            case "NegativeBrushes":
            case "NegativeOutlines":
                this.vu((ad, ae) => {
                    if (ad.cz) {
                        ad.jh(this.ov != null ? ArrayExtension.getModulus$1(Brush.$, this.ov, ae) : null, this.ow != null ? ArrayExtension.getModulus$1(Brush.$, this.ow, ae) : null);
                    }
                }, true);
                break;
            case "OverlayBrushes":
                this.f9((ad, ae) => {
                    if (!(this.rf(ad))) {
                        return;
                    }
                    ad.k4 = this.ox != null ? ArrayExtension.getModulus$1(Brush.$, this.ox, ae) : null;
                });
                break;
            case "OverlayOutlines":
                this.f9((ad, ae) => {
                    if (!(this.rf(ad))) {
                        return;
                    }
                    ad.k7 = this.oy != null ? ArrayExtension.getModulus$1(Brush.$, this.oy, ae) : null;
                });
                break;
            case "OverlayThickness":
                this.f9((ad, ae) => {
                    if (!(this.rf(ad))) {
                        return;
                    }
                    ad.ey = this.sg;
                });
                break;
            case "VolumeBrushes":
                this.vv((ad, ae) => ad.k4 = ArrayExtension.getModulus$1(Brush.$, this.oz, ae));
                break;
            case "VolumeOutlines":
                this.vv((ad, ae) => ad.k7 = ArrayExtension.getModulus$1(Brush.$, this.o0, ae));
                break;
            case "VolumeThickness":
                this.vv((ad, ae) => ad.ey = this.si);
                break;
            case "IndicatorNegativeBrushes":
                this.vs((ad, ae) => this.qa.setNegativeBrush(ad, ArrayExtension.getModulus$1(Brush.$, this.ou, ae)));
                break;
            case "IndicatorBrushes":
                this.vs((ad, ae) => ad.k4 = ArrayExtension.getModulus$1(Brush.$, this.ot, ae));
                break;
            case "IndicatorThickness":
                this.vs((ad, ae) => ad.ey = this.sf);
                break;
            case "IndicatorDisplayTypes":
                this.vs((ad, ae) => this.qa.setDisplayType(ad, ArrayExtension.getModulus$1(IndicatorDisplayType_$type, this.qc, ae)));
                break;
            case "IndicatorPeriod":
                this.vs((ad, ae) => {
                    if (this.s4 == -2147483648 || !(this.rd(ad))) {
                        return;
                    }
                    this.qa.setPeriodOverride(ad, this.s4);
                });
                break;
            case "IndicatorMultiplier":
                this.vs((ad, ae) => {
                    if (!(this.rf(ad))) {
                        return;
                    }
                    this.qf.setMultiplier(ad, this.se);
                });
                break;
            case "IndicatorSmoothingPeriod":
                this.vs((ad, ae) => {
                    if (!(this.rd(ad))) {
                        return;
                    }
                    this.qa.setSmoothingPeriod(ad, this.s7);
                });
                break;
            case "IndicatorShortPeriod":
                this.vs((ad, ae) => {
                    if (!(this.rd(ad))) {
                        return;
                    }
                    this.qa.setShortPeriod(ad, this.s5);
                });
                break;
            case "IndicatorLongPeriod":
                this.vs((ad, ae) => {
                    if (!(this.rd(ad))) {
                        return;
                    }
                    this.qa.setLongPeriod(ad, this.s3);
                });
                break;
            case "IndicatorSignalPeriod":
                this.vs((ad, ae) => {
                    if (!(this.rd(ad))) {
                        return;
                    }
                    this.qa.setSignalPeriod(ad, this.s6);
                });
                break;
            case "RangeSelectorOptions":
                let ad = typeCast(FinancialChartRangeSelectorOptionCollection.$, b);
                if (ad != null) {
                    ad.collectionChanged = delegateRemove(ad.collectionChanged, runOn(this, this.v3));
                }
                if (this.pi != null) {
                    let ae = this.pi;
                    ae.collectionChanged = delegateCombine(ae.collectionChanged, runOn(this, this.v3));
                }
                this.wo();
                break;
            case "CustomIndicatorNames":
                let af = typeCast(ObservableCollection$1.$.specialize(String_$type), b);
                if (af != null) {
                    af.collectionChanged = delegateRemove(af.collectionChanged, runOn(this, this.vz));
                }
                if (this.customIndicatorNames != null) {
                    let ag = this.customIndicatorNames;
                    ag.collectionChanged = delegateCombine(ag.collectionChanged, runOn(this, this.vz));
                }
                this.wy();
                break;
            case "Legend":
                if (this.zoomSliderChart != null) {
                    this.zoomSliderChart.legend = null;
                }
                break;
            case "Brushes":
                this.vu((ah, ai) => ah.k4 = this.j != null ? ArrayExtension.getModulus$1(Brush.$, this.j, ai) : null, true);
                break;
            case "Outlines":
                this.vu((ah, ai) => ah.k7 = this.n != null ? ArrayExtension.getModulus$1(Brush.$, this.n, ai) : null, true);
                break;
            case "XAxisLabelVisibility":
                if (this.oj == 0 && this.zoomSliderXAxis != null) {
                    this.mz(this.zoomSliderXAxis, (ah) => ah.visibility = 1);
                }
                this.vo();
                break;
            case "YAxisLabelVisibility":
                if (this.ok != 0 || this.zoomSliderAxis == null) {
                    break;
                }
                this.mz(this.zoomSliderAxis, (ah) => ah.visibility = 1);
                break;
            case "ZoomSliderXAxisMajorStroke":
                this.zoomSliderXAxis.ii = this.w9;
                break;
            case "ZoomSliderXAxisMajorStrokeThickness":
                this.zoomSliderXAxis.d2 = this.sp;
                break;
            case "FinancialChartLegend":
            case "IsLegendVisible":
                let ah = this.re && this.legend == null;
                {
                    let ai = (aj, ak) => {
                        if (aj.isUsableInLegend) {
                            aj.legend = ah ? this.p4 : null;
                        }
                    };
                    this.vu(ai, false);
                    this.vt(ai);
                }
                break;
            case "ZoomSliderAxis":
                let aj = typeCast(Axis.$, b);
                if (aj != null && this.zoomSliderChart != null) {
                    this.zoomSliderChart.axes.remove(aj);
                }
                if (this.zoomSliderAxis != null) {
                    if (this.zoomSliderChart != null) {
                        this.zoomSliderChart.axes.add(this.zoomSliderAxis);
                        this.wl();
                        this.zoomSliderXAxis.itemsSource = this.xAxis.itemsSource;
                    }
                    if (this.bh) {
                        this.m3(this.zoomSliderAxis);
                    }
                    this.mz(this.zoomSliderAxis, (ak) => ak.visibility = 1);
                }
                break;
            case "XAxisMinimumValue":
                this.m5((ak) => {
                    let al = typeCast(ITimeRangeSelectorAxis_$type, ak);
                    if (al != null) {
                        al.minimumValue = this.r2;
                    }
                });
                break;
            case "XAxisMaximumValue":
                this.m5((ak) => {
                    let al = typeCast(ITimeRangeSelectorAxis_$type, ak);
                    if (al != null) {
                        al.maximumValue = this.r1;
                    }
                });
                break;
        }
    }
    get rn() {
        return this.q7;
    }
    set rn(a) {
        let b = this.rn;
        if (a != b) {
            this.q7 = a;
            this.g1("YAxisIsLogarithmic", b, this.rn);
        }
    }
    get s9() {
        return this.s1;
    }
    set s9(a) {
        let b = this.s9;
        if (a != b) {
            this.s1 = a;
            this.g1("YAxisLogarithmBase", b, this.s9);
        }
    }
    get sl() {
        return this.r8;
    }
    set sl(a) {
        let b = this.sl;
        if (a != b) {
            this.r8 = a;
            this.g1("YAxisInterval", b, this.sl);
        }
    }
    get sn() {
        return this.sa;
    }
    set sn(a) {
        let b = this.sn;
        if (a != b) {
            this.sa = a;
            this.g1("YAxisMinimumValue", b, this.sn);
        }
    }
    get sm() {
        return this.r9;
    }
    set sm(a) {
        let b = this.sm;
        if (a != b) {
            this.r9 = a;
            this.g1("YAxisMaximumValue", b, this.sm);
        }
    }
    get so() {
        return this.sb;
    }
    set so(a) {
        let b = this.so;
        if (a != b) {
            this.sb = a;
            this.g1("YAxisMinorInterval", b, this.so);
        }
    }
    get sk() {
        return this.yAxis.j3;
    }
    get sj() {
        return this.yAxis.j2;
    }
    get pr() {
        return this.pp;
    }
    set pr(a) {
        let b = this.pr;
        if (a != b) {
            this.pp = a;
            this.g1("VolumeType", enumGetBox(FinancialChartVolumeType_$type, b), enumGetBox(FinancialChartVolumeType_$type, this.pr));
        }
    }
    get pt() {
        return this.ps;
    }
    set pt(a) {
        let b = this.pt;
        if (a != b) {
            this.ps = a;
            this.g1("XAxisMode", enumGetBox(FinancialChartXAxisMode_$type, b), enumGetBox(FinancialChartXAxisMode_$type, this.pt));
        }
    }
    get pv() {
        return this.pu;
    }
    set pv(a) {
        let b = this.pv;
        if (a != b) {
            this.pu = a;
            this.g1("YAxisMode", enumGetBox(FinancialChartYAxisMode_$type, b), enumGetBox(FinancialChartYAxisMode_$type, this.pv));
        }
    }
    get ri() {
        return this.q4;
    }
    set ri(a) {
        let b = this.ri;
        if (a != b) {
            this.q4 = a;
            this.g1("IsToolbarVisible", b, this.ri);
        }
    }
    get pm() {
        return this.pl;
    }
    set pm(a) {
        let b = this.pm;
        if (a != b) {
            this.pl = a;
            this.g1("ChartType", enumGetBox(FinancialChartType_$type, b), enumGetBox(FinancialChartType_$type, this.pm));
        }
    }
    get_ni() {
        return super.get_ni();
    }
    set_ni(a) {
        super.set_ni(a);
    }
    get rj() {
        return this.q5;
    }
    set rj(a) {
        let b = this.rj;
        if (a != b) {
            this.q5 = a;
            this.g1("IsWindowSyncedToVisibleRange", b, this.rj);
        }
    }
    get p0() {
        return this.pz;
    }
    set p0(a) {
        let b = this.p0;
        if (a != b) {
            this.pz = a;
            this.g1("IndicatorTypes", b, this.p0);
        }
    }
    get p2() {
        return this.p1;
    }
    set p2(a) {
        let b = this.p2;
        if (a != b) {
            this.p1 = a;
            this.g1("OverlayTypes", b, this.p2);
        }
    }
    get indicatorCharts() {
        return this.rs;
    }
    set indicatorCharts(a) {
        let b = this.indicatorCharts;
        if (a != b) {
            this.rs = a;
            this.g1("IndicatorCharts", b, this.indicatorCharts);
        }
    }
    get rm() {
        return this.q6;
    }
    set rm(a) {
        let b = this.rm;
        if (a != b) {
            this.q6 = a;
            this.g1("YAxisAbbreviateLargeNumbers", b, this.rm);
        }
    }
    get_cu() {
        return super.get_cu();
    }
    set_cu(a) {
        super.set_cu(a);
    }
    get_ie() {
        return super.get_ie();
    }
    set_ie(a) {
        super.set_ie(a);
    }
    get zoomSliderChart() {
        return this.qy;
    }
    set zoomSliderChart(a) {
        let b = this.zoomSliderChart;
        if (a != b) {
            this.qy = a;
            this.g1("ZoomSliderChart", b, this.zoomSliderChart);
        }
    }
    get zoomSliderAxis() {
        return this.qj;
    }
    set zoomSliderAxis(a) {
        let b = this.zoomSliderAxis;
        if (a != b) {
            this.qj = a;
            this.g1("ZoomSliderAxis", b, this.zoomSliderAxis);
        }
    }
    get zoomSliderXAxis() {
        return this.o4;
    }
    set zoomSliderXAxis(a) {
        let b = this.zoomSliderXAxis;
        if (a != b) {
            this.o4 = a;
            this.g1("ZoomSliderXAxis", b, this.zoomSliderXAxis);
        }
    }
    get py() {
        return this.pw;
    }
    set py(a) {
        let b = this.py;
        if (a != b) {
            this.pw = a;
            this.g1("ZoomSliderType", enumGetBox(FinancialChartZoomSliderType_$type, b), enumGetBox(FinancialChartZoomSliderType_$type, this.py));
        }
    }
    get ov() {
        return this.on;
    }
    set ov(a) {
        let b = this.ov;
        if (a != b) {
            this.on = a;
            this.g1("NegativeBrushes", b, this.ov);
        }
    }
    get ow() {
        return this.oo;
    }
    set ow(a) {
        let b = this.ow;
        if (a != b) {
            this.oo = a;
            this.g1("NegativeOutlines", b, this.ow);
        }
    }
    get ox() {
        return this.op;
    }
    set ox(a) {
        let b = this.ox;
        if (a != b) {
            this.op = a;
            this.g1("OverlayBrushes", b, this.ox);
        }
    }
    get oy() {
        return this.oq;
    }
    set oy(a) {
        let b = this.oy;
        if (a != b) {
            this.oq = a;
            this.g1("OverlayOutlines", b, this.oy);
        }
    }
    get o0() {
        return this.os;
    }
    set o0(a) {
        let b = this.o0;
        if (a != b) {
            this.os = a;
            this.g1("VolumeOutlines", b, this.o0);
        }
    }
    get oz() {
        return this.or;
    }
    set oz(a) {
        let b = this.oz;
        if (a != b) {
            this.or = a;
            this.g1("VolumeBrushes", b, this.oz);
        }
    }
    get ou() {
        return this.om;
    }
    set ou(a) {
        let b = this.ou;
        if (a != b) {
            this.om = a;
            this.g1("IndicatorNegativeBrushes", b, this.ou);
        }
    }
    get ot() {
        return this.ol;
    }
    set ot(a) {
        let b = this.ot;
        if (a != b) {
            this.ol = a;
            this.g1("IndicatorBrushes", b, this.ot);
        }
    }
    get si() {
        return this.r7;
    }
    set si(a) {
        let b = this.si;
        if (a != b) {
            this.r7 = a;
            this.g1("VolumeThickness", b, this.si);
        }
    }
    get sg() {
        return this.r5;
    }
    set sg(a) {
        let b = this.sg;
        if (a != b) {
            this.r5 = a;
            this.g1("OverlayThickness", b, this.sg);
        }
    }
    get sf() {
        return this.r4;
    }
    set sf(a) {
        let b = this.sf;
        if (a != b) {
            this.r4 = a;
            this.g1("IndicatorThickness", b, this.sf);
        }
    }
    get qc() {
        return this.qb;
    }
    set qc(a) {
        let b = this.qc;
        if (a != b) {
            this.qb = a;
            this.g1("IndicatorDisplayTypes", b, this.qc);
        }
    }
    get s4() {
        return this.sx;
    }
    set s4(a) {
        let b = this.s4;
        if (a != b) {
            this.sx = a;
            this.g1("IndicatorPeriod", b, this.s4);
        }
    }
    get se() {
        return this.r3;
    }
    set se(a) {
        let b = this.se;
        if (a != b) {
            this.r3 = a;
            this.g1("IndicatorMultiplier", b, this.se);
        }
    }
    get s7() {
        return this.s0;
    }
    set s7(a) {
        let b = this.s7;
        if (a != b) {
            this.s0 = a;
            this.g1("IndicatorSmoothingPeriod", b, this.s7);
        }
    }
    get s5() {
        return this.sy;
    }
    set s5(a) {
        let b = this.s5;
        if (a != b) {
            this.sy = a;
            this.g1("IndicatorShortPeriod", b, this.s5);
        }
    }
    get s3() {
        return this.sw;
    }
    set s3(a) {
        let b = this.s3;
        if (a != b) {
            this.sw = a;
            this.g1("IndicatorLongPeriod", b, this.s3);
        }
    }
    get s6() {
        return this.sz;
    }
    set s6(a) {
        let b = this.s6;
        if (a != b) {
            this.sz = a;
            this.g1("IndicatorSignalPeriod", b, this.s6);
        }
    }
    get pi() {
        return this.pg;
    }
    set pi(a) {
        let b = this.pi;
        if (a != b) {
            this.pg = a;
            this.g1("RangeSelectorOptions", b, this.pi);
        }
    }
    get customIndicatorNames() {
        return this.pc;
    }
    set customIndicatorNames(a) {
        let b = this.pc;
        if (a != b) {
            this.pc = a;
            this.g1("CustomIndicatorNames", b, this.customIndicatorNames);
        }
    }
    get_bw() {
        return super.get_bw();
    }
    set_bw(a) {
        super.set_bw(a);
    }
    get w9() {
        return this.w8;
    }
    set w9(a) {
        let b = this.w9;
        if (a != b) {
            this.w8 = a;
            this.g1("ZoomSliderXAxisMajorStroke", b, this.w9);
        }
    }
    get sp() {
        return this.sc;
    }
    set sp(a) {
        let b = this.sp;
        if (a != b) {
            this.sc = a;
            this.g1("ZoomSliderXAxisMajorStrokeThickness", b, this.sp);
        }
    }
    get re() {
        return this.q3;
    }
    set re(a) {
        let b = this.re;
        if (a != b) {
            this.q3 = a;
            this.g1("IsLegendVisible", b, this.re);
        }
    }
    get p4() {
        return this.p3;
    }
    set p4(a) {
        let b = this.p4;
        if (a != b) {
            this.p3 = a;
            this.g1("FinancialChartLegend", b, this.p4);
        }
    }
    get xAxis() {
        return this.o3;
    }
    set xAxis(a) {
        let b = this.xAxis;
        if (a != b) {
            this.o3 = a;
            this.g1("XAxis", b, this.xAxis);
        }
    }
    get yAxis() {
        return this.qi;
    }
    set yAxis(a) {
        let b = this.yAxis;
        if (a != b) {
            this.qi = a;
            this.g1("YAxis", b, this.yAxis);
        }
    }
    get volumeAxis() {
        return this.qh;
    }
    set volumeAxis(a) {
        let b = this.volumeAxis;
        if (a != b) {
            this.qh = a;
            this.g1("VolumeAxis", b, this.volumeAxis);
        }
    }
    get volumeXAxis() {
        return this.o2;
    }
    set volumeXAxis(a) {
        let b = this.volumeXAxis;
        if (a != b) {
            this.o2 = a;
            this.g1("VolumeXAxis", b, this.volumeXAxis);
        }
    }
    get indicatorAxis() {
        return this.qg;
    }
    set indicatorAxis(a) {
        let b = this.indicatorAxis;
        if (a != b) {
            this.qg = a;
            this.g1("IndicatorAxis", b, this.indicatorAxis);
        }
    }
    get indicatorXAxis() {
        return this.o1;
    }
    set indicatorXAxis(a) {
        let b = this.indicatorXAxis;
        if (a != b) {
            this.o1 = a;
            this.g1("IndicatorXAxis", b, this.indicatorXAxis);
        }
    }
    get ru() {
        return this.rr;
    }
    set ru(a) {
        let b = this.ru;
        if (a != b) {
            this.rr = a;
            this.g1("IndicatorAxes", b, this.ru);
        }
    }
    get rw() {
        return this.rt;
    }
    set rw(a) {
        let b = this.rw;
        if (a != b) {
            this.rt = a;
            this.g1("IndicatorXAxes", b, this.rw);
        }
    }
    get r2() {
        return this.r0;
    }
    set r2(a) {
        let b = this.r2;
        if (+a != +b) {
            this.r0 = a;
            this.g1("XAxisMinimumValue", b, this.r2);
        }
    }
    get r1() {
        return this.rz;
    }
    set r1(a) {
        let b = this.r1;
        if (+a != +b) {
            this.rz = a;
            this.g1("XAxisMaximumValue", b, this.r1);
        }
    }
    get xAxisBreaks() {
        return this.rx;
    }
}
FinancialChart.$t = markType(FinancialChart, 'FinancialChart', XYChart.$);
FinancialChart.s8 = 0;
/**
 * @hidden
 */
export class FinancialChartToolbar extends TemplatedHtmlControl {
    constructor(a, b) {
        super("financialChartToolbar", b, "");
        this.o = null;
        this.r = null;
        this.z = null;
        this.p = null;
        this.w = null;
        this.aq = dateMaxValue();
        this.ap = dateMinValue();
        this.ao = dateMaxValue();
        this.an = dateMinValue();
        this.rangeChanged = null;
        this.chartTypeChanged = null;
        this.trendLineTypeChanged = null;
        this.volumeTypeChanged = null;
        this.indicatorTypesChanged = null;
        this.overlayTypesChanged = null;
        this.x = 0;
        this.ah = 0;
        this.ab = 0;
        this.ad = null;
        this.af = null;
        this.t = null;
        this.ak = false;
        this.o = a;
        this.w = new FinancialChartToolbarViewModel();
        this.s = new FinancialChartRangeSelector(a != null ? a.tk : null);
        this.aa = new FinancialChartTypePicker(a != null ? a.th : null);
        this.q = new FinancialChartIndicatorMenu(1, a != null ? a.ti : null);
    }
    bg(a) {
        let b = a.querySelectorAll(".financialChartToolbarArea");
        if (b != null && b.length > 0) {
            this.g = b[0];
            if (!this.al) {
                this.g.hide();
            }
        }
        let c = a.querySelectorAll(".financialChartIndicatorMenu");
        if (c != null && c.length > 0) {
            this.q.g = c[0];
        }
        let d = a.querySelectorAll(".financialChartRangeSelector");
        if (d != null && d.length > 0) {
            this.s.g = d[0];
        }
        let e = a.querySelectorAll(".financialChartTypePicker");
        if (e != null && e.length > 0) {
            this.aa.g = e[0];
        }
    }
    bl(a) {
        this.q.u(a);
        this.aa.u(a);
    }
    get s() {
        return this.r;
    }
    set s(a) {
        let b = this.s;
        if (a != b) {
            this.r = a;
            this.bm("RangeSelector", b, this.s);
        }
    }
    get aa() {
        return this.z;
    }
    set aa(a) {
        let b = this.aa;
        if (a != b) {
            this.z = a;
            this.bm("ChartTypePicker", b, this.aa);
        }
    }
    get q() {
        return this.p;
    }
    set q(a) {
        let b = this.q;
        if (a != b) {
            this.p = a;
            this.bm("IndicatorMenu", b, this.q);
        }
    }
    bm(a, b, c) {
        switch (a) {
            case "RangeStart":
                if (this.s != null) {
                    this.s.t = this.as;
                }
                break;
            case "RangeEnd":
                if (this.s != null) {
                    this.s.s = this.ar;
                }
                break;
            case "SelectionStart":
                if (this.s != null) {
                    this.s.v = this.au;
                }
                break;
            case "SelectionEnd":
                if (this.s != null) {
                    this.s.u = this.at;
                }
                break;
            case "RangeSelector":
                let d = b;
                if (d != null) {
                    d.rangeChanged = delegateRemove(d.rangeChanged, runOn(this, this.bj));
                }
                if (this.s != null) {
                    this.s.t = this.as;
                    this.s.s = this.ar;
                    this.s.o = this.u;
                    this.s.v = this.au;
                    this.s.u = this.at;
                    let e = this.s;
                    e.rangeChanged = delegateCombine(e.rangeChanged, runOn(this, this.bj));
                }
                break;
            case "ChartTypePicker":
                let f = b;
                if (f != null) {
                    f.selectionChanged = delegateRemove(f.selectionChanged, runOn(this, this.bi));
                }
                if (this.aa != null) {
                    this.aa.w = this.y;
                    let g = this.aa;
                    g.selectionChanged = delegateCombine(g.selectionChanged, runOn(this, this.bi));
                }
                break;
            case "ChartType":
                if (this.aa != null) {
                    this.aa.w = this.y;
                }
                break;
            case "RangeSelectorOptions":
                if (this.s != null) {
                    this.s.o = this.u;
                }
                break;
            case "IsToolbarVisible":
                if (this.g == null) {
                    break;
                }
                if (this.al) {
                    this.g.show();
                }
                else {
                    this.g.hide();
                }
                break;
            case "IndicatorMenu":
                let h = typeCast(FinancialChartIndicatorMenu.$, b);
                if (h != null) {
                    h.selectionChanged = delegateRemove(h.selectionChanged, runOn(this, this.bh));
                }
                if (this.q != null) {
                    let i = this.q;
                    i.selectionChanged = delegateCombine(i.selectionChanged, runOn(this, this.bh));
                }
                break;
            case "IndicatorTypes":
                if (this.indicatorTypesChanged != null) {
                    this.indicatorTypesChanged(this, EventArgs.empty);
                }
                break;
            case "OverlayTypes":
                if (this.overlayTypesChanged != null) {
                    this.overlayTypesChanged(this, EventArgs.empty);
                }
                break;
            case "TrendLineType":
                if (this.trendLineTypeChanged != null) {
                    this.trendLineTypeChanged(this, EventArgs.empty);
                }
                break;
            case "VolumeType":
                if (this.volumeTypeChanged != null) {
                    this.volumeTypeChanged(this, EventArgs.empty);
                }
                break;
        }
    }
    bh(a, b) {
        this.ag = this.q.z;
        this.ae = this.q.y;
        this.ai = this.q.aa;
        this.ac = this.q.x;
    }
    bi(a, b) {
        this.y = this.aa.w;
        if (this.chartTypeChanged != null) {
            this.chartTypeChanged(this, b);
        }
    }
    bj(a, b) {
        this.au = this.s.v;
        this.at = this.s.u;
        if (this.rangeChanged != null) {
            this.rangeChanged(this, b);
        }
    }
    v() {
        let a = new FinancialChartToolbarContext();
        a.container = this.d;
        a.vm = this.w;
        a.vm.financialChartIndicatorMenu = ((() => {
            let $ret = new FinancialChartIndicatorMenuContext();
            $ret.vm = this.q.p;
            return $ret;
        })());
        a.vm.financialChartTypePicker = ((() => {
            let $ret = new FinancialChartTypePickerContext();
            $ret.vm = this.aa.p;
            return $ret;
        })());
        a.vm.financialChartRangeSelector = ((() => {
            let $ret = new FinancialChartRangeSelectorContext();
            $ret.vm = this.s.q;
            return $ret;
        })());
        return a;
    }
    l(a) {
        super.l(a);
        if (a == null) {
            return;
        }
        this.bk();
    }
    bk() {
    }
    get au() {
        return this.aq;
    }
    set au(a) {
        let b = this.au;
        if (+a != +b) {
            this.aq = a;
            this.bm("SelectionStart", b, this.au);
        }
    }
    get at() {
        return this.ap;
    }
    set at(a) {
        let b = this.at;
        if (+a != +b) {
            this.ap = a;
            this.bm("SelectionEnd", b, this.at);
        }
    }
    get as() {
        return this.ao;
    }
    set as(a) {
        let b = this.as;
        if (+a != +b) {
            this.ao = a;
            this.bm("RangeStart", b, this.as);
        }
    }
    get ar() {
        return this.an;
    }
    set ar(a) {
        let b = this.ar;
        if (+a != +b) {
            this.an = a;
            this.bm("RangeEnd", b, this.ar);
        }
    }
    bn() {
        if (this.s != null) {
            this.s.ab();
        }
    }
    get y() {
        return this.x;
    }
    set y(a) {
        let b = this.y;
        if (a != b) {
            this.x = a;
            this.bm("ChartType", enumGetBox(FinancialChartType_$type, b), enumGetBox(FinancialChartType_$type, this.y));
        }
    }
    get ai() {
        return this.ah;
    }
    set ai(a) {
        let b = this.ai;
        if (a != b) {
            this.ah = a;
            this.bm("TrendLineType", enumGetBox(TrendLineType_$type, b), enumGetBox(TrendLineType_$type, this.ai));
        }
    }
    get ac() {
        return this.ab;
    }
    set ac(a) {
        let b = this.ac;
        if (a != b) {
            this.ab = a;
            this.bm("VolumeType", enumGetBox(FinancialChartVolumeType_$type, b), enumGetBox(FinancialChartVolumeType_$type, this.ac));
        }
    }
    get ae() {
        return this.ad;
    }
    set ae(a) {
        let b = this.ae;
        if (!FinancialChartToolbar.am(b, a)) {
            this.ad = a;
            this.bm("IndicatorTypes", b, this.ae);
        }
    }
    get ag() {
        return this.af;
    }
    set ag(a) {
        let b = this.ag;
        if (!FinancialChartToolbar.am(b, a)) {
            this.af = a;
            this.bm("OverlayTypes", b, this.ag);
        }
    }
    static am(a, b) {
        if (a == b) {
            return true;
        }
        if ((a == null && b != null) || (a != null && b == null)) {
            return false;
        }
        if (a.count != b.count) {
            return false;
        }
        for (let c of fromEn(a)) {
            if (!b.contains(c)) {
                return false;
            }
        }
        return true;
    }
    aj() {
        return ((() => {
            let $ret = new FinancialChartToolbarVisualData();
            $ret.rangeSelector = this.s == null ? null : this.s.r();
            return $ret;
        })());
    }
    get u() {
        return this.t;
    }
    set u(a) {
        let b = this.u;
        if (a != b) {
            this.t = a;
            this.bm("RangeSelectorOptions", b, this.u);
        }
    }
    get al() {
        return this.ak;
    }
    set al(a) {
        let b = this.al;
        if (a != b) {
            this.ak = a;
            this.bm("IsToolbarVisible", b, this.al);
        }
    }
}
FinancialChartToolbar.$t = markType(FinancialChartToolbar, 'FinancialChartToolbar', TemplatedHtmlControl.$);
/**
 * @hidden
 */
export class FinancialChartRangeSelector extends TemplatedHtmlControl {
    constructor(a) {
        super("financialChartRangeSelector", a, "");
        this.q = null;
        this.rangeChanged = null;
        this.p = null;
        this.p = new FinancialChartRangeSelectorView();
        this.q = new FinancialChartRangeSelectorViewModel();
        this.q.onRangeChanged = runOn(this, this.z);
    }
    get v() {
        return this.q.selectionStart;
    }
    set v(a) {
        this.q.selectionStart = a;
    }
    get u() {
        return this.q.selectionEnd;
    }
    set u(a) {
        this.q.selectionEnd = a;
    }
    get t() {
        return this.q.rangeStart;
    }
    set t(a) {
        this.q.rangeStart = a;
    }
    get s() {
        return this.q.rangeEnd;
    }
    set s(a) {
        this.q.rangeEnd = a;
    }
    ab() {
        if (this.q.onPropertiesChanged()) {
            this.p.e(this.q);
        }
    }
    l(a) {
        super.l(a);
        this.p.d(this.d, this.q);
    }
    z() {
        if (this.rangeChanged != null) {
            this.rangeChanged(this, EventArgs.empty);
        }
    }
    r() {
        return this.p.c(this.g);
    }
    get o() {
        return this.q.options;
    }
    set o(a) {
        let b = !FinancialChartToolbar.am(this.q.options, a);
        if (b) {
            this.q.options = a;
            this.aa();
        }
    }
    aa() {
        if (this.q != null) {
            this.q.af();
        }
        if (this.p != null) {
            this.p.g(this.o);
        }
    }
}
FinancialChartRangeSelector.$t = markType(FinancialChartRangeSelector, 'FinancialChartRangeSelector', TemplatedHtmlControl.$);
FinancialChartRangeSelector.x = 0;
