// igniteui-react-charts/ES2015/FinancialLegend_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { LegendBase } from "./LegendBase";
import { LegendBaseView } from "./LegendBaseView";
import { fromEn, delegateCombine, markType } from "igniteui-react-core/ES2015/type";
import { FinancialLegendViewManager } from "./FinancialLegendViewManager";
/**
 * @hidden
 */
export class FinancialLegend extends LegendBase {
    constructor() {
        super();
        this.ab = FinancialLegend.$;
        let a = this.at;
        a.collectionChanged = delegateCombine(a.collectionChanged, (a, b) => {
            if (b.oldItems != null) {
                for (let c of fromEn(b.oldItems)) {
                    this.a9.al(c);
                }
            }
            if (b.newItems != null) {
                for (let d of fromEn(b.newItems)) {
                    this.a9.x(d);
                }
            }
        });
    }
    aj() {
        return new FinancialLegendView(this);
    }
    get a9() {
        return this.ak;
    }
    addChildInOrder(a, b) {
        this.at.add(a);
    }
    get_isFinancial() {
        return true;
    }
}
FinancialLegend.$t = markType(FinancialLegend, 'FinancialLegend', LegendBase.$);
/**
 * @hidden
 */
export class FinancialLegendView extends LegendBaseView {
    constructor(a) {
        super(a);
        this.c = new FinancialLegendViewManager(this);
    }
}
FinancialLegendView.$t = markType(FinancialLegendView, 'FinancialLegendView', LegendBaseView.$);
