// igniteui-react-charts/ES2015/FinancialPriceSeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType, getInstanceType } from "igniteui-react-core/ES2015/type";
import { IFinancialPriceSeriesProxy_$type } from "./IFinancialPriceSeriesProxy";
import { stringEndsWith } from "igniteui-react-core/ES2015/string";
/**
 * @hidden
 */
export class FinancialPriceSeriesProxy extends Base {
    getCloseMemberPath(a) {
        return a.ob;
    }
    getHighMemberPath(a) {
        return a.oe;
    }
    getLowMemberPath(a) {
        return a.oj;
    }
    getOpenMemberPath(a) {
        return a.on;
    }
    getVolumeMemberPath(a) {
        return a.or;
    }
    matchesType(a) {
        if (!a.cr) {
            return false;
        }
        if (stringEndsWith(getInstanceType(a).typeName, "Indicator") || stringEndsWith(getInstanceType(a).typeName, "Overlay")) {
            return false;
        }
        return true;
    }
    setCategoryMode(a, b) {
        a.me = b;
    }
    setCloseMemberPath(a, b) {
        a.ob = b;
    }
    setDisplayType(a, b) {
        a.pj = b;
    }
    setHighMemberPath(a, b) {
        a.oe = b;
    }
    setLowMemberPath(a, b) {
        a.oj = b;
    }
    setOpenMemberPath(a, b) {
        a.on = b;
    }
    setVolumeMemberPath(a, b) {
        a.or = b;
    }
    setXAxis(a, b) {
        a.l7 = b;
    }
    setYAxis(a, b) {
        a.ms = b;
    }
}
FinancialPriceSeriesProxy.$t = markType(FinancialPriceSeriesProxy, 'FinancialPriceSeriesProxy', Base.$, [IFinancialPriceSeriesProxy_$type]);
