// igniteui-react-charts/ES2015/FullStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FullStochasticOscillatorIndicatorStrategy } from "./FullStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
import { Brush } from "igniteui-react-core/ES2015/Brush";
/**
 * @hidden
 */
export class FullStochasticOscillatorIndicator extends StrategyBasedIndicator {
    constructor() {
        super();
        this.trendLineBrush = ((() => {
            let $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        this.trendLineType = 10;
        this.trendLinePeriod = 3;
    }
    get_qb() {
        return new FullStochasticOscillatorIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return FullStochasticOscillatorIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(FullStochasticOscillatorIndicator.q1));
    }
    set qw(a) {
        this.h(FullStochasticOscillatorIndicator.q1, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
    get smoothingPeriod() {
        return typeGetValue(this.c(FullStochasticOscillatorIndicator.q2));
    }
    set smoothingPeriod(a) {
        this.h(FullStochasticOscillatorIndicator.q2, a);
    }
    qj() {
        return this.smoothingPeriod;
    }
    get qy() {
        return typeGetValue(this.c(FullStochasticOscillatorIndicator.q3));
    }
    set qy(a) {
        this.h(FullStochasticOscillatorIndicator.q3, a);
    }
    qg() {
        return this.qy;
    }
    pr() {
        return this.qy;
    }
}
FullStochasticOscillatorIndicator.$t = markType(FullStochasticOscillatorIndicator, 'FullStochasticOscillatorIndicator', StrategyBasedIndicator.$);
FullStochasticOscillatorIndicator.q1 = StrategyBasedIndicator.qt(14, FullStochasticOscillatorIndicator.$);
FullStochasticOscillatorIndicator.q2 = StrategyBasedIndicator.qu(3, FullStochasticOscillatorIndicator.$, "SmoothingPeriod");
FullStochasticOscillatorIndicator.q3 = StrategyBasedIndicator.qu(3, FullStochasticOscillatorIndicator.$, "TriggerPeriod");
