// igniteui-react-charts/ES2015/HighDensityScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Series } from "./Series";
import { NumericXAxis } from "./NumericXAxis";
import { DependencyProperty } from "igniteui-react-core/ES2015/DependencyProperty";
import { NumericYAxis } from "./NumericYAxis";
import { Color } from "igniteui-react-core/ES2015/Color";
import { ScatterFrame } from "./ScatterFrame";
import { KDTree2D } from "./KDTree2D";
import { Type, typeGetValue, runOn, delegateRemove, typeCast, delegateCombine, markType, Nullable$1 } from "igniteui-react-core/ES2015/type";
import { Rect } from "igniteui-react-core/ES2015/Rect";
import { IFastItemsSource_$type } from "igniteui-react-core/ES2015/IFastItemsSource";
import { SeriesView } from "./SeriesView";
import { List$1 } from "igniteui-react-core/ES2015/List$1";
import { KDPointData } from "./KDPointData";
import { ScalerParams } from "./ScalerParams";
import { ProgressiveLoadStatusEventArgs } from "./ProgressiveLoadStatusEventArgs";
import { SearchArgs } from "./SearchArgs";
import { KDTreeNode2D } from "./KDTreeNode2D";
import { AxisRange } from "./AxisRange";
import { PropertyMetadata } from "igniteui-react-core/ES2015/PropertyMetadata";
import { truncate, isNaN_, isInfinity, logBase, intDivide } from "igniteui-react-core/ES2015/number";
import { boxArray$1, arrayShallowClone } from "igniteui-react-core/ES2015/array";
import { dateNow } from "igniteui-react-core/ES2015/date";
import { RenderingContext } from "igniteui-react-core/ES2015/RenderingContext";
import { CanvasViewRenderer } from "igniteui-react-core/ES2015/CanvasViewRenderer";
import { Rectangle } from "igniteui-react-core/ES2015/Rectangle";
/**
 * @hidden
 */
export class HighDensityScatterSeries extends Series {
    constructor() {
        super();
        this.mc = null;
        this.me = null;
        this.nk = 0;
        this.mg = false;
        this.m0 = 0;
        this.mz = 0;
        this.my = 0;
        this.mx = 255;
        this.mw = 0;
        this.mv = 0;
        this.m3 = 255;
        this.m2 = 0;
        this.m1 = 0;
        this.l4 = null;
        this.l9 = null;
        this.ma = null;
        this.l5 = null;
        this.l7 = null;
        this.mi = false;
        this.nl = 0;
        this.nn = 0;
        this.nh = 0;
        this.ng = 0;
        this.progressiveLoadStatusChanged = null;
        this.mn = null;
        this.mt = 0;
        this.ms = 0;
        this.mu = 0;
        this.m6 = 0;
        this.m4 = 0;
        this.m5 = 0;
        this.m7 = 0;
        this.m9 = 0;
        this.nb = 0;
        this.na = 0;
        this.m8 = 0;
        this.mp = 0;
        this.mq = 0;
        this.mr = 0;
        this.mo = 0;
        this.nj = 0;
        this.ni = 0;
        this.l1 = null;
        this.nm = 0;
        this.mh = false;
        this._itemIndexes = null;
        this.l2 = null;
        this.lz = null;
        this.mb = null;
        this.ab = HighDensityScatterSeries.$;
        this.c7 = false;
    }
    get_c4() {
        return true;
    }
    get md() {
        return this.mc;
    }
    set md(a) {
        if (this.mc != a) {
            let b = this.md;
            this.mc = a;
            this.raisePropertyChanged("XColumn", b, this.md);
        }
    }
    fs(a, b) {
        let c = this.gu(b);
        if (c == HighDensityScatterSeries.$$p[8]) {
            return this.ft(a, b, this.n2);
        }
        if (c == HighDensityScatterSeries.$$p[10]) {
            return this.ft(a, b, this.n6);
        }
        return super.fs(a, b);
    }
    get mf() {
        return this.me;
    }
    set mf(a) {
        if (this.me != a) {
            let b = this.mf;
            this.me = a;
            this.raisePropertyChanged("YColumn", b, this.mf);
        }
    }
    ok() {
        if (this.l4 == null) {
            return;
        }
        this.l4.s();
        let a = this.l4;
        a.progressiveThunkCompleted = delegateRemove(a.progressiveThunkCompleted, runOn(this, this.n8));
        this.l4 = null;
    }
    iy(a, b, c, d) {
        super.iy(a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                if (this.bs != null) {
                    let e = this.getEffectiveViewport1(this.bi);
                    this.mh = !e.isEmpty;
                }
                break;
            case "FastItemsSource":
                this.ok();
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    c.deregisterColumn(this.md);
                    c.deregisterColumn(this.mf);
                    this.md = null;
                    this.mf = null;
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    this.md = this.bw(this.n2);
                    this.mf = this.bw(this.n6);
                }
                if ((this.l8 != null && !this.l8.ck()) || (this.l6 != null && !this.l6.ck())) {
                    this.jc(false);
                }
                this.ik();
                break;
            case HighDensityScatterSeries.$$p[7]:
                let f = this.l5;
                this.l5 = this.l6;
                this.ok();
                if (f != null) {
                    f.rangeChanged = delegateRemove(f.rangeChanged, runOn(this, this.ob));
                }
                if (this.l5 != null) {
                    let g = this.l5;
                    g.rangeChanged = delegateCombine(g.rangeChanged, runOn(this, this.ob));
                }
                if (c != null) {
                    c.bv(this);
                }
                if (d != null) {
                    d.ch(this);
                }
                if ((this.l6 != null && !this.l6.ck()) || (d == null && c != null)) {
                    this.jc(false);
                }
                break;
            case HighDensityScatterSeries.$$p[9]:
                let h = this.l7;
                this.l7 = this.l8;
                this.ok();
                if (h != null) {
                    h.rangeChanged = delegateRemove(h.rangeChanged, runOn(this, this.ob));
                }
                if (this.l7 != null) {
                    let i = this.l7;
                    i.rangeChanged = delegateCombine(i.rangeChanged, runOn(this, this.ob));
                }
                if (c != null) {
                    c.bv(this);
                }
                if (d != null) {
                    d.ch(this);
                }
                if ((this.l8 != null && !this.l8.ck()) || (d == null && c != null)) {
                    this.jc(false);
                }
                break;
            case "MouseOverEnabled":
                this.mi = this.c7;
                this.jc(false);
                break;
            case HighDensityScatterSeries.$$p[8]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.md);
                    this.md = this.bw(this.n2);
                }
                break;
            case "XColumn":
                this.ok();
                if (this.l6 != null && !this.l6.ck()) {
                    this.jc(false);
                }
                this.ik();
                break;
            case HighDensityScatterSeries.$$p[10]:
                if (this.bz != null) {
                    this.bz.deregisterColumn(this.mf);
                    this.mf = this.bw(this.n6);
                }
                break;
            case "YColumn":
                this.ok();
                if (this.l8 != null && !this.l8.ck()) {
                    this.jc(false);
                }
                this.ik();
                break;
            case HighDensityScatterSeries.$$p[6]:
                this.ok();
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[3]:
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[1]:
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[0]:
            case HighDensityScatterSeries.$$p[2]:
                if (!this.l3.by(this.oz, this.oy)) {
                    return;
                }
                this.m0 = this.oz.o;
                this.mz = this.oz.n;
                this.my = this.oz.m;
                this.mx = this.oy.o;
                this.mw = this.oy.n;
                this.mv = this.oy.m;
                this.m3 = this.mx - this.m0;
                this.m2 = this.mw - this.mz;
                this.m1 = this.mv - this.my;
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[4]:
                this.nk = Math.max(typeGetValue(d) - 1, 0);
                this.mg = this.nk > 0;
                this.jc(false);
                this.ij();
                break;
            case HighDensityScatterSeries.$$p[5]:
                this.jc(false);
                this.ij();
                break;
        }
    }
    mk(a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    }
    mj(a) {
        if (typeCast(NumericXAxis.$, a) !== null) {
            return true;
        }
        return false;
    }
    ob(a, b) {
        this.ok();
        this.jc(false);
    }
    dw(a, b, c) {
        let d = super.dw(a, b, c);
        if (this.l8 == null || this.l6 == null || this.l6.bf == null || this.l8.bf == null || this.mf == null || this.md == null || this.mf.count < 1 || this.md.count < 1 || this.mf.count != this.md.count || this.l6.j3 == this.l6.j2 || this.l8.j3 == this.l8.j2 || this.lx.isEmpty) {
            d = false;
        }
        return d;
    }
    get nr() {
        return this.nl;
    }
    set nr(a) {
        let b = this.nl;
        this.nl = a;
        this.raisePropertyChanged("ProgressiveStatus", b, this.nl);
    }
    jd(a) {
        super.jd(a);
        this.od(a, this.bi);
    }
    od(a, b) {
        if (this.cc(b)) {
            return;
        }
        if (this.l4 == null && !this.mm) {
            let c = new List$1(KDPointData.$, 2, this.md.count);
            let d = this.l6;
            let e = this.l8;
            let f = new Rect(0, 0, 0, 1, 1);
            let g = new Rect(0, 0, 0, 1, 1);
            let h = new ScalerParams(1, f, g, d.b8);
            let i = new ScalerParams(1, f, g, e.b8);
            let j;
            let k;
            for (let l = 0; l < this.md.count; l++) {
                j = d.dv(this.md.item(l), h);
                k = e.dv(this.mf.item(l), i);
                if (!isNaN_(j) && !isNaN_(k) && !isInfinity(j) && !isInfinity(k)) {
                    c.add(((() => {
                        let $ret = new KDPointData();
                        $ret.a = j;
                        $ret.b = k;
                        $ret.c = l;
                        return $ret;
                    })()));
                }
            }
            if (this.ml) {
                this.ng = 1;
                this.nh = truncate(logBase(c.count, 2)) + 3;
                this.nr = truncate(((this.ng / this.nh) * 100));
                if (this.progressiveLoadStatusChanged != null) {
                    this.progressiveLoadStatusChanged(this, new ProgressiveLoadStatusEventArgs(this.nr));
                }
                this.l4 = KDTree2D.b(c.toArray(), 1);
                let m = this.l4;
                m.progressiveThunkCompleted = delegateCombine(m.progressiveThunkCompleted, runOn(this, this.n8));
                if (!this.l4.f(this.e4)) {
                    let n = this.l4;
                    n.progressiveThunkCompleted = delegateRemove(n.progressiveThunkCompleted, runOn(this, this.n8));
                }
                this.ij();
            }
            else {
                let o = dateNow();
                this.l4 = new KDTree2D(0, c.toArray(), 1);
                let p = dateNow();
            }
        }
        if (this.ml && !this.mm) {
            this.oe(b);
        }
        else {
            this.of(b);
        }
    }
    hy(a, b, c, d) {
        super.hy(a, b, c, d);
        this.ok();
        let e = false;
        if (this.l6 != null && !this.l6.ck()) {
            e = true;
        }
        if (this.l8 != null && !this.l8.ck()) {
            e = true;
        }
        if (e) {
            this.jc(false);
        }
        this.ik();
    }
    oa(a) {
        let b = a;
        let c = this.nj * this.ni;
        if (this.mi) {
            if (this._itemIndexes == null || this._itemIndexes.length != c) {
                this._itemIndexes = new Array(c);
                for (let d = 0; d < c; d++) {
                    this._itemIndexes[d] = 0;
                }
            }
            else {
                for (let e = 0; e < c; e++) {
                    this._itemIndexes[e] = 0;
                }
            }
        }
    }
    of(a) {
        let b;
        let c;
        let d = null;
        if (a.n) {
            d = this._itemIndexes;
        }
        let e = a.getViewInfo(c, b);
        c = e.p0;
        b = e.p1;
        let f = a;
        let g = this.getEffectiveViewport1(a);
        this.l9 = new ScalerParams(0, b, c, this.l5.b8, g);
        this.ma = new ScalerParams(0, b, c, this.l7.b8, g);
        this.n9(a);
        this.oa(a);
        let h = this.nj * this.ni;
        if (this.l2 == null || this.l2.length != h) {
            this.l2 = new Array(h);
            this.lz = new Array(h);
        }
        let i = this.l2.length;
        let j = this.l2;
        let k = this.lz;
        for (let l = 0; l < i; l++) {
            j[l] = 0;
            k[l] = 0;
        }
        this.nn = truncate(Math.round(this.eu));
        this.l1 = f.bv();
        let m = this.l1.length;
        let n = this.l1;
        for (let o = 0; o < m; o++) {
            n[o] = 0;
        }
        let p = this.ne;
        let q = this.nd;
        if (isNaN_(p) || isInfinity(p)) {
            p = 0;
        }
        if (isNaN_(q) || isInfinity(q)) {
            q = 50;
        }
        this.mt = Math.min(p, q);
        this.ms = Math.max(p, q);
        if (this.mt < 0) {
            this.mt = 0;
        }
        if (this.ms < 0) {
            this.ms = 0;
        }
        this.mu = this.ms - this.mt;
        if (this.mm) {
            this.oc(a);
        }
        else {
            this.om(a);
        }
        f.b6();
        if (a.n) {
            this._itemIndexes = d;
            this.l3.b7();
        }
    }
    oe(a) {
        if (this.l4 == null) {
            return;
        }
        this.of(a);
    }
    n8(a, b) {
        this.l3.b5(() => {
            if (this.ng < this.nh - 1) {
                this.ng++;
            }
            this.nr = truncate(((this.ng / this.nh) * 100));
            if (this.progressiveLoadStatusChanged != null) {
                this.progressiveLoadStatusChanged(this, new ProgressiveLoadStatusEventArgs(this.nr));
            }
            this.oe(this.bi);
            this.ij();
            if (this.l4 != null) {
                if (!this.l4.f(this.e4)) {
                    let c = this.l4;
                    c.progressiveThunkCompleted = delegateRemove(c.progressiveThunkCompleted, runOn(this, this.n8));
                    this.nr = 100;
                    if (this.progressiveLoadStatusChanged != null) {
                        this.progressiveLoadStatusChanged(this, new ProgressiveLoadStatusEventArgs(this.nr));
                    }
                }
            }
        });
    }
    oc(a) {
        let b = arrayShallowClone(this.md.asArray());
        this.l6.gm(boxArray$1(b), 0, b.length, this.l9);
        let c = arrayShallowClone(this.mf.asArray());
        this.nm = 0;
        let d = this.nk;
        let e = this.mg;
        let f = this.l2.length;
        this.l8.gm(boxArray$1(c), 0, c.length, this.ma);
        let g = 255 << 24 | truncate(this.m0) << 16 | truncate(this.mz) << 8 | truncate(this.my);
        for (let h = 0; h < b.length; h++) {
            let i = truncate(b[h]);
            let j = truncate(c[h]);
            i = truncate(Math.floor(i));
            j = truncate(Math.floor(j));
            if (i < 0 || i >= this.nj || j < 0 || j >= this.ni) {
                continue;
            }
            if (e) {
                let k = i - d;
                let l = i + d;
                l = l > (this.nj - 1) ? (this.nj - 1) : l;
                k = k < 0 ? 0 : k;
                let m = j + d;
                let n = j - d;
                for (let o = k; o <= l; o++) {
                    for (let p = n; p <= m; p++) {
                        let q = (p * this.nj) + o;
                        if (q < 0 || q > f - 1) {
                            continue;
                        }
                        this.l2[q] = this.l2[q] + 1;
                        this.lz[q] = 1;
                        this.nm++;
                        if (this.mi) {
                            this._itemIndexes[q] = h + 1;
                        }
                    }
                }
            }
            else {
                let r = (j * this.nj) + i;
                this.l2[r] = this.l2[r] + 1;
                this.lz[r] = 1;
                this.nm++;
                if (this.mi) {
                    this._itemIndexes[r] = h + 1;
                }
            }
        }
        this.og();
    }
    om(a) {
        let b;
        let c;
        let d = a.getViewInfo(b, c);
        b = d.p0;
        c = d.p1;
        if (this.l4 == null) {
            return;
        }
        if (this.mn == null) {
            this.mn = new List$1(KDTreeNode2D.$, 2, truncate(Math.round(b.width * b.height)));
        }
        else {
            this.mn.clear();
        }
        let e = c.left;
        let f = c.right;
        let g = c.top;
        let h = c.bottom;
        this.na = c.top;
        this.m8 = c.height;
        this.m9 = c.left;
        this.nb = c.width;
        this.m6 = b.top;
        this.m4 = b.height;
        this.m5 = b.left;
        this.m7 = b.width;
        let i = this.getEffectiveViewport1(a);
        this.mp = (i.left - this.m5) / this.m7;
        this.mq = (i.top - this.m6) / this.m4;
        let j = (i.right - this.m5) / this.m7;
        this.mr = j - this.mp;
        let k = (i.bottom - this.m6) / this.m4;
        this.mo = k - this.mq;
        if (this.mh) {
            e = (e - this.mp) / this.mr;
            f = (f - this.mp) / this.mr;
            g = (g - this.mq) / this.mo;
            h = (h - this.mq) / this.mo;
        }
        let l = this.eu / b.width * c.width;
        let m = this.eu / b.height * c.height;
        let n = Math.min(l, m);
        let o = ((() => {
            let $ret = new SearchArgs();
            $ret.c = e;
            $ret.a = f;
            $ret.d = g;
            $ret.b = h;
            $ret.e = l;
            $ret.f = m;
            $ret.g = 0x7FFFFFFF;
            return $ret;
        })());
        this.l4.t(this.mn, o, 0, 1, 0, 1);
        let p;
        this.nm = 0;
        for (let q = 0; q < this.mn.count; q++) {
            p = this.mn._inner[q];
            this.oh(p);
        }
        this.og();
    }
    oh(a) {
        if (a.g) {
            return;
        }
        let b = a.e != null && a.e.a;
        let c = a.a == null ? 0 : a.a.length;
        let d = (a.h - c) + 1;
        if (b && d > 0) {
        }
        else {
            d = 1;
        }
        this.oj(a.b, d, b, a.e);
        if (c > 0 && !b) {
            let e;
            for (let f = 0; f < c; f++) {
                e = a.a[f];
                this.oj(e, d, false, a.e);
            }
        }
        if (a.e != null) {
            a.e.a = false;
        }
    }
    c9(a) {
        let b = this.bs != null ? this.bs.mq : Rect.empty;
        let c = this.bs != null ? this.bs.m3 : Rect.empty;
        let d = new Rect(0, 0, 0, 1, 1);
        let e = this.lr(this.bi);
        let f = new ScalerParams(0, d, d, this.l6.b8, e);
        let g = new ScalerParams(0, d, d, this.l8.b8, e);
        let h = !b.isEmpty && !c.isEmpty && this.bz != null ? this.bz.indexOf(a) : -1;
        let i = this.l6 != null && this.md != null && h < this.md.count ? this.l6.dv(this.md.item(h), f) : NaN;
        let j = this.l8 != null && this.mf != null && h < this.mf.count ? this.l8.dv(this.mf.item(h), g) : NaN;
        if (!isNaN_(i)) {
            if (i < b.left + 0.1 * b.width) {
                i = i + 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
            if (i > b.right - 0.1 * b.width) {
                i = i - 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
        }
        if (!isNaN_(j)) {
            if (j < b.top + 0.1 * b.height) {
                j = j + 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
            if (j > b.bottom - 0.1 * b.height) {
                j = j - 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
        }
        if (this.bn != null) {
            this.bn.ap(this.bs, b, true);
        }
        return h >= 0;
    }
    oj(a, b, c, d) {
        let e = this.np(b);
        let f = a.c;
        let g = this.nk;
        if (c) {
            let h = d.d;
            let i = d.b;
            let j = d.e;
            let k = d.c;
            if (this.mh) {
                h = this.mp + this.mr * h;
                i = this.mp + this.mr * i;
                j = this.mq + this.mo * j;
                k = this.mq + this.mo * k;
            }
            let l = truncate((this.m5 + this.m7 * (h - this.m9) / this.nb));
            let m = truncate((this.m5 + this.m7 * (i - this.m9) / this.nb));
            let n = truncate((this.m6 + this.m4 * (j - this.na) / this.m8));
            let o = truncate((this.m6 + this.m4 * (k - this.na) / this.m8));
            l = truncate(Math.floor(l));
            m = truncate(Math.floor(m));
            n = truncate(Math.floor(n));
            o = truncate(Math.floor(o));
            if (this.mg) {
                l -= g;
                m += g;
                n -= g;
                o += g;
            }
            let p = ((m - l) + 1) * ((o - n) + 1);
            let q = b / p;
            let r = q;
            for (let s = l; s <= m; s++) {
                for (let t = n; t <= o; t++) {
                    this.oi(f, s, t, q, r);
                }
            }
        }
        else {
            let u = a.a;
            let v = a.b;
            if (this.mh) {
                u = this.mp + this.mr * u;
                v = this.mq + this.mo * v;
            }
            let w = truncate((this.m5 + this.m7 * (u - this.m9) / this.nb));
            let x = truncate((this.m6 + this.m4 * (v - this.na) / this.m8));
            w = truncate(Math.floor(w));
            x = truncate(Math.floor(x));
            if (this.mg) {
                let y = w - g;
                let z = w + g;
                let aa = x + g;
                let ab = x - g;
                let ac = ((z - y) + 1) * ((aa - ab) + 1);
                let ad = b / ac;
                let ae = ad;
                for (let af = y; af <= z; af++) {
                    for (let ag = ab; ag <= aa; ag++) {
                        this.oi(f, af, ag, ad, ae);
                    }
                }
            }
            else {
                this.oi(f, w, x, b, 1);
            }
        }
    }
    no(a, b) {
        return truncate((Math.floor(255 * b))) << 24 | truncate(Math.floor(b * (this.m0 + a * this.m3))) << 16 | truncate(Math.floor(b * (this.mz + a * this.m2))) << 8 | truncate(Math.floor(b * (this.my + a * this.m1)));
    }
    np(a) {
        return 255 << 24 | truncate(Math.floor((this.m0 + a * this.m3))) << 16 | truncate(Math.floor((this.mz + a * this.m2))) << 8 | truncate(Math.floor((this.my + a * this.m1)));
    }
    nc(a) {
        return (a >> 24 & 255) / 255;
    }
    oi(a, b, c, d, e) {
        if (b < 0 || b >= this.nj || c < 0 || c >= this.ni) {
            return;
        }
        let f = (c * this.nj) + b;
        this.l2[f] = this.l2[f] + d;
        this.lz[f] = this.lz[f] + e;
        if (this.mi) {
            this._itemIndexes[f] = a + 1;
        }
        this.nm += truncate(d);
    }
    og() {
        let a = 0;
        let b = 0;
        let c = this.mu;
        let d = this.mt;
        let e = this.ms;
        let f = this.l2.length;
        let g = this.l2;
        let h = this.lz;
        let i = this.l1;
        let j = NaN;
        let k = NaN;
        let l = 0;
        let m = 0;
        for (let n = 0; n < f; n++) {
            a = g[n];
            if (a != 0) {
                b = h[n];
                if (a >= e) {
                    a = 1;
                }
                else if (a <= d) {
                    a = 0;
                }
                else {
                    a = (a - d) / (c);
                }
                if (b > 1) {
                    b = 1;
                }
                if (b < 0.2) {
                    b = 0.2;
                }
                if (j != a || k != b) {
                    l = this.no(a, b);
                    j = a;
                    k = b;
                }
                i[m] = l >> 16 & 255;
                i[m + 1] = l >> 8 & 255;
                i[m + 2] = l & 255;
                i[m + 3] = l >> 24 & 255;
            }
            m += 4;
        }
    }
    n9(a) {
        let b = a;
        b.b3();
    }
    bf() {
        return new HighDensityScatterSeriesView(this);
    }
    it(a) {
        super.it(a);
        this.l3 = a;
    }
    an(a) {
        if (a != null && a == this.l6 && this.md != null) {
            return new AxisRange(this.md.minimum, this.md.maximum);
        }
        if (a != null && a == this.l8 && this.mf != null) {
            return new AxisRange(this.mf.minimum, this.mf.maximum);
        }
        return null;
    }
    jr(a, b) {
        if (this.bs != null) {
            let c = this.getEffectiveViewport1(this.bi);
            this.mh = !c.isEmpty;
        }
        this.jc(false);
    }
    js(a, b) {
        this.jc(false);
    }
    fp(a) {
        if (!this.mi || this._itemIndexes == null || this.bs == null || this.bz == null || this._itemIndexes.length != (this.nj * this.ni)) {
            return null;
        }
        let b = this.bs.mq;
        let c = (a.x - b.left) / b.width;
        let d = (a.y - b.top) / b.height;
        let e = truncate(Math.round(this.lx.left + (this.lx.width * c)));
        let f = truncate(Math.round(this.lx.top + (this.lx.height * d)));
        let g = this.ns(e, f);
        if (g < 0 || g > this._itemIndexes.length - 1) {
            return null;
        }
        let h = this._itemIndexes[g] - 1;
        if (h < 0 || h > this.bz.count) {
            return null;
        }
        return this.bz.item(h);
    }
    ns(a, b) {
        let c = (this.nj * b) + a;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        let d = 1;
        c = (this.nj * (b - d)) + a;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b + d)) + a;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b)) + a - 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b)) + a + 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b - d)) + a - 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b + d)) + a + 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b - d)) + a + 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        c = (this.nj * (b + d)) + a - 1;
        if (c > 0 && c < this._itemIndexes.length && this._itemIndexes[c] > 0) {
            return c;
        }
        return 0;
    }
    ol(a, b, c) {
        this.l1 = a;
        this.nj = b;
        this.ni = c;
    }
    get_dq() {
        return true;
    }
    set_dq(a) {
        super.set_dq(a);
    }
    hv(a, b) {
        super.hv(a, b);
        this.l3.b4();
        if (this.bh != null) {
            this.bh.b4();
        }
        this.ij();
    }
    je(a, b) {
        super.je(a, b);
        if (!this.dl) {
            this.bi.az(b);
            return;
        }
        this.bi.az(b);
        if (this.cc(this.bh)) {
            return;
        }
        let c = typeCast(HighDensityScatterSeriesView.$, this.bh);
        let d = new ScatterFrame();
        this.od(false, c);
        this.dl = false;
    }
    jb(a, b, c, d, e) {
        super.jb(a, b, c, d, e);
        let f = this.a0.a$h.item(d);
        let g = f;
        f.ay(c);
        if (this.cc(g)) {
            return;
        }
        if (this.mb == null) {
            this.mb = new ScatterFrame();
        }
        this.od(false, f);
    }
    ht() {
        super.ht();
        this.l6 = null;
        this.l8 = null;
    }
}
HighDensityScatterSeries.$t = markType(HighDensityScatterSeries, 'HighDensityScatterSeries', Series.$);
Type.dep(DependencyProperty, PropertyMetadata, HighDensityScatterSeries, 'raisePropertyChanged', ['HeatMaximumColor:oy:on', [Color.$, null], 'HeatMaximum:nd:oo', [1, 50], 'HeatMinimumColor:oz:op', [Color.$, null], 'HeatMinimum:ne:oq', [1, 0], 'PointExtent:nq:or', [1, 1], 'ProgressiveLoad:ml:os', [0, true], 'UseBruteForce:mm:ot', [0, false], 'XAxis:l6:ou', [NumericXAxis.$, null], 'XMemberPath:n2:ov', [2, null], 'YAxis:l8:ow', [NumericYAxis.$, null], 'YMemberPath:n6:ox', [2, null]]);
/**
 * @hidden
 */
export class HighDensityScatterSeriesView extends SeriesView {
    constructor(a) {
        super(a);
        this.bw = null;
        this.bu = null;
        this.b2 = 0;
        this.b1 = 0;
        this.bz = null;
        this.b0 = null;
        this.bx = null;
        this.bw = a;
    }
    ap() {
        super.ap();
        if (!this.n) {
            this.e.eu = 4;
            this.bw.oz = Color.u(255, 0, 0, 0);
            this.bw.oy = Color.u(255, 255, 0, 0);
        }
    }
    b7() {
        this.bw.ol(this.bu, this.b2, this.b1);
    }
    b3() {
        if (this.bz == null) {
            this.bz = this.e.e4.createElement("canvas");
            let a = this.e.e4.get2DCanvasContext(this.bz);
            this.bx = new RenderingContext(new CanvasViewRenderer(), a);
        }
        let b;
        if (this.m) {
            b = this.br;
        }
        else {
            b = this.e.bs.mv();
        }
        let c = truncate(Math.round(b.left));
        let d = truncate(Math.round(b.top));
        let e = truncate(Math.round(b.width));
        let f = truncate(Math.round(b.height));
        let g = this.b2;
        let h = this.b1;
        this.b2 = e;
        this.b1 = f;
        if (g != this.b2 || h != this.b1 || this.bu == null) {
            this.bz.setAttribute("width", e.toString());
            this.bz.setAttribute("height", f.toString());
            this.b0 = this.bx.g().getImageData(0, 0, e, f);
            let imageDataArray_ = this.b0.data;
            this.bu = imageDataArray_;
        }
        this.bw.ol(this.bu, this.b2, this.b1);
    }
    bv() {
        return this.bu;
    }
    b6() {
        let a = truncate(Math.round(this.br.left));
        let b = truncate(Math.round(this.br.left));
        let c = truncate(Math.round(this.br.width));
        let d = truncate(Math.round(this.br.height));
        this.bx.g().putImageData(this.b0, 0, 0);
        this.an();
    }
    b5(a) {
        if (this.e.e4 != null) {
            this.e.e4.setTimeout(a, 0);
        }
        else {
            window.setTimeout(a, 0);
        }
    }
    a3(a, b) {
        super.a3(a, b);
        if (b) {
            if (!this.bw.c7) {
                return;
            }
            let c = new Rectangle();
            c.n = this.br.left;
            c.o = this.br.top;
            c.width = this.br.width;
            c.height = this.br.height;
            c._fill = this.bl();
            a.v(c);
            return;
        }
        if (this.bz == null) {
            return;
        }
        let d = truncate(Math.round(this.br.left));
        let e = truncate(Math.round(this.br.top));
        let f = truncate(Math.round(this.br.width));
        let g = truncate(Math.round(this.br.height));
        this.h.n(this.bz.getNativeElement(), this.e._opacity, d, e, f, g, d, e, f, g);
    }
    by(a, b) {
        return Color.e(Nullable$1.toNullable(Color.$, a), Nullable$1.toNullable(Color.$, null)) && Color.e(Nullable$1.toNullable(Color.$, b), Nullable$1.toNullable(Color.$, null));
    }
    b4() {
        if (this.bu != null) {
            for (let a = 0; a < this.bu.length; a++) {
                this.bu[a] = 0;
            }
            this.b6();
        }
    }
    ai(a) {
        super.ai(a);
        if (this.bu == null) {
            this.e.a2.exportSeriesPixelsData(a, null, this.b2, this.b1);
            return;
        }
        let b = new Array(intDivide(this.bu.length, 4));
        let c;
        for (let d = 0; d < intDivide(this.bu.length, 4); d++) {
            c = d * 4;
            b[d] = this.bu[c + 3] << 24 | this.bu[c] << 16 | this.bu[c + 1] << 8 | this.bu[c + 2];
        }
        this.e.a2.exportSeriesPixelsData(a, b, this.b2, this.b1);
    }
}
HighDensityScatterSeriesView.$t = markType(HighDensityScatterSeriesView, 'HighDensityScatterSeriesView', SeriesView.$);
