// igniteui-react-charts/ES2015/HorizontalRangeCategorySeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, typeCast, markType } from "igniteui-react-core/ES2015/type";
import { IHorizontalRangeCategorySeriesProxy_$type } from "./IHorizontalRangeCategorySeriesProxy";
import { HorizontalRangeCategorySeries } from "./HorizontalRangeCategorySeries";
/**
 * @hidden
 */
export class HorizontalRangeCategorySeriesProxy extends Base {
    matchesType(a) {
        return typeCast(HorizontalRangeCategorySeries.$, a) !== null;
    }
    setHighMemberPath(a, b) {
        a.o2 = b;
    }
    setLowMemberPath(a, b) {
        a.o5 = b;
    }
    setXAxis(a, b) {
        a.xAxis = b;
    }
    setYAxis(a, b) {
        a.yAxis = b;
    }
}
HorizontalRangeCategorySeriesProxy.$t = markType(HorizontalRangeCategorySeriesProxy, 'HorizontalRangeCategorySeriesProxy', Base.$, [IHorizontalRangeCategorySeriesProxy_$type]);
