// igniteui-react-charts/ES2015/igr-value-overlay.js
import { IgrAxis } from "./igr-axis";
import { IgrSeries } from "./igr-series";
import { ValueOverlay } from "./ValueOverlay";
import { ensureBool, brushToString, stringToBrush, toPoint, fromPoint } from "igniteui-react-core/ES2015/componentUtil";
/**
 * Represents the class of the value overlay. The value overlay is a line or circle representing a value on an axis.
*/
export class IgrValueOverlay extends IgrSeries {
    constructor(props) {
        super(props);
        this._axisName = null;
    }
    createImplementation() {
        return new ValueOverlay();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    /**
     * Gets or sets the axis used by the value overlay.
    */
    get axis() {
        if (this.i.lz == null) {
            return null;
        }
        if (!this.i.lz.externalObject) {
            let e = IgrAxis._createFromInternal(this.i.lz);
            if (e) {
                e._implementation = this.i.lz;
            }
            this.i.lz.externalObject = e;
        }
        return this.i.lz.externalObject;
    }
    set axis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.lz = null : this.i.lz = v.i;
    }
    /**
    * Gets or sets the name to use to resolve axis from markup.
    */
    get axisName() {
        return this._axisName;
    }
    set axisName(v) {
        this._axisName = v;
    }
    /**
     * Gets or sets the value of the overlay.
    */
    get value() {
        return this.i.mh;
    }
    set value(v) {
        this.i.mh = +v;
    }
    /**
     * Gets or sets whether to draw annotations over the axes where the crosshair meets with them.
    */
    get isAxisAnnotationEnabled() {
        return this.i.l3;
    }
    set isAxisAnnotationEnabled(v) {
        this.i.l3 = ensureBool(v);
    }
    /**
     * Sets or gets a function which takes an object that produces a formatted label for the axis annotation.
    */
    get axisAnnotationFormatLabel() {
        return this.i.l1;
    }
    set axisAnnotationFormatLabel(v) {
        this.i.l1 = v;
    }
    /**
     * Gets or sets the color to use for the axis annotation text. Leave unset for an automatic value.
    */
    get axisAnnotationTextColor() {
        return brushToString(this.i.nf);
    }
    set axisAnnotationTextColor(v) {
        this.i.nf = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for the axis annotation backing. Leave unset for an automatic value.
    */
    get axisAnnotationBackground() {
        return brushToString(this.i.nd);
    }
    set axisAnnotationBackground(v) {
        this.i.nd = stringToBrush(v);
    }
    /**
     * Gets or sets the precision to use displaying values for interpolated positions.
    */
    get axisAnnotationInterpolatedValuePrecision() {
        return this.i.mi;
    }
    set axisAnnotationInterpolatedValuePrecision(v) {
        this.i.mi = +v;
    }
    /**
     * Gets or sets the color to use for the axis annotation outline. Leave unset for an automatic value.
    */
    get axisAnnotationOutline() {
        return brushToString(this.i.ne);
    }
    set axisAnnotationOutline(v) {
        this.i.ne = stringToBrush(v);
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get axisAnnotationPaddingLeft() {
        return this.i.l8;
    }
    set axisAnnotationPaddingLeft(v) {
        this.i.l8 = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get axisAnnotationPaddingTop() {
        return this.i.ma;
    }
    set axisAnnotationPaddingTop(v) {
        this.i.ma = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get axisAnnotationPaddingRight() {
        return this.i.l9;
    }
    set axisAnnotationPaddingRight(v) {
        this.i.l9 = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get axisAnnotationPaddingBottom() {
        return this.i.l7;
    }
    set axisAnnotationPaddingBottom(v) {
        this.i.l7 = +v;
    }
    /**
     * Gets or sets the stroke thickness for the axis annotation backing. Leave unset for an automatic value.
    */
    get axisAnnotationStrokeThickness() {
        return this.i.mb;
    }
    set axisAnnotationStrokeThickness(v) {
        this.i.mb = +v;
    }
    /**
     * Gets whether the series is a value overlay.
    */
    get isValueOverlay() {
        return this.i.c6;
    }
    get labelResolved() {
        return this.i.mw;
    }
    bindAxes(axes) {
        super.bindAxes(axes);
        for (let i = 0; i < axes.length; i++) {
            if (this.axisName && this.axisName.length > 0 &&
                axes[i].name == this.axisName) {
                this.axis = axes[i];
            }
        }
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.axis && this.axis._styling) {
            this.axis._styling(container, component, this);
        }
        this._inStyling = false;
    }
    getSeriesValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.en(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.lk(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    /**
     * Gets the item item index associated with the specified world position
    
    * @param world
    */
    getItemIndex(world) {
        let iv = this.i.e8(toPoint(world));
        return (iv);
    }
    /**
     * Gets the item that is the best match for the specified world coordinates.
    
    * @param world  * The world coordinates to use.
    */
    getItem(world) {
        let iv = this.i.fp(toPoint(world));
        return (iv);
    }
    /**
     * Requests that the provided item should be brought into view if possible.
    
    * @param item  * The item to attempt to bring into view.
    */
    scrollIntoView(item) {
        let iv = this.i.c9(item);
        return (iv);
    }
    /**
     * Gets the label for a data item.
    
    * @param value  * The unscaled value to get a label for.
    */
    getLabel(value) {
        let iv = this.i.mu(value);
        return (iv);
    }
}
