// igniteui-react-charts/ES2015/igr-stacked-column-series.js
import { IgrHorizontalStackedSeriesBase } from "./igr-horizontal-stacked-series-base";
import { StackedColumnSeries } from "./StackedColumnSeries";
/**
 * Represents a IgxDataChartComponent stacked column series.
*/
export class IgrStackedColumnSeries extends IgrHorizontalStackedSeriesBase {
    constructor(props) {
        super(props);
    }
    createImplementation() {
        return new StackedColumnSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    /**
     * Gets or sets the x-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusX() {
        return this.i.p2;
    }
    set radiusX(v) {
        this.i.p2 = +v;
    }
    /**
     * Gets or sets the y-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusY() {
        return this.i.p3;
    }
    set radiusY(v) {
        this.i.p3 = +v;
    }
    /**
     * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
    
    */
    getItemSpan() {
        let iv = this.i.eg();
        return (iv);
    }
}
