// igniteui-react-charts/ES2015/igr-assigning-category-style-event-args.js
import { AssigningCategoryStyleEventArgsBase } from "./igr-assigning-category-style-event-args-base";
import { AssigningCategoryStyleEventArgs as AssigningCategoryStyleEventArgs_internal } from "./AssigningCategoryStyleEventArgs";
/**
 * Represents event arguments class for the AssigningCategoryStyleEvent
*
* The `AssigningCategoryStyle` event fired to allow you to override the style of items in a category or financial series
*
* ```ts
* <IgrDataChart
*     dataSource={this.state.dataSource}
*     width="700px"
*     height="500px">
*
*     <IgrCategoryXAxis name="xAxis" label="Year" />
*     <IgrNumericYAxis  name="yAxis" />
*
*     <IgrRangeColumnSeries
*         name="series1"
*         xAxisName="xAxis"
*         yAxisName="yAxis"
*         highMemberPath="High"
*         lowMemberPath="Low" />
*  </IgrDataChart>
* ```
*
* ```ts
* series.assigningCategoryStyle= this.chart_AssigningCategoryStyle;
* chart_AssigningCategoryStyle(sender :any,args: AssigningCategoryStyleEventArgs )
*   {
*
*
*   }
* ```
*/
export class AssigningCategoryStyleEventArgs extends AssigningCategoryStyleEventArgsBase {
    createImplementation() {
        return new AssigningCategoryStyleEventArgs_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Gets or sets stroke thickness to use for the current item.
    */
    get strokeThickness() {
        return this.i.r;
    }
    set strokeThickness(v) {
        this.i.r = +v;
    }
    /**
     * Gets or sets corner radius X to use for the current item, if applicable.
    *
    * Use the `radiusX` property to round the corners of the column.
    */
    get radiusX() {
        return this.i.p;
    }
    set radiusX(v) {
        this.i.p = +v;
    }
    /**
     * Gets or sets corner radius Y to use for the current item, if applicable.
    *
    * Use the `radiusY` property for the y-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusY() {
        return this.i.q;
    }
    set radiusY(v) {
        this.i.q = +v;
    }
}
