// igniteui-react-charts/ES2015/igr-moving-average-convergence-divergence-indicator.js
import { IgrStrategyBasedIndicator } from "./igr-strategy-based-indicator";
import { MovingAverageConvergenceDivergenceIndicator } from "./MovingAverageConvergenceDivergenceIndicator";
/**
 * Represents a IgxDataChartComponent Moving Average Convergence/Divergence (MACD) indicator series.
 * Default required members: High, Low, Close
*
* Represents a Ignite UIDataChart Moving Average Convergence/Divergence (MACD) indicator series.
*
* Example:
*
* ```ts
*  <IgrDataChart
*     dataSource={this.state.dataSource} >
*
*     <IgrCategoryXAxis name="xAxis" />
*     <IgrNumericYAxis name="yAxis" />
*
*     <IgrMovingAverageConvergenceDivergenceIndicator
* 	name="series1"
* 	openMemberPath="open"
* 	volumeMemberPath="Volume"
* 	highMemberPath="High"
* 	lowMemberPath="Low"
* 	closeMemberPath="close"/>
*  </IgrDataChart>
* ```
*/
export class IgrMovingAverageConvergenceDivergenceIndicator extends IgrStrategyBasedIndicator {
    createImplementation() {
        return new MovingAverageConvergenceDivergenceIndicator();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets the short moving average period for the current MovingAverageConvergenceDivergenceIndicator object.
     * The typical, and initial, value for short MACD periods is 10.
    *
    * Gets or sets the short moving average period for the current MovingAverageConvergenceDivergenceIndicator object.
    *
    * Example:
    *
    * ```ts
    * this.shortPeriod="30"
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrCategoryXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrMovingAverageConvergenceDivergenceIndicator
    * 	name="series1"
    * 	openMemberPath="open"
    * 	volumeMemberPath="Volume"
    * 	highMemberPath="High"
    * 	lowMemberPath="Low"
    * 	closeMemberPath="close"
    * 	shortPeriod="30"/>
    *  </IgrDataChart>
    * ```
    */
    get shortPeriod() {
        return this.i.shortPeriod;
    }
    set shortPeriod(v) {
        this.i.shortPeriod = +v;
    }
    /**
     * Gets or sets the long moving average period for the current MovingAverageConvergenceDivergenceIndicator object.
     * The typical, and initial, value for long MACD periods is 30.
    *
    * Gets or sets the long moving average period for the current MovingAverageConvergenceDivergenceIndicator object.
    *
    * Example:
    *
    * ```ts
    * this.longPeriod="30"
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrCategoryXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrMovingAverageConvergenceDivergenceIndicator
    * 	name="series1"
    * 	openMemberPath="open"
    * 	volumeMemberPath="Volume"
    * 	highMemberPath="High"
    * 	lowMemberPath="Low"
    * 	closeMemberPath="close"
    * 	longPeriod="30"/>
    *  </IgrDataChart>
    * ```
    */
    get longPeriod() {
        return this.i.longPeriod;
    }
    set longPeriod(v) {
        this.i.longPeriod = +v;
    }
    /**
     * Gets or sets the long moving average period for the current MovingAverageConvergenceDivergenceIndicator object.
     * The typical, and initial, value for long PVO periods is 30.
    *
    * Gets or sets the long moving average period for the current MovingAverageConvergenceDivergenceIndicator object.
    *
    * Example:
    *
    * ```ts
    * this.signalPeriod="30"
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrCategoryXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrMovingAverageConvergenceDivergenceIndicator
    * 	name="series1"
    * 	openMemberPath="open"
    * 	volumeMemberPath="Volume"
    * 	highMemberPath="High"
    * 	lowMemberPath="Low"
    * 	closeMemberPath="close"
    * 	signalPeriod="30"/>
    *  </IgrDataChart>
    * ```
    */
    get signalPeriod() {
        return this.i.signalPeriod;
    }
    set signalPeriod(v) {
        this.i.signalPeriod = +v;
    }
}
