// igniteui-react-charts/ES2015/igr-anchored-radial-series.js
import { TrendLineType_$type } from "igniteui-react-core/ES2015/TrendLineType";
import { IgrRadialBase } from "./igr-radial-base";
import { ensureEnum, brushToString, stringToBrush, toPoint, fromPoint } from "igniteui-react-core/ES2015/componentUtil";
/**
 * Represents the base class for all IgxDataChartComponent anchored radial category series.
*/
export class IgrAnchoredRadialSeries extends IgrRadialBase {
    constructor(props) {
        super(props);
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    /**
     * Gets or sets the item path that provides the values for the current series.
    */
    get valueMemberPath() {
        return this.i.n7;
    }
    set valueMemberPath(v) {
        this.i.n7 = v;
    }
    /**
     * Gets or sets the trend type for the current series object.
    */
    get trendLineType() {
        return this.i.nt;
    }
    set trendLineType(v) {
        this.i.nt = ensureEnum(TrendLineType_$type, v);
    }
    /**
     * Gets or sets the brush that specifies how to the current series
     * object's Trend line is drawn.
    */
    get trendLineBrush() {
        return brushToString(this.i.ol);
    }
    set trendLineBrush(v) {
        this.i.ol = stringToBrush(v);
    }
    /**
     * Gets the effective TrendLineBrush for this series.
    */
    get actualTrendLineBrush() {
        return brushToString(this.i.ok);
    }
    set actualTrendLineBrush(v) {
        this.i.ok = stringToBrush(v);
    }
    /**
     * Gets or sets the thickness of the current series object's trend line.
    */
    get trendLineThickness() {
        return this.i.n3;
    }
    set trendLineThickness(v) {
        this.i.n3 = +v;
    }
    /**
     * Gets or sets the trend line period for the current series.
     * The typical, and initial, value for bollinger band periods is 20.
    */
    get trendLinePeriod() {
        return this.i.n4;
    }
    set trendLinePeriod(v) {
        this.i.n4 = +v;
    }
    /**
     * Sets or Gets the Z index of the trendline.
    */
    get trendLineZIndex() {
        return this.i.n5;
    }
    set trendLineZIndex(v) {
        this.i.n5 = +v;
    }
    getItemValue(item, memberPathName) {
        let iv = this.i.fs(item, memberPathName);
        return (iv);
    }
    getPreviousOrExactIndex(world, skipUnknowns) {
        let iv = this.i.fe(toPoint(world), skipUnknowns);
        return (iv);
    }
    getNextOrExactIndex(world, skipUnknowns) {
        let iv = this.i.fc(toPoint(world), skipUnknowns);
        return (iv);
    }
    getSeriesValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.en(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.lk(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    /**
     * Returns the offset value for this series if grouped on a category axis.
    
    */
    getOffsetValue() {
        let iv = this.i.getOffsetValue();
        return (iv);
    }
    /**
     * Returns the width of the category grouping this series is in.
    
    */
    getCategoryWidth() {
        let iv = this.i.getCategoryWidth();
        return (iv);
    }
    /**
     * Scrolls the specified item into the view.
    
    * @param item  * The item to scroll into view.
    */
    scrollIntoView(item) {
        let iv = this.i.c9(item);
        return (iv);
    }
}
