// igniteui-react-charts/ES2015/igr-financial-series.js
import { delegateCombine, delegateRemove } from "igniteui-react-core/ES2015/type";
import { IgrCategoryAxisBase } from "./igr-category-axis-base";
import { IgrNumericYAxis } from "./igr-numeric-y-axis";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { AssigningCategoryStyleEventArgs } from "./igr-assigning-category-style-event-args";
import { FinancialEventArgs } from "./igr-financial-event-args";
import { IgrSeries } from "./igr-series";
import { ensureBool, brushToString, stringToBrush, ensureEnum, toPoint, fromPoint } from "igniteui-react-core/ES2015/componentUtil";
/**
 * Represents the base class for all financial indicator and overlay series.
*
* The `IgrFinancialSeries` represents the base class for all financial indicator and overlay series.
*
* ```ts
* <IgrDataChart
*     dataSource={this.state.dataSource}
*     width="700px"
*     height="500px">
*
*
*     <IgrCategoryXAxis name="xAxis" label="Date" />
*     <IgrNumericYAxis  name="yAxis" />
*
*
*     <IgrFinancialPriceSeries
*     name="series2"
*     xAxisName="xAxis"
*     yAxisName="yAxis"
*     displayType="Candlestick"
*     lowMemberPath="Low"
*     highMemberPath="High"
*     openMemberPath="Open"
*     closeMemberPath="Close"
*     volumeMemberPath="Volume" />
*
*  </IgrDataChart>
* ```
*/
export class IgrFinancialSeries extends IgrSeries {
    constructor(props) {
        super(props);
        this._xAxisName = null;
        this._yAxisName = null;
        this._assigningCategoryStyle = null;
        this._assigningCategoryStyle_wrapped = null;
        this._typical = null;
        this._typical_wrapped = null;
        this._typicalBasedOn = null;
        this._typicalBasedOn_wrapped = null;
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    /**
     * Gets whether the current series is a financial type series.
    */
    get isFinancial() {
        return this.i.cr;
    }
    /**
     * Gets or sets the brush to use for negative portions of the series.
    *
    * The `negativeBrush` property is used to brush the negative portions of the series.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume"
    * 	negativeBrush="Red" />
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * series.negativeBrush="red";
    * ```
    */
    get negativeBrush() {
        return brushToString(this.i.pd);
    }
    set negativeBrush(v) {
        this.i.pd = stringToBrush(v);
    }
    /**
     * Gets or sets the effective x-axis for the current FinancialSeries object.
    *
    * Use `xAxis` property to set effective x-axis for the current FinancialSeries object.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume" />
    *  </IgrDataChart>
    * ```
    */
    get xAxis() {
        if (this.i.l7 == null) {
            return null;
        }
        if (!this.i.l7.externalObject) {
            let e = IgrCategoryAxisBase._createFromInternal(this.i.l7);
            if (e) {
                e._implementation = this.i.l7;
            }
            this.i.l7.externalObject = e;
        }
        return this.i.l7.externalObject;
    }
    set xAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.l7 = null : this.i.l7 = v.i;
    }
    /**
    * Gets or sets the name to use to resolve xAxis from markup.
    */
    get xAxisName() {
        return this._xAxisName;
    }
    set xAxisName(v) {
        this._xAxisName = v;
    }
    /**
     * Gets or sets the effective y-axis for the current FinancialSeries object.
    *
    * Use `yAxis` property to set the effective y-axis for the current FinancialSeries object.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume" />
    *
    *  </IgrDataChart>
    * ```
    */
    get yAxis() {
        if (this.i.ms == null) {
            return null;
        }
        if (!this.i.ms.externalObject) {
            let e = IgrNumericYAxis._createFromInternal(this.i.ms);
            if (e) {
                e._implementation = this.i.ms;
            }
            this.i.ms.externalObject = e;
        }
        return this.i.ms.externalObject;
    }
    set yAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.ms = null : this.i.ms = v.i;
    }
    /**
    * Gets or sets the name to use to resolve yAxis from markup.
    */
    get yAxisName() {
        return this._yAxisName;
    }
    set yAxisName(v) {
        this._yAxisName = v;
    }
    /**
     * Gets or sets the open mapping property for the current series object.
    *
    * The `openMemberPath` property is used for open mapping of the current series object.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume" />
    *
    *  </IgrDataChart>
    * ```
    */
    get openMemberPath() {
        return this.i.on;
    }
    set openMemberPath(v) {
        this.i.on = v;
    }
    /**
     * Gets or sets the high mapping property for the current series object.
    *
    * Use `highMemberPath` propert for high mapping of the current series object.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume" />
    *
    *  </IgrDataChart>
    * ```
    */
    get highMemberPath() {
        return this.i.oe;
    }
    set highMemberPath(v) {
        this.i.oe = v;
    }
    /**
     * Gets or sets the low mapping property for the current series object.
    *
    * The `lowMemberPath` is used for low mapping property of the current series object.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume" />
    *
    *  </IgrDataChart>
    * ```
    */
    get lowMemberPath() {
        return this.i.oj;
    }
    set lowMemberPath(v) {
        this.i.oj = v;
    }
    /**
     * Gets or sets the close mapping property for the current series object.
    *
    * Use `closeMemberPath` property for the close mapping of the current series object.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume" />
    *
    *  </IgrDataChart>
    * ```
    */
    get closeMemberPath() {
        return this.i.ob;
    }
    set closeMemberPath(v) {
        this.i.ob = v;
    }
    /**
     * Gets or sets the volume mapping property for the current series object.
    *
    * The `volumeMemberPath` property is used for the volume mapping property of the current series object.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume" />
    *  </IgrDataChart>
    * ```
    */
    get volumeMemberPath() {
        return this.i.or;
    }
    set volumeMemberPath(v) {
        this.i.or = v;
    }
    /**
     * Gets or sets whether this category series should allow custom style overrides of its individual visuals.
    *
    * The `isCustomCategoryStyleAllowed` property is used to check if the category series should allow custom style.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume"
    * 	isCustomCategoryStyleAllowed="True" />
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * series.isCustomCategoryStyleAllowed= true;
    * ```
    */
    get isCustomCategoryStyleAllowed() {
        return this.i.nd;
    }
    set isCustomCategoryStyleAllowed(v) {
        this.i.nd = ensureBool(v);
    }
    /**
     * Gets or sets the method by which to animate the data into the chart when the chart data source is swapped.
     * Note: Transitions are not currently supported for stacked series.
    *
    * The `transitionInMode` property is used to gets or sets the method by which to animate the data into the chart when the chart data source is swapped.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume"
    * 	transitionInMode="Auto"/>
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * series.transitionInMode =CategoryTransitionInMode.Auto;
    * ```
    */
    get transitionInMode() {
        return this.i.mh;
    }
    set transitionInMode(v) {
        this.i.mh = ensureEnum(CategoryTransitionInMode_$type, v);
    }
    /**
     * Gets or sets whether the series should transition into the plot area when a new data source is assigned.
     * Note: Transitions are not currently supported for stacked series.
    *
    * The `isTransitionInEnabled` property is used to check if the series should transition into the plot area when a new data source is assigned.
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *
    *     <IgrFinancialPriceSeries
    *     name="series2"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Candlestick"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume"
    * 	isTransitionInEnabled="true" />
    *
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * series.isTransitionInEnabled= true;
    * ```
    */
    get isTransitionInEnabled() {
        return this.i.ne;
    }
    set isTransitionInEnabled(v) {
        this.i.ne = ensureBool(v);
    }
    /**
     * Overridden by derived series classes to indicate when negative colors are supported or not.
    */
    get isNegativeColorSupported() {
        return this.i.cz;
    }
    bindAxes(axes) {
        super.bindAxes(axes);
        for (let i = 0; i < axes.length; i++) {
            if (this.xAxisName && this.xAxisName.length > 0 &&
                axes[i].name == this.xAxisName) {
                this.xAxis = axes[i];
            }
        }
        for (let i = 0; i < axes.length; i++) {
            if (this.yAxisName && this.yAxisName.length > 0 &&
                axes[i].name == this.yAxisName) {
                this.yAxis = axes[i];
            }
        }
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.xAxis && this.xAxis._styling) {
            this.xAxis._styling(container, component, this);
        }
        if (this.yAxis && this.yAxis._styling) {
            this.yAxis._styling(container, component, this);
        }
        this._inStyling = false;
    }
    getItemValue(item, memberPathName) {
        let iv = this.i.fs(item, memberPathName);
        return (iv);
    }
    getPreviousOrExactIndex(world, skipUnknowns) {
        let iv = this.i.fe(toPoint(world), skipUnknowns);
        return (iv);
    }
    getNextOrExactIndex(world, skipUnknowns) {
        let iv = this.i.fc(toPoint(world), skipUnknowns);
        return (iv);
    }
    /**
     * Returns the offset value for this series if grouped on a category axis.
    
    *
    * The `getOffsetValue` method returns the offset value for this series if grouped on a category axis.
    *
    * ```ts
    * number offset= series.getOffsetValue();
    * ```
    */
    getOffsetValue() {
        let iv = this.i.getOffsetValue();
        return (iv);
    }
    /**
     * Returns the width of the category grouping this series is in.
    
    *
    * The `getCategoryWidth` method returns the width of the category grouping this series is in.
    *
    * ```ts
    * number width = series.getCategoryWidth();
    * ```
    */
    getCategoryWidth() {
        let iv = this.i.getCategoryWidth();
        return (iv);
    }
    getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.lk(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.en(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesHighValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.ej(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesLowValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.el(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesCloseValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.nz(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesOpenValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.n0(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesVolumeValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.n1(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesHighValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.lg(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesLowValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.li(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesOpenValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.pf(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesCloseValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.pe(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesVolumeValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.pg(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    /**
     * Determine if object can be used as YAxis
    
    * @param axis  * The object to check
    *
    * Use `canUseAsYAxis` method to determine if the object can be used as YAxis.
    *
    * ```ts
    *   boolean canY =series.canUseAsYAxis(yAxis);
    * ```
    */
    canUseAsYAxis(axis) {
        let iv = this.i.m9(axis);
        return (iv);
    }
    /**
     * Determine if object can be used as XAxis
    
    * @param axis  * The object to check
    *
    * Use `canUseAsXAxis` method to determine if the object can be used as XAxis.
    *
    * ```ts
    *   boolean canX =series.canUseAsXAxis(xAxis);
    * ```
    */
    canUseAsXAxis(axis) {
        let iv = this.i.m8(axis);
        return (iv);
    }
    /**
     * Gets the precise item index, if possible, based on the closeness to the previous or next whole integer. If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
    
    * @param world  * The world position for which to return the index.
    *
    * The `getExactItemIndex` method gets the precise item index, if possible, based on the closeness to the previous or next whole integer.
    *
    * If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
    *
    * ```ts
    * number index = series.getExactItemIndex(world);
    * ```
    */
    getExactItemIndex(world) {
        let iv = this.i.eb(toPoint(world));
        return (iv);
    }
    /**
     * Gets the index of the item that resides at the provided world coordinates.
    
    * @param world  * The world coordinates of the requested item.
    *
    * The `getItemIndex` method gets the index of the item that resides at the provided world coordinates.
    *
    * ```ts
    * number itemindex= series.getItemIndex(world);
    * ```
    */
    getItemIndex(world) {
        let iv = this.i.e8(toPoint(world));
        return (iv);
    }
    /**
     * Gets the item that is the best match for the specified world coordinates.
    
    * @param world  * The world coordinates to use.
    *
    * The `getItem` method gets the item that is the best match for the specified world coordinates.
    */
    getItem(world) {
        let iv = this.i.fp(toPoint(world));
        return (iv);
    }
    setNegativeColors(negativeBrush, negativeOutline) {
        this.i.jh(stringToBrush(negativeBrush), stringToBrush(negativeOutline));
    }
    /**
     * Event raised when Assigning Category Style
    *
    * The `AssigningCategoryStyle` event raised when assigning Category Style.
    *
    * ```ts
    * series.assigningCategoryStyle= this.chart_AssigningCategoryStyle;
    * chart_AssigningCategoryStyle(sender :any,args: AssigningCategoryStyleEventArgs )
    *   {
    *
    *
    *   }
    * ```
    */
    get assigningCategoryStyle() {
        return this._assigningCategoryStyle;
    }
    set assigningCategoryStyle(ev) {
        if (this._assigningCategoryStyle_wrapped !== null) {
            this.i.assigningCategoryStyle = delegateRemove(this.i.assigningCategoryStyle, this._assigningCategoryStyle_wrapped);
            this._assigningCategoryStyle_wrapped = null;
            this._assigningCategoryStyle = null;
        }
        this._assigningCategoryStyle = ev;
        this._assigningCategoryStyle_wrapped = (o, e) => {
            let outerArgs = new AssigningCategoryStyleEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeAssigningCategoryStyle) {
                this.beforeAssigningCategoryStyle(this, outerArgs);
            }
            if (this._assigningCategoryStyle) {
                this._assigningCategoryStyle(this, outerArgs);
            }
        };
        this.i.assigningCategoryStyle = delegateCombine(this.i.assigningCategoryStyle, this._assigningCategoryStyle_wrapped);
    }
    /**
     * Handle this event in order to perform a custom typical price calculation.
    *
    * Use `Typical` event to perform a custom typical price calculation.
    *
    * ```ts
    * series.typical= this.chart_typical;
    *
    *   chart_typical(sender :any,args: FinancialEventArgs )
    *   {
    *
    *
    *   }
    * ```
    */
    get typical() {
        return this._typical;
    }
    set typical(ev) {
        if (this._typical_wrapped !== null) {
            this.i.typical = delegateRemove(this.i.typical, this._typical_wrapped);
            this._typical_wrapped = null;
            this._typical = null;
        }
        this._typical = ev;
        this._typical_wrapped = (o, e) => {
            let outerArgs = new FinancialEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeTypical) {
                this.beforeTypical(this, outerArgs);
            }
            if (this._typical) {
                this._typical(this, outerArgs);
            }
        };
        this.i.typical = delegateCombine(this.i.typical, this._typical_wrapped);
    }
    /**
     * Handle this event in order to specify which columns the Typical price calculation is based on.
    *
    * Use `TypicalBasedOn` event to specify which columns the Typical price calculation is based on.
    *
    * ```ts
    *    series.typicalBasedOn= this.chart_typicalBasedOn;
    *     chart_typicalBasedOn(sender :any,args: FinancialEventArgs )
    *   {
    *
    *   }
    * ```
    */
    get typicalBasedOn() {
        return this._typicalBasedOn;
    }
    set typicalBasedOn(ev) {
        if (this._typicalBasedOn_wrapped !== null) {
            this.i.typicalBasedOn = delegateRemove(this.i.typicalBasedOn, this._typicalBasedOn_wrapped);
            this._typicalBasedOn_wrapped = null;
            this._typicalBasedOn = null;
        }
        this._typicalBasedOn = ev;
        this._typicalBasedOn_wrapped = (o, e) => {
            let outerArgs = new FinancialEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeTypicalBasedOn) {
                this.beforeTypicalBasedOn(this, outerArgs);
            }
            if (this._typicalBasedOn) {
                this._typicalBasedOn(this, outerArgs);
            }
        };
        this.i.typicalBasedOn = delegateCombine(this.i.typicalBasedOn, this._typicalBasedOn_wrapped);
    }
}
