// igniteui-react-charts/ES2015/igr-scatter-polygon-series.js
import { Style } from "igniteui-react-core/ES2015/igr-style";
import { MarkerType_$type } from "./MarkerType";
import { CollisionAvoidanceType_$type } from "./CollisionAvoidanceType";
import { IgrShapeSeriesBase } from "./igr-shape-series-base";
import { ScatterPolygonSeries } from "./ScatterPolygonSeries";
import { ensureEnum, brushToString, stringToBrush } from "igniteui-react-core/ES2015/componentUtil";
/**
 * Series class which renders polygons as Paths based on lists of points in the ItemsSource.
*/
export class IgrScatterPolygonSeries extends IgrShapeSeriesBase {
    createImplementation() {
        return new ScatterPolygonSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Returns whether the current series supports visual markers.
    */
    get hasMarkers() {
        return this.i.hasMarkers;
    }
    /**
     * The StyleSelector which is used to select a style for each Shape.
    */
    get shapeStyleSelector() {
        return this.i.nh;
    }
    set shapeStyleSelector(v) {
        this.i.nh = v;
    }
    /**
     * The default style to apply to all Shapes in the series.
    */
    get shapeStyle() {
        if (this.i.oe == null) {
            return null;
        }
        if (!this.i.oe.externalObject) {
            let e = new Style();
            e._implementation = this.i.oe;
            this.i.oe.externalObject = e;
        }
        return this.i.oe.externalObject;
    }
    set shapeStyle(v) {
        v == null ? this.i.oe = null : this.i.oe = v.i;
    }
    /**
     * Gets or sets the marker type for the current series object.
     * If the MarkerTemplate property is set, the setting of the MarkerType property will be ignored.
    */
    get markerType() {
        return this.i.nc;
    }
    set markerType(v) {
        this.i.nc = ensureEnum(MarkerType_$type, v);
    }
    /**
     * Gets or sets the MarkerTemplate for the current series object.
    */
    get markerTemplate() {
        return this.i.nw;
    }
    set markerTemplate(v) {
        this.i.nw = v;
    }
    /**
     * Gets the effective marker template for the current series object.
    */
    get actualMarkerTemplate() {
        return this.i.nv;
    }
    set actualMarkerTemplate(v) {
        this.i.nv = v;
    }
    /**
     * Gets or sets the brush that specifies how the current series object's marker interiors are painted.
    */
    get markerBrush() {
        return brushToString(this.i.oa);
    }
    set markerBrush(v) {
        this.i.oa = stringToBrush(v);
    }
    /**
     * Gets the effective marker brush for the current series object.
    */
    get actualMarkerBrush() {
        return brushToString(this.i.n8);
    }
    set actualMarkerBrush(v) {
        this.i.n8 = stringToBrush(v);
    }
    /**
     * Gets or sets the brush that specifies how the current series object's marker outlines are painted.
    */
    get markerOutline() {
        return brushToString(this.i.ob);
    }
    set markerOutline(v) {
        this.i.ob = stringToBrush(v);
    }
    /**
     * Gets the effective marker outline for the current series object.
    */
    get actualMarkerOutline() {
        return brushToString(this.i.n9);
    }
    set actualMarkerOutline(v) {
        this.i.n9 = stringToBrush(v);
    }
    /**
     * The desired behavior for markers in this series which are placed too close together for the current view, resulting in a collision.
    */
    get markerCollisionAvoidance() {
        return this.i.nb;
    }
    set markerCollisionAvoidance(v) {
        this.i.nb = ensureEnum(CollisionAvoidanceType_$type, v);
    }
}
