// igniteui-react-charts/ES2015/igr-item-tool-tip-layer.js
import { IgrAnnotationLayer } from "./igr-annotation-layer";
import { ItemToolTipLayer } from "./ItemToolTipLayer";
import { ensureBool } from "igniteui-react-core/ES2015/componentUtil";
/**
 * Represents an annotation layer that displays tooltips for all target series individually.
*/
export class IgrItemToolTipLayer extends IgrAnnotationLayer {
    createImplementation() {
        return new ItemToolTipLayer();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the default tooltip behaviors for the chart are disabled if this layer is present.
    */
    get isDefaultTooltipBehaviorDisabled() {
        return this.i.cn;
    }
    /**
     * Gets or sets whether to use value interpolation when drawing the tooltips.
    */
    get useInterpolation() {
        return this.i.m7;
    }
    set useInterpolation(v) {
        this.i.m7 = ensureBool(v);
    }
    /**
     * Gets or sets whether to skip past unknown values when searching for series values.
    */
    get skipUnknownValues() {
        return this.i.m6;
    }
    set skipUnknownValues(v) {
        this.i.m6 = ensureBool(v);
    }
}
