// igniteui-react-charts/ES2015/igr-scatter-contour-series.js
import { IgrValueBrushScale } from "./igr-value-brush-scale";
import { IgrContourValueResolver } from "./igr-contour-value-resolver";
import { IgrScatterTriangulationSeries } from "./igr-scatter-triangulation-series";
import { ScatterContourSeries } from "./ScatterContourSeries";
/**
 * Series class for rendering isarithmic contours based on a triangulation of X+Y+Value points in the ItemsSource.
*/
export class IgrScatterContourSeries extends IgrScatterTriangulationSeries {
    createImplementation() {
        return new ScatterContourSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or set the name of the property from which to retrieve the numeric values from the ItemsSource items.
    */
    get valueMemberPath() {
        return this.i.no;
    }
    set valueMemberPath(v) {
        this.i.no = v;
    }
    /**
     * Gets or sets the ValueBrushScale to use when determining Brushes for each contour line, based on the values found in ValueMemberPath.
    */
    get fillScale() {
        if (this.i.nj == null) {
            return null;
        }
        if (!this.i.nj.externalObject) {
            let e = IgrValueBrushScale._createFromInternal(this.i.nj);
            if (e) {
                e._implementation = this.i.nj;
            }
            this.i.nj.externalObject = e;
        }
        return this.i.nj.externalObject;
    }
    set fillScale(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.nj = null : this.i.nj = v.i;
    }
    /**
     * Gets or set the ContourValueResolver used to determine the numeric values of contours.
    */
    get valueResolver() {
        if (this.i.nh == null) {
            return null;
        }
        if (!this.i.nh.externalObject) {
            let e = IgrContourValueResolver._createFromInternal(this.i.nh);
            if (e) {
                e._implementation = this.i.nh;
            }
            this.i.nh.externalObject = e;
        }
        return this.i.nh.externalObject;
    }
    set valueResolver(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.nh = null : this.i.nh = v.i;
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.fillScale && this.fillScale._styling) {
            this.fillScale._styling(container, component, this);
        }
        if (this.valueResolver && this.valueResolver._styling) {
            this.valueResolver._styling(container, component, this);
        }
        this._inStyling = false;
    }
    getItemValue(item, memberPathName) {
        let iv = this.i.fs(item, memberPathName);
        return (iv);
    }
}
