// igniteui-react-charts/ES2015/igr-category-item-highlight-layer.js
import { IgrSeries } from "./igr-series";
import { CategoryItemHighlightType_$type } from "./CategoryItemHighlightType";
import { IgrAnnotationLayer } from "./igr-annotation-layer";
import { CategoryItemHighlightLayer } from "./CategoryItemHighlightLayer";
import { ensureBool, ensureEnum } from "igniteui-react-core/ES2015/componentUtil";
/**
 * Represents an annotation layer that highlights items in a series that use a category axis
 * either by drawing a banded shape at their position, or by rendering a marker at their position.
 * Depending on the type of series, the default highlight will be affected. To override
 * the type of highlight used, you can set the HighlightType property.
*/
export class IgrCategoryItemHighlightLayer extends IgrAnnotationLayer {
    createImplementation() {
        return new CategoryItemHighlightLayer();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets the series to target this annotation to. If null, this annotation targets all series simultaneously.
    */
    get targetSeries() {
        if (this.i.m6 == null) {
            return null;
        }
        if (!this.i.m6.externalObject) {
            let e = IgrSeries._createFromInternal(this.i.m6);
            if (e) {
                e._implementation = this.i.m6;
            }
            this.i.m6.externalObject = e;
        }
        return this.i.m6.externalObject;
    }
    set targetSeries(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.m6 = null : this.i.m6 = v.i;
    }
    /**
     * Gets or sets whether to use value interpolation when drawing a line through the best value for the pointer position.
    */
    get useInterpolation() {
        return this.i.na;
    }
    set useInterpolation(v) {
        this.i.na = ensureBool(v);
    }
    /**
     * Gets or sets which type of highlight shape to use when highlighting items.
    */
    get highlightType() {
        return this.i.m4;
    }
    set highlightType(v) {
        this.i.m4 = ensureEnum(CategoryItemHighlightType_$type, v);
    }
    /**
     * Gets or sets the template to use for marker visuals for the current series object.
    */
    get markerTemplate() {
        return this.i.no;
    }
    set markerTemplate(v) {
        this.i.no = v;
    }
    /**
     * Gets or sets the width to use for the highlight region if highlighting items in a grid aligned series (line, spline, etc), with a banded shape.
    */
    get bandHighlightWidth() {
        return this.i.nc;
    }
    set bandHighlightWidth(v) {
        this.i.nc = +v;
    }
    /**
     * Gets or sets whether to skip unknown values when searching for series values.
    */
    get skipUnknownValues() {
        return this.i.m9;
    }
    set skipUnknownValues(v) {
        this.i.m9 = ensureBool(v);
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.targetSeries && this.targetSeries._styling) {
            this.targetSeries._styling(container, component, this);
        }
        this._inStyling = false;
    }
}
