// igniteui-react-charts/ES2015/igr-scatter-polyline-series.js
import { Style } from "igniteui-react-core/ES2015/igr-style";
import { IgrShapeSeriesBase } from "./igr-shape-series-base";
import { ScatterPolylineSeries } from "./ScatterPolylineSeries";
/**
 * Series class used for displaying multiple polylines for datasources which contain multiple lists of points.
*/
export class IgrScatterPolylineSeries extends IgrShapeSeriesBase {
    createImplementation() {
        return new ScatterPolylineSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * The StyleSelector which is used to select a style for each Shape.
    */
    get shapeStyleSelector() {
        return this.i.na;
    }
    set shapeStyleSelector(v) {
        this.i.na = v;
    }
    /**
     * The default style to apply to all Shapes in the series.
    */
    get shapeStyle() {
        if (this.i.nf == null) {
            return null;
        }
        if (!this.i.nf.externalObject) {
            let e = new Style();
            e._implementation = this.i.nf;
            this.i.nf.externalObject = e;
        }
        return this.i.nf.externalObject;
    }
    set shapeStyle(v) {
        v == null ? this.i.nf = null : this.i.nf = v.i;
    }
}
