// igniteui-react-charts/ES2015/IndicatorProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, typeCast, markType, getInstanceType } from "igniteui-react-core/ES2015/type";
import { IIndicatorProxy_$type } from "./IIndicatorProxy";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { stringEndsWith } from "igniteui-react-core/ES2015/string";
/**
 * @hidden
 */
export class IndicatorProxy extends Base {
    matchesType(a) {
        if (stringEndsWith(getInstanceType(a).typeName, "Indicator")) {
            return true;
        }
        return false;
    }
    setCloseMemberPath(a, b) {
        a.ob = b;
    }
    setDisplayType(a, b) {
        a.pj = b;
    }
    setHighMemberPath(a, b) {
        a.oe = b;
    }
    setLongPeriod(a, b) {
        if (typeCast(IShortPeriodLongPeriod_$type, a) !== null) {
            a.longPeriod = b;
        }
    }
    setLowMemberPath(a, b) {
        a.oj = b;
    }
    setNegativeBrush(a, b) {
        a.pd = b;
    }
    setOpenMemberPath(a, b) {
        a.on = b;
    }
    setPeriodOverride(a, b) {
        if (typeCast(StrategyBasedIndicator.$, a) !== null) {
            a.qq(b);
        }
    }
    setShortPeriod(a, b) {
        if (typeCast(IShortPeriodLongPeriod_$type, a) !== null) {
            a.shortPeriod = b;
        }
    }
    setSignalPeriod(a, b) {
        if (getInstanceType(a).typeName == "MovingAverageConvergenceDivergenceIndicator") {
            let s_ = a;
            let p_ = b;
            s_['signalPeriod'] = p_;
        }
    }
    setSmoothingPeriod(a, b) {
        if (getInstanceType(a).typeName == "FullStochasticOscillatorIndicator") {
            let s_ = a;
            let p_ = b;
            s_['smoothingPeriod'] = p_;
        }
    }
    setVolumeMemberPath(a, b) {
        a.or = b;
    }
    setXAxis(a, b) {
        a.l7 = b;
    }
    setYAxis(a, b) {
        a.ms = b;
    }
    handleCustomIndicatorEvent(a, b) {
        if (getInstanceType(a).typeName == "CustomIndicator") {
            let s_ = a;
            let p_ = b;
            s_['indicator'] = p_;
        }
    }
}
IndicatorProxy.$t = markType(IndicatorProxy, 'IndicatorProxy', Base.$, [IIndicatorProxy_$type]);
