// igniteui-react-charts/ES2015/MarketFacilitationIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ItemwiseStrategyBasedIndicator } from "./ItemwiseStrategyBasedIndicator";
import { MarketFacilitationIndexIndicatorStrategy } from "./MarketFacilitationIndexIndicatorStrategy";
import { markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class MarketFacilitationIndexIndicator extends ItemwiseStrategyBasedIndicator {
    constructor() {
        super();
        this.pj = 1;
    }
    get_qx() {
        return new MarketFacilitationIndexIndicatorStrategy();
    }
    get qx() {
        return this.get_qx();
    }
    get_qo() {
        return MarketFacilitationIndexIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
}
MarketFacilitationIndexIndicator.$t = markType(MarketFacilitationIndexIndicator, 'MarketFacilitationIndexIndicator', ItemwiseStrategyBasedIndicator.$);
