// igniteui-react-charts/ES2015/MoneyFlowIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MoneyFlowIndexIndicatorStrategy } from "./MoneyFlowIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class MoneyFlowIndexIndicator extends StrategyBasedIndicator {
    constructor() {
        super();
        this.pj = 1;
    }
    get_qb() {
        return new MoneyFlowIndexIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return MoneyFlowIndexIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(MoneyFlowIndexIndicator.qx));
    }
    set qw(a) {
        this.h(MoneyFlowIndexIndicator.qx, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
}
MoneyFlowIndexIndicator.$t = markType(MoneyFlowIndexIndicator, 'MoneyFlowIndexIndicator', StrategyBasedIndicator.$);
MoneyFlowIndexIndicator.qx = StrategyBasedIndicator.qt(14, MoneyFlowIndexIndicator.$);
