// igniteui-react-charts/ES2015/OverlayProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType, getInstanceType } from "igniteui-react-core/ES2015/type";
import { IOverlayProxy_$type } from "./IOverlayProxy";
import { stringEndsWith } from "igniteui-react-core/ES2015/string";
/**
 * @hidden
 */
export class OverlayProxy extends Base {
    matchesType(a) {
        if (stringEndsWith(getInstanceType(a).typeName, "Overlay")) {
            return true;
        }
        return false;
    }
    setCloseMemberPath(a, b) {
        a.ob = b;
    }
    setHighMemberPath(a, b) {
        a.oe = b;
    }
    setLowMemberPath(a, b) {
        a.oj = b;
    }
    setMultiplier(a, b) {
        if (getInstanceType(a).typeName == "BollingerBandsOverlay") {
            let s_ = a;
            let p_ = b;
            s_['multiplier'] = p_;
        }
    }
    setOpenMemberPath(a, b) {
        a.on = b;
    }
    setVolumeMemberPath(a, b) {
        a.or = b;
    }
    setXAxis(a, b) {
        a.l7 = b;
    }
    setYAxis(a, b) {
        a.ms = b;
    }
}
OverlayProxy.$t = markType(OverlayProxy, 'OverlayProxy', Base.$, [IOverlayProxy_$type]);
