// igniteui-react-charts/ES2015/PercentagePriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentagePriceOscillatorIndicatorStrategy } from "./PercentagePriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class PercentagePriceOscillatorIndicator extends StrategyBasedIndicator {
    get_qb() {
        return new PercentagePriceOscillatorIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return PercentagePriceOscillatorIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get shortPeriod() {
        return typeGetValue(this.c(PercentagePriceOscillatorIndicator.qx));
    }
    set shortPeriod(a) {
        this.h(PercentagePriceOscillatorIndicator.qx, a);
    }
    qj() {
        return this.shortPeriod;
    }
    get longPeriod() {
        return typeGetValue(this.c(PercentagePriceOscillatorIndicator.qw));
    }
    set longPeriod(a) {
        this.h(PercentagePriceOscillatorIndicator.qw, a);
    }
    qg() {
        return this.longPeriod;
    }
}
PercentagePriceOscillatorIndicator.$t = markType(PercentagePriceOscillatorIndicator, 'PercentagePriceOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
PercentagePriceOscillatorIndicator.qx = StrategyBasedIndicator.qv(10, PercentagePriceOscillatorIndicator.$);
PercentagePriceOscillatorIndicator.qw = StrategyBasedIndicator.qr(30, PercentagePriceOscillatorIndicator.$);
