// igniteui-react-charts/ES2015/PieChartViewManager_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Point_$type, runOn, delegateCombine, Type, markType, PointUtil, EventArgs, typeCast, INotifyPropertyChanged_$type, IEnumerable_$type, PropertyChangedEventArgs, IList_$type, Number_$type, enumGetBox, EnumUtil, Boolean_$type, delegateRemove, Delegate_$type, fromEnum, fromEn } from "igniteui-react-core/ES2015/type";
import { Rect } from "igniteui-react-core/ES2015/Rect";
import { RenderingContext } from "igniteui-react-core/ES2015/RenderingContext";
import { PieChartStyleInfo } from "./PieChartStyleInfo";
import { BrushCollection } from "igniteui-react-core/ES2015/BrushCollection";
import { BrushUtil } from "igniteui-react-core/ES2015/BrushUtil";
import { Brush } from "igniteui-react-core/ES2015/Brush";
import { FontUtil } from "igniteui-react-core/ES2015/FontUtil";
import { DeviceUtils } from "igniteui-react-core/ES2015/DeviceUtils";
import { Path } from "igniteui-react-core/ES2015/Path";
import { CanvasViewRenderer } from "igniteui-react-core/ES2015/CanvasViewRenderer";
import { DOMEventProxy } from "igniteui-react-core/ES2015/DOMEventProxy";
import { PieChartVisualData } from "./PieChartVisualData";
import { Style } from "igniteui-react-core/ES2015/Style";
import { isNaN_, isInfinity } from "igniteui-react-core/ES2015/number";
import { Control } from "igniteui-react-core/ES2015/Control";
import { Line } from "igniteui-react-core/ES2015/Line";
import { LabelsPosition_$type } from "./LabelsPosition";
import { TextBlock } from "igniteui-react-core/ES2015/TextBlock";
import { GeometryUtil } from "igniteui-react-core/ES2015/GeometryUtil";
import { PathFigure } from "igniteui-react-core/ES2015/PathFigure";
import { BezierSegment } from "igniteui-react-core/ES2015/BezierSegment";
import { PathGeometry } from "igniteui-react-core/ES2015/PathGeometry";
import { List$1 } from "igniteui-react-core/ES2015/List$1";
import { Visibility_$type } from "igniteui-react-core/ES2015/Visibility";
import { LeaderLineType_$type } from "./LeaderLineType";
import { stringEmpty } from "igniteui-react-core/ES2015/string";
import { ILegendOwner_$type } from "igniteui-react-core/ES2015/ILegendOwner";
import { ILegendSeries_$type } from "igniteui-react-core/ES2015/ILegendSeries";
import { ILegendTemplateProvider_$type } from "igniteui-react-core/ES2015/ILegendTemplateProvider";
import { FastItemsSource } from "igniteui-react-core/ES2015/FastItemsSource";
import { DependencyProperty } from "igniteui-react-core/ES2015/DependencyProperty";
import { UIElement } from "igniteui-react-core/ES2015/UIElement";
import { OthersCategoryType_$type } from "./OthersCategoryType";
import { IndexCollection } from "./IndexCollection";
import { NotifyCollectionChangedEventArgs } from "igniteui-react-core/ES2015/NotifyCollectionChangedEventArgs";
import { IChartLegend_$type } from "igniteui-react-core/ES2015/IChartLegend";
import { SweepDirection_$type } from "igniteui-react-core/ES2015/SweepDirection";
import { DataTemplate } from "igniteui-react-core/ES2015/DataTemplate";
import { SliceSelectionMode_$type } from "./SliceSelectionMode";
import { ObjectCollection } from "igniteui-react-core/ES2015/ObjectCollection";
import { PropertyUpdatedEventArgs } from "igniteui-react-core/ES2015/PropertyUpdatedEventArgs";
import { Pool$1 } from "igniteui-react-core/ES2015/Pool$1";
import { SelectedItemChangingEventArgs } from "./SelectedItemChangingEventArgs";
import { SelectedItemsChangingEventArgs } from "./SelectedItemsChangingEventArgs";
import { SelectedItemChangedEventArgs } from "./SelectedItemChangedEventArgs";
import { SelectedItemsChangedEventArgs } from "./SelectedItemsChangedEventArgs";
import { MathUtil } from "igniteui-react-core/ES2015/MathUtil";
import { DataContext } from "igniteui-react-core/ES2015/DataContext";
import { Size } from "igniteui-react-core/ES2015/Size";
import { ContentControl } from "igniteui-react-core/ES2015/ContentControl";
import { PieSliceVisualData } from "./PieSliceVisualData";
import { RectData } from "igniteui-react-core/ES2015/RectData";
import { MouseEventArgs } from "igniteui-react-core/ES2015/MouseEventArgs";
import { PropertyMetadata } from "igniteui-react-core/ES2015/PropertyMetadata";
import { FastItemObjectColumn } from "igniteui-react-core/ES2015/FastItemObjectColumn";
import { ArcSegment } from "igniteui-react-core/ES2015/ArcSegment";
import { LineSegment } from "igniteui-react-core/ES2015/LineSegment";
import { EllipseGeometry } from "igniteui-react-core/ES2015/EllipseGeometry";
import { PathSegmentCollection } from "igniteui-react-core/ES2015/PathSegmentCollection";
import { PathFigureCollection } from "igniteui-react-core/ES2015/PathFigureCollection";
import { GeometryGroup } from "igniteui-react-core/ES2015/GeometryGroup";
import { RectUtil } from "igniteui-react-core/ES2015/RectUtil";
import { PrimitiveAppearanceData } from "igniteui-react-core/ES2015/PrimitiveAppearanceData";
import { LabelAppearanceData } from "igniteui-react-core/ES2015/LabelAppearanceData";
import { PointData } from "igniteui-react-core/ES2015/PointData";
import { AppearanceHelper } from "igniteui-react-core/ES2015/AppearanceHelper";
import { PathVisualData } from "igniteui-react-core/ES2015/PathVisualData";
import { PieLegendTemplates } from "./PieLegendTemplates";
import { MouseButtonEventArgs } from "igniteui-react-core/ES2015/MouseButtonEventArgs";
import { TranslateTransform } from "igniteui-react-core/ES2015/TranslateTransform";
/**
 * @hidden
 */
export class PieChartViewManager extends Base {
    constructor(a) {
        super();
        this.o = null;
        this.q = null;
        this.t = null;
        this.s = null;
        this.r = null;
        this.k = 0;
        this.j = 0;
        this.l = 1;
        this.ao = null;
        this.e = null;
        this.g = null;
        this.f = null;
        this.h = null;
        this.a = null;
        this.n = null;
        this.c = null;
        this._toolTipObject = null;
        this.v = null;
        this.i = true;
        this.a = a;
    }
    ab() {
        let a = this.o.rootWrapper.width();
        let b = this.o.rootWrapper.height();
        this.ad(a, b);
    }
    ad(a, b) {
        this.c.bd = new Rect(0, 0, 0, a, b);
        let c = this.a.a.cd;
        if (c == this.l && a == this.k && b == this.j) {
            return;
        }
        this.k = a;
        this.j = b;
        this.l = c;
        this.a.a8();
    }
    aa() {
        if (this.n == null) {
            this.n = (a) => {
                let b = this.a.a.cq;
                let c = this.a.a.cd;
                if (!isNaN_(b) && b != 0) {
                    return;
                }
                if (window.devicePixelRatio != undefined && window.devicePixelRatio != c) {
                    this.a.a.cd = window.devicePixelRatio;
                }
            };
            window.addEventListener("resize", this.n, false);
        }
        this.n(null);
    }
    x(a, b) {
        let c = this.a.a.cd;
        if (a != this.k || b != this.j || c != this.l) {
            this.k = a;
            this.j = b;
            this.l = this.a.a.cd;
            this.an();
        }
    }
    an() {
        this.ao = new Rect(0, 0, 0, this.k, this.j);
        let a = this.k;
        let b = this.j;
        let c = this.l;
        if (c != 1) {
            a = Math.round(this.k * c);
            b = Math.round(this.j * c);
        }
        this.q.setAttribute("width", a.toString());
        this.q.setAttribute("height", b.toString());
        this.q.setStyleProperty("width", this.k.toString() + "px");
        this.q.setStyleProperty("height", this.j.toString() + "px");
        this.s.setAttribute("width", a.toString());
        this.s.setAttribute("height", b.toString());
        this.s.setStyleProperty("width", this.k.toString() + "px");
        this.s.setStyleProperty("height", this.j.toString() + "px");
        this.r.setAttribute("width", a.toString());
        this.r.setAttribute("height", b.toString());
        this.r.setStyleProperty("width", this.k.toString() + "px");
        this.r.setStyleProperty("height", this.j.toString() + "px");
        this.t.setAttribute("width", a.toString());
        this.t.setAttribute("height", b.toString());
        this.t.setStyleProperty("width", this.k.toString() + "px");
        this.t.setStyleProperty("height", this.j.toString() + "px");
        return this.ao;
    }
    d() {
        return this.a.b.c;
    }
    am() {
        let a = DOMEventProxy.cv(this.o.rootWrapper);
        return { $type: Point_$type, x: a.left, y: a.top };
    }
    ag(a) {
        if (this.o != null) {
            this.o.setTimeout(a, 0);
        }
        else {
            window.setTimeout(a, 0);
        }
    }
    b() {
        let a = new PieChartStyleInfo();
        let b;
        let c;
        let d = BrushUtil.d("chart", this.o, b, c);
        b = d.p2;
        c = d.p3;
        a.a = b;
        a.b = c;
        this.o.startCSSQuery();
        a.f = ((() => {
            let $ret = new Brush();
            $ret.fill = this.o.getCssDefaultPropertyValue("ui-chart-innerlabels", "color");
            return $ret;
        })());
        a.g = ((() => {
            let $ret = new Brush();
            $ret.fill = this.o.getCssDefaultPropertyValue("ui-chart-outerlabels", "color");
            return $ret;
        })());
        this.o.endCSSQuery();
        a.c = FontUtil.getFont(this.o);
        a.e = DeviceUtils.g(6);
        a.i = ((() => {
            let $ret = new Brush();
            $ret.fill = "#00000033";
            return $ret;
        })());
        a.d = 1;
        return a;
    }
    ah(a) {
        a._stroke = ((() => {
            let $ret = new Brush();
            $ret.fill = "#000000";
            return $ret;
        })());
        a._fill = ((() => {
            let $ret = new Brush();
            $ret.fill = "#222222";
            return $ret;
        })());
    }
    ac(a) {
        if (a == null) {
            if (this.c != null) {
                this.c.ao();
            }
            this.o = null;
            this.q = null;
            this.s = null;
            this.r = null;
            this.t = null;
            this.e = null;
            this.g = null;
            this.f = null;
            this.h = null;
            this.c = null;
            if (this.n != null) {
                window.removeEventListener("resize", this.n, false);
                this.n = null;
            }
            return;
        }
        let b = a;
        this.o = b;
        this.a.a9();
        this.a.bi();
        this.o.rootWrapper.setStyleProperty("position", "relative");
        this.o.rootWrapper.addClass("ui-corner-all");
        this.o.rootWrapper.addClass("ui-widget-content");
        this.q = this.o.createElement("canvas");
        this.q.setStyleProperty("position", "absolute");
        this.q.setStyleProperty("top", "0px");
        this.q.setStyleProperty("left", "0px");
        this.s = this.o.createElement("canvas");
        this.s.setStyleProperty("position", "absolute");
        this.s.setStyleProperty("top", "0px");
        this.s.setStyleProperty("left", "0px");
        this.r = this.o.createElement("canvas");
        this.r.setStyleProperty("position", "absolute");
        this.r.setStyleProperty("top", "0px");
        this.r.setStyleProperty("left", "0px");
        this.t = this.o.createElement("canvas");
        this.t.setStyleProperty("position", "absolute");
        this.t.setStyleProperty("top", "0px");
        this.t.setStyleProperty("left", "0px");
        this.o.append(this.q);
        this.o.append(this.s);
        this.o.append(this.r);
        this.o.append(this.t);
        this.e = new RenderingContext(new CanvasViewRenderer(), this.o.get2DCanvasContext(this.q));
        this.g = new RenderingContext(new CanvasViewRenderer(), this.o.get2DCanvasContext(this.s));
        this.f = new RenderingContext(new CanvasViewRenderer(), this.o.get2DCanvasContext(this.r));
        this.h = new RenderingContext(new CanvasViewRenderer(), this.o.get2DCanvasContext(this.t));
        this.c = new DOMEventProxy(this.t, this.o, false);
        let c = this.c;
        c.onMouseOver = delegateCombine(c.onMouseOver, runOn(this.a, this.a.aj));
        let d = this.c;
        d.onMouseLeave = delegateCombine(d.onMouseLeave, runOn(this.a, this.a.ai));
        let e = this.c;
        e.onMouseDown = delegateCombine(e.onMouseDown, runOn(this.a, this.a.ah));
        let f = this.c;
        f.onMouseUp = delegateCombine(f.onMouseUp, runOn(this.a, this.a.ak));
        this.c.ab = this.a.a.bv;
        this.ad(this.o.rootWrapper.width(), this.o.rootWrapper.height());
    }
    w(a) {
        this.o.append(a);
    }
    ae(a) {
        if (this.c == null) {
            return;
        }
        this.c.ab = a;
    }
    u() {
        return this.o;
    }
    y(a) {
        a.width = this.k;
        a.height = this.j;
    }
    aj(a) {
        this.v = null;
        this._toolTipObject = null;
        if (typeof a === 'string') {
            this.v = a;
        }
        else {
            this._toolTipObject = a;
            if (this._toolTipObject != null) {
                this._toolTipObject.setStyleProperty("position", "absolute");
                this._toolTipObject.setStyleProperty("white-space", "nowrap");
                this._toolTipObject.setStyleProperty("top", "0");
                this._toolTipObject.setStyleProperty("left", "0");
                this._toolTipObject.setStyleProperty("z-index", "10000");
            }
        }
    }
    ai(pos_, context_) {
        if (this._toolTipObject != null) {
            if (this.i) {
                this.w(this._toolTipObject);
                this.i = false;
            }
            context_.hideOthers = true;
            if (this._toolTipObject.updateToolTip) {
                this._toolTipObject.updateToolTip(context_);
            }
            ;
            this._toolTipObject.setOffset(pos_.x, pos_.y);
        }
    }
    z(a) {
        if (this._toolTipObject != null) {
            let context_ = a;
            if (this._toolTipObject.hideToolTip) {
                this._toolTipObject.hideToolTip(context_);
            }
            ;
        }
    }
    ak(style_) {
        let a = null;
        let fillColor_ = null;
        if (style_ != null) {
            if (style_.fill) {
                fillColor_ = style_.fill;
            }
            ;
            if (fillColor_ != null) {
                a = Brush.create(fillColor_);
            }
        }
        return a;
    }
    al(style_) {
        let a = null;
        let outlineColor_ = null;
        if (style_ != null) {
            if (style_.stroke) {
                outlineColor_ = style_.stroke;
            }
            ;
            if (outlineColor_ != null) {
                a = Brush.create(outlineColor_);
            }
        }
        return a;
    }
    m(style_) {
        let actualOutlineThickness_ = NaN;
        if (style_ != null) {
            if (style_.strokeThickness) {
                actualOutlineThickness_ = style_.strokeThickness;
            }
            ;
        }
        return actualOutlineThickness_;
    }
    af() {
    }
}
PieChartViewManager.$t = markType(PieChartViewManager, 'PieChartViewManager');
/**
 * @hidden
 */
export class LabelClickEventArgs extends EventArgs {
    constructor(a, b) {
        super();
        this._slice = null;
        this.c = true;
        this.slice = a;
    }
    get slice() {
        return this._slice;
    }
    set slice(a) {
        this._slice = a;
    }
    get item() {
        if (this.slice != null) {
            return this.slice.dataContext;
        }
        else {
            return null;
        }
    }
    get allowSliceClick() {
        return this.c;
    }
    set allowSliceClick(a) {
        this.c = a;
    }
}
LabelClickEventArgs.$t = markType(LabelClickEventArgs, 'LabelClickEventArgs', EventArgs.$);
/**
 * @hidden
 */
export class SliceClickEventArgs extends EventArgs {
    constructor(a, b) {
        super();
        this._slice = null;
        this.d = false;
        this.c = false;
        this._originalEvent = null;
        this.slice = a;
        this.originalEvent = b;
        if (a == null) {
            return;
        }
        this.d = a.isSelected;
        this.c = a.isExploded;
    }
    get slice() {
        return this._slice;
    }
    set slice(a) {
        this._slice = a;
    }
    get isSelected() {
        return this.d;
    }
    set isSelected(a) {
        this.d = a;
        this.slice.ai.fi(this.slice, a);
    }
    get isExploded() {
        return this.c;
    }
    set isExploded(a) {
        this.c = a;
        this.slice.isExploded = a;
    }
    get isOthersSlice() {
        return this.slice.isOthersSlice;
    }
    get dataContext() {
        if (this.slice != null) {
            return this.slice.dataContext;
        }
        return null;
    }
    get originalEvent() {
        return this._originalEvent;
    }
    set originalEvent(a) {
        this._originalEvent = a;
    }
}
SliceClickEventArgs.$t = markType(SliceClickEventArgs, 'SliceClickEventArgs', EventArgs.$);
/**
 * @hidden
 */
export class PieLabel extends Control {
    constructor() {
        super();
        this.ah = null;
        this.ao = null;
        this.ai = 0;
        this.aj = null;
        this.ap = null;
        this.aq = null;
        this.ag = 0;
    }
    ak(a) {
        if (a.a.ao == 3) {
            this.ao = a.bz(this);
            let b = typeCast(TextBlock.$, this.aj);
            let c = b.ak;
            a.a.ff();
            let d = this.ah.isExploded ? this.ah.b5 : this.ah.b7;
            let e = GeometryUtil.k(d, this.ai, this.ah.a2 + a.a.cm);
            let f = a.b0.width;
            if (a.y(b) + e.x > f && e.x < f) {
                while (c.length > 0 && a.y(b) + e.x > f) {
                    c = c.substr(0, c.length - 1);
                    b.ak = c;
                }
                if (c.length > 3) {
                    c = c.substr(0, c.length - 3);
                    c = c + "...";
                }
                else {
                    if (a.y("...") < this.ao.width) {
                        c = "...";
                    }
                    else {
                        c = stringEmpty();
                    }
                }
            }
            let g = new TextBlock();
            g.ak = c;
            this.aj = g;
        }
    }
    al() {
    }
    am() {
        this.ap._visibility = this.ah.ai.g8 != 0 ? this.ah.ai.g8 : this.ah.aj._visibility;
        let a;
        let b = this.ah.ai.cn;
        let c = GeometryUtil.k(this.ah.b6(), this.ai, this.ah.a2);
        this.ap.aj = c.x;
        this.ap.al = c.y;
        if (this.ai < 90 || this.ai >= 270) {
            a = { $type: Point_$type, x: this.ao.left, y: (this.ao.top + this.ao.bottom) / 2 };
        }
        else {
            a = { $type: Point_$type, x: this.ao.right, y: (this.ao.top + this.ao.bottom) / 2 };
        }
        if (b > 0) {
            let d = GeometryUtil.f(c, a);
            if (b < d) {
                a = GeometryUtil.m(a, c, b);
            }
            else {
                a = c;
            }
        }
        this.ap.ak = a.x;
        this.ap.am = a.y;
    }
    an() {
        this.aq._visibility = this.ah.ai.g8 != 0 ? this.ah.ai.g8 : this.ah.aj._visibility;
        let a = this.ah.b6();
        let b = this.aq.aj.b._inner[0];
        let c = b._segments._inner[0];
        let d = this.ah.ai.cn;
        let e = (this.ao.top + this.ao.bottom) / 2;
        let f = GeometryUtil.k(a, this.ai, this.ah.a2);
        let g = GeometryUtil.k(a, this.ai, this.ah.a2 + this.ah.ai.cm / 2);
        let h;
        let i;
        let j;
        let k = e;
        if (this.ai < 90 && this.ai >= 0) {
            h = { $type: Point_$type, x: this.ao.left - 30, y: k };
            j = { $type: Point_$type, x: this.ao.left, y: e };
        }
        else if (this.ai < 180 && this.ai >= 90) {
            h = { $type: Point_$type, x: this.ao.right + 30, y: k };
            j = { $type: Point_$type, x: this.ao.right, y: e };
        }
        else if (this.ai < 270 && this.ai >= 180) {
            h = { $type: Point_$type, x: this.ao.right + 30, y: k };
            j = { $type: Point_$type, x: this.ao.right, y: e };
        }
        else {
            h = { $type: Point_$type, x: this.ao.left - 30, y: k };
            j = { $type: Point_$type, x: this.ao.left, y: e };
        }
        if (d > 0) {
            let l = GeometryUtil.f(f, j);
            if (d < l) {
                i = GeometryUtil.m(j, f, d);
                let m = g.x - f.x;
                let n = g.y - f.y;
                g = GeometryUtil.m(g, f, d);
                let o = g.x - f.x;
                let p = g.y - f.y;
                if ((m > 0 && o < 0) || (m < 0 && o > 0) || (n > 0 && p < 0) || (n < 0 && p > 0)) {
                    g = f;
                }
                h = GeometryUtil.m(h, f, d);
                let q = GeometryUtil.f(f, i) / GeometryUtil.f(f, j);
                h.x = f.x + (h.x - f.x) * q;
                h.y = f.y + (h.y - f.y) * q;
            }
            else {
                this.aq._visibility = 1;
                return;
            }
        }
        else {
            i = j;
        }
        if (this.ai < 180 && this.ai >= 0) {
            if (g.y > h.y) {
                g.y = f.y;
            }
        }
        else {
            if (g.y < h.y) {
                g.y = f.y;
            }
        }
        b._startPoint = f;
        if (this.ah.ai.ap == 1) {
            c.e = f;
        }
        else {
            c.e = g;
        }
        c.f = h;
        c.g = i;
    }
}
PieLabel.$t = markType(PieLabel, 'PieLabel', Control.$);
/**
 * @hidden
 */
export class PieChartBase extends Control {
    constructor() {
        super();
        this.as = null;
        this.bh = false;
        this.bk = false;
        this.bj = false;
        this.bl = false;
        this.bi = false;
        this.bo = false;
        this.bn = false;
        this.bm = false;
        this.aw = null;
        this.b5 = null;
        this.b7 = null;
        this.b6 = null;
        this.b8 = null;
        this.cv = null;
        this.a9 = null;
        this.cx = 0;
        this.cp = 0;
        this.cu = 0;
        this.b2 = null;
        this.b1 = null;
        this.b0 = null;
        this.cf = 0;
        this.bz = null;
        this.av = null;
        this.a7 = null;
        this.a8 = null;
        this.aj = new IndexCollection();
        this.ak = new IndexCollection();
        this.ca = NaN;
        this.b9 = 1;
        this._selectedItems = null;
        this.labelClick = null;
        this.selectedItemChanging = null;
        this.selectedItemsChanging = null;
        this.selectedItemChanged = null;
        this.selectedItemsChanged = null;
        this.sliceClick = null;
        this.bg = null;
        this.bf = null;
        this.g3 = null;
        this.propertyChanged = null;
        this.propertyUpdated = null;
        this.c1 = null;
        this.g3 = Rect.empty;
        let a = this.ar();
        this.e4(a);
        a.az();
        this.ab = PieChartBase.$;
        this.b2 = new List$1(Number_$type, 0);
        this.b1 = new List$1(Number_$type, 0);
        this.b0 = new List$1(Base.$, 0);
        this.aw = (b, c) => this.e8(b, c.propertyName, c.oldValue, c.newValue);
        this.b5 = (b, c) => {
            this.ff();
            this.fe();
        };
        this.b6 = (b, c) => {
            if (this.bp) {
                this.e7();
                this.e6();
                this.ff();
                this.fd();
            }
        };
        this.b7 = (b, c) => this.fh(c);
        this.b8 = (b, c) => this.e0(c);
        this.cv = (b, c) => this.ei(c.action, c.position, c.count, c.propertyName);
        this.propertyUpdated = delegateCombine(this.propertyUpdated, this.aw);
        let b = this.selectedSlices;
        b.collectionChanged = delegateCombine(b.collectionChanged, this.b7);
        this.bl = true;
        let c = this.explodedSlices;
        c.collectionChanged = delegateCombine(c.collectionChanged, this.b6);
        this.bi = true;
        this.selectedItems = new ObjectCollection(0);
        let d = this.selectedItems;
        d.collectionChanged = delegateCombine(d.collectionChanged, this.b8);
        this.bg = ((() => {
            let $ret = new Pool$1(Slice.$);
            $ret.create = runOn(this.as, this.as.g);
            $ret.activate = runOn(this.as, this.as.bc);
            $ret.disactivate = runOn(this.as, this.as.be);
            $ret.destroy = runOn(this.as, this.as.bd);
            return $ret;
        })());
        this.bf = ((() => {
            let $ret = new Pool$1(PieLabel.$);
            $ret.create = runOn(this.as, this.as.d);
            $ret.activate = runOn(this.as, this.as.ar);
            $ret.disactivate = runOn(this.as, this.as.at);
            $ret.destroy = runOn(this.as, this.as.as);
            return $ret;
        })());
        this.av = new PieChartSliceSelector(this);
        FastItemObjectColumn.m();
    }
    ar() {
        return new PieChartBaseView(this);
    }
    e4(a) {
        this.as = a;
    }
    get be() {
        return this.a9;
    }
    set be(a) {
        if (this.a9 != a) {
            let b = this.a9;
            this.a9 = a;
            this.fa("ValueColumn", b, this.a9);
        }
    }
    get cc() {
        return Math.abs(this.cl) > 1 ? this.cl : this.cl * 100;
    }
    get cl() {
        return this.c(PieChartBase.f0);
    }
    set cl(a) {
        let b = a;
        if (isNaN_(a) || isInfinity(a)) {
            b = 0;
        }
        if (a < 0) {
            b = 0;
        }
        if (a > 100) {
            b = 100;
        }
        this.h(PieChartBase.f0, b);
    }
    get ba() {
        return this.a7;
    }
    set ba(a) {
        if (this.a7 != a) {
            let b = this.ba;
            this.a7 = a;
            this.fa("LabelColumn", b, this.ba);
        }
    }
    get bb() {
        return this.a8;
    }
    set bb(a) {
        if (this.a8 != a) {
            let b = this.bb;
            this.a8 = a;
            this.fa("LegendLabelColumn", b, this.bb);
        }
    }
    get cb() {
        let a = this.cg;
        if (isNaN_(a) || isInfinity(a) || a < 0) {
            return 0;
        }
        if (a > 1) {
            return 1;
        }
        return a;
    }
    get ce() {
        let a = this.cr;
        if (isNaN_(a) || isInfinity(a) || a < 0) {
            return 0;
        }
        if (a > 1) {
            return 1;
        }
        return a;
    }
    get explodedSlices() {
        return this.aj;
    }
    set explodedSlices(a) {
        let b = this.aj;
        if (b != null) {
            let c = this.aj;
            c.collectionChanged = delegateRemove(c.collectionChanged, this.b6);
        }
        this.bi = false;
        this.aj = a;
        if (this.aj != null) {
            let d = this.aj;
            d.collectionChanged = delegateCombine(d.collectionChanged, this.b6);
            this.bi = true;
        }
        if (this.bp) {
            this.e7();
            this.e6();
            this.ff();
            this.fd();
        }
    }
    get selectedSlices() {
        return this.ak;
    }
    set selectedSlices(a) {
        let b = this.ak;
        b.collectionChanged = delegateRemove(b.collectionChanged, this.b7);
        this.bl = false;
        this.ak = a;
        if (this.ak != null) {
            let c = this.ak;
            c.collectionChanged = delegateCombine(c.collectionChanged, this.b7);
            this.bl = true;
        }
        if (this.bq) {
            this.e7();
            this.e6();
            this.ff();
            this.fd();
            this.fe();
        }
    }
    get fs() {
        return typeCast(DataTemplate.$, this.c(PieChartBase.f8));
    }
    set fs(a) {
        this.h(PieChartBase.f8, a);
    }
    bs() {
        return this.at != null;
    }
    bt() {
        return this.au != null;
    }
    get cq() {
        return this.ca;
    }
    set cq(a) {
        let b = this.ca;
        this.ca = a;
        this.fa("PixelScalingRatio", b, this.ca);
    }
    get cd() {
        return this.b9;
    }
    set cd(a) {
        let b = this.b9;
        this.b9 = a;
        this.fa("ActualPixelScalingRatio", b, this.b9);
    }
    get selectedItem() {
        return this.c(PieChartBase.gn);
    }
    set selectedItem(a) {
        this.h(PieChartBase.gn, a);
    }
    get selectedItems() {
        return this._selectedItems;
    }
    set selectedItems(a) {
        this._selectedItems = a;
    }
    ew(a, b) {
        if (this.labelClick != null) {
            this.labelClick(a, b);
        }
    }
    ey(a, b) {
        if (this.selectedItemChanging != null) {
            this.selectedItemChanging(a, b);
        }
    }
    e1(a, b) {
        if (this.selectedItemsChanging != null) {
            this.selectedItemsChanging(a, b);
        }
    }
    ex(a, b) {
        if (this.selectedItemChanged != null) {
            this.selectedItemChanged(a, b);
        }
    }
    ez(a, b) {
        if (this.selectedItemsChanged != null) {
            this.selectedItemsChanged(a, b);
        }
    }
    e3(a, b) {
        if (this.sliceClick != null) {
            this.sliceClick(a, b);
        }
    }
    ek(a, b) {
        if (!this.bi) {
            return;
        }
        if (!this.explodedSlices.contains(a.a5) && b) {
            this.explodedSlices.add(a.a5);
        }
        if (this.explodedSlices.contains(a.a5) && !b) {
            this.explodedSlices.remove(a.a5);
        }
    }
    fi(a, b) {
        this.bn = true;
        if (!this.selectedSlices.contains(a.a5) && b) {
            this.selectedSlices.add(a.a5);
        }
        if (this.selectedSlices.contains(a.a5) && !b) {
            this.selectedSlices.remove(a.a5);
        }
        this.bn = false;
    }
    fj(a) {
        this.as.ba(a);
    }
    cy(a) {
        return this.as.ad(a);
    }
    g2(a, b) {
        let c = new Rect(0, 0, 0, 0, 0);
        return c;
    }
    br(a, b) {
        let c = a.ah;
        if (c == null) {
            return false;
        }
        let d = c.b6();
        let e = this.gx == 1 ? c.a3 : c.ax;
        let f = this.gx == 1 ? c.ax : c.a3;
        let g = false;
        let h = false;
        let i = GeometryUtil.k(d, e, c.a2);
        let j = GeometryUtil.k(d, f, c.a2);
        e = PieChartBase.ch(i.x, d.x, i.y, d.y);
        f = PieChartBase.ch(j.x, d.x, j.y, d.y);
        if (this.bu(c)) {
            h = true;
        }
        let k;
        k = MathUtil.e(a.ao.right - d.x, a.ao.top - d.y);
        if (k > c.a2) {
            return false;
        }
        k = MathUtil.e(a.ao.right - d.x, a.ao.bottom - d.y);
        if (k > c.a2) {
            return false;
        }
        k = MathUtil.e(a.ao.left - d.x, a.ao.top - d.y);
        if (k > c.a2) {
            return false;
        }
        k = MathUtil.e(a.ao.left - d.x, a.ao.bottom - d.y);
        if (k > c.a2) {
            return false;
        }
        if (h) {
            return true;
        }
        if (e > f) {
            e = e - 360;
            g = true;
        }
        let l;
        l = PieChartBase.ch(a.ao.right, d.x, a.ao.top, d.y);
        if (g && l > 180 && l < 360) {
            l = l - 360;
        }
        if (l < e || l > f) {
            return false;
        }
        l = PieChartBase.ch(a.ao.right, d.x, a.ao.bottom, d.y);
        if (g && l > 180 && l < 360) {
            l = l - 360;
        }
        if (l < e || l > f) {
            return false;
        }
        l = PieChartBase.ch(a.ao.left, d.x, a.ao.top, d.y);
        if (g && l > 180 && l < 360) {
            l = l - 360;
        }
        if (l < e || l > f) {
            return false;
        }
        l = PieChartBase.ch(a.ao.left, d.x, a.ao.bottom, d.y);
        if (g && l > 180 && l < 360) {
            l = l - 360;
        }
        if (l < e || l > f) {
            return false;
        }
        return true;
    }
    static cs(a) {
        let b = Math.round(a * Math.pow(10, 5)) / Math.pow(10, 5);
        return b;
    }
    bu(a) {
        return PieChartBase.cs(Math.abs(a.ax - a.a3)) == 360;
    }
    static ch(a, b, c, d) {
        let e = MathUtil.e(a - b, c - d);
        let f = Math.asin((c - d) / e) * 180 / Math.PI;
        if (a < b) {
            f = 180 - f;
        }
        if (a > b) {
            f = 360 + f;
        }
        if (f == 360) {
            f = 0;
        }
        return GeometryUtil.g(f);
    }
    static fo(a) {
        let b = new Array(a.count);
        for (let c = 0; c < a.count; c++) {
            let d = a._inner[c];
            let e = d.ah;
            let f = GeometryUtil.k(e.b7, d.ai, e.a2);
            b[c] = f.y;
        }
        for (let g = 0; g < a.count; g++) {
            for (let h = g + 1; h < a.count; h++) {
                if (b[g] >= b[h]) {
                    let i = a._inner[g];
                    a._inner[g] = a._inner[h];
                    a._inner[h] = i;
                    let j = b[g];
                    b[g] = b[h];
                    b[h] = j;
                }
            }
        }
    }
    fg(a) {
        if (a.count == 0) {
            return;
        }
        let b = this.as.b0.width;
        let c = this.as.b0.height;
        let d = a.count;
        let e = a._inner[0].ah.a2;
        let f = a._inner[0].ah.b6();
        let g = true;
        let h = 0;
        let i = Number.POSITIVE_INFINITY;
        let j = Number.NEGATIVE_INFINITY;
        for (let k = 0; k < d - 1; k++) {
            let l = a._inner[k];
            let m = a._inner[k + 1];
            if (l.ao.intersectsWith(m.ao)) {
                h++;
            }
        }
        let n = 0;
        for (let o of fromEnum(a)) {
            i = Math.min(i, o.ao.height);
            j = Math.max(j, o.ao.height);
            n += o.ao.height;
        }
        if (n > c) {
            g = false;
        }
        if (g && h > 0) {
            for (let p = 0; p < d - 1; p++) {
                for (let q = p + 1; q < d; q++) {
                    let r = a._inner[p];
                    let s = a._inner[q];
                    if (r.ao.intersectsWith(s.ao)) {
                        let t = s.ao;
                        t.y = Math.min(r.ao.bottom + 0.01, c - i);
                        let u = this.cm + e;
                        let v = Math.abs(f.y - (t.y + i / 2));
                        let w = Math.sqrt(Math.abs(u * u - v * v));
                        let x = GeometryUtil.g(s.ai);
                        if (x > 90 && x < 270) {
                            w = (t.width + w) * -1;
                        }
                        t.x = f.x + w;
                        s.ao = t;
                    }
                }
            }
            for (let y = d - 1; y > 0; y--) {
                for (let z = y - 1; z >= 0; z--) {
                    let aa = a._inner[y];
                    let ab = a._inner[z];
                    if (aa.ao.intersectsWith(ab.ao)) {
                        let ac = ab.ao;
                        ac.y = Math.max(aa.ao.top - i - 0.01, 0);
                        let ad = this.cm + e;
                        let ae = Math.abs(f.y - (ac.y + i / 2));
                        let af = Math.sqrt(Math.abs(ad * ad - ae * ae));
                        let ag = GeometryUtil.g(ab.ai);
                        if (ag > 90 && ag < 270) {
                            af = (ac.width + af) * -1;
                        }
                        ac.x = f.x + af;
                        ab.ao = ac;
                    }
                }
            }
        }
        for (let ah of fromEnum(a)) {
            let ai = ah.ao;
            if (ai.left > b || ai.right < 0) {
                ah._visibility = 1;
                if (this.ap == 0) {
                    ah.ap._visibility = 1;
                }
                else {
                    ah.aq._visibility = 1;
                }
            }
            else if (ai.left < 0) {
                let aj = Math.abs(ai.x);
                ai.x = 0;
                if (aj > ai.width) {
                    ai.width = 0;
                }
                else {
                }
                ah.ao = ai;
            }
            else if (ai.right > b) {
                let ak = ai.right - b;
                if (ak > ai.width) {
                    ai.width = 0;
                }
                else {
                    ai.width = ai.width - ak;
                }
                ah.ao = ai;
            }
        }
    }
    ei(a, b, c, d) {
        this.fc();
    }
    fc() {
        this.e5();
        this.e7();
        this.e6();
        this.ff();
        this.fd();
        this.fe();
        this.as.bp();
    }
    e5() {
        this.cu = 0;
        this.cp = 0;
        this.b2.clear();
        this.b1.clear();
        this.b0.clear();
        if (this.itemsSource == null || this.a5 == null) {
            return;
        }
        if (this.be == null || this.be.count == 0) {
            return;
        }
        for (let a of fromEnum(this.be)) {
            if (isNaN_(a) || isInfinity(a) || a <= 0) {
                continue;
            }
            this.cu = this.cu + a;
        }
        for (let b = 0; b < this.be.count; b++) {
            let c = this.be.item(b);
            let d = c <= 0;
            if (isNaN_(c) || isInfinity(c) || d) {
                continue;
            }
            let e = this.aq == 1 ? c / this.cu : c;
            let f = this.aq == 1 ? this.co / 100 : this.co;
            let g = true;
            if (e <= f && g) {
                this.cp = this.cp + c;
                this.b1.add(b);
                this.b0.add1(this.a5.item(b));
            }
            else {
                this.b2.add(b);
            }
        }
    }
    e7() {
        if (this.itemsSource == null || this.a5 == null) {
            this.bg.count = 0;
            return;
        }
        let a = this.b2.count;
        let b = this.b1.count > 0;
        let c = PieChartBase.cs(this.cf);
        let d = PieChartBase.cs(this.cf);
        if (b) {
            a++;
        }
        this.cx = a;
        for (let e = 0; e < a; e++) {
            let f = false;
            let g;
            if (e == a - 1 && b) {
                g = this.cp;
                f = true;
            }
            else {
                g = this.be.item(this.b2._inner[e]);
            }
            if (this.gx == 1) {
                d += PieChartBase.cs(Math.abs(g) * 360 / this.cu);
            }
            else {
                d -= PieChartBase.cs(Math.abs(g) * 360 / this.cu);
            }
            let h = this.bg.item(e);
            h.aw = true;
            h.a3 = c;
            h.ax = d;
            h.a1 = h.a0 = this.cc;
            h.isOthersSlice = f;
            h.ay = this.cb;
            h.a5 = e;
            h.dataContext = f ? this.b0 : this.a5.item(this.b2._inner[e]);
            this.av.j(h);
            h.isExploded = this.bi ? this.explodedSlices.contains(e) : false;
            c = d;
            h.aw = false;
        }
        this.bg.count = a;
    }
    e6() {
        if (this.itemsSource == null || this.a5 == null) {
            this.bf.count = 0;
            return;
        }
        if (this.ba == null || this.ba.count == 0 || this.ao == 0) {
            this.bf.count = 0;
        }
        let a = this.b2.count;
        if (this.b1.count > 0) {
            a++;
        }
        this.as.au();
        let b = 0;
        for (let c = 0; c < a; c++) {
            let d = stringEmpty();
            let e = false;
            let f;
            let g = null;
            if (c == a - 1 && this.b1.count > 0) {
                f = this.dz;
                e = true;
                if (this.bs()) {
                    let h = ((() => {
                        let $ret = new PieSliceDataContext();
                        $ret.series = this;
                        $ret.item = e ? this.b0 : this.a5.item(this.b2._inner[c]);
                        $ret.percentValue = this.ck(this.bg.item(c));
                        $ret.isOthersSlice = e;
                        return $ret;
                    })());
                    if (f != null) {
                        h.itemLabel = f.toString();
                        h.legendLabel = f.toString();
                    }
                    let i = h;
                    i = i.flatten();
                    f = this.at(i);
                }
                if (null != f) {
                    d = f.toString();
                }
                else {
                    d = this.dz;
                }
                d = this.dz;
            }
            else {
                if (this.ba != null && this.ba.count >= this.b2.count) {
                    f = this.ba.item(this.b2._inner[c]);
                }
                else {
                    f = null;
                }
                g = (this.bb != null && this.bb.count > 0) ? this.bb.item(this.b2._inner[c]) : null;
                if (this.bs()) {
                    let j = ((() => {
                        let $ret = new PieSliceDataContext();
                        $ret.series = this;
                        $ret.item = e ? this.b0 : this.a5.item(this.b2._inner[c]);
                        $ret.itemLabel = f != null ? f.toString() : null;
                        $ret.legendLabel = g != null ? g.toString() : f != null ? f.toString() : null;
                        $ret.percentValue = this.ck(this.bg.item(c));
                        $ret.isOthersSlice = e;
                        return $ret;
                    })());
                    let k = j;
                    k = k.flatten();
                    let formatedLabel_ = this.at(k);
                    if (null == formatedLabel_ && this.ba != null && this.ba.count > 0) {
                        f = this.ba.item(this.b2._inner[c]);
                    }
                    else {
                        f = formatedLabel_;
                    }
                }
                let label_ = f;
                let errorMsg_ = this.as.af;
                if (label_ === undefined) {
                    throw new Error(errorMsg_);
                }
                ;
                if (null != f) {
                    d = f.toString();
                }
                else {
                    d = null;
                }
            }
            if (this.ba != null && this.ba.count > 0) {
                let l = this.bf.item(c);
                let m = this.bg.item(c);
                m.aj = l;
                l.ai = GeometryUtil.g((m.a3 + m.ax) / 2);
                l.ah = m;
                l.aj = ((() => {
                    let $ret = new TextBlock();
                    $ret.ak = d;
                    return $ret;
                })());
                if (this.fs == null) {
                    l.dataContext = e ? this.b0 : this.a5.item(this.b2._inner[c]);
                    l.ak(this.as);
                }
                else {
                }
                l._visibility = 0;
                l.ao = this.as.bz(l);
                b++;
            }
            if (this.bb != null && this.bb.count > 0) {
                let n = this.bg.item(c);
                n.bf = g != null ? g.toString() : null;
            }
        }
        this.bf.count = b;
    }
    ff() {
        if (this.itemsSource == null || this.a5 == null) {
            return;
        }
        let a = this.as.b1();
        this.g3 = new Rect(0, 0, 0, a.width, a.height);
        let b = this.as.aa();
        let c = this.as.z(this.g6);
        b = isNaN_(b) ? 0 : b / 2;
        c = isNaN_(c) ? 0 : c / 2;
        b = Math.max(c, b);
        let d = Math.min(a.width / 2, a.height / 2);
        let e = { $type: Point_$type, x: a.width / 2, y: a.height / 2 };
        let f = d * this.ce;
        let g = f + b;
        if (g > d) {
            let h = g - d;
            f = f - h;
        }
        for (let i of fromEnum(this.bg.a)) {
            let j = GeometryUtil.j(a.width, a.height, true, (i.a3 + i.ax) / 2, f * this.cb);
            i.aw = true;
            i.a1 = i.a0 = this.cc;
            i.a2 = f;
            i.ay = this.cb;
            i.b7 = e;
            i.b5 = j;
            this.fj(i);
            i.aw = false;
        }
        this.as.bp();
    }
    fd() {
        let a = this.as.b0.width;
        let b = this.as.b0.height;
        if (b == 0 || a == 0) {
            return;
        }
        if (this.bf.a.count == 0 || this.ao == 0) {
            this.bf.count = 0;
            return;
        }
        let c = new List$1(PieLabel.$, 0);
        let d = new List$1(PieLabel.$, 0);
        let e = new List$1(PieLabel.$, 0);
        let f = new List$1(PieLabel.$, 0);
        for (let g of fromEnum(this.bf.a)) {
            let h = g.ah;
            if (h == null) {
                continue;
            }
            let i = h.b6();
            let j = g.ao.width;
            let k = g.ao.height;
            if (this.ao == 1 || this.ao == 4) {
                let l = h.a2 * h.a1 / 100;
                let m = GeometryUtil.k(i, g.ai, h.a2 - (h.a2 - l) / 2);
                g.ao = new Rect(0, m.x - j / 2, m.y - k / 2, j, k);
                let n = this.br(g, m);
                if (n || this.ao == 1) {
                    if (this.gv == null) {
                        h.f(Slice.bp);
                    }
                    else {
                        h.b2 = this.gv;
                    }
                    h.au = true;
                    e.add(g);
                    g.ag = 1;
                    if (g.ah.ai.ap == 0) {
                        g.ap._visibility = 1;
                    }
                    else {
                        g.aq._visibility = 1;
                    }
                    if (!n && this.ao == 1) {
                        g._visibility = 1;
                    }
                    else {
                        g._visibility = 0;
                    }
                    continue;
                }
            }
            if (this.ao == 2 || this.ao == 4) {
                let o = PieChartBase.ci(i, h.a2, g);
                let p = GeometryUtil.k(i, g.ai, o);
                g.ao = new Rect(0, p.x - j / 2, p.y - k / 2, j, k);
                let q = this.br(g, p);
                if (q || this.ao == 2) {
                    if (this.gv == null) {
                        h.f(Slice.bp);
                    }
                    else {
                        h.b2 = this.gv;
                    }
                    h.au = true;
                    f.add(g);
                    g.ag = 2;
                    if (g.ah.ai.ap == 0) {
                        g.ap._visibility = 1;
                    }
                    else {
                        g.aq._visibility = 1;
                    }
                    if (!q && this.ao == 2) {
                        g._visibility = 1;
                    }
                    else {
                        g._visibility = 0;
                    }
                    continue;
                }
            }
            if (this.gw == null) {
                h.f(Slice.bp);
            }
            else {
                h.b2 = this.gw;
            }
            h.au = false;
            let r = GeometryUtil.k(i, g.ai, h.a2 + this.cm);
            g.ag = 3;
            g._visibility = 0;
            g.al();
            if (g.ai < 90 && g.ai >= 0) {
                g.ao = new Rect(0, r.x, r.y, j, k);
                c.add(g);
            }
            else if (g.ai < 180 && g.ai >= 90) {
                g.ao = new Rect(0, r.x - j, r.y, j, k);
                d.add(g);
            }
            else if (g.ai < 270 && g.ai >= 180) {
                g.ao = new Rect(0, r.x - j, r.y - k, j, k);
                d.add(g);
            }
            else {
                g.ao = new Rect(0, r.x, r.y - k, j, k);
                c.add(g);
            }
            if (g.ao.y < 0) {
                g.ao = new Rect(0, g.ao.x, 0, g.ao.width, g.ao.height);
            }
            if (g.ao.bottom > b) {
                g.ao = new Rect(0, g.ao.x, b - g.ao.height, g.ao.width, g.ao.height);
            }
        }
        PieChartBase.fo(c);
        this.fg(c);
        PieChartBase.fo(d);
        this.fg(d);
        for (let s of fromEnum(e)) {
            this.as.bk(s, s.ao.x, s.ao.y);
        }
        for (let t of fromEnum(f)) {
            this.as.bk(t, t.ao.x, t.ao.y);
        }
        for (let u of fromEnum(c)) {
            this.as.bk(u, u.ao.x, u.ao.y);
            if (u.ah.ai.ap == 0) {
                u.am();
            }
            else {
                u.an();
            }
        }
        for (let v of fromEnum(d)) {
            this.as.bk(v, v.ao.x, v.ao.y);
            if (v.ah.ai.ap == 0) {
                v.am();
            }
            else {
                v.an();
            }
        }
        this.as.bp();
    }
    static ci(a, b, c) {
        let d = c.ai * Math.PI / 180;
        let e = 1;
        let f = GeometryUtil.k({ $type: Point_$type, x: 0, y: 0 }, c.ai, e);
        let g = PieChartBase.gz(c);
        let h = Math.sqrt(g.x * g.x + g.y * g.y);
        let i = { $type: Point_$type, x: g.x / h, y: g.y / h };
        let j = PieChartBase.cj(f, i);
        let k = h / b;
        let l = j * k;
        let m = l;
        let n = GeometryUtil.e({ $type: Point_$type, x: 0, y: 0 }, f);
        let o = GeometryUtil.e({ $type: Point_$type, x: 0, y: 0 }, i);
        if (o < n) {
            m = -m;
        }
        let p = GeometryUtil.k(a, (d + m) * 180 / Math.PI, b);
        let q = PieChartBase.gy(c, p);
        let r = Math.sqrt((q.x - a.x) * (q.x - a.x) + (q.y - a.y) * (q.y - a.y));
        let s = r - 5;
        return s;
    }
    static gy(a, b) {
        let c = GeometryUtil.g(a.ai);
        let d = a.ao.width / 2;
        let e = a.ao.height / 2;
        if (c >= 0 && c <= 90) {
            return { $type: Point_$type, x: b.x - d, y: b.y - e };
        }
        if (c >= 90 && c <= 180) {
            return { $type: Point_$type, x: b.x + d, y: b.y - e };
        }
        if (c >= 180 && c <= 270) {
            return { $type: Point_$type, x: b.x + d, y: b.y + e };
        }
        return { $type: Point_$type, x: b.x - d, y: b.y + e };
    }
    static cj(a, b) {
        let c = a.x * b.x + a.y * b.y;
        let d = Math.sqrt(a.x * a.x + a.y * a.y);
        let e = Math.sqrt(b.x * b.x + b.y * b.y);
        let f = Math.acos(c / (d * e));
        return f;
    }
    static gz(a) {
        let b = GeometryUtil.g(a.ai);
        let c = a.ao.width / 2;
        let d = a.ao.height / 2;
        if (b >= 0 && b <= 90) {
            return { $type: Point_$type, x: c, y: d };
        }
        if (b >= 90 && b <= 180) {
            return { $type: Point_$type, x: -c, y: d };
        }
        if (b >= 180 && b <= 270) {
            return { $type: Point_$type, x: -c, y: -d };
        }
        return { $type: Point_$type, x: c, y: -d };
    }
    fe() {
        if (this.legend == null || !this.legend.isItemwise) {
            return;
        }
        if ((this.ba == null || this.ba.count == 0) && (this.bb == null || this.bb.count == 0)) {
            this.legend.clearLegendItemsForSeries(this);
            return;
        }
        this.bz = new List$1(UIElement.$, 0);
        for (let a of fromEnum(this.bg.a)) {
            let b = new ContentControl();
            let c = this.cy(a);
            let d = this.as.br(a);
            let e = this.as.bs(a);
            b.content = ((() => {
                let $ret = new PieSliceDataContext();
                $ret.series = this;
                $ret.slice = a;
                $ret.item = a.dataContext;
                $ret.itemBrush = d;
                $ret.outline = e;
                $ret.thickness = a.a4;
                $ret.itemLabel = c != null ? c.toString() : null;
                $ret.legendLabel = a.isOthersSlice ? this.dz : a.bf != null ? a.bf : (c != null ? c.toString() : null);
                $ret.percentValue = this.ck(a);
                $ret.isOthersSlice = a.isOthersSlice;
                return $ret;
            })());
            let f = b.content;
            if (this.bt()) {
                let g = this.au(b.content);
                if (g != null) {
                    f.legendLabel = g.toString();
                }
            }
            if (this.legendItemTemplate != null) {
                b.ah = this.legendItemTemplate;
            }
            this.bz.add(b);
        }
        if (this.legend.isItemwise) {
            let h = this.legend;
            h.createItemwiseLegendItems(this.bz, this);
        }
    }
    a4() {
        let a = new PieChartVisualData();
        a.viewport = RectData.b(this.g3);
        this.as.an(a);
        for (let b = 0; b < this.bg.count; b++) {
            let c = this.bg.item(b).am();
            a.slices.add(c);
            if (this.bg.item(b).isOthersSlice) {
                a.othersSlice = this.bg.item(b).am();
            }
        }
        a.others = this.b0;
        a.name = this.name;
        a.radiusFactor = this.ce;
        a.leaderLineMargin = this.cn;
        return a;
    }
    c8() {
        let a = this.a4();
        a.scaleByViewport();
        return a.serialize();
    }
    bc(a) {
        let b = null;
        return this.a5.registerColumn(a, b, false);
    }
    bd(a) {
        let b = null;
        return this.a5.registerColumnObject(a, b, false);
    }
    e8(a, b, c, d) {
        switch (b) {
            case PieChartBase.$$p[8]:
                this.a5 = ((() => {
                    let $ret = new FastItemsSource();
                    $ret.e = d;
                    return $ret;
                })());
                break;
            case PieChartBase.$$p[4]:
                if (typeCast(FastItemsSource.$, c) != null) {
                    let e = c;
                    e.event = delegateRemove(e.event, this.cv);
                    this.bj = false;
                    c.deregisterColumn(this.be);
                    c.deregisterColumn(this.ba);
                    this.be = null;
                    this.ba = null;
                    this.selectedItem = null;
                    this.selectedItems.clear();
                    if (this.selectedSlices != null) {
                        this.selectedSlices.clear();
                    }
                    (typeCast(FastItemsSource.$, c)).deregisterColumn(this.bb);
                    this.bb = null;
                }
                if (typeCast(FastItemsSource.$, d) != null) {
                    let f = d;
                    f.event = delegateCombine(f.event, this.cv);
                    this.bj = true;
                    this.be = this.bc(this.eg);
                    this.ba = this.bd(this.di);
                    this.bb = this.bd(this.dv);
                    this.e7();
                    if (this.selectedItem != null || this.selectedItems.count > 0) {
                        this.fq();
                    }
                    else if (this.selectedSlices != null && this.selectedSlices.count > 0) {
                        this.fr();
                    }
                }
                this.fc();
                break;
            case PieChartBase.$$p[34]:
                if (this.a5 != null) {
                    this.a5.deregisterColumn(this.be);
                    this.be = this.bc(this.eg);
                }
                break;
            case PieChartBase.$$p[20]:
                if (this.a5 != null) {
                    this.a5.deregisterColumn(this.bb);
                    this.bb = this.bd(this.dv);
                    this.e6();
                    this.fd();
                    this.fe();
                }
                break;
            case PieChartBase.$$p[6]:
                this.fe();
                break;
            case PieChartBase.$$p[5]:
            case PieChartBase.$$p[11]:
                if (this.a5 != null) {
                    this.a5.deregisterColumn(this.ba);
                    this.ba = this.bd(this.di);
                    this.e6();
                    this.fd();
                    this.fe();
                }
                break;
            case PieChartBase.$$p[30]:
                this.cf = d;
                this.e7();
                this.e6();
                this.ff();
                this.fd();
                break;
            case PieChartBase.$$p[32]:
                this.as.bn(this.c0);
                break;
            case PieChartBase.$$p[21]:
                let g = typeCast(IChartLegend_$type, c);
                if (g != null) {
                    g.clearLegendItemsForSeries(this);
                }
                this.fe();
                break;
            case PieChartBase.$$p[18]:
            case PieChartBase.$$p[19]:
                this.fe();
                break;
            case PieChartBase.$$p[27]:
            case PieChartBase.$$p[3]:
            case PieChartBase.$$p[31]:
                this.e7();
                this.e6();
                this.ff();
                this.fd();
                break;
            case PieChartBase.$$p[22]:
            case PieChartBase.$$p[28]:
                this.ff();
                this.fe();
                break;
            case PieChartBase.$$p[2]:
                if (c != null) {
                    let h = c;
                    h.collectionChanged = delegateRemove(h.collectionChanged, this.b5);
                    this.bh = false;
                }
                if (d != null) {
                    let i = d;
                    i.collectionChanged = delegateCombine(i.collectionChanged, this.b5);
                    this.bh = true;
                }
                this.ff();
                this.fe();
                break;
            case PieChartBase.$$p[26]:
                if (c != null) {
                    let j = c;
                    j.collectionChanged = delegateRemove(j.collectionChanged, this.b5);
                    this.bk = false;
                }
                if (d != null) {
                    let k = d;
                    k.collectionChanged = delegateCombine(k.collectionChanged, this.b5);
                    this.bk = true;
                }
                this.ff();
                this.fe();
                break;
            case "ValueColumn":
            case PieChartBase.$$p[24]:
            case PieChartBase.$$p[25]:
                this.fc();
                this.fq();
                this.fr();
                break;
            case PieChartBase.$$p[0]:
            case PieChartBase.$$p[1]:
                this.fc();
                break;
            case PieChartBase.$$p[13]:
                this.e6();
                this.fd();
                this.as.bp();
                break;
            case PieChartBase.$$p[10]:
            case PieChartBase.$$p[12]:
            case PieChartBase.$$p[9]:
                this.e6();
                this.fd();
                break;
            case PieChartBase.$$p[23]:
                this.e6();
                this.fd();
                this.fe();
                break;
            case PieChartBase.$$p[17]:
                this.fd();
                break;
            case PieChartBase.$$p[15]:
                this.fd();
                break;
            case "TextStyle":
                this.as.bf();
                break;
            case "PixelScalingRatio":
                if (isNaN_(DeviceUtils.a(this.cq))) {
                    this.cd = 1;
                }
                else {
                    this.cd = DeviceUtils.a(this.cq);
                }
                break;
            case "ActualPixelScalingRatio":
                this.as.a1();
                break;
            case "LabelTemplate":
                this.e6();
                this.fd();
                break;
            case PieChartBase.$$p[16]:
            case PieChartBase.$$p[14]:
                this.bf.c(runOn(this.as, this.as.bj));
                this.fd();
                break;
            case PieChartBase.$$p[7]:
                this.as.a0();
                break;
            case "InnerExtent":
                this.fc();
                break;
            case "SelectedItem":
                if (!this.bn) {
                    this.e7();
                    this.ff();
                    this.fe();
                }
                break;
            case PieChartBase.$$p[29]:
                if (EnumUtil.getEnumValue(SliceSelectionMode_$type, d) == 1 && this.selectedItems != null && this.selectedItems.count > 1) {
                    this.selectedItems.clear();
                }
                else if (EnumUtil.getEnumValue(SliceSelectionMode_$type, d) == 2 && this.selectedItem != null && this.selectedItems != null) {
                    if (!this.selectedItems.contains1(this.selectedItem)) {
                        if (typeCast(PieSliceOthersContext.$, this.selectedItem) !== null) {
                            let l = typeCast(PieSliceOthersContext.$, this.selectedItem);
                            for (let m of fromEn(l.data)) {
                                this.selectedItems.add1(m);
                            }
                        }
                        else {
                            this.selectedItems.add1(this.selectedItem);
                        }
                        this.fr();
                    }
                }
                this.fq();
                this.e7();
                this.ff();
                break;
        }
    }
    fh(a) {
        if (this.bo) {
            return;
        }
        this.fr();
        if (this.bq) {
            this.e7();
            this.ff();
            this.fe();
        }
    }
    e0(a) {
        if (a.action == 4 && a.newItems == null && a.oldItems == null) {
            return;
        }
        if (this.bn) {
            return;
        }
        this.fq();
        if (this.bq) {
            this.e7();
            this.ff();
            this.fe();
        }
        this.ez(this, new SelectedItemsChangedEventArgs(a.oldItems, a.newItems));
    }
    fa(a, b, c) {
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
        if (this.propertyUpdated != null) {
            this.propertyUpdated(this, new PropertyUpdatedEventArgs(a, b, c));
        }
    }
    fr() {
        if (this.itemsSource == null || this.bg.a.count == 0) {
            return;
        }
        this.bn = true;
        let a = this.selectedSlices.count;
        if (a == 0) {
            if (this.selectedItems.count > 0) {
                this.selectedItems.clear();
            }
            if (this.selectedItem != null) {
                this.selectedItem = null;
            }
        }
        if (this.a3 == 1) {
            if (a != 0) {
                let b = this.bg.a._inner[this.selectedSlices._inner[a - 1]];
                if (b.isOthersSlice) {
                    this.selectedItem = ((() => {
                        let $ret = new PieSliceOthersContext();
                        $ret.data = typeCast(IList_$type, b.dataContext);
                        $ret.slice = b;
                        return $ret;
                    })());
                }
                else {
                    this.selectedItem = b.dataContext;
                }
            }
        }
        else if (this.a3 == 2) {
            if (a != 0) {
                this.selectedItems.clear();
                for (let c of fromEnum(this.selectedSlices)) {
                    let d = this.bg.a._inner[c];
                    if (d.isOthersSlice) {
                        let e = typeCast(IEnumerable_$type, d.dataContext);
                        for (let f of fromEn(e)) {
                            this.av.e(f);
                        }
                    }
                    else {
                        this.av.e(d.dataContext);
                    }
                }
                this.selectedItem = this.bg.a._inner[this.selectedSlices._inner[a - 1]].dataContext;
            }
        }
        this.bn = false;
    }
    fq() {
        if (this.itemsSource == null || this.bg.a.count == 0) {
            return;
        }
        this.bo = true;
        this.selectedSlices.clear();
        for (let a of fromEnum(this.bg.a)) {
            if (this.av.d(a)) {
                this.selectedSlices.add(a.a5);
            }
        }
        this.bo = false;
    }
    fm(a, b) {
        this.fn(a, b, false);
    }
    fn(a, b, c) {
        if (this.bq) {
            if (this.a3 == 1) {
                this.av.k(a);
            }
            else if (this.a3 == 2) {
                this.av.i(a);
            }
        }
        let d = new SliceClickEventArgs(a, this.as.ae(b));
        this.e3(this, d);
        if (!c) {
            this.as.bl(a, b);
        }
    }
    fl(a) {
        this.as.ak(a);
    }
    bw(a, b) {
        let c = new LabelClickEventArgs(a, this.as.ae(b));
        this.ew(this, c);
        return c.allowSliceClick;
    }
    em(a, b) {
        this.as.bl(a, b);
    }
    ep(a, b, c) {
        this.as.bl(a, b);
    }
    eo(a, b) {
        this.ep(a, b, false);
    }
    en(a, b) {
        this.as.al();
    }
    e2() {
        this.fc();
    }
    provideContainer(a) {
        this.as.ax(a);
    }
    ev(a, b) {
        this.as.ay(a, b);
    }
    notifyContainerResized() {
        this.as.aw();
    }
    g1() {
        return this.as.by();
    }
    g0() {
        return this.as.bv();
    }
    destroy() {
        this.fb();
        this.as.ax(null);
    }
    el() {
        this.as.ao();
    }
    get eb() {
        return this.c1;
    }
    set eb(a) {
        let b = this.c1;
        this.c1 = a;
        this.fa("TextStyle", b, a);
    }
    get container() {
        return this;
    }
    get isStacked() {
        return false;
    }
    get isVertical() {
        return false;
    }
    get hasMarkers() {
        return false;
    }
    get isUsableInLegend() {
        return true;
    }
    get hasSubItems() {
        return false;
    }
    ck(a) {
        if (a == null || this.be == null || this.b2 == null) {
            return NaN;
        }
        if (a.isOthersSlice) {
            return this.cp / this.cu * 100;
        }
        else {
            return this.be.item(this.b2._inner[a.a5]) / this.cu * 100;
        }
    }
    eu(source_, a, b, c) {
        if (source_.dataView && source_.dataSource) {
            source_ = source_.dataView();
        }
        ;
        if (!this.bx(source_)) {
            return;
        }
        let d = this.a5;
        if (d == null) {
            return;
        }
        d.handleCollectionChanged(new NotifyCollectionChangedEventArgs(2, 2, c, b, a));
    }
    eq(source_) {
        if (source_.dataView && source_.dataSource) {
            source_ = source_.dataView();
        }
        ;
        if (!this.bx(source_)) {
            return;
        }
        let a = this.a5;
        a.handleCollectionChanged(new NotifyCollectionChangedEventArgs(0, 4));
    }
    bx(a) {
        return PieChartBase.by(a, this.itemsSource);
    }
    static by(a, b) {
        return a == b;
    }
    es(source_, a, b) {
        if (source_.dataView && source_.dataSource) {
            source_ = source_.dataView();
        }
        ;
        if (!this.bx(source_)) {
            return;
        }
        let c = this.a5;
        if (c == null) {
            return;
        }
        c.handleCollectionChanged(new NotifyCollectionChangedEventArgs(1, 0, b, a));
    }
    et(source_, a, b) {
        if (source_.dataView && source_.dataSource) {
            source_ = source_.dataView();
        }
        ;
        if (!this.bx(source_)) {
            return;
        }
        let c = this.a5;
        if (c == null) {
            return;
        }
        c.handleCollectionChanged(new NotifyCollectionChangedEventArgs(1, 1, b, a));
    }
    fk(source_) {
        if (source_.dataView && source_.dataSource) {
            source_ = source_.dataView();
        }
        ;
        this.itemsSource = source_;
    }
    fb() {
        this.itemsSource = null;
    }
    fp() {
        this.as.bf();
    }
    a6() {
        return FontUtil.toFontInfo(this.as.c.o, this.eb);
    }
    resolveLegendIndex() {
        return -1;
    }
    forSubItems(a) {
    }
}
PieChartBase.$t = markType(PieChartBase, 'PieChartBase', Control.$, [INotifyPropertyChanged_$type, ILegendOwner_$type, ILegendSeries_$type, ILegendTemplateProvider_$type]);
PieChartBase.f0 = DependencyProperty.i("InnerExtent", Number_$type, PieChartBase.$, new PropertyMetadata(2, 0, (a, b) => a.fa("InnerExtent", b.oldValue, b.newValue)));
PieChartBase.f8 = DependencyProperty.i("LabelTemplate", DataTemplate.$, PieChartBase.$, new PropertyMetadata(2, null, (a, b) => a.fa("LabelTemplate", b.oldValue, b.newValue)));
PieChartBase.gn = DependencyProperty.i("SelectedItem", Base.$, PieChartBase.$, new PropertyMetadata(1, (a, b) => {
    let c = typeCast(PieChartBase.$, a);
    if (c.bm) {
        return;
    }
    let d = new SelectedItemChangingEventArgs(b.oldValue, b.newValue);
    c.ey(c, d);
    if (!d.cancel) {
        if (!c.bn) {
            c.fq();
        }
        c.fa("SelectedItem", b.oldValue, b.newValue);
        c.ex(a, new SelectedItemChangedEventArgs(b.oldValue, b.newValue));
    }
    else {
        c.bm = true;
        c.selectedItem = b.oldValue;
        c.bm = false;
    }
}));
Type.dep(DependencyProperty, PropertyMetadata, PieChartBase, 'fa', ['AllowSliceExplosion:bp:ft', [0, true], 'AllowSliceSelection:bq:fu', [0, true], 'Brushes:ag:fv', [BrushCollection.$, null], 'ExplodedRadius:cg:fw', [1, 0.2], 'FastItemsSource:a5:fx', [FastItemsSource.$], 'FormatLabel:at:fy', [Delegate_$type], 'FormatLegendLabel:au:fz', [Delegate_$type], 'IsSurfaceInteractionDisabled:bv:f1', [0, false], 'ItemsSource::f2', [IEnumerable_$type, null], 'LabelExtent:cm:f3', [1, 10], 'LabelInnerColor:gv:f4', [Brush.$, null], 'LabelMemberPath:di:f5', [2], 'LabelOuterColor:gw:f6', [Brush.$, null], 'LabelsPosition:ao:f7', [LabelsPosition_$type, enumGetBox(LabelsPosition_$type, 1)], 'LeaderLineMargin:cn:f9', [1, 6], 'LeaderLineStyle:g4:ga', [Style.$, null], 'LeaderLineType:ap:gb', [LeaderLineType_$type, enumGetBox(LeaderLineType_$type, 0)], 'LeaderLineVisibility:g8:gc', [Visibility_$type, enumGetBox(Visibility_$type, 0)], 'LegendItemBadgeTemplate::gd', [DataTemplate.$, null], 'LegendItemTemplate::ge', [DataTemplate.$, null], 'LegendLabelMemberPath:dv:gf', [2], 'Legend::gg', [IChartLegend_$type, null], 'OthersCategoryStyle:g5:gh', [Style.$, null], 'OthersCategoryText:dz:gi', [2, "Others"], 'OthersCategoryThreshold:co:gj', [1, 3], 'OthersCategoryType:aq:gk', [OthersCategoryType_$type, enumGetBox(OthersCategoryType_$type, 1)], 'Outlines:ah:gl', [BrushCollection.$, null], 'RadiusFactor:cr:gm', [1, 0.9], 'SelectedStyle:g6:go', [Style.$, null], 'SelectionMode:a3:gp', [SliceSelectionMode_$type, enumGetBox(SliceSelectionMode_$type, 1)], 'StartAngle:ct:gq', [1, 0], 'SweepDirection:gx:gr', [SweepDirection_$type, enumGetBox(SweepDirection_$type, 1)], 'ToolTip:c0:gs', [4, null], 'ToolTipStyle:g7:gt', [Style.$, null], 'ValueMemberPath:eg:gu', [2, null]]);
/**
 * @hidden
 */
export class PieChartSliceSelector extends Base {
    constructor(a) {
        super();
        this.a = null;
        this.a = a;
    }
    d(a) {
        let b = a.dataContext;
        if (this.a.a3 == 1) {
            if (a.isOthersSlice) {
                let c = typeCast(IList_$type, b);
                if (c != null && c.contains(this.a.selectedItem)) {
                    return true;
                }
                if (typeCast(PieSliceOthersContext.$, this.a.selectedItem) !== null) {
                    return true;
                }
            }
            if (b == this.a.selectedItem) {
                return true;
            }
        }
        else if (this.a.a3 == 2) {
            if (this.c(a)) {
                return true;
            }
        }
        else if (this.a.a3 == 0) {
            if (this.a.selectedSlices.contains(a.a5)) {
                return true;
            }
        }
        return false;
    }
    c(a) {
        let b = a.dataContext;
        if (a.isOthersSlice) {
            for (let c of fromEn(b)) {
                if (this.a.selectedItems.contains1(c)) {
                    return true;
                }
            }
        }
        else {
            if (this.a.selectedItems.contains1(b)) {
                return true;
            }
        }
        return false;
    }
    h(a) {
        let b = a.dataContext;
        if (!a.isOthersSlice) {
            if (this.a.selectedItem != b) {
                return b;
            }
            else {
                return null;
            }
        }
        else {
            if (typeCast(PieSliceOthersContext.$, this.a.selectedItem) !== null) {
                return null;
            }
            else {
                return ((() => {
                    let $ret = new PieSliceOthersContext();
                    $ret.data = typeCast(IList_$type, a.dataContext);
                    $ret.slice = a;
                    return $ret;
                })());
            }
        }
    }
    f(a, b) {
        let c = false;
        if (b) {
            if (this.a.selectedItems.contains1(a)) {
                c = this.g(a);
            }
        }
        else {
            if (!this.a.selectedItems.contains1(a)) {
                c = this.b(a);
            }
            else {
                c = this.g(a);
            }
        }
        return c;
    }
    e(a) {
        return this.f(a, false);
    }
    g(a) {
        let b = new SelectedItemsChangingEventArgs(this.a.selectedItems, ((() => {
            let $ret = new List$1(Base.$, 0);
            return $ret;
        })()));
        this.a.e1(this, b);
        if (!b.cancel) {
            this.a.selectedItems.remove1(a);
        }
        return b.cancel;
    }
    b(a) {
        let b = new SelectedItemsChangingEventArgs(this.a.selectedItems, ((() => {
            let $ret = new List$1(Base.$, 0);
            $ret.add1(a);
            return $ret;
        })()));
        this.a.e1(this, b);
        if (!b.cancel) {
            this.a.selectedItems.add1(a);
        }
        return b.cancel;
    }
    k(a) {
        this.a.selectedItem = this.h(a);
    }
    i(a) {
        let b = new List$1(Base.$, 1, this.a.selectedItems);
        if (a.isOthersSlice) {
            let c = typeCast(IList_$type, a.dataContext);
            let d = 0;
            for (let e of fromEn(c)) {
                if (this.a.selectedItems.contains1(e)) {
                    d++;
                }
            }
            let f = d > 0 && d < c.count;
            for (let g of fromEn(c)) {
                this.f(g, f);
            }
            if (this.c(a)) {
                this.a.selectedItem = ((() => {
                    let $ret = new PieSliceOthersContext();
                    $ret.data = typeCast(IList_$type, a.dataContext);
                    $ret.slice = a;
                    return $ret;
                })());
            }
            else {
                this.a.selectedItem = this.a.selectedItems.count > 0 ? this.a.selectedItems._inner[this.a.selectedItems.count - 1] : null;
            }
        }
        else {
            this.e(a.dataContext);
            this.a.selectedItem = this.a.selectedItems.count > 0 ? this.a.selectedItems._inner[this.a.selectedItems.count - 1] : null;
        }
    }
    j(a) {
        if (a != null) {
            a.av = true;
            a.isSelected = this.d(a);
            a.av = false;
        }
    }
}
PieChartSliceSelector.$t = markType(PieChartSliceSelector, 'PieChartSliceSelector');
/**
 * @hidden
 */
export class PieSliceOthersContext extends Base {
    constructor() {
        super(...arguments);
        this._data = null;
        this._slice = null;
    }
    get data() {
        return this._data;
    }
    set data(a) {
        this._data = a;
    }
    get slice() {
        return this._slice;
    }
    set slice(a) {
        this._slice = a;
    }
}
PieSliceOthersContext.$t = markType(PieSliceOthersContext, 'PieSliceOthersContext');
/**
 * @hidden
 */
export class Slice extends ContentControl {
    constructor() {
        super();
        this.al = null;
        this.b8 = null;
        this.ai = null;
        this.aj = null;
        this.bf = null;
        this.b9 = null;
        this.b3 = null;
        this.aq = false;
        this.au = false;
        this.av = false;
        this.an = false;
        this.b0 = null;
        this.b1 = null;
        let a = this.ak();
        this.bl(a);
        a.b();
        this.av = false;
    }
    ak() {
        return new SliceView(this);
    }
    bl(a) {
        this.al = a;
    }
    get isSelected() {
        return this.c(Slice.bv);
    }
    set isSelected(a) {
        this.h(Slice.bv, a);
    }
    get isExploded() {
        return this.c(Slice.bt);
    }
    set isExploded(a) {
        this.h(Slice.bt, a);
    }
    get isOthersSlice() {
        return this.c(Slice.bu);
    }
    set isOthersSlice(a) {
        this.h(Slice.bu, a);
    }
    get b2() {
        return this.c(Slice.bp);
    }
    set b2(a) {
        this.h(Slice.bp, a);
    }
    ac() {
        super.ac();
        this.al.c();
    }
    az() {
        let a = Math.abs(PieChartBase.cs(this.ax - this.a3)) == 360;
        if (a) {
            return this.a2;
        }
        return this.a2;
    }
    cb() {
        let a = (this.ai != null && this.ai.bp) ? true : false;
        let b = this.az();
        if (this.isExploded && a && this.ai.cx > 1) {
            let c = new Rect(0, this.b5.x - b, this.b5.y - b, b * 2, b * 2);
            return c;
        }
        return new Rect(0, this.b7.x - b, this.b7.y - b, b * 2, b * 2);
    }
    b6() {
        if (this.isExploded && this.ai != null && this.ai.bp && this.ai.cx > 1) {
            if (this.aq) {
                return this.b3;
            }
            return this.b5;
        }
        return this.b7;
    }
    get aw() {
        return this.an;
    }
    set aw(a) {
        let b = this.an;
        this.an = a;
        if (b && !this.an) {
            this.bk();
        }
    }
    bk() {
        if (this.aw) {
            return;
        }
        let a = this.al.f();
        let b = this.ai.g3;
        if (isNaN_(this.a3) || isNaN_(this.ax) || isNaN_(this.a2) || isNaN_(this.ay) || this.a5 < 0 || this.ai == null || b.width == 0 || b.height == 0 || (this.b7.x == 0 && this.b7.y == 0) || (this.b5.x == 0 && this.b5.y == 0) || this.a3 == this.ax) {
            return;
        }
        this.b8 = this.cb();
        let c = GeometryUtil.d(this.b8);
        let d = this.b8.height / 2;
        let e = this.az();
        let f = RectUtil.l(this.b8);
        let g = Slice.b4(MathUtil.k(this.a3), c, f, d, 100);
        let h = Slice.b4(MathUtil.k(this.ax), c, f, d, 100);
        let i = Slice.b4(MathUtil.k(this.ax), c, f, this.a2, this.a0);
        let j = Slice.b4(MathUtil.k(this.a3), c, f, this.a2, this.a1);
        let k = Math.abs(PieChartBase.cs(this.ax - this.a3)) == 360;
        if (k) {
            if (this.a0 == 0) {
                let l = ((() => {
                    let $ret = new EllipseGeometry();
                    $ret.d = f;
                    $ret.b = this.a2;
                    $ret.c = this.a2;
                    return $ret;
                })());
                a.aj = l;
                return;
            }
            else if (this.a0 > 0) {
                let m = new Size(1, this.a2, this.a2);
                let n = new Size(1, this.a2 * this.a0 / 100, this.a2 * this.a0 / 100);
                let o = ((() => {
                    let $ret = PointUtil.create();
                    $ret.x = f.x - this.a2;
                    $ret.y = f.y;
                    return $ret;
                })());
                let p = ((() => {
                    let $ret = PointUtil.create();
                    $ret.x = f.x + this.a2;
                    $ret.y = f.y;
                    return $ret;
                })());
                let q = ((() => {
                    let $ret = PointUtil.create();
                    $ret.x = f.x - this.a2 * this.a0 / 100;
                    $ret.y = f.y;
                    return $ret;
                })());
                let r = ((() => {
                    let $ret = PointUtil.create();
                    $ret.x = f.x + this.a2 * this.a0 / 100;
                    $ret.y = f.y;
                    return $ret;
                })());
                let s = ((() => {
                    let $ret = new PathFigure();
                    $ret.startPoint = o;
                    return $ret;
                })());
                let t = new PathSegmentCollection();
                let u = ((() => {
                    let $ret = new ArcSegment();
                    $ret.f = m;
                    $ret.d = 1;
                    $ret.e = p;
                    return $ret;
                })());
                t.add(u);
                s._segments = t;
                let v = ((() => {
                    let $ret = new PathFigure();
                    $ret.startPoint = r;
                    return $ret;
                })());
                let w = new PathSegmentCollection();
                let x = ((() => {
                    let $ret = new ArcSegment();
                    $ret.f = n;
                    $ret.d = 0;
                    $ret.e = q;
                    return $ret;
                })());
                w.add(x);
                v._segments = w;
                let y = ((() => {
                    let $ret = new PathFigure();
                    $ret.startPoint = o;
                    return $ret;
                })());
                let z = new PathSegmentCollection();
                let aa = ((() => {
                    let $ret = new ArcSegment();
                    $ret.f = m;
                    $ret.d = 0;
                    $ret.e = p;
                    return $ret;
                })());
                z.add(aa);
                y._segments = z;
                let ab = ((() => {
                    let $ret = new PathFigure();
                    $ret.startPoint = r;
                    return $ret;
                })());
                let ac = new PathSegmentCollection();
                let ad = ((() => {
                    let $ret = new ArcSegment();
                    $ret.f = n;
                    $ret.d = 1;
                    $ret.e = q;
                    return $ret;
                })());
                ac.add(ad);
                ab._segments = ac;
                let ae = ((() => {
                    let $ret = new PathFigureCollection();
                    $ret.add(s);
                    $ret.add(v);
                    $ret.add(y);
                    $ret.add(ab);
                    return $ret;
                })());
                let af = ((() => {
                    let $ret = new PathGeometry();
                    $ret.b = ae;
                    return $ret;
                })());
                let ag = new GeometryGroup();
                ag.c.add(af);
                a.aj = ag;
                return;
            }
            else {
                let ah = ((() => {
                    let $ret = new GeometryGroup();
                    $ret.b = 0;
                    return $ret;
                })());
                ah.c.add(((() => {
                    let $ret = new EllipseGeometry();
                    $ret.d = f;
                    $ret.b = this.a2;
                    $ret.c = this.a2;
                    return $ret;
                })()));
                ah.c.add(((() => {
                    let $ret = new EllipseGeometry();
                    $ret.d = f;
                    $ret.b = this.a2 * this.a0 / 100;
                    $ret.c = this.a2 * this.a0 / 100;
                    return $ret;
                })()));
                a.aj = ah;
                return;
            }
        }
        let ai = new PathGeometry();
        a.aj = ai;
        let aj = ((() => {
            let $ret = new PathFigure();
            $ret.isClosed = true;
            return $ret;
        })());
        ai.b = new PathFigureCollection();
        ai.b.add(aj);
        aj._startPoint = g;
        let ak = Math.abs(this.ax - this.a3) > 180;
        let al = new ArcSegment();
        al.e = h;
        al.f = new Size(1, this.b8.width / 2, this.b8.height / 2);
        al.b = ak;
        if (this.ax > this.a3) {
            al.d = 1;
        }
        else {
            al.d = 0;
        }
        aj._segments = new PathSegmentCollection();
        aj._segments.add(al);
        let am = new LineSegment(1);
        am.c = i;
        aj._segments.add(am);
        let an = Math.sqrt(Math.pow(i.x - h.x, 2) + Math.pow(i.y - h.y, 2));
        let ao = e - an;
        let ap = new ArcSegment();
        ap.e = j;
        if (ao < 0) {
            ao = 0.1;
        }
        ap.f = new Size(1, ao, ao);
        ap.b = al.b;
        ap.d = 0;
        aj._segments.add(ap);
        let aq = this.ca(a, g, h, this.a3, this.ax, f, e);
        let ar = true;
        if (ar && this.b8.height > 0 && this.b8.width > 0 && !b.containsRect(aq)) {
            let as = aq;
            let at = new Rect(0, 0, 0, b.width, b.height);
            at.intersect(as);
            let au = GeometryUtil.g((this.a3 + this.ax) / 2);
            let av = au / 180 * Math.PI;
            let aw = Math.abs((as.height - at.height) / Math.sin(av));
            let ax = Math.abs((as.width - at.width) / Math.cos(av));
            if (isNaN_(aw) || isInfinity(aw)) {
                aw = 0;
            }
            if (isNaN_(ax) || isInfinity(ax)) {
                ax = 0;
            }
            let ay = Math.max(aw, ax);
            let az = GeometryUtil.j(this.ai.g3.width, this.ai.g3.height, true, au, this.a2 * this.ai.cb - ay);
            this.al.d(az.x - this.b5.x, az.y - this.b5.y);
            this.aq = true;
            this.b3 = az;
            this.b9 = new Rect(0, as.x - (az.x - this.b5.x), as.y - (az.y - this.b5.y), as.width, as.height);
        }
        else {
            this.aq = false;
            this.al.e();
        }
    }
    ap(a) {
        let b = this.ai.g3;
        if (this._visibility == 1) {
            return false;
        }
        if (isNaN_(this.a3) || isNaN_(this.ax) || isNaN_(this.a2) || isNaN_(this.ay) || this.a5 < 0 || this.ai == null || b.width == 0 || b.height == 0 || (this.b7.x == 0 && this.b7.y == 0) || (this.b5.x == 0 && this.b5.y == 0)) {
            return false;
        }
        let c = this.az();
        let d = ((this.a0 / 100) * c);
        let e = c;
        let f = this.b7;
        if (this.isExploded && this.ai.bp && this.ai.cx > 1) {
            f = this.b5;
            if (this.aq) {
                f = this.b3;
            }
        }
        let g = d * d;
        let h = e * e;
        let i = Math.pow(a.x - f.x, 2) + Math.pow(a.y - f.y, 2);
        if (i < g || i > h) {
            return false;
        }
        let j = Math.atan2(a.y - f.y, a.x - f.x);
        j = j * 180 / Math.PI;
        return this.ao(j);
    }
    ao(a) {
        a = GeometryUtil.g(a);
        let b = Math.min(this.a3, this.ax);
        let c = Math.max(this.a3, this.ax);
        return (a > b && a < c) || (a - 360 > b && a - 360 < c) || (a + 360 > b && a + 360 < c);
    }
    ca(a, b, c, d, e, f, g) {
        let h = new List$1(Point_$type, 0);
        h.add(b);
        h.add(c);
        h.add(f);
        d = GeometryUtil.g(d);
        e = GeometryUtil.g(e);
        if ((0 > d && 0 < e) || (360 > d && 360 < e) || (e < d)) {
            h.add(GeometryUtil.k(f, 0, g));
        }
        if (90 > d && 90 < e) {
            h.add(GeometryUtil.k(f, 90, g));
        }
        if (180 > d && 180 < e) {
            h.add(GeometryUtil.k(f, 180, g));
        }
        if (270 > d && 270 < e) {
            h.add(GeometryUtil.k(f, 270, g));
        }
        let i = 1.7976931348623157E+308;
        let j = 1.7976931348623157E+308;
        let k = -1.7976931348623157E+308;
        let l = -1.7976931348623157E+308;
        for (let m of fromEnum(h)) {
            i = Math.min(i, m.x);
            j = Math.min(j, m.y);
            k = Math.max(k, m.x);
            l = Math.max(l, m.y);
        }
        return new Rect(0, i, j, k - i, l - j);
    }
    am() {
        let a = new PieSliceVisualData();
        a.appearance = new PrimitiveAppearanceData();
        a.labelAppearance = new LabelAppearanceData();
        a.leaderLineAppearance = new PrimitiveAppearanceData();
        a.startAngle = this.a3;
        a.endAngle = this.ax;
        a.explodedOrigin = PointData.b(this.b5);
        a.explodedRadius = this.ay;
        a.origin = PointData.b(this.b7);
        a.radius = this.az();
        a.index = this.a5;
        a.innerExtentEnd = this.a0;
        a.innerExtentStart = this.a1;
        a.isExploded = this.isExploded;
        a.isOthersSlice = this.isOthersSlice;
        a.isSelected = this.isSelected;
        if (this.aj != null) {
            a.labelAppearance.angle = this.aj.ai;
            a.labelBounds = RectData.b(this.aj.ao);
            a.labelAngle = this.aj.ai;
            if (typeCast(TextBlock.$, this.aj.aj) !== null) {
                let b = typeCast(TextBlock.$, this.aj.aj);
                a.label = b.ak;
                let c = null;
                a.labelAppearance.labelBrush = AppearanceHelper.b(b.am);
                a.labelAppearance.text = b.ak;
                if (this.ai != null) {
                    c = this.ai.a6();
                    if (c == null) {
                        c = this.ai.as.b.c;
                    }
                }
                a.labelAppearance = AppearanceHelper.c(b, c);
                a.labelAppearance.visibility = this.aj._visibility == 0;
            }
            else {
                a.label = this.aj.aj.toString();
            }
            if (this.aj.ap != null) {
                a.leaderLineAppearance.fill = AppearanceHelper.b(this.aj.ap._fill);
                a.leaderLineAppearance.opacity = this.aj.ap._opacity;
                a.leaderLineAppearance.stroke = AppearanceHelper.b(this.aj.ap._stroke);
                a.leaderLineAppearance.strokeThickness = this.aj.ap.ac;
                a.leaderLineAppearance.isVisible = this.aj.ap._visibility == 0;
                a.leaderLinePath = new PathVisualData(2, "leaderLine", this.aj.ap);
            }
            else if (this.aj.aq != null) {
                a.leaderLineAppearance.fill = AppearanceHelper.b(this.aj.aq._fill);
                a.leaderLineAppearance.opacity = this.aj.aq._opacity;
                a.leaderLineAppearance.stroke = AppearanceHelper.b(this.aj.aq._stroke);
                a.leaderLineAppearance.strokeThickness = this.aj.aq.ac;
                a.leaderLineAppearance.isVisible = this.aj.aq._visibility == 0;
                a.leaderLinePath = new PathVisualData(1, "leaderLine", this.aj.aq);
            }
        }
        if (this.al != null) {
            let d = this.al.f();
            a.slicePath = new PathVisualData(1, "slicePath", d);
            a.appearance.fill = AppearanceHelper.b(d._fill);
            a.appearance.opacity = d._opacity;
            a.appearance.stroke = AppearanceHelper.b(d._stroke);
            a.appearance.strokeThickness = d.ac;
            a.appearance.isVisible = d._visibility == 0;
        }
        return a;
    }
    static b4(a, b, c, d, e) {
        if (d == 0) {
            return c;
        }
        let f = Math.cos(a);
        let g = Math.sin(a);
        let h = Math.sqrt(d * d / (1 - (b * Math.pow(f, 2))));
        h *= (e / 100);
        return { $type: Point_$type, x: h * f + c.x, y: h * g + c.y };
    }
}
Slice.$t = markType(Slice, 'Slice', ContentControl.$);
Slice.bv = DependencyProperty.i("IsSelected", Boolean_$type, Slice.$, new PropertyMetadata(2, false, (a, b) => {
    let c = typeCast(Slice.$, a);
    if (!c.av) {
        let d = b.newValue;
        if (c.ai != null) {
            c.ai.fi(c, d);
        }
    }
}));
Slice.bt = DependencyProperty.i("IsExploded", Boolean_$type, Slice.$, new PropertyMetadata(2, false, (a, b) => {
    let c = typeCast(Slice.$, a);
    let d = b.newValue;
    if (c.ai != null) {
        c.ai.ek(c, d);
    }
}));
Slice.bu = DependencyProperty.i("IsOtherSlice", Boolean_$type, Slice.$, new PropertyMetadata(2, false, (a, b) => a.bk()));
Slice.bp = DependencyProperty.i("Foreground", Brush.$, Slice.$, new PropertyMetadata(1, null));
Type.dep(DependencyProperty, PropertyMetadata, Slice, 'bk', ['EndAngle:ax:bm', [1, 0], 'ExplodedOrigin:b5:bn', [Point_$type, PointUtil.create()], 'ExplodedRadius:ay:bo', [1, 0], 'Index:a5:bq', [1, -1], 'InnerExtentEnd:a0:br', [1, 0], 'InnerExtentStart:a1:bs', [1, 0], 'Origin:b7:bw', [Point_$type, PointUtil.create()], 'Radius:a2:bx', [1, 0], 'StartAngle:a3:by', [1, 0], 'StrokeThickness:a4:bz', [1, 1]]);
/**
 * @hidden
 */
export class PieChartBaseView extends Base {
    constructor(a) {
        super();
        this.a = null;
        this.bq = null;
        this.b = null;
        this.h = null;
        this.j = null;
        this.i = null;
        this.k = null;
        this.c = null;
        this.b0 = null;
        this.l = false;
        this.bx = Rect.empty;
        this.f = null;
        this.u = null;
        this.t = null;
        this.v = 0;
        this.bw = null;
        this.ag = null;
        this.s = false;
        this.e = null;
        this.a = a;
        this.c = new PieChartViewManager(this);
    }
    az() {
        this.e = new PieSliceDataContext();
        this.e.series = this.a;
        this.b0 = Rect.empty;
        this.u = new List$1(Slice.$, 0);
        this.t = new List$1(PieLabel.$, 0);
        this.a.legendItemTemplate = ((() => {
            let $ret = new DataTemplate();
            $ret.render = PieLegendTemplates.f;
            $ret.measure = PieLegendTemplates.a;
            return $ret;
        })());
        this.a.legendItemBadgeTemplate = ((() => {
            let $ret = new DataTemplate();
            $ret.render = PieLegendTemplates.e;
            $ret.measure = PieLegendTemplates.d;
            return $ret;
        })());
        this.c.aa();
    }
    a9() {
        let a = this.c.b();
        if (this.a.ag == null) {
            this.a.ag = a.a;
        }
        if (this.a.ah == null) {
            this.a.ah = a.b;
        }
        if (this.a.g6 == null) {
            this.a.g6 = ((() => {
                let $ret = new Style();
                $ret.fill = a.h != null ? a.h.color.colorString : null;
                $ret.stroke = a.i != null ? a.i.color.colorString : null;
                $ret.strokeThickness = a.e;
                $ret.opacity = a.d;
                return $ret;
            })());
        }
        this.b = a;
        if (this.a.eb != null) {
            this.b.c = this.a.a6();
        }
    }
    ax(a) {
        if (a == null) {
            this.l = false;
            this.c.ac(null);
            this.h = null;
            this.j = null;
            this.i = null;
            this.k = null;
            return;
        }
        this.c.ac(a);
        this.h = this.c.e;
        this.j = this.c.g;
        this.i = this.c.f;
        this.k = this.c.h;
    }
    by() {
        return this.b0;
    }
    bv() {
        return this.c.am();
    }
    a8() {
        this.b0 = this.c.an();
        this.a.e2();
    }
    av() {
        if (!this.l) {
            this.l = true;
            this.c.ag(runOn(this, this.bg));
        }
    }
    bg() {
        if (this.l) {
            this.l = false;
            this.a3();
        }
    }
    a3() {
        this.am();
        this.c.x(this.b0.width, this.b0.height);
        if (this.k != null && this.k.d) {
            if (this.a.cd != 1) {
                this.k.z();
                this.k.aa(this.a.cd, this.a.cd);
            }
            if (!this.bx.isEmpty) {
                this.k.k(this.bx.left, this.bx.top, this.bx.width, this.bx.height);
            }
        }
        if (this.i != null && this.i.d) {
            if (this.a.cd != 1) {
                this.i.z();
                this.i.aa(this.a.cd, this.a.cd);
            }
            if (!this.bx.isEmpty) {
                this.i.k(this.bx.left, this.bx.top, this.bx.width, this.bx.height);
            }
        }
        if (this.j != null && this.j.d) {
            if (this.a.cd != 1) {
                this.j.z();
                this.j.aa(this.a.cd, this.a.cd);
            }
            if (!this.bx.isEmpty) {
                this.j.k(this.bx.left, this.bx.top, this.bx.width, this.bx.height);
            }
        }
        if (this.h != null && this.h.d) {
            if (this.a.cd != 1) {
                this.h.z();
                this.h.aa(this.a.cd, this.a.cd);
            }
            if (!this.bx.isEmpty) {
                this.h.k(this.bx.left, this.bx.top, this.bx.width, this.bx.height);
            }
        }
        this.a4();
        if (!this.bx.isEmpty && this.a.cd != 1) {
            if (this.k != null && this.k.d) {
                this.k.y();
            }
            if (this.i != null && this.i.d) {
                this.i.y();
            }
            if (this.j != null && this.j.d) {
                this.j.y();
            }
            if (this.h != null && this.h.d) {
                this.h.y();
            }
        }
    }
    a4() {
        this.bx = this.b0;
        if (this.j != null && this.j.d) {
            this.a7();
            this.a5();
        }
        if (this.i != null && this.i.d) {
            let a = this.b.c;
            if (this.a.eb != null) {
                a = this.a.a6();
            }
            this.i.ac(a);
            for (let b of fromEnum(this.t)) {
                if (b._visibility == 0) {
                    let c = b.aj;
                    if (c != null) {
                        {
                            if (b.ah.b2 == null) {
                                if (b.ah.au) {
                                    c.am = this.b.f;
                                }
                                else {
                                    c.am = this.b.g;
                                }
                            }
                            else {
                                c.am = b.ah.b2;
                            }
                            this.i.w(c);
                        }
                    }
                    let d = b.ah.b0;
                    if (b.ah.y != null) {
                        let e = new Path();
                        this.j.i(e, b.ah.y);
                        if (e._fill != null) {
                            d = e._fill;
                        }
                    }
                    if (this.a.ap != 0 && b.aq != null) {
                        b.aq._stroke = d;
                        b.aq.ac = 1;
                        if (this.a.g4 != null) {
                            this.i.i(b.aq, this.a.g4);
                        }
                        this.i.s(b.aq);
                    }
                    if (this.a.ap == 0 && b.ap != null) {
                        b.ap._stroke = d;
                        b.ap.ac = 1;
                        if (this.a.g4 != null) {
                            this.i.i(b.ap, this.a.g4);
                        }
                        this.i.r(b.ap);
                    }
                }
            }
        }
    }
    a6(a) {
        if (a._visibility != 0) {
            return;
        }
        let b = a.al.f();
        this.c.ah(b);
        b.ac = 1;
        b._opacity = a._opacity;
        if (a.b0 != null) {
            b._fill = a.b0;
        }
        if (a.b1 != null) {
            b._stroke = a.b1;
        }
        if (a.y != null) {
            this.j.i(b, a.y);
        }
        if (b.j != null) {
            this.j.z();
            this.j.j(b.j);
        }
        let c = b._fill;
        let d = b._stroke;
        let e = this.m(a, b, this.j);
        this.j.s(b);
        if (e) {
            b._fill = c;
            b._stroke = d;
        }
        if (b.j != null) {
            this.j.y();
        }
    }
    a7() {
        for (let a of fromEnum(this.u)) {
            if (!a.isSelected) {
                this.a6(a);
            }
        }
    }
    a5() {
        for (let a of fromEnum(this.u)) {
            if (a.isSelected) {
                this.a6(a);
            }
        }
    }
    m(a, b, c) {
        return false;
    }
    aj(a, b, c) {
        let d = this.f;
        this.bh(a);
        if (this.f == null && d != null) {
            let e = ((() => {
                let $ret = new MouseEventArgs();
                $ret.position = a;
                return $ret;
            })());
            this.a.en(this.f, e);
        }
        else if (this.f != null) {
            let f = ((() => {
                let $ret = new MouseEventArgs();
                $ret.position = a;
                return $ret;
            })());
            this.a.eo(this.f, f);
        }
    }
    ai(a) {
        let b = this.f;
        this.bh(a);
        let c = ((() => {
            let $ret = new MouseEventArgs();
            $ret.position = a;
            return $ret;
        })());
        this.a.en(b, c);
    }
    ah(a) {
        this.bh(a);
    }
    ak(a) {
        this.ap(a);
    }
    ap(a) {
        let b = ((() => {
            let $ret = new MouseButtonEventArgs();
            $ret.position = a;
            return $ret;
        })());
        let c = true;
        let d = null;
        for (let e of fromEnum(this.t)) {
            let f = e.ao.containsPoint(a);
            if (f) {
                d = e;
                c = this.a.bw(e.ah, b);
                break;
            }
        }
        this.bh(a);
        if (this.f == null && c && d != null) {
            this.a.fm(d.ah, b);
        }
        else if (this.f != null && c) {
            this.a.fm(this.f, b);
        }
    }
    bh(a) {
        this.f = null;
        for (let b = 0; b < this.u.count; b++) {
            if (this.u._inner[b]._visibility == 0 && this.u._inner[b].ap(a)) {
                this.f = this.u._inner[b];
                break;
            }
        }
    }
    n(a) {
        for (let b of fromEnum(this.t)) {
            let c = b.ao.containsPoint(a);
            if (c) {
                return true;
            }
        }
        return false;
    }
    p(a) {
        for (let b = 0; b < this.u.count; b++) {
            if (this.u._inner[b]._visibility == 0 && this.u._inner[b].ap(a)) {
                return true;
            }
        }
        return false;
    }
    o(a) {
        return this.n(a) || this.p(a);
    }
    bb() {
        this.a.e2();
    }
    al() {
        this.aq();
    }
    bl(a, b) {
        let c = b;
        this.bm(c.position, a, b);
    }
    g() {
        let a = new Slice();
        a.ai = this.a;
        this.u.add(a);
        return a;
    }
    bc(a) {
        a._visibility = 0;
    }
    be(a) {
        a._visibility = 1;
    }
    bd(a) {
        a.ai = null;
        this.u.remove(a);
    }
    d() {
        let a = new PieLabel();
        if (this.a.ap == 0) {
            a.ap = new Line();
        }
        else {
            a.aq = new Path();
            let b = new PathGeometry();
            b.b.add(new PathFigure());
            b.b._inner[0]._segments.add(new BezierSegment(0));
            a.aq.aj = b;
        }
        this.t.add(a);
        return a;
    }
    ar(a) {
        a._visibility = 0;
    }
    at(a) {
        a._visibility = 1;
    }
    as(a) {
        this.t.remove(a);
    }
    bj(a) {
        if ((this.a.ap != 0 && a.aq != null) || (this.a.ap == 0 && a.ap != null)) {
            return;
        }
        if (this.a.ap == 0) {
            a.aq = null;
            a.ap = new Line();
        }
        else {
            a.ap = null;
            a.aq = new Path();
            let b = new PathGeometry();
            let c = new PathFigure();
            c._startPoint = { $type: Point_$type, x: NaN, y: NaN };
            b.b.add(c);
            b.b._inner[0]._segments.add(new BezierSegment(0));
            a.aq.aj = b;
        }
    }
    ba(a) {
        if (this.a.g5 != null && a.isOthersSlice) {
            a.b0 = this.bt(a.a5);
            a.b1 = this.bu(a.a5);
            a.y = this.a.g5;
        }
        else if (a.isSelected && this.a.bq && this.a.g6 != null) {
            a.y = this.a.g6;
        }
        else {
            a.y = null;
            a.b0 = this.bt(a.a5);
            a.b1 = this.bu(a.a5);
        }
    }
    bt(a) {
        if (this.a.ag != null && a >= 0 && this.a.ag.count > 0) {
            return this.a.ag.item(a % this.a.ag.count);
        }
        return null;
    }
    bu(a) {
        if (this.a.ah != null && a >= 0 && this.a.ah.count > 0) {
            return this.a.ah.item(a % this.a.ah.count);
        }
        return null;
    }
    ad(a) {
        if (a == null || a.aj == null) {
            return this.a.di;
        }
        let b = a.aj;
        let c = b.aj;
        if (typeCast(TextBlock.$, c) !== null) {
            return c.ak;
        }
        return c;
    }
    y(a) {
        let b = typeCast(TextBlock.$, a);
        if (b != null && b.ak != null && this.i != null) {
            return this.i.e(b) + 0;
        }
        return 0;
    }
    bi() {
        this.v = FontUtil.getCurrentFontHeight(this.c.o, this.c.d());
    }
    x(a) {
        return this.v + 0;
    }
    bz(a) {
        let b = a.aj;
        if (typeCast(TextBlock.$, b) !== null) {
            let c = this.y(b);
            let d = this.x(b);
            return new Rect(0, 0, 0, c, d);
        }
        return Rect.empty;
    }
    b1() {
        return new Size(1, this.b0.width, this.b0.height);
    }
    bk(a, b, c) {
        let d = typeCast(TextBlock.$, a.aj);
        if (d != null) {
            d.n = b;
            d.o = c;
        }
        this.av();
    }
    bn(a) {
        this.bo(a);
    }
    bp() {
        this.av();
    }
    au() {
        this.am();
    }
    am() {
        let a = this.b.c;
        if (this.a.eb != null) {
            a = this.a.a6();
        }
        if (this.i != null) {
            this.i.ac(a);
        }
    }
    ay(a, b) {
        this.c.ad(a, b);
    }
    aw() {
        this.c.ab();
    }
    get af() {
        return this.c != null && this.c.o != null ? this.c.o.getResourceString("invalidLabelBinding") : "There is no such value for the labels to bind.";
    }
    aq() {
        if (this.s == true) {
            this.s = false;
            this.c.z(this.e);
        }
    }
    bo(a) {
        this.c.aj(a);
    }
    bm(a, b, c) {
        this.s = true;
        this.bw = a;
        if (b == null) {
            this.ag = null;
            this.e.item = null;
        }
        else {
            this.ag = b.dataContext;
            this.e.item = b.dataContext;
            this.e.slice = b;
        }
        let d = this.a.g0();
        let pos_ = { $type: Point_$type, x: a.x + 15 + d.x, y: a.y + 15 + d.y };
        let context_ = this.e;
        this.c.ai(pos_, context_);
    }
    bf() {
        this.a9();
        this.bi();
        this.a.fc();
    }
    a0() {
        this.c.ae(this.a.bv);
    }
    ac() {
        return this.c.u();
    }
    ao() {
        if (this.l) {
            this.bg();
        }
    }
    r() {
        return this.ac() != null;
    }
    an(a) {
        this.c.y(a);
    }
    q() {
        return true;
    }
    a2() {
    }
    br(a) {
        let b = this.c.ak(a.y);
        if (b != null) {
            return b;
        }
        return a.b0;
    }
    bs(a) {
        let b = this.c.al(a.y);
        if (b != null) {
            return b;
        }
        return a.b1;
    }
    w(a) {
        return this.c.m(a.y);
    }
    z(a) {
        return this.c.m(a);
    }
    aa() {
        for (let a of fromEnum(this.u)) {
            if (!a.isSelected) {
                return a.a4;
            }
        }
        return NaN;
    }
    a1() {
        this.c.af();
        this.av();
    }
    ae(a) {
        return this.c.c.am;
    }
}
PieChartBaseView.$t = markType(PieChartBaseView, 'PieChartBaseView');
/**
 * @hidden
 */
export class PieSliceDataContext extends DataContext {
    constructor() {
        super(...arguments);
        this._slice = null;
        this._percentValue = 0;
        this._isOthersSlice = false;
    }
    get slice() {
        return this._slice;
    }
    set slice(a) {
        this._slice = a;
    }
    get percentValue() {
        return this._percentValue;
    }
    set percentValue(a) {
        this._percentValue = a;
    }
    get isOthersSlice() {
        return this._isOthersSlice;
    }
    set isOthersSlice(a) {
        this._isOthersSlice = a;
    }
    flatten() {
        let ret_ = super.flatten();
        let percentValue_ = this.percentValue;
        let isOthersSlice_ = this.isOthersSlice;
        ret_.percentValue = percentValue_;
        ret_.isOthersSlice = isOthersSlice_;
        if (isOthersSlice_ && this.item != null) {
            let items_ = [];
            for (let currItem_ of fromEn(this.item)) {
                items_.push(currItem_);
            }
            ret_.item = items_;
        }
        return ret_;
    }
}
PieSliceDataContext.$t = markType(PieSliceDataContext, 'PieSliceDataContext', DataContext.$);
/**
 * @hidden
 */
export class SliceView extends Base {
    constructor(a) {
        super();
        this.a = null;
        this.g = null;
        this.a = a;
    }
    b() {
        this.g = new Path();
        this.a.content = this.g;
    }
    c() {
        this.a.bk();
    }
    f() {
        return this.g;
    }
    d(a, b) {
        this.g.j = ((() => {
            let $ret = new TranslateTransform();
            $ret.j = a;
            $ret.k = b;
            return $ret;
        })());
    }
    e() {
        this.g.j = null;
    }
}
SliceView.$t = markType(SliceView, 'SliceView');
