// igniteui-react-charts/ES2015/PolarScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PolarBase } from "./PolarBase";
import { markType } from "igniteui-react-core/ES2015/type";
import { PolarBaseView } from "./PolarBaseView";
import { Color } from "igniteui-react-core/ES2015/Color";
/**
 * @hidden
 */
export class PolarScatterSeries extends PolarBase {
    constructor() {
        super();
        this.ab = PolarScatterSeries.$;
    }
    bf() {
        return new PolarScatterSeriesView(this);
    }
    it(a) {
        super.it(a);
        this.oy = a;
    }
}
PolarScatterSeries.$t = markType(PolarScatterSeries, 'PolarScatterSeries', PolarBase.$);
/**
 * @hidden
 */
export class PolarScatterSeriesView extends PolarBaseView {
    constructor(a) {
        super(a);
        this.cy = null;
        this.cy = a;
    }
    ap() {
        super.ap();
        if (!this.n) {
            this.e.bj = 1;
        }
    }
    ac() {
        let a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 3;
        this.e.ew = 2;
        this.e.ex = 2;
        this.e.du = false;
    }
}
PolarScatterSeriesView.$t = markType(PolarScatterSeriesView, 'PolarScatterSeriesView', PolarBaseView.$);
