// igniteui-react-charts/ES2015/PolarSplineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { DependencyProperty } from "igniteui-react-core/ES2015/DependencyProperty";
import { Type, markType } from "igniteui-react-core/ES2015/type";
import { PropertyMetadata } from "igniteui-react-core/ES2015/PropertyMetadata";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Path } from "igniteui-react-core/ES2015/Path";
import { PathGeometry } from "igniteui-react-core/ES2015/PathGeometry";
import { PathFigure } from "igniteui-react-core/ES2015/PathFigure";
import { Extensions } from "igniteui-react-core/ES2015/Extensions";
import { Numeric } from "igniteui-react-core/ES2015/Numeric";
import { Color } from "igniteui-react-core/ES2015/Color";
/**
 * @hidden
 */
export class PolarSplineSeries extends PolarLineSeriesBase {
    constructor() {
        super();
        this.ab = PolarSplineSeries.$;
    }
    bf() {
        return new PolarSplineSeriesView(this);
    }
    it(a) {
        super.it(a);
        this.pc = a;
    }
    iy(a, b, c, d) {
        super.iy(a, b, c, d);
        switch (b) {
            case PolarSplineSeries.$$p[0]:
                this.jc(false);
                break;
        }
    }
    pa(a) {
        let b = a;
        b.c0();
    }
    pb(a, b) {
        this.pa(b);
        let c = b;
        c.c1(a.m, this.pd);
    }
}
PolarSplineSeries.$t = markType(PolarSplineSeries, 'PolarSplineSeries', PolarLineSeriesBase.$);
Type.dep(DependencyProperty, PropertyMetadata, PolarSplineSeries, 'raisePropertyChanged', ['Stiffness:pd:pf', [1, 0.5]]);
/**
 * @hidden
 */
export class PolarSplineSeriesView extends PolarLineSeriesBaseView {
    constructor(a) {
        super(a);
        this.cz = null;
        this.c2 = ((() => {
            let $ret = new Path();
            $ret.aj = new PathGeometry();
            return $ret;
        })());
        this.cz = a;
    }
    ap() {
        super.ap();
        let a = new PathGeometry();
        a.b.add(((() => {
            let $ret = new PathFigure();
            return $ret;
        })()));
        this.c2.aj = a;
    }
    c0() {
        Extensions.e(this.c2.aj);
        this.an();
    }
    c1(a, b) {
        this.c2.aj.b = Numeric.h(a.count, (c) => a._inner[c].x, (c) => a._inner[c].y, b);
        this.an();
    }
    a8() {
        super.a8();
        this.c2._stroke = this.e.k2;
        this.c2.ac = this.e.ey;
        this.c2.ai = this.e.la;
        this.c2.ad = this.e.lb;
    }
    a9() {
        super.a9();
        let a = this.bl();
        this.c2._stroke = a;
        this.c2.ac = this.e.ey + 3;
    }
    a3(a, b) {
        super.a3(a, b);
        a.s(this.c2);
    }
    ai(a) {
        super.ai(a);
        this.e.a2.exportPathData(a, this.c2, "polarShape", ["Main"]);
    }
    ac() {
        let a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 3;
        this.e.ew = 1;
        this.e.ex = 4;
        this.e.du = false;
    }
}
PolarSplineSeriesView.$t = markType(PolarSplineSeriesView, 'PolarSplineSeriesView', PolarLineSeriesBaseView.$);
