// igniteui-react-charts/ES2015/RateOfChangeAndMomentumIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RateOfChangeAndMomentumIndicatorStrategy } from "./RateOfChangeAndMomentumIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class RateOfChangeAndMomentumIndicator extends StrategyBasedIndicator {
    get_qb() {
        return new RateOfChangeAndMomentumIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return RateOfChangeAndMomentumIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(RateOfChangeAndMomentumIndicator.qx));
    }
    set qw(a) {
        this.h(RateOfChangeAndMomentumIndicator.qx, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
}
RateOfChangeAndMomentumIndicator.$t = markType(RateOfChangeAndMomentumIndicator, 'RateOfChangeAndMomentumIndicator', StrategyBasedIndicator.$);
RateOfChangeAndMomentumIndicator.qx = StrategyBasedIndicator.qt(14, RateOfChangeAndMomentumIndicator.$);
