// igniteui-react-charts/ES2015/RelativeStrengthIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RelativeStrengthIndexIndicatorStrategy } from "./RelativeStrengthIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class RelativeStrengthIndexIndicator extends StrategyBasedIndicator {
    constructor() {
        super();
        this.pj = 1;
    }
    get_qb() {
        return new RelativeStrengthIndexIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return RelativeStrengthIndexIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(RelativeStrengthIndexIndicator.qx));
    }
    set qw(a) {
        this.h(RelativeStrengthIndexIndicator.qx, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
}
RelativeStrengthIndexIndicator.$t = markType(RelativeStrengthIndexIndicator, 'RelativeStrengthIndexIndicator', StrategyBasedIndicator.$);
RelativeStrengthIndexIndicator.qx = StrategyBasedIndicator.qt(14, RelativeStrengthIndexIndicator.$);
