// igniteui-react-charts/ES2015/RingSeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RingSeriesBase } from "./RingSeriesBase";
import { Ring } from "./Ring";
import { ArcItem } from "./ArcItem";
import { fromEnum, markType } from "igniteui-react-core/ES2015/type";
import { RingCollection } from "./RingCollection";
/**
 * @hidden
 */
export class RingSeries extends RingSeriesBase {
    constructor() {
        super();
        this.dr = null;
        this.dr = ((() => {
            let $ret = new Ring();
            $ret.c = this;
            $ret.e = false;
            return $ret;
        })());
        let a = ((() => {
            let $ret = new ArcItem();
            $ret.k = this.a4;
            $ret.d = this.dr;
            $ret.n = this.b7;
            $ret.c = this.al;
            $ret.j = this.a2;
            return $ret;
        })());
        this.dr.g.add(a);
        this.ab = RingSeries.$;
    }
    cj(a, b, c, d) {
        super.cj(a, b, c, d);
        if (b == RingSeriesBase.$$p[2]) {
            for (let e = 0; e < this.dr.ringControl.arcs.count; e++) {
                this.dr.ringControl.arcs.item(e).at = d;
            }
        }
        if (b == RingSeriesBase.$$p[3]) {
            for (let f = 0; f < this.dr.ringControl.arcs.count; f++) {
                this.dr.ringControl.arcs.item(f).au = d;
            }
        }
        if (b == RingSeriesBase.$$p[14]) {
            for (let g = 0; g < this.dr.ringControl.arcs.count; g++) {
                this.dr.ringControl.arcs.item(g).dv = this.bp;
            }
        }
        if (b == RingSeriesBase.$$p[0]) {
            this.ch();
        }
        if (b == RingSeriesBase.$$p[23]) {
            if (this.dr.g != null && this.dr.g.count > 0) {
                this.dr.g._inner[0].k = this.a4;
            }
        }
        if (b == RingSeriesBase.$$p[26]) {
            if (this.dr.g != null && this.dr.g.count > 0) {
                this.dr.g._inner[0].n = this.b7;
                this.ci();
            }
        }
        if (this.dr != null && this.as.e()) {
            this.dr.o();
            if (this.dr.f()) {
                this.dr.c.as.l();
            }
        }
    }
    aq() {
        let a = new RingCollection();
        if (this.dr.g._inner[0].g.count > 0) {
            a.add(this.dr);
        }
        return a;
    }
    ci() {
        if (this.dr != null) {
            this.dr.g._inner[0].h = this.itemsSource;
            this.dr.g._inner[0].p(this.a4);
            if (this.at != null) {
                this.at.bv();
                this.at.b0();
            }
        }
    }
    ch() {
        if (this.dr != null) {
            this.dr.g._inner[0].a = this.ag;
        }
    }
    cn() {
        if (this.dr != null) {
            this.width = this.dr.q.width;
            this.height = this.dr.q.height;
            this.as.m(this.dr.p.x, this.dr.p.y);
        }
    }
    cm() {
        if (this.dr != null) {
            for (let a of fromEnum(this.dr.ringControl.arcs.a)) {
                a.fe();
            }
        }
    }
    cd(a, b) {
        super.cd(a, b);
        if (this.dr != null) {
            this.dr.l(this.itemsSource, a, b);
        }
    }
    cc() {
        super.cc();
        if (this.dr != null) {
            this.dr.k(this.itemsSource);
        }
    }
    ce(a, b) {
        super.ce(a, b);
        if (this.dr != null) {
            this.dr.m(this.itemsSource, a, b);
        }
    }
    cf(a, b, c) {
        super.cf(a, b, c);
        if (this.dr != null) {
            this.dr.n(this.itemsSource, a, b, c);
        }
    }
}
RingSeries.$t = markType(RingSeries, 'RingSeries', RingSeriesBase.$);
