// igniteui-react-charts/ES2015/SlowStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { SlowStochasticOscillatorIndicatorStrategy } from "./SlowStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
import { Brush } from "igniteui-react-core/ES2015/Brush";
/**
 * @hidden
 */
export class SlowStochasticOscillatorIndicator extends StrategyBasedIndicator {
    constructor() {
        super();
        this.trendLineBrush = ((() => {
            let $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        this.trendLineType = 10;
        this.trendLinePeriod = 3;
    }
    get_qb() {
        return new SlowStochasticOscillatorIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return SlowStochasticOscillatorIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(SlowStochasticOscillatorIndicator.qx));
    }
    set qw(a) {
        this.h(SlowStochasticOscillatorIndicator.qx, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
}
SlowStochasticOscillatorIndicator.$t = markType(SlowStochasticOscillatorIndicator, 'SlowStochasticOscillatorIndicator', StrategyBasedIndicator.$);
SlowStochasticOscillatorIndicator.qx = StrategyBasedIndicator.qt(14, SlowStochasticOscillatorIndicator.$);
