// igniteui-react-charts/ES2015/SplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SplineSeriesBase } from "./SplineSeriesBase";
import { CategoryLineRasterizer } from "./CategoryLineRasterizer";
import { typeCast, runOn, markType } from "igniteui-react-core/ES2015/type";
import { Path } from "igniteui-react-core/ES2015/Path";
import { ScalerParams } from "./ScalerParams";
import { SplineSeriesBaseView } from "./SplineSeriesBaseView";
import { ISortingAxis_$type } from "./ISortingAxis";
import { Color } from "igniteui-react-core/ES2015/Color";
/**
 * @hidden
 */
export class SplineAreaSeries extends SplineSeriesBase {
    constructor() {
        super();
        this.qq = null;
        this.qq = new CategoryLineRasterizer();
        this.ab = SplineAreaSeries.$;
    }
    bf() {
        return new SplineAreaSeriesView(this);
    }
    nf() {
        return 1;
    }
    it(a) {
        super.it(a);
        this.qr = a;
    }
    preferredCategoryMode(a) {
        return 0;
    }
    hv(a, b) {
        super.hv(a, b);
        let c = b;
        c.c8();
    }
    dh(a, b) {
        let c = this.lk(this.ln(a), true, false);
        if (this.p7(a, c, b)) {
            return true;
        }
        if (this.dg(a, b)) {
            return true;
        }
        if (this.l3(a, b)) {
            return true;
        }
        return false;
    }
    on(a, b) {
        super.on(a, b);
        let c = this.ne.cn.d;
        let d = typeCast(SplineAreaSeriesView.$, b);
        let e = a.f;
        let f = this.getEffectiveViewport1(b);
        this.nd.x(this, this.ny(), this.px, runOn(this, this.mz), this.n6(b), this.n7(b));
        let g = false;
        let h = this.nd.c;
        if (h != null) {
            g = true;
        }
        if (g) {
            let i = new ScalerParams(0, b.bs, b.br, this.px.b8, f);
            this.oj(e, -1, this.valueColumn.count, this.px, i, b.n);
        }
        let j = d.df;
        let k = d.dg;
        let l = d.di;
        let m = d.dk;
        this.nd.ae(j, true, false, true, true);
        this.nd.ae(k, true, false, true, true);
        this.nd.ae(l, false, true, false, false);
        this.nd.ae(m, false, true, false, false);
        l._opacity = this.nd.i * this.d6;
        m._opacity = 0.5 * this.nd.i * this.d6;
        if (b.cr(a)) {
            d.da(a.f.count, e, true, c, this.eu, (n, o, p, q, r) => this.ph(n, a.f.count, b), 1, a.f.count == 1 ? null : this.ow(e, e.count - 1, b.br, b.bs, true));
            b.cu(a);
        }
        let n = this.fetchYAxis();
    }
    get useHighMarkerFidelity() {
        return this.n1;
    }
}
SplineAreaSeries.$t = markType(SplineAreaSeries, 'SplineAreaSeries', SplineSeriesBase.$);
/**
 * @hidden
 */
export class SplineAreaSeriesView extends SplineSeriesBaseView {
    constructor(a) {
        super(a);
        this.c7 = null;
        this.dh = new Path();
        this.dl = new Path();
        this.dj = new Path();
        this.dm = new Path();
        this.db = new Path();
        this.dd = new Path();
        this.dc = new Path();
        this.de = new Path();
        this.c7 = a;
    }
    ap() {
        super.ap();
        if (!this.n) {
            this.e.bj = 1;
        }
    }
    get df() {
        return this.dl;
    }
    get dg() {
        return this.dm;
    }
    get di() {
        return this.dh;
    }
    get dk() {
        return this.dj;
    }
    c8() {
        this.dh.aj = null;
        this.dj.aj = null;
        this.dl.aj = null;
        this.dm.aj = null;
        this.an();
    }
    c9(a, b, c, d, e, f, g) {
        this.da(a, b, c, d, e, f, g, null);
    }
    da(a, b, c, d, e, f, g, h) {
        this.c7.qq.a = typeCast(ISortingAxis_$type, this.cp.fetchXAxis()) !== null ? true : false;
        this.c7.qq.j(this.dh, this.dl, this.dj, this.dm, a, b, c, d, e, f, g, h);
        this.an();
    }
    a9() {
        super.a9();
        this.db.aj = this.dh.aj;
        this.dd.aj = this.dl.aj;
        this.dc.aj = this.dj.aj;
        this.de.aj = this.dm.aj;
        let a = this.bl();
        this.db._fill = a;
        this.db._opacity = 1;
        this.dc._fill = a;
        this.dc._opacity = 1;
        this.dd._stroke = a;
        this.dd.ac = this.e.ey + 3;
        this.de._stroke = a;
        this.de.ac = this.e.ey + 3;
    }
    a3(a, b) {
        super.a3(a, b);
        if (b) {
            a.s(this.db);
            a.s(this.dc);
            a.s(this.dd);
            a.s(this.de);
        }
        else {
            a.s(this.dh);
            a.s(this.dj);
            a.s(this.dl);
            a.s(this.dm);
        }
    }
    ai(a) {
        super.ai(a);
        this.e.a2.exportPathData(a, this.dl, "lowerShape", ["Lower"]);
        this.e.a2.exportPathData(a, this.dm, "upperShape", ["Upper", "Main"]);
        this.e.a2.exportPathData(a, this.dj, "translucentShape", ["Translucent"]);
        this.e.a2.exportPathData(a, this.dh, "fillShape", ["Fill"]);
    }
    ac() {
        let a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.k9 = a;
        this.e.ev = 5;
        this.e.ew = 1;
        this.e.ex = -3;
    }
}
SplineAreaSeriesView.$t = markType(SplineAreaSeriesView, 'SplineAreaSeriesView', SplineSeriesBaseView.$);
