// igniteui-react-charts/ES2015/StandardDeviationIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StandardDeviationIndicatorStrategy } from "./StandardDeviationIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class StandardDeviationIndicator extends StrategyBasedIndicator {
    get_qb() {
        return new StandardDeviationIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return StandardDeviationIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(StandardDeviationIndicator.qx));
    }
    set qw(a) {
        this.h(StandardDeviationIndicator.qx, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
}
StandardDeviationIndicator.$t = markType(StandardDeviationIndicator, 'StandardDeviationIndicator', StrategyBasedIndicator.$);
StandardDeviationIndicator.qx = StrategyBasedIndicator.qt(20, StandardDeviationIndicator.$);
