// igniteui-react-charts/ES2015/TRIXIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { TRIXIndicatorStrategy } from "./TRIXIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core/ES2015/type";
import { Brush } from "igniteui-react-core/ES2015/Brush";
/**
 * @hidden
 */
export class TRIXIndicator extends StrategyBasedIndicator {
    constructor() {
        super();
        this.pj = 0;
        this.trendLineBrush = ((() => {
            let $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        this.trendLineType = 9;
        this.trendLinePeriod = 9;
    }
    get_qb() {
        return new TRIXIndicatorStrategy();
    }
    get qb() {
        return this.get_qb();
    }
    get_qo() {
        return TRIXIndicator.$;
    }
    get qo() {
        return this.get_qo();
    }
    get qw() {
        return typeGetValue(this.c(TRIXIndicator.qx));
    }
    set qw(a) {
        this.h(TRIXIndicator.qx, a);
    }
    qh() {
        return this.qw;
    }
    qq(a) {
        this.qw = a;
    }
}
TRIXIndicator.$t = markType(TRIXIndicator, 'TRIXIndicator', StrategyBasedIndicator.$);
TRIXIndicator.qx = StrategyBasedIndicator.qt(15, TRIXIndicator.$);
