// igniteui-react-charts/ES2015/XamPieChart_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PieChartBase } from "./PieChartBase";
import { PieChartBaseView } from "./PieChartBaseView";
import { markType } from "igniteui-react-core/ES2015/type";
/**
 * @hidden
 */
export class XamPieChart extends PieChartBase {
    constructor() {
        super();
        this.g9 = null;
        this.ab = XamPieChart.$;
    }
    ar() {
        return new XamPieChartView(this);
    }
    e4(a) {
        super.e4(a);
        this.g9 = a;
    }
}
XamPieChart.$t = markType(XamPieChart, 'XamPieChart', PieChartBase.$);
/**
 * @hidden
 */
export class XamPieChartView extends PieChartBaseView {
    constructor(a) {
        super(a);
        this.b2 = null;
        this.b2 = a;
    }
}
XamPieChartView.$t = markType(XamPieChartView, 'XamPieChartView', PieChartBaseView.$);
